<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (1.8.0_144) on Tue Sep 18 12:11:30 PDT 2018 -->
<title>Uses of Class biolockj.module.BioModuleImpl</title>
<meta name="date" content="2018-09-18">
<link rel="stylesheet" type="text/css" href="../../../stylesheet.css" title="Style">
<script type="text/javascript" src="../../../script.js"></script>
</head>
<body>
<script type="text/javascript"><!--
    try {
        if (location.href.indexOf('is-external=true') == -1) {
            parent.document.title="Uses of Class biolockj.module.BioModuleImpl";
        }
    }
    catch(err) {
    }
//-->
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar.top">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.top.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../overview-summary.html">Overview</a></li>
<li><a href="../package-summary.html">Package</a></li>
<li><a href="../../../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">Class</a></li>
<li class="navBarCell1Rev">Use</li>
<li><a href="../package-tree.html">Tree</a></li>
<li><a href="../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../index-files/index-1.html">Index</a></li>
<li><a href="../../../help-doc.html">Help</a></li>
<li>&nbsp;&nbsp;</li>
<li><a href="https://github.com/msioda/BioLockJ/wiki" target="_top"><strong>RETURN to WIKI</strong></a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li>Prev</li>
<li>Next</li>
</ul>
<ul class="navList">
<li><a href="../../../index.html?biolockj/module/class-use/BioModuleImpl.html" target="_top">Frames</a></li>
<li><a href="BioModuleImpl.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip.navbar.top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<div class="header">
<h2 title="Uses of Class biolockj.module.BioModuleImpl" class="title">Uses of Class<br>biolockj.module.BioModuleImpl</h2>
</div>
<div class="classUseContainer">
<ul class="blockList">
<li class="blockList">
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing packages, and an explanation">
<caption><span>Packages that use <a href="../../../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Package</th>
<th class="colLast" scope="col">Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><a href="#biolockj.module">biolockj.module</a></td>
<td class="colLast">
<div class="block">Pipeline modules are loaded based on the <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s listed in the <a href="../../../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>
 file.<br>
 All modules (i.e., implement <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>) are contained within this package or its
 sub-packages.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><a href="#biolockj.module.classifier">biolockj.module.classifier</a></td>
<td class="colLast">
<div class="block">This package contains Classifier specific <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s that build bash scripts to generate
 taxonomy reports from the sequences files.<br></div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><a href="#biolockj.module.classifier.r16s">biolockj.module.classifier.r16s</a></td>
<td class="colLast">
<div class="block">This package contains RDP and QIIME classifier modules that assign taxonomy to 16s sequences.<br>
 Output file formats are classifier specific so classifier requires a unique post-requisite
 <a href="../../../biolockj/module/implicit/parser/ParserModule.html" title="interface in biolockj.module.implicit.parser"><code>ParserModule</code></a>.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><a href="#biolockj.module.classifier.wgs">biolockj.module.classifier.wgs</a></td>
<td class="colLast">
<div class="block">This package contains Kraken, MetaPhlAn, and SLIMM classifier modules that assign taxonomy to WGS data.<br>
 As with all BioLockJ classifiers, these modules generate raw classifier output files from the sequence data.<br>
 Include the corresponding <a href="../../../biolockj/module/implicit/parser/ParserModule.html" title="interface in biolockj.module.implicit.parser"><code>ParserModule</code></a> in the <a href="../../../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> file to
 generate standardized OTU abundance tables.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><a href="#biolockj.module.implicit">biolockj.module.implicit</a></td>
<td class="colLast">
<div class="block">The modules in this package are implicitly added to pipelines as needed.<br>
 These modules cannot be directly added to any pipeline unless overridden via project.allowImplicitModules=Y.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><a href="#biolockj.module.implicit.parser">biolockj.module.implicit.parser</a></td>
<td class="colLast">
<div class="block">This package contains Parser BioModules in the r16s and WGS sub-packages that are paired with a
 <a href="../../../biolockj/module/classifier/ClassifierModule.html" title="interface in biolockj.module.classifier"><code>ClassifierModule</code></a> via
 <a href="../../../biolockj/module/BioModule.html#getPostRequisiteModules--"><code>BioModule.getPostRequisiteModules()</code></a> to run immediately after the
 classifier.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><a href="#biolockj.module.implicit.parser.r16s">biolockj.module.implicit.parser.r16s</a></td>
<td class="colLast">
<div class="block">This package contains Parser <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s that convert the 16S taxonomy reports generated by
 16S classifiers (such as RDP and QIIME) into standardized OTU abundance tables.<br></div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><a href="#biolockj.module.implicit.parser.wgs">biolockj.module.implicit.parser.wgs</a></td>
<td class="colLast">
<div class="block">This package contains Parser <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s that convert the WGS taxonomy reports generated by
 WGS classifiers (such as Kraken, MetaPhlAn, or SLIMM) into standardized OTU abundance tables.<br></div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><a href="#biolockj.module.implicit.qiime">biolockj.module.implicit.qiime</a></td>
<td class="colLast">
<div class="block">This package contains <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s that are implicitly added to QIIME pipeline as needed.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><a href="#biolockj.module.r">biolockj.module.r</a></td>
<td class="colLast">
<div class="block">This package contains <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s that build pipeline reports from the standard OTU abundance
 tables by generating R scripts to produce the summary statistics and data visualizations output to PDF files.<br></div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><a href="#biolockj.module.report">biolockj.module.report</a></td>
<td class="colLast">
<div class="block">This package contains <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s that normalize OTU abundance tables output by Parser
 modules, merges them with the metadata, and generates various reports and notifications.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><a href="#biolockj.module.seq">biolockj.module.seq</a></td>
<td class="colLast">
<div class="block"><a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s used to prepare sequence files or update the metadata prior to classification.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList">
<ul class="blockList">
<li class="blockList"><a name="biolockj.module">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a> in <a href="../../../biolockj/module/package-summary.html">biolockj.module</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing subclasses, and an explanation">
<caption><span>Subclasses of <a href="../../../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a> in <a href="../../../biolockj/module/package-summary.html">biolockj.module</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/JavaModuleImpl.html" title="class in biolockj.module">JavaModuleImpl</a></span></code>
<div class="block">Superclass for Java BioModules that will be called in separate instances of the application.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/ScriptModuleImpl.html" title="class in biolockj.module">ScriptModuleImpl</a></span></code>
<div class="block">Superclass for Java BioModules that will be called in separate instances of the application.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module.classifier">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a> in <a href="../../../biolockj/module/classifier/package-summary.html">biolockj.module.classifier</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing subclasses, and an explanation">
<caption><span>Subclasses of <a href="../../../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a> in <a href="../../../biolockj/module/classifier/package-summary.html">biolockj.module.classifier</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/classifier/ClassifierModuleImpl.html" title="class in biolockj.module.classifier">ClassifierModuleImpl</a></span></code>
<div class="block">This is the superclass for all WGS and 16S biolockj.module.classifier BioModules.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module.classifier.r16s">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a> in <a href="../../../biolockj/module/classifier/r16s/package-summary.html">biolockj.module.classifier.r16s</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing subclasses, and an explanation">
<caption><span>Subclasses of <a href="../../../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a> in <a href="../../../biolockj/module/classifier/r16s/package-summary.html">biolockj.module.classifier.r16s</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/classifier/r16s/QiimeClosedRefClassifier.html" title="class in biolockj.module.classifier.r16s">QiimeClosedRefClassifier</a></span></code>
<div class="block">This BioModule executes the QIIME script pick_closed_reference_otus.py on a FastA sequence files.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/classifier/r16s/QiimeDeNovoClassifier.html" title="class in biolockj.module.classifier.r16s">QiimeDeNovoClassifier</a></span></code>
<div class="block">This module runs the QIIME pick_de_novo_otus.py script on FastA sequence files in a single script so it is important
 to allocate sufficient job resources if running in a clustered environment.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/classifier/r16s/QiimeOpenRefClassifier.html" title="class in biolockj.module.classifier.r16s">QiimeOpenRefClassifier</a></span></code>
<div class="block">This module runs the QIIME pick_open_reference_otus.py script on FastA sequence files in a single script so it is
 important to allocate sufficient job resources if running in a clustered environment.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/classifier/r16s/RdpClassifier.html" title="class in biolockj.module.classifier.r16s">RdpClassifier</a></span></code>
<div class="block">This BioModule uses RDP to assign taxonomy to 16s sequences.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module.classifier.wgs">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a> in <a href="../../../biolockj/module/classifier/wgs/package-summary.html">biolockj.module.classifier.wgs</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing subclasses, and an explanation">
<caption><span>Subclasses of <a href="../../../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a> in <a href="../../../biolockj/module/classifier/wgs/package-summary.html">biolockj.module.classifier.wgs</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/classifier/wgs/KrakenClassifier.html" title="class in biolockj.module.classifier.wgs">KrakenClassifier</a></span></code>
<div class="block">This BioModule assigns taxonomy to WGS sequences and translates the results into mpa-format.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/classifier/wgs/MetaphlanClassifier.html" title="class in biolockj.module.classifier.wgs">MetaphlanClassifier</a></span></code>
<div class="block">This BioModule builds the bash scripts used to execute metaphlan2.py to classify WGS sequences.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/classifier/wgs/SlimmClassifier.html" title="class in biolockj.module.classifier.wgs">SlimmClassifier</a></span></code>
<div class="block">This BioModule builds the scripts used to call SLIMM for classification of WGS data.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module.implicit">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a> in <a href="../../../biolockj/module/implicit/package-summary.html">biolockj.module.implicit</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing subclasses, and an explanation">
<caption><span>Subclasses of <a href="../../../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a> in <a href="../../../biolockj/module/implicit/package-summary.html">biolockj.module.implicit</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/Demultiplexer.html" title="class in biolockj.module.implicit">Demultiplexer</a></span></code>
<div class="block">This BioModule splits multiplexed data into a separate file or pair of files (for paired reads) for each sample.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/ImportMetadata.html" title="class in biolockj.module.implicit">ImportMetadata</a></span></code>
<div class="block">This BioModule validates the contents/format of the project metadata file and the related Config properties.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/RegisterNumReads.html" title="class in biolockj.module.implicit">RegisterNumReads</a></span></code>
<div class="block">This BioModule parses sequence file to count the number of reads per sample.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module.implicit.parser">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a> in <a href="../../../biolockj/module/implicit/parser/package-summary.html">biolockj.module.implicit.parser</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing subclasses, and an explanation">
<caption><span>Subclasses of <a href="../../../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a> in <a href="../../../biolockj/module/implicit/parser/package-summary.html">biolockj.module.implicit.parser</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/parser/ParserModuleImpl.html" title="class in biolockj.module.implicit.parser">ParserModuleImpl</a></span></code>
<div class="block">Parser <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s read <a href="../../../biolockj/module/classifier/ClassifierModule.html" title="interface in biolockj.module.classifier"><code>ClassifierModule</code></a> output to build
 standardized OTU abundance tables.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module.implicit.parser.r16s">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a> in <a href="../../../biolockj/module/implicit/parser/r16s/package-summary.html">biolockj.module.implicit.parser.r16s</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing subclasses, and an explanation">
<caption><span>Subclasses of <a href="../../../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a> in <a href="../../../biolockj/module/implicit/parser/r16s/package-summary.html">biolockj.module.implicit.parser.r16s</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/parser/r16s/QiimeParser.html" title="class in biolockj.module.implicit.parser.r16s">QiimeParser</a></span></code>
<div class="block">This BioModules parses QiimeClassifier output reports to build standard OTU abundance tables.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/parser/r16s/RdpParser.html" title="class in biolockj.module.implicit.parser.r16s">RdpParser</a></span></code>
<div class="block">This BioModule parses RDP output files to build standard OTU abundance tables.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module.implicit.parser.wgs">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a> in <a href="../../../biolockj/module/implicit/parser/wgs/package-summary.html">biolockj.module.implicit.parser.wgs</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing subclasses, and an explanation">
<caption><span>Subclasses of <a href="../../../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a> in <a href="../../../biolockj/module/implicit/parser/wgs/package-summary.html">biolockj.module.implicit.parser.wgs</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/parser/wgs/KrakenParser.html" title="class in biolockj.module.implicit.parser.wgs">KrakenParser</a></span></code>
<div class="block">This BioModules parses KrakenClassifier output reports to build standard OTU abundance tables.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/parser/wgs/MetaphlanParser.html" title="class in biolockj.module.implicit.parser.wgs">MetaphlanParser</a></span></code>
<div class="block">This BioModules parses MetaphlanClassifier output reports to build standard OTU abundance tables.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/parser/wgs/SlimmParser.html" title="class in biolockj.module.implicit.parser.wgs">SlimmParser</a></span></code>
<div class="block">This BioModule parses SlimmClassifier output reports to build standard OTU abundance tables.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module.implicit.qiime">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a> in <a href="../../../biolockj/module/implicit/qiime/package-summary.html">biolockj.module.implicit.qiime</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing subclasses, and an explanation">
<caption><span>Subclasses of <a href="../../../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a> in <a href="../../../biolockj/module/implicit/qiime/package-summary.html">biolockj.module.implicit.qiime</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/qiime/BuildQiimeMapping.html" title="class in biolockj.module.implicit.qiime">BuildQiimeMapping</a></span></code>
<div class="block">This BioModule converts the metadata file into a tab delimited QIIME mapping file (if provided).<br>
 The QIIME mapping file is validated by calling QIIME script validate_mapping_file.py<br></div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/qiime/MergeOtuTables.html" title="class in biolockj.module.implicit.qiime">MergeOtuTables</a></span></code>
<div class="block">This BioModule will run immediately after QiimeClosedRefClassifier if multiple otu_table.biom files were created.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime">QiimeClassifier</a></span></code>
<div class="block">This BioModule generates the bash script used to create QIIME summary scripts, taxonomy-level reports, and add alpha
 diversity metrics (if configured) to the metadata file.<br>
 For a complete list of available metrics, see:
 <a href= "http://scikit-bio.org/docs/latest/generated/skbio.diversity.alpha.html" target=
 "_top">http://scikit-bio.org/docs/latest/generated/skbio.diversity.alpha.html</a><br></div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module.r">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a> in <a href="../../../biolockj/module/r/package-summary.html">biolockj.module.r</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing subclasses, and an explanation">
<caption><span>Subclasses of <a href="../../../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a> in <a href="../../../biolockj/module/r/package-summary.html">biolockj.module.r</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/r/BuildMdsPlots.html" title="class in biolockj.module.r">BuildMdsPlots</a></span></code>
<div class="block">This BioModule is used to build the R script used to generate MDS plots for each report field and each taxonomy level
 configured.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/r/BuildOtuPlots.html" title="class in biolockj.module.r">BuildOtuPlots</a></span></code>
<div class="block">This BioModule is used to build the R script used to generate OTU-metadata box-plots and scatter-plots for each
 report field and taxonomy level.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/r/BuildPvalHistograms.html" title="class in biolockj.module.r">BuildPvalHistograms</a></span></code>
<div class="block">This BioModule is used to build the R script used to generate p-value histograms for each report field and each
 taxonomy level configured.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/r/CalculateStats.html" title="class in biolockj.module.r">CalculateStats</a></span></code>
<div class="block">This BioModule is used to build the R script used to generate taxonomy statistics and plots.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/r/R_Module.html" title="class in biolockj.module.r">R_Module</a></span></code>
<div class="block">This BioModule is the superclass for R script generating modules.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module.report">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a> in <a href="../../../biolockj/module/report/package-summary.html">biolockj.module.report</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing subclasses, and an explanation">
<caption><span>Subclasses of <a href="../../../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a> in <a href="../../../biolockj/module/report/package-summary.html">biolockj.module.report</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/AddMetadataToOtuTables.html" title="class in biolockj.module.report">AddMetadataToOtuTables</a></span></code>
<div class="block">This BioModule is used to add metadata columns to the OTU abundance tables.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/Email.html" title="class in biolockj.module.report">Email</a></span></code>
<div class="block">This BioModule is used to email the user the pipeline execution status and summary.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/JsonReport.html" title="class in biolockj.module.report">JsonReport</a></span></code>
<div class="block">This BioModule is used to build a JSON file (summary.json) from pipeline OTU-metadata tables.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/Normalizer.html" title="class in biolockj.module.report">Normalizer</a></span></code>
<div class="block">This utility is used to normalize and/or log-transform the raw OTU counts using the formulas:
 
 Normalized OTU count formula = (RC/n)*((SUM(x))/N)+1
 Relative abundance formula = Log(log_base) [ (RC/n)*((SUM(x))/N)+1 ]
 
 The code implementation supports (log_base = e) and (log_base = 10) which is configured via
 <a href="../../../biolockj/Config.html#REPORT_LOG_BASE"><code>Config.REPORT_LOG_BASE</code></a> property.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module.seq">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a> in <a href="../../../biolockj/module/seq/package-summary.html">biolockj.module.seq</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing subclasses, and an explanation">
<caption><span>Subclasses of <a href="../../../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a> in <a href="../../../biolockj/module/seq/package-summary.html">biolockj.module.seq</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/seq/AwkFastaConverter.html" title="class in biolockj.module.seq">AwkFastaConverter</a></span></code>
<div class="block">This BioModule uses awk and gzip to convert input sequence files into a decompressed fasta file format.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/seq/Gunzipper.html" title="class in biolockj.module.seq">Gunzipper</a></span></code>
<div class="block">This BioModule uses gzip to decompress input sequence files.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/seq/Multiplexer.html" title="class in biolockj.module.seq">Multiplexer</a></span></code>
<div class="block">This BioModule will merge sequence files into a single combined sequence file, with either the sample ID or an
 identifying barcode (if defined in the metatata) is stored in the sequence header.<br>
 BioLockJ is designed to run on demultiplexed data so this must be the last module to run before the Summary module.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/seq/PearMergeReads.html" title="class in biolockj.module.seq">PearMergeReads</a></span></code>
<div class="block">This BioModule will merge forward and reverse fastq files using PEAR.<br>
 For more informations, see the online PEAR manual:
 <a href="https://sco.h-its.org/exelixis/web/software/pear/doc.html" target=
 "_top">https://sco.h-its.org/exelixis/web/software/pear/doc.html</a></div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/seq/Rarefier.html" title="class in biolockj.module.seq">Rarefier</a></span></code>
<div class="block">This BioModule imposes a minimum and/or maximum number of reads per sample.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/seq/SeqFileValidator.html" title="class in biolockj.module.seq">SeqFileValidator</a></span></code>
<div class="block">This BioModule validates fasta/fastq file formats are valid and enforces min/max read lengths.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/seq/TrimPrimers.html" title="class in biolockj.module.seq">TrimPrimers</a></span></code>
<div class="block">This BioModule removes sequence primers from demultiplexed files.<br>
 The primers are defined using regular expressions in a separate file.</div>
</td>
</tr>
</tbody>
</table>
</li>
</ul>
</li>
</ul>
</div>
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar.bottom">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.bottom.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../overview-summary.html">Overview</a></li>
<li><a href="../package-summary.html">Package</a></li>
<li><a href="../../../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">Class</a></li>
<li class="navBarCell1Rev">Use</li>
<li><a href="../package-tree.html">Tree</a></li>
<li><a href="../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../index-files/index-1.html">Index</a></li>
<li><a href="../../../help-doc.html">Help</a></li>
<li>&nbsp;&nbsp;</li>
<li><a href="https://github.com/msioda/BioLockJ/wiki" target="_top"><strong>RETURN to WIKI</strong></a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li>Prev</li>
<li>Next</li>
</ul>
<ul class="navList">
<li><a href="../../../index.html?biolockj/module/class-use/BioModuleImpl.html" target="_top">Frames</a></li>
<li><a href="BioModuleImpl.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip.navbar.bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
