<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (1.8.0_144) on Wed Feb 27 01:20:49 EST 2019 -->
<title>Uses of Interface biolockj.module.BioModule</title>
<meta name="date" content="2019-02-27">
<link rel="stylesheet" type="text/css" href="../../../stylesheet.css" title="Style">
<script type="text/javascript" src="../../../script.js"></script>
</head>
<body>
<script type="text/javascript"><!--
    try {
        if (location.href.indexOf('is-external=true') == -1) {
            parent.document.title="Uses of Interface biolockj.module.BioModule";
        }
    }
    catch(err) {
    }
//-->
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar.top">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.top.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../overview-summary.html">Overview</a></li>
<li><a href="../package-summary.html">Package</a></li>
<li><a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">Class</a></li>
<li class="navBarCell1Rev">Use</li>
<li><a href="../package-tree.html">Tree</a></li>
<li><a href="../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../index-files/index-1.html">Index</a></li>
<li><a href="../../../help-doc.html">Help</a></li>
<li>&nbsp;&nbsp;</li>
<li><a href="https://github.com/msioda/BioLockJ/wiki" target="_top"><strong>RETURN to WIKI</strong></a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li>Prev</li>
<li>Next</li>
</ul>
<ul class="navList">
<li><a href="../../../index.html?biolockj/module/class-use/BioModule.html" target="_top">Frames</a></li>
<li><a href="BioModule.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip.navbar.top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<div class="header">
<h2 title="Uses of Interface biolockj.module.BioModule" class="title">Uses of Interface<br>biolockj.module.BioModule</h2>
</div>
<div class="classUseContainer">
<ul class="blockList">
<li class="blockList">
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing packages, and an explanation">
<caption><span>Packages that use <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Package</th>
<th class="colLast" scope="col">Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><a href="#biolockj">biolockj</a></td>
<td class="colLast">
<div class="block">The root biolockj package contains core classes used by all BioLockJ pipelines.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><a href="#biolockj.module">biolockj.module</a></td>
<td class="colLast">
<div class="block">Pipeline modules are loaded based on the <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s listed in the <a href="../../../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>
 file.<br>
 All modules (i.e., implement <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>) are contained within this package or its
 sub-packages.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><a href="#biolockj.module.classifier">biolockj.module.classifier</a></td>
<td class="colLast">
<div class="block">This package contains Classifier specific <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s that build bash scripts to generate
 taxonomy reports from the sequences files.<br></div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><a href="#biolockj.module.classifier.r16s">biolockj.module.classifier.r16s</a></td>
<td class="colLast">
<div class="block">This package contains RDP and QIIME classifier modules that assign taxonomy to 16s sequences.<br>
 Output file formats are classifier specific so classifier requires a unique post-requisite
 <a href="../../../biolockj/module/implicit/parser/ParserModule.html" title="interface in biolockj.module.implicit.parser"><code>ParserModule</code></a>.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><a href="#biolockj.module.classifier.wgs">biolockj.module.classifier.wgs</a></td>
<td class="colLast">
<div class="block">This package contains Kraken, Kraken2, Humann2, MetaPhlAn classifier modules that assign taxonomy to WGS data.<br>
 As with all BioLockJ classifiers, these modules generate raw classifier output files from the sequence data.<br>
 Include the corresponding <a href="../../../biolockj/module/implicit/parser/ParserModule.html" title="interface in biolockj.module.implicit.parser"><code>ParserModule</code></a> in the <a href="../../../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> file to
 generate standardized OTU abundance tables.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><a href="#biolockj.module.implicit">biolockj.module.implicit</a></td>
<td class="colLast">
<div class="block">The modules in this package are implicitly added to pipelines as needed.<br>
 These modules cannot be directly added to any pipeline unless overridden via project.disableImplicitModules=Y.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><a href="#biolockj.module.implicit.parser">biolockj.module.implicit.parser</a></td>
<td class="colLast">
<div class="block">This package contains Parser BioModules in the r16s and WGS sub-packages that are paired with a
 <a href="../../../biolockj/module/classifier/ClassifierModule.html" title="interface in biolockj.module.classifier"><code>ClassifierModule</code></a> via
 <a href="../../../biolockj/module/BioModule.html#getPostRequisiteModules--"><code>getPostRequisiteModules()</code></a> to run immediately after the
 classifier.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><a href="#biolockj.module.implicit.parser.r16s">biolockj.module.implicit.parser.r16s</a></td>
<td class="colLast">
<div class="block">This package contains Parser <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s that convert the 16S taxonomy reports generated by
 16S classifiers (such as RDP and QIIME) into standardized OTU abundance tables.<br></div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><a href="#biolockj.module.implicit.parser.wgs">biolockj.module.implicit.parser.wgs</a></td>
<td class="colLast">
<div class="block">This package contains Parser <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s that convert the WGS taxonomy reports generated by
 WGS classifiers (such as Kraken, Kraken2 Metaphlan2, or Humann2) into standardized OTU abundance tables.<br></div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><a href="#biolockj.module.implicit.qiime">biolockj.module.implicit.qiime</a></td>
<td class="colLast">
<div class="block">This package contains <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s that are implicitly added to QIIME pipeline as needed.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><a href="#biolockj.module.report">biolockj.module.report</a></td>
<td class="colLast">
<div class="block">This package contains <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s that normalize OTU abundance tables output by Parser
 modules, merges them with the metadata, and generates various reports and notifications.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><a href="#biolockj.module.report.humann2">biolockj.module.report.humann2</a></td>
<td class="colLast">&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><a href="#biolockj.module.report.otu">biolockj.module.report.otu</a></td>
<td class="colLast">&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><a href="#biolockj.module.report.r">biolockj.module.report.r</a></td>
<td class="colLast">
<div class="block">This package contains <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s that build pipeline reports from the standard OTU abundance
 tables by generating R scripts to produce the summary statistics and data visualizations output to PDF files.<br></div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><a href="#biolockj.module.report.taxa">biolockj.module.report.taxa</a></td>
<td class="colLast">&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><a href="#biolockj.module.seq">biolockj.module.seq</a></td>
<td class="colLast">
<div class="block"><a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s used to prepare sequence files or update the metadata prior to classification.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><a href="#biolockj.util">biolockj.util</a></td>
<td class="colLast">
<div class="block">Static utilities centralize and organize reusable core methods.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList">
<ul class="blockList">
<li class="blockList"><a name="biolockj">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a> in <a href="../../../biolockj/package-summary.html">biolockj</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../biolockj/package-summary.html">biolockj</a> that return types with arguments of type <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/util/List.html?is-external=true" title="class or interface in java.util">List</a>&lt;<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">BioModuleFactory.</span><code><span class="memberNameLink"><a href="../../../biolockj/BioModuleFactory.html#buildPipeline--">buildPipeline</a></span>()</code>
<div class="block">Build all modules for the pipeline.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/util/List.html?is-external=true" title="class or interface in java.util">List</a>&lt;<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">Pipeline.</span><code><span class="memberNameLink"><a href="../../../biolockj/Pipeline.html#getModules--">getModules</a></span>()</code>
<div class="block">Return a list of <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s constructed by the <a href="../../../biolockj/BioModuleFactory.html" title="class in biolockj"><code>BioModuleFactory</code></a></div>
</td>
</tr>
</tbody>
</table>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../biolockj/package-summary.html">biolockj</a> with parameters of type <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>protected static void</code></td>
<td class="colLast"><span class="typeNameLabel">Pipeline.</span><code><span class="memberNameLink"><a href="../../../biolockj/Pipeline.html#deleteIncompleteModule-biolockj.module.BioModule-">deleteIncompleteModule</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">Delete and recreate incomplete module directory.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static void</code></td>
<td class="colLast"><span class="typeNameLabel">Pipeline.</span><code><span class="memberNameLink"><a href="../../../biolockj/Pipeline.html#executeModule-biolockj.module.BioModule-">executeModule</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">Execute a single pipeline module.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static boolean</code></td>
<td class="colLast"><span class="typeNameLabel">Config.</span><code><span class="memberNameLink"><a href="../../../biolockj/Config.html#getBoolean-biolockj.module.BioModule-java.lang.String-">getBoolean</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
          <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;property)</code>
<div class="block">Parse property value (Y or N) to return boolean, if not found, return false;</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Double.html?is-external=true" title="class or interface in java.lang">Double</a></code></td>
<td class="colLast"><span class="typeNameLabel">Config.</span><code><span class="memberNameLink"><a href="../../../biolockj/Config.html#getDoubleVal-biolockj.module.BioModule-java.lang.String-">getDoubleVal</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
            <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;property)</code>
<div class="block">Parse property for numeric (double) value</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a></code></td>
<td class="colLast"><span class="typeNameLabel">Config.</span><code><span class="memberNameLink"><a href="../../../biolockj/Config.html#getExe-biolockj.module.BioModule-java.lang.String-">getExe</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
      <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;property)</code>
<div class="block">Get exe.* property name.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a></code></td>
<td class="colLast"><span class="typeNameLabel">Config.</span><code><span class="memberNameLink"><a href="../../../biolockj/Config.html#getExeParams-biolockj.module.BioModule-java.lang.String-">getExeParams</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
            <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;property)</code>
<div class="block">Call this function to get the parameters configured for this property.<br>
 Make sure the last character for non-null resultss is an empty character for use in bash scripts calling the
 corresponding executable.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/io/File.html?is-external=true" title="class or interface in java.io">File</a></code></td>
<td class="colLast"><span class="typeNameLabel">Config.</span><code><span class="memberNameLink"><a href="../../../biolockj/Config.html#getExistingDir-biolockj.module.BioModule-java.lang.String-">getExistingDir</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
              <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;property)</code>
<div class="block">Get a valid File directory or return null</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/io/File.html?is-external=true" title="class or interface in java.io">File</a></code></td>
<td class="colLast"><span class="typeNameLabel">Config.</span><code><span class="memberNameLink"><a href="../../../biolockj/Config.html#getExistingFile-biolockj.module.BioModule-java.lang.String-">getExistingFile</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
               <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;property)</code>
<div class="block">Get a valid File or return null.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/util/List.html?is-external=true" title="class or interface in java.util">List</a>&lt;<a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">Config.</span><code><span class="memberNameLink"><a href="../../../biolockj/Config.html#getList-biolockj.module.BioModule-java.lang.String-">getList</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
       <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;property)</code>
<div class="block">Parse comma delimited property value to return list</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>protected static <a href="https://docs.oracle.com/javase/8/docs/api/java/io/File.html?is-external=true" title="class or interface in java.io">File</a></code></td>
<td class="colLast"><span class="typeNameLabel">Pipeline.</span><code><span class="memberNameLink"><a href="../../../biolockj/Pipeline.html#getMetadata-biolockj.module.BioModule-">getMetadata</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;bioModule)</code>
<div class="block">If the bioModule is complete and contains a metadata file in its output directory, return the metadata file,
 since it must be a new version.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a></code></td>
<td class="colLast"><span class="typeNameLabel">Config.</span><code><span class="memberNameLink"><a href="../../../biolockj/Config.html#getModuleProp-biolockj.module.BioModule-java.lang.String-">getModuleProp</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
             <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;prop)</code>
<div class="block">Return module specific property if configured, otherwise use the given prop.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a></code></td>
<td class="colLast"><span class="typeNameLabel">Config.</span><code><span class="memberNameLink"><a href="../../../biolockj/Config.html#getNonNegativeInteger-biolockj.module.BioModule-java.lang.String-">getNonNegativeInteger</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
                     <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;property)</code>
<div class="block">Parse property as non-negative integer value</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Double.html?is-external=true" title="class or interface in java.lang">Double</a></code></td>
<td class="colLast"><span class="typeNameLabel">Config.</span><code><span class="memberNameLink"><a href="../../../biolockj/Config.html#getPositiveDoubleVal-biolockj.module.BioModule-java.lang.String-">getPositiveDoubleVal</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
                    <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;property)</code>
<div class="block">Parse property as positive double value</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a></code></td>
<td class="colLast"><span class="typeNameLabel">Config.</span><code><span class="memberNameLink"><a href="../../../biolockj/Config.html#getPositiveInteger-biolockj.module.BioModule-java.lang.String-">getPositiveInteger</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
                  <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;property)</code>
<div class="block">Parse property as positive integer value</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>protected <a href="https://docs.oracle.com/javase/8/docs/api/java/util/List.html?is-external=true" title="class or interface in java.util">List</a>&lt;<a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">BioModuleFactory.</span><code><span class="memberNameLink"><a href="../../../biolockj/BioModuleFactory.html#getPostRequisites-biolockj.module.BioModule-">getPostRequisites</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">This method returns all module post-requisites (including post-requisites of the post-requisites).</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>protected <a href="https://docs.oracle.com/javase/8/docs/api/java/util/List.html?is-external=true" title="class or interface in java.util">List</a>&lt;<a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">BioModuleFactory.</span><code><span class="memberNameLink"><a href="../../../biolockj/BioModuleFactory.html#getPreRequisites-biolockj.module.BioModule-">getPreRequisites</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">This method returns all module prerequisites (including prerequisites of the prerequisites).</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/util/Set.html?is-external=true" title="class or interface in java.util">Set</a>&lt;<a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">Config.</span><code><span class="memberNameLink"><a href="../../../biolockj/Config.html#getSet-biolockj.module.BioModule-java.lang.String-">getSet</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
      <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;property)</code>
<div class="block">Parse comma-separated property value to build an unordered Set</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a></code></td>
<td class="colLast"><span class="typeNameLabel">Config.</span><code><span class="memberNameLink"><a href="../../../biolockj/Config.html#getString-biolockj.module.BioModule-java.lang.String-">getString</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
         <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;property)</code>
<div class="block">Get property value as String.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/util/Set.html?is-external=true" title="class or interface in java.util">Set</a>&lt;<a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">Config.</span><code><span class="memberNameLink"><a href="../../../biolockj/Config.html#getTreeSet-biolockj.module.BioModule-java.lang.String-">getTreeSet</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
          <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;property)</code>
<div class="block">Parse comma-separated property value to build an ordered Set</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>protected static void</code></td>
<td class="colLast"><span class="typeNameLabel">Pipeline.</span><code><span class="memberNameLink"><a href="../../../biolockj/Pipeline.html#refreshOutputMetadata-biolockj.module.BioModule-">refreshOutputMetadata</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">Call <a href="../../../biolockj/util/MetaUtil.html" title="class in biolockj.util"><code>MetaUtil</code></a> to refresh the metadata cache if a new metadata file was output by the
 bioModule .</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>protected static void</code></td>
<td class="colLast"><span class="typeNameLabel">Pipeline.</span><code><span class="memberNameLink"><a href="../../../biolockj/Pipeline.html#refreshRCacheIfNeeded-biolockj.module.BioModule-">refreshRCacheIfNeeded</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">Refresh R cache if about to run the 1st R module.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static boolean</code></td>
<td class="colLast"><span class="typeNameLabel">Config.</span><code><span class="memberNameLink"><a href="../../../biolockj/Config.html#requireBoolean-biolockj.module.BioModule-java.lang.String-">requireBoolean</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
              <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;property)</code>
<div class="block">Required to return a valid boolean  or </div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Double.html?is-external=true" title="class or interface in java.lang">Double</a></code></td>
<td class="colLast"><span class="typeNameLabel">Config.</span><code><span class="memberNameLink"><a href="../../../biolockj/Config.html#requireDoubleVal-biolockj.module.BioModule-java.lang.String-">requireDoubleVal</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
                <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;property)</code>
<div class="block">Requires valid double value</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/io/File.html?is-external=true" title="class or interface in java.io">File</a></code></td>
<td class="colLast"><span class="typeNameLabel">Config.</span><code><span class="memberNameLink"><a href="../../../biolockj/Config.html#requireExistingDir-biolockj.module.BioModule-java.lang.String-">requireExistingDir</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
                  <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;property)</code>
<div class="block">Requires valid existing directory.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/util/List.html?is-external=true" title="class or interface in java.util">List</a>&lt;<a href="https://docs.oracle.com/javase/8/docs/api/java/io/File.html?is-external=true" title="class or interface in java.io">File</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">Config.</span><code><span class="memberNameLink"><a href="../../../biolockj/Config.html#requireExistingDirs-biolockj.module.BioModule-java.lang.String-">requireExistingDirs</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
                   <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;property)</code>
<div class="block">Requires valid list of file directories</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/io/File.html?is-external=true" title="class or interface in java.io">File</a></code></td>
<td class="colLast"><span class="typeNameLabel">Config.</span><code><span class="memberNameLink"><a href="../../../biolockj/Config.html#requireExistingFile-biolockj.module.BioModule-java.lang.String-">requireExistingFile</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
                   <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;property)</code>
<div class="block">Require valid existing file</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a></code></td>
<td class="colLast"><span class="typeNameLabel">Config.</span><code><span class="memberNameLink"><a href="../../../biolockj/Config.html#requireInteger-biolockj.module.BioModule-java.lang.String-">requireInteger</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
              <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;property)</code>
<div class="block">Requires valid integer value</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/util/List.html?is-external=true" title="class or interface in java.util">List</a>&lt;<a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">Config.</span><code><span class="memberNameLink"><a href="../../../biolockj/Config.html#requireList-biolockj.module.BioModule-java.lang.String-">requireList</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
           <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;property)</code>
<div class="block">Require valid list property</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Double.html?is-external=true" title="class or interface in java.lang">Double</a></code></td>
<td class="colLast"><span class="typeNameLabel">Config.</span><code><span class="memberNameLink"><a href="../../../biolockj/Config.html#requirePositiveDouble-biolockj.module.BioModule-java.lang.String-">requirePositiveDouble</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
                     <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;property)</code>
<div class="block">Require valid positive double value</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Integer.html?is-external=true" title="class or interface in java.lang">Integer</a></code></td>
<td class="colLast"><span class="typeNameLabel">Config.</span><code><span class="memberNameLink"><a href="../../../biolockj/Config.html#requirePositiveInteger-biolockj.module.BioModule-java.lang.String-">requirePositiveInteger</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
                      <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;property)</code>
<div class="block">Require valid positive integer value</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/util/Set.html?is-external=true" title="class or interface in java.util">Set</a>&lt;<a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">Config.</span><code><span class="memberNameLink"><a href="../../../biolockj/Config.html#requireSet-biolockj.module.BioModule-java.lang.String-">requireSet</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
          <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;property)</code>
<div class="block">Require valid Set value</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a></code></td>
<td class="colLast"><span class="typeNameLabel">Config.</span><code><span class="memberNameLink"><a href="../../../biolockj/Config.html#requireString-biolockj.module.BioModule-java.lang.String-">requireString</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
             <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;property)</code>
<div class="block">Require valid String value</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a> in <a href="../../../biolockj/module/package-summary.html">biolockj.module</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing subinterfaces, and an explanation">
<caption><span>Subinterfaces of <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a> in <a href="../../../biolockj/module/package-summary.html">biolockj.module</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Interface and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>interface&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/JavaModule.html" title="interface in biolockj.module">JavaModule</a></span></code>
<div class="block">Classes that implement this interface are pure Java modules.<br></div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>interface&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/ScriptModule.html" title="interface in biolockj.module">ScriptModule</a></span></code>
<div class="block">Classes that implement this interface are <br></div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>interface&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/SeqModule.html" title="interface in biolockj.module">SeqModule</a></span></code>
<div class="block">Classes that implement this interface requires sequence files for input.<br></div>
</td>
</tr>
</tbody>
</table>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing classes, and an explanation">
<caption><span>Classes in <a href="../../../biolockj/module/package-summary.html">biolockj.module</a> that implement <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a></span></code>
<div class="block">Superclass for standard BioModules (classifiers, parsers, etc).</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/JavaModuleImpl.html" title="class in biolockj.module">JavaModuleImpl</a></span></code>
<div class="block">Superclass for Java BioModules that will be called in separate instances of the application.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/ScriptModuleImpl.html" title="class in biolockj.module">ScriptModuleImpl</a></span></code>
<div class="block">Superclass for Java BioModules that will be called in separate instances of the application.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/SeqModuleImpl.html" title="class in biolockj.module">SeqModuleImpl</a></span></code>
<div class="block">Superclass for SeqModules that take sequence files as input for pre-processing prior to classification.</div>
</td>
</tr>
</tbody>
</table>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../biolockj/module/package-summary.html">biolockj.module</a> with parameters of type <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><span class="typeNameLabel">BioModuleImpl.</span><code><span class="memberNameLink"><a href="../../../biolockj/module/BioModuleImpl.html#compareTo-biolockj.module.BioModule-">compareTo</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="typeNameLabel">BioModuleImpl.</span><code><span class="memberNameLink"><a href="../../../biolockj/module/BioModuleImpl.html#isValidInputModule-biolockj.module.BioModule-">isValidInputModule</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">In the early stages of the pipeline, starting with the very 1st module
 <a href="../../../biolockj/module/implicit/ImportMetadata.html" title="class in biolockj.module.implicit"><code>ImportMetadata</code></a>, most modules expect sequence files as input.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="typeNameLabel">SeqModuleImpl.</span><code><span class="memberNameLink"><a href="../../../biolockj/module/SeqModuleImpl.html#isValidInputModule-biolockj.module.BioModule-">isValidInputModule</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="typeNameLabel">JavaModuleImpl.</span><code><span class="memberNameLink"><a href="../../../biolockj/module/JavaModuleImpl.html#isValidInputModule-biolockj.module.BioModule-">isValidInputModule</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">If module is a <a href="../../../biolockj/module/SeqModule.html" title="interface in biolockj.module"><code>SeqModule</code></a> input must contain sequence data.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="typeNameLabel">BioModule.</span><code><span class="memberNameLink"><a href="../../../biolockj/module/BioModule.html#isValidInputModule-biolockj.module.BioModule-">isValidInputModule</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;previousModule)</code>
<div class="block">BioModules <a href="../../../biolockj/module/BioModule.html#getInputFiles--"><code>getInputFiles()</code></a> method typically, but not always, return the previousModule output files.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module.classifier">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a> in <a href="../../../biolockj/module/classifier/package-summary.html">biolockj.module.classifier</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing subinterfaces, and an explanation">
<caption><span>Subinterfaces of <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a> in <a href="../../../biolockj/module/classifier/package-summary.html">biolockj.module.classifier</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Interface and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>interface&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/classifier/ClassifierModule.html" title="interface in biolockj.module.classifier">ClassifierModule</a></span></code>
<div class="block">Classifier <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s build one or more bash scripts to call the application on sequence
 files.</div>
</td>
</tr>
</tbody>
</table>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing classes, and an explanation">
<caption><span>Classes in <a href="../../../biolockj/module/classifier/package-summary.html">biolockj.module.classifier</a> that implement <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/classifier/ClassifierModuleImpl.html" title="class in biolockj.module.classifier">ClassifierModuleImpl</a></span></code>
<div class="block">This is the superclass for all WGS and 16S biolockj.module.classifier BioModules.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module.classifier.r16s">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a> in <a href="../../../biolockj/module/classifier/r16s/package-summary.html">biolockj.module.classifier.r16s</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing classes, and an explanation">
<caption><span>Classes in <a href="../../../biolockj/module/classifier/r16s/package-summary.html">biolockj.module.classifier.r16s</a> that implement <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/classifier/r16s/QiimeClosedRefClassifier.html" title="class in biolockj.module.classifier.r16s">QiimeClosedRefClassifier</a></span></code>
<div class="block">This BioModule executes the QIIME script pick_closed_reference_otus.py on a FastA sequence files.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/classifier/r16s/QiimeDeNovoClassifier.html" title="class in biolockj.module.classifier.r16s">QiimeDeNovoClassifier</a></span></code>
<div class="block">This module runs the QIIME pick_de_novo_otus.py script on FastA sequence files in a single script so it is important
 to allocate sufficient job resources if running in a clustered environment.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/classifier/r16s/QiimeOpenRefClassifier.html" title="class in biolockj.module.classifier.r16s">QiimeOpenRefClassifier</a></span></code>
<div class="block">This module runs the QIIME pick_open_reference_otus.py script on FastA sequence files in a single script so it is
 important to allocate sufficient job resources if running in a clustered environment.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/classifier/r16s/RdpClassifier.html" title="class in biolockj.module.classifier.r16s">RdpClassifier</a></span></code>
<div class="block">This BioModule uses RDP to assign taxonomy to 16s sequences.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module.classifier.wgs">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a> in <a href="../../../biolockj/module/classifier/wgs/package-summary.html">biolockj.module.classifier.wgs</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing classes, and an explanation">
<caption><span>Classes in <a href="../../../biolockj/module/classifier/wgs/package-summary.html">biolockj.module.classifier.wgs</a> that implement <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/classifier/wgs/Humann2Classifier.html" title="class in biolockj.module.classifier.wgs">Humann2Classifier</a></span></code>
<div class="block">This BioModule runs biobakery humann2 program to generate the HMP Unified Metabolic Analysis Network<br>
 HUMAnN is a pipeline for efficiently and accurately profiling the presence/absence and abundance of microbial
 pathways in a community from metagenomic or metatranscriptomic sequencing data (typically millions of short DNA/RNA
 reads).</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/classifier/wgs/Kraken2Classifier.html" title="class in biolockj.module.classifier.wgs">Kraken2Classifier</a></span></code>
<div class="block">This BioModule assigns taxonomy to WGS sequences and translates the results into mpa-format.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/classifier/wgs/KrakenClassifier.html" title="class in biolockj.module.classifier.wgs">KrakenClassifier</a></span></code>
<div class="block">This BioModule assigns taxonomy to WGS sequences and translates the results into mpa-format.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/classifier/wgs/Metaphlan2Classifier.html" title="class in biolockj.module.classifier.wgs">Metaphlan2Classifier</a></span></code>
<div class="block">This BioModule builds the bash scripts used to execute metaphlan2.py to classify WGS sequences.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module.implicit">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a> in <a href="../../../biolockj/module/implicit/package-summary.html">biolockj.module.implicit</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing classes, and an explanation">
<caption><span>Classes in <a href="../../../biolockj/module/implicit/package-summary.html">biolockj.module.implicit</a> that implement <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/Demultiplexer.html" title="class in biolockj.module.implicit">Demultiplexer</a></span></code>
<div class="block">This BioModule splits multiplexed data into a separate file or pair of files (for paired reads) for each sample.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/ImportMetadata.html" title="class in biolockj.module.implicit">ImportMetadata</a></span></code>
<div class="block">This BioModule validates the contents/format of the project metadata file and the related Config properties.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/RegisterNumReads.html" title="class in biolockj.module.implicit">RegisterNumReads</a></span></code>
<div class="block">This BioModule parses sequence file to count the number of reads per sample.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module.implicit.parser">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a> in <a href="../../../biolockj/module/implicit/parser/package-summary.html">biolockj.module.implicit.parser</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing subinterfaces, and an explanation">
<caption><span>Subinterfaces of <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a> in <a href="../../../biolockj/module/implicit/parser/package-summary.html">biolockj.module.implicit.parser</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Interface and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>interface&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/parser/ParserModule.html" title="interface in biolockj.module.implicit.parser">ParserModule</a></span></code>
<div class="block">This interface defines the required methods to parse ClassifierModule output.</div>
</td>
</tr>
</tbody>
</table>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing classes, and an explanation">
<caption><span>Classes in <a href="../../../biolockj/module/implicit/parser/package-summary.html">biolockj.module.implicit.parser</a> that implement <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/parser/ParserModuleImpl.html" title="class in biolockj.module.implicit.parser">ParserModuleImpl</a></span></code>
<div class="block">Parser <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s read <a href="../../../biolockj/module/classifier/ClassifierModule.html" title="interface in biolockj.module.classifier"><code>ClassifierModule</code></a> output to build
 standardized OTU count tables.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module.implicit.parser.r16s">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a> in <a href="../../../biolockj/module/implicit/parser/r16s/package-summary.html">biolockj.module.implicit.parser.r16s</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing classes, and an explanation">
<caption><span>Classes in <a href="../../../biolockj/module/implicit/parser/r16s/package-summary.html">biolockj.module.implicit.parser.r16s</a> that implement <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/parser/r16s/QiimeParser.html" title="class in biolockj.module.implicit.parser.r16s">QiimeParser</a></span></code>
<div class="block">This BioModules parses QiimeClassifier output reports to build standard OTU abundance tables.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/parser/r16s/RdpParser.html" title="class in biolockj.module.implicit.parser.r16s">RdpParser</a></span></code>
<div class="block">This BioModule parses RDP output files to build standard OTU abundance tables.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module.implicit.parser.wgs">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a> in <a href="../../../biolockj/module/implicit/parser/wgs/package-summary.html">biolockj.module.implicit.parser.wgs</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing classes, and an explanation">
<caption><span>Classes in <a href="../../../biolockj/module/implicit/parser/wgs/package-summary.html">biolockj.module.implicit.parser.wgs</a> that implement <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/parser/wgs/Humann2Parser.html" title="class in biolockj.module.implicit.parser.wgs">Humann2Parser</a></span></code>
<div class="block">This BioModules parses Humann2Classifier output reports to build standard OTU abundance tables.<br>
 Samples IDs are found in the column headers starting with the 2nd column.<br>
 The count type depends on the HumanN2 config properties.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/parser/wgs/Kraken2Parser.html" title="class in biolockj.module.implicit.parser.wgs">Kraken2Parser</a></span></code>
<div class="block">This BioModules parses KrakenClassifier output reports to build standard OTU abundance tables.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/parser/wgs/KrakenParser.html" title="class in biolockj.module.implicit.parser.wgs">KrakenParser</a></span></code>
<div class="block">This BioModules parses KrakenClassifier output reports to build standard OTU abundance tables.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/parser/wgs/Metaphlan2Parser.html" title="class in biolockj.module.implicit.parser.wgs">Metaphlan2Parser</a></span></code>
<div class="block">This BioModules parses Metaphlan2Classifier output reports to build standard OTU abundance tables.</div>
</td>
</tr>
</tbody>
</table>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../biolockj/module/implicit/parser/wgs/package-summary.html">biolockj.module.implicit.parser.wgs</a> with parameters of type <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="typeNameLabel">Humann2Parser.</span><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/parser/wgs/Humann2Parser.html#isValidInputModule-biolockj.module.BioModule-">isValidInputModule</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>&nbsp;</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module.implicit.qiime">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a> in <a href="../../../biolockj/module/implicit/qiime/package-summary.html">biolockj.module.implicit.qiime</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing classes, and an explanation">
<caption><span>Classes in <a href="../../../biolockj/module/implicit/qiime/package-summary.html">biolockj.module.implicit.qiime</a> that implement <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/qiime/BuildQiimeMapping.html" title="class in biolockj.module.implicit.qiime">BuildQiimeMapping</a></span></code>
<div class="block">This BioModule converts the metadata file into a tab delimited QIIME mapping file (if provided).<br>
 The QIIME mapping file is validated by calling QIIME script validate_mapping_file.py<br></div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/qiime/MergeQiimeOtuTables.html" title="class in biolockj.module.implicit.qiime">MergeQiimeOtuTables</a></span></code>
<div class="block">This BioModule will run immediately after QiimeClosedRefClassifier if multiple otu_table.biom files were created.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime">QiimeClassifier</a></span></code>
<div class="block">This BioModule generates the bash script used to create QIIME summary scripts, taxonomy-level reports, and add alpha
 diversity metrics (if configured) to the metadata file.<br>
 For a complete list of available metrics, see:
 <a href= "http://scikit-bio.org/docs/latest/generated/skbio.diversity.alpha.html" target=
 "_top">http://scikit-bio.org/docs/latest/generated/skbio.diversity.alpha.html</a><br></div>
</td>
</tr>
</tbody>
</table>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../biolockj/module/implicit/qiime/package-summary.html">biolockj.module.implicit.qiime</a> with parameters of type <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="typeNameLabel">QiimeClassifier.</span><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/qiime/QiimeClassifier.html#isValidInputModule-biolockj.module.BioModule-">isValidInputModule</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">If superclass is fed by another QiimeClassifier, it must be a subclass with biom output.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module.report">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a> in <a href="../../../biolockj/module/report/package-summary.html">biolockj.module.report</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing classes, and an explanation">
<caption><span>Classes in <a href="../../../biolockj/module/report/package-summary.html">biolockj.module.report</a> that implement <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/Email.html" title="class in biolockj.module.report">Email</a></span></code>
<div class="block">This BioModule is used to email the user the pipeline execution status and summary.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/JsonReport.html" title="class in biolockj.module.report">JsonReport</a></span></code>
<div class="block">This BioModule is used to build a JSON file (summary.json) compiled from all OTUs in the dataset.</div>
</td>
</tr>
</tbody>
</table>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../biolockj/module/report/package-summary.html">biolockj.module.report</a> with parameters of type <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="typeNameLabel">JsonReport.</span><code><span class="memberNameLink"><a href="../../../biolockj/module/report/JsonReport.html#isValidInputModule-biolockj.module.BioModule-">isValidInputModule</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>&nbsp;</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module.report.humann2">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a> in <a href="../../../biolockj/module/report/humann2/package-summary.html">biolockj.module.report.humann2</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing classes, and an explanation">
<caption><span>Classes in <a href="../../../biolockj/module/report/humann2/package-summary.html">biolockj.module.report.humann2</a> that implement <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/humann2/AddMetadataToPathwayTables.html" title="class in biolockj.module.report.humann2">AddMetadataToPathwayTables</a></span></code>
<div class="block">This BioModule is used to add metadata columns to the OTU abundance tables.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/humann2/Humann2CountModule.html" title="class in biolockj.module.report.humann2">Humann2CountModule</a></span></code>
<div class="block">TBD</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/humann2/RemoveLowPathwayCounts.html" title="class in biolockj.module.report.humann2">RemoveLowPathwayCounts</a></span></code>
<div class="block">This BioModule</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/humann2/RemoveScarcePathwayCounts.html" title="class in biolockj.module.report.humann2">RemoveScarcePathwayCounts</a></span></code>
<div class="block">This BioModule</div>
</td>
</tr>
</tbody>
</table>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../biolockj/module/report/humann2/package-summary.html">biolockj.module.report.humann2</a> with parameters of type <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>protected boolean</code></td>
<td class="colLast"><span class="typeNameLabel">Humann2CountModule.</span><code><span class="memberNameLink"><a href="../../../biolockj/module/report/humann2/Humann2CountModule.html#isHumann2CountModule-biolockj.module.BioModule-">isHumann2CountModule</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">Check the module to determine if it generated OTU count files.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="typeNameLabel">Humann2CountModule.</span><code><span class="memberNameLink"><a href="../../../biolockj/module/report/humann2/Humann2CountModule.html#isValidInputModule-biolockj.module.BioModule-">isValidInputModule</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>&nbsp;</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module.report.otu">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a> in <a href="../../../biolockj/module/report/otu/package-summary.html">biolockj.module.report.otu</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing classes, and an explanation">
<caption><span>Classes in <a href="../../../biolockj/module/report/otu/package-summary.html">biolockj.module.report.otu</a> that implement <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/otu/CompileOtuCounts.html" title="class in biolockj.module.report.otu">CompileOtuCounts</a></span></code>
<div class="block">This BioModule compiles the counts from all OTU count files into a single summary OTU count file containing OTU
 counts for the entire dataset.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/otu/OtuCountModule.html" title="class in biolockj.module.report.otu">OtuCountModule</a></span></code>
<div class="block">OtuCount modules reads OTU count assignment tables (1 file/sample) with 2 columns.<br>
 Col1: Full OTU pathway spanning top to bottom level Col2: Count (# of reads) for the sample.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/otu/RarefyOtuCounts.html" title="class in biolockj.module.report.otu">RarefyOtuCounts</a></span></code>
<div class="block">This BioModule applies a mean iterative post-OTU classification rarefication algorithm so that each output sample
 will have approximately the same number of OTUs.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/otu/RemoveLowOtuCounts.html" title="class in biolockj.module.report.otu">RemoveLowOtuCounts</a></span></code>
<div class="block">This BioModule</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/otu/RemoveScarceOtuCounts.html" title="class in biolockj.module.report.otu">RemoveScarceOtuCounts</a></span></code>
<div class="block">This BioModule removes scarce OTUs not found in enough samples.<br>
 The OTU must be found in a configurable percentage of samples.</div>
</td>
</tr>
</tbody>
</table>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../biolockj/module/report/otu/package-summary.html">biolockj.module.report.otu</a> with parameters of type <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>protected boolean</code></td>
<td class="colLast"><span class="typeNameLabel">OtuCountModule.</span><code><span class="memberNameLink"><a href="../../../biolockj/module/report/otu/OtuCountModule.html#isOtuModule-biolockj.module.BioModule-">isOtuModule</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">Check the module to determine if it generated OTU count files.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="typeNameLabel">OtuCountModule.</span><code><span class="memberNameLink"><a href="../../../biolockj/module/report/otu/OtuCountModule.html#isValidInputModule-biolockj.module.BioModule-">isValidInputModule</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>&nbsp;</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module.report.r">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a> in <a href="../../../biolockj/module/report/r/package-summary.html">biolockj.module.report.r</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing classes, and an explanation">
<caption><span>Classes in <a href="../../../biolockj/module/report/r/package-summary.html">biolockj.module.report.r</a> that implement <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/r/R_CalculateStats.html" title="class in biolockj.module.report.r">R_CalculateStats</a></span></code>
<div class="block">This BioModule is used to build the R script used to generate taxonomy statistics and plots.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/r/R_Module.html" title="class in biolockj.module.report.r">R_Module</a></span></code>
<div class="block">This BioModule is the superclass for R script generating modules.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/r/R_PlotEffectSize.html" title="class in biolockj.module.report.r">R_PlotEffectSize</a></span></code>
<div class="block">This BioModule is used to run the R script used to generate OTU-metadata fold-change-barplots for each binary report
 field.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/r/R_PlotMds.html" title="class in biolockj.module.report.r">R_PlotMds</a></span></code>
<div class="block">This BioModule is used to build the R script used to generate MDS plots for each report field and each taxonomy level
 configured.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/r/R_PlotOtus.html" title="class in biolockj.module.report.r">R_PlotOtus</a></span></code>
<div class="block">This BioModule is used to build the R script used to generate OTU-metadata box-plots and scatter-plots for each
 report field and taxonomy level.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/r/R_PlotPvalHistograms.html" title="class in biolockj.module.report.r">R_PlotPvalHistograms</a></span></code>
<div class="block">This BioModule is used to build the R script used to generate p-value histograms for each report field and each
 taxonomy level configured.</div>
</td>
</tr>
</tbody>
</table>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../biolockj/module/report/r/package-summary.html">biolockj.module.report.r</a> with parameters of type <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/io/File.html?is-external=true" title="class or interface in java.io">File</a></code></td>
<td class="colLast"><span class="typeNameLabel">R_CalculateStats.</span><code><span class="memberNameLink"><a href="../../../biolockj/module/report/r/R_CalculateStats.html#getStatsFile-biolockj.module.BioModule-java.lang.String-java.lang.Boolean-java.lang.Boolean-">getStatsFile</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
            <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;level,
            <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Boolean.html?is-external=true" title="class or interface in java.lang">Boolean</a>&nbsp;isParametric,
            <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Boolean.html?is-external=true" title="class or interface in java.lang">Boolean</a>&nbsp;isAdjusted)</code>
<div class="block">Get the stats file for the given fileType and taxonomy level.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module.report.taxa">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a> in <a href="../../../biolockj/module/report/taxa/package-summary.html">biolockj.module.report.taxa</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing classes, and an explanation">
<caption><span>Classes in <a href="../../../biolockj/module/report/taxa/package-summary.html">biolockj.module.report.taxa</a> that implement <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/taxa/AddMetadataToTaxaTables.html" title="class in biolockj.module.report.taxa">AddMetadataToTaxaTables</a></span></code>
<div class="block">This BioModule is used to add metadata columns to the OTU abundance tables.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/taxa/BuildTaxaTables.html" title="class in biolockj.module.report.taxa">BuildTaxaTables</a></span></code>
<div class="block">Many R BioModules expect separate tables containing log-normalized taxa counts for each taxonomy level.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/taxa/LogTransformTaxaTables.html" title="class in biolockj.module.report.taxa">LogTransformTaxaTables</a></span></code>
<div class="block">This utility is used to log-transform the raw OTU counts on Log10 or Log-e scales.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/taxa/NormalizeTaxaTables.html" title="class in biolockj.module.report.taxa">NormalizeTaxaTables</a></span></code>
<div class="block">This utility is used to normalize and/or log-transform the raw OTU counts using the formulas:
 
 Normalized OTU count formula = (RC/n)*((SUM(x))/N)+1
 Relative abundance formula = Log(log_base) [ (RC/n)*((SUM(x))/N)+1 ]
 
 The code implementation supports (log_base = e) and (log_base = 10) which is configured via
 <a href="../../../biolockj/Constants.html#REPORT_LOG_BASE"><code>Constants.REPORT_LOG_BASE</code></a> property.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/taxa/TaxaCountModule.html" title="class in biolockj.module.report.taxa">TaxaCountModule</a></span></code>
<div class="block">TBD</div>
</td>
</tr>
</tbody>
</table>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../biolockj/module/report/taxa/package-summary.html">biolockj.module.report.taxa</a> with parameters of type <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="typeNameLabel">TaxaCountModule.</span><code><span class="memberNameLink"><a href="../../../biolockj/module/report/taxa/TaxaCountModule.html#isTaxaModule-biolockj.module.BioModule-">isTaxaModule</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">Check the module output directory for taxonomy table files generated by BioLockJ.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><span class="typeNameLabel">TaxaCountModule.</span><code><span class="memberNameLink"><a href="../../../biolockj/module/report/taxa/TaxaCountModule.html#isValidInputModule-biolockj.module.BioModule-">isValidInputModule</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>&nbsp;</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module.seq">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a> in <a href="../../../biolockj/module/seq/package-summary.html">biolockj.module.seq</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing classes, and an explanation">
<caption><span>Classes in <a href="../../../biolockj/module/seq/package-summary.html">biolockj.module.seq</a> that implement <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/seq/AwkFastaConverter.html" title="class in biolockj.module.seq">AwkFastaConverter</a></span></code>
<div class="block">This BioModule uses awk and gzip to convert input sequence files into a decompressed fasta file format.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/seq/Gunzipper.html" title="class in biolockj.module.seq">Gunzipper</a></span></code>
<div class="block">This BioModule uses gzip to decompress input sequence files.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/seq/KneadData.html" title="class in biolockj.module.seq">KneadData</a></span></code>
<div class="block">This BioModule runs biobakery kneaddata program to remove contaminated DNA.<br>
 Multiple contaminent DNA databases can be used to filter reads simultaniously.<br>
 Common contaminents include Human, Viral, and Plasmid DNA.<br></div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/seq/Multiplexer.html" title="class in biolockj.module.seq">Multiplexer</a></span></code>
<div class="block">This BioModule will merge sequence files into a single combined sequence file, with either the sample ID or an
 identifying bar-code (if defined in the metatata) is stored in the sequence header.<br>
 BioLockJ is designed to run on demultiplexed data so this must be the last module to run in its branch.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/seq/PearMergeReads.html" title="class in biolockj.module.seq">PearMergeReads</a></span></code>
<div class="block">This BioModule will merge forward and reverse fastq files using PEAR.<br>
 For more informations, see the online PEAR manual:
 <a href="https://sco.h-its.org/exelixis/web/software/pear/doc.html" target=
 "_top">https://sco.h-its.org/exelixis/web/software/pear/doc.html</a></div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/seq/RarefySeqs.html" title="class in biolockj.module.seq">RarefySeqs</a></span></code>
<div class="block">This BioModule imposes a minimum and/or maximum number of reads per sample.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/seq/SeqFileValidator.html" title="class in biolockj.module.seq">SeqFileValidator</a></span></code>
<div class="block">This BioModule validates fasta/fastq file formats are valid and enforces min/max read lengths.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/seq/TrimPrimers.html" title="class in biolockj.module.seq">TrimPrimers</a></span></code>
<div class="block">This BioModule removes sequence primers from demultiplexed files.<br>
 The primers are defined using regular expressions in a separate file.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.util">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a> in <a href="../../../biolockj/util/package-summary.html">biolockj.util</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../biolockj/util/package-summary.html">biolockj.util</a> that return <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>static <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></code></td>
<td class="colLast"><span class="typeNameLabel">ModuleUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/ModuleUtil.html#getModule-biolockj.module.BioModule-java.lang.String-boolean-">getModule</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
         <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;className,
         boolean&nbsp;checkAhead)</code>
<div class="block">Get a module with given className unless a classifier module is found 1st.<br>
 Use checkAhead parameter to determine if we look forward or backwards starting from the given module.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></code></td>
<td class="colLast"><span class="typeNameLabel">ModuleUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/ModuleUtil.html#getModule-java.lang.String-">getModule</a></span>(<a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;className)</code>
<div class="block">Construct a BioModule based on its className.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></code></td>
<td class="colLast"><span class="typeNameLabel">ModuleUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/ModuleUtil.html#getNextModule-biolockj.module.BioModule-">getNextModule</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">BioModules are run in the order configured.<br>
 Return the module configured to run after the given module.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></code></td>
<td class="colLast"><span class="typeNameLabel">ModuleUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/ModuleUtil.html#getPreviousModule-biolockj.module.BioModule-">getPreviousModule</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">BioModules are run in the order configured.<br>
 Return the module configured to run before the given module.</div>
</td>
</tr>
</tbody>
</table>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../biolockj/util/package-summary.html">biolockj.util</a> that return types with arguments of type <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>protected static <a href="https://docs.oracle.com/javase/8/docs/api/java/util/List.html?is-external=true" title="class or interface in java.util">List</a>&lt;<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">DownloadUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/DownloadUtil.html#getDownloadModules--">getDownloadModules</a></span>()</code>
<div class="block">Get the modules to download.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/util/List.html?is-external=true" title="class or interface in java.util">List</a>&lt;<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">ModuleUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/ModuleUtil.html#getModules-biolockj.module.BioModule-java.lang.Boolean-">getModules</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
          <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Boolean.html?is-external=true" title="class or interface in java.lang">Boolean</a>&nbsp;checkAhead)</code>
<div class="block">Return pipeline modules after the given module if checkAhead = TRUE<br>
 Otherwise return pipeline modules before the given module.<br>
 If returning the prior modules, return the pipeline modules in reverse order, so the 1st item in the list is the
 module immediately preceding the given module.</div>
</td>
</tr>
</tbody>
</table>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Methods in <a href="../../../biolockj/util/package-summary.html">biolockj.util</a> with parameters of type <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/util/List.html?is-external=true" title="class or interface in java.util">List</a>&lt;<a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">DockerUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/DockerUtil.html#buildSpawnDockerContainerFunction-biolockj.module.BioModule-">buildSpawnDockerContainerFunction</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">Build the <a href="../../../biolockj/util/DockerUtil.html#SPAWN_DOCKER_CONTAINER">"spawnDockerContainer"</a> method, which takes container name, in/out port, and optionally script
 path parameters.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static void</code></td>
<td class="colLast"><span class="typeNameLabel">RMetaUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/RMetaUtil.html#classifyReportableMetadata-biolockj.module.BioModule-">classifyReportableMetadata</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">Classify and verify the R filter and reportable metadata fields listed in the <a href="../../../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> file.<br>
 All metadata fields are reported unless specific fields are listed in:
 <a href="../../../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../../../biolockj/util/RMetaUtil.html#R_REPORT_FIELDS">"r.reportFields"</a>.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a></code></td>
<td class="colLast"><span class="typeNameLabel">ModuleUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/ModuleUtil.html#displayID-biolockj.module.BioModule-">displayID</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">Return the module ID as a 2 digit display number (add leading zero if needed).</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/util/Set.html?is-external=true" title="class or interface in java.util">Set</a>&lt;<a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">RMetaUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/RMetaUtil.html#getBinaryFields-biolockj.module.BioModule-">getBinaryFields</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">Get the <a href="../../../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../../../biolockj/util/RMetaUtil.html#BINARY_FIELDS">"R_internal.binaryFields"</a> fields containing only 2 non-numeric values.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="../../../biolockj/module/classifier/ClassifierModule.html" title="interface in biolockj.module.classifier">ClassifierModule</a></code></td>
<td class="colLast"><span class="typeNameLabel">ModuleUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/ModuleUtil.html#getClassifier-biolockj.module.BioModule-boolean-">getClassifier</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
             boolean&nbsp;checkAhead)</code>
<div class="block">Get a classifier module<br>
 Use checkAhead parameter to determine if we look forward or backwards starting from the given module.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a></code></td>
<td class="colLast"><span class="typeNameLabel">RuntimeParamUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/RuntimeParamUtil.html#getDirectModuleParam-biolockj.module.BioModule-">getDirectModuleParam</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">Direct module parameters contain 2 parts separated by a colon: (pipeline directory name):(module name)</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a></code></td>
<td class="colLast"><span class="typeNameLabel">DockerUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/DockerUtil.html#getDockerImage-biolockj.module.BioModule-">getDockerImage</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">Return the name of the Docker image needed for the given module.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a></code></td>
<td class="colLast"><span class="typeNameLabel">SummaryUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/SummaryUtil.html#getInputSummary-biolockj.module.BioModule-">getInputSummary</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">Build a summary of the input files for the given module</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></code></td>
<td class="colLast"><span class="typeNameLabel">ModuleUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/ModuleUtil.html#getModule-biolockj.module.BioModule-java.lang.String-boolean-">getModule</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
         <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;className,
         boolean&nbsp;checkAhead)</code>
<div class="block">Get a module with given className unless a classifier module is found 1st.<br>
 Use checkAhead parameter to determine if we look forward or backwards starting from the given module.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a></code></td>
<td class="colLast"><span class="typeNameLabel">SummaryUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/SummaryUtil.html#getModuleRunTime-biolockj.module.BioModule-">getModuleRunTime</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">Return duration module ran based on modified data of started file, formatted for display (as hours, minutes,
 seconds).</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/util/List.html?is-external=true" title="class or interface in java.util">List</a>&lt;<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&gt;</code></td>
<td class="colLast"><span class="typeNameLabel">ModuleUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/ModuleUtil.html#getModules-biolockj.module.BioModule-java.lang.Boolean-">getModules</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
          <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Boolean.html?is-external=true" title="class or interface in java.lang">Boolean</a>&nbsp;checkAhead)</code>
<div class="block">Return pipeline modules after the given module if checkAhead = TRUE<br>
 Otherwise return pipeline modules before the given module.<br>
 If returning the prior modules, return the pipeline modules in reverse order, so the 1st item in the list is the
 module immediately preceding the given module.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></code></td>
<td class="colLast"><span class="typeNameLabel">ModuleUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/ModuleUtil.html#getNextModule-biolockj.module.BioModule-">getNextModule</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">BioModules are run in the order configured.<br>
 Return the module configured to run after the given module.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a></code></td>
<td class="colLast"><span class="typeNameLabel">SummaryUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/SummaryUtil.html#getOutputDirSummary-biolockj.module.BioModule-">getOutputDirSummary</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">Return summary of <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a> output directory, with metrics:
 
 Number of output files
 Mean output file size
 Path of new metadata file if any created
 </div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></code></td>
<td class="colLast"><span class="typeNameLabel">ModuleUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/ModuleUtil.html#getPreviousModule-biolockj.module.BioModule-">getPreviousModule</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">BioModules are run in the order configured.<br>
 Return the module configured to run before the given module.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a></code></td>
<td class="colLast"><span class="typeNameLabel">MetaUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/MetaUtil.html#getSystemMetaCol-biolockj.module.BioModule-java.lang.String-">getSystemMetaCol</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
                <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;col)</code>
<div class="block">Return a system generated metadata column name based on the module status.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static boolean</code></td>
<td class="colLast"><span class="typeNameLabel">ModuleUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/ModuleUtil.html#hasExecuted-biolockj.module.BioModule-">hasExecuted</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">Return TRUE if module has executed.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static boolean</code></td>
<td class="colLast"><span class="typeNameLabel">ModuleUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/ModuleUtil.html#isComplete-biolockj.module.BioModule-">isComplete</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">Return TRUE if module completed successfully.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static boolean</code></td>
<td class="colLast"><span class="typeNameLabel">ModuleUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/ModuleUtil.html#isFirstRModule-biolockj.module.BioModule-">isFirstRModule</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">Test if module is the first <a href="../../../biolockj/module/report/r/R_Module.html" title="class in biolockj.module.report.r"><code>R_Module</code></a> configured in the pipeline.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static boolean</code></td>
<td class="colLast"><span class="typeNameLabel">ModuleUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/ModuleUtil.html#isIncomplete-biolockj.module.BioModule-">isIncomplete</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">Return TRUE if module started execution but is not complete.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static boolean</code></td>
<td class="colLast"><span class="typeNameLabel">ModuleUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/ModuleUtil.html#isMetadataModule-biolockj.module.BioModule-">isMetadataModule</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">Method determines if the given module is a metadata-module (which does not use/modify sequence data.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static boolean</code></td>
<td class="colLast"><span class="typeNameLabel">RMetaUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/RMetaUtil.html#isMetaMergeModule-biolockj.module.BioModule-">isMetaMergeModule</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">Check module for metadata merged output in hte module output directory.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static boolean</code></td>
<td class="colLast"><span class="typeNameLabel">PathwayUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/PathwayUtil.html#isPathwayModule-biolockj.module.BioModule-">isPathwayModule</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">Check the module to determine if it generated OTU count files.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static boolean</code></td>
<td class="colLast"><span class="typeNameLabel">SeqUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/SeqUtil.html#isSeqModule-biolockj.module.BioModule-">isSeqModule</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">Check the module to determine if it generated sequence file output.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static void</code></td>
<td class="colLast"><span class="typeNameLabel">ModuleUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/ModuleUtil.html#markComplete-biolockj.module.BioModule-">markComplete</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">Method creates a file named <a href="../../../biolockj/Constants.html#BLJ_COMPLETE">"biolockjComplete"</a> in module root directory to document module
 has completed successfully.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static void</code></td>
<td class="colLast"><span class="typeNameLabel">ModuleUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/ModuleUtil.html#markStarted-biolockj.module.BioModule-">markStarted</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">Method creates a file named <a href="../../../biolockj/Constants.html#BLJ_STARTED">"biolockjStarted"</a> in module root directory to document module
 has completed successfully.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static boolean</code></td>
<td class="colLast"><span class="typeNameLabel">RMetaUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/RMetaUtil.html#reportAllFields-biolockj.module.BioModule-">reportAllFields</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">The override property: <a href="../../../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../../../biolockj/util/RMetaUtil.html#R_REPORT_FIELDS">"r.reportFields"</a> can be used to list the metadata
 reportable fields for use in the R modules.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static void</code></td>
<td class="colLast"><span class="typeNameLabel">SummaryUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/SummaryUtil.html#reportSuccess-biolockj.module.BioModule-">reportSuccess</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">After each module completes, this method is called to track the execution summary.<br>
 If module is null, the pipeline is complete/successful.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/io/File.html?is-external=true" title="class or interface in java.io">File</a></code></td>
<td class="colLast"><span class="typeNameLabel">ModuleUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/ModuleUtil.html#requireSubDir-biolockj.module.BioModule-java.lang.String-">requireSubDir</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
             <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;subDirName)</code>
<div class="block">Get BioModule subdirectory File object with given name.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>protected static void</code></td>
<td class="colLast"><span class="typeNameLabel">SummaryUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/SummaryUtil.html#resetModuleSummary-biolockj.module.BioModule-">resetModuleSummary</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">Modules can be forced to reset to incomplete status.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static boolean</code></td>
<td class="colLast"><span class="typeNameLabel">ModuleUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/ModuleUtil.html#subDirExists-biolockj.module.BioModule-java.lang.String-">subDirExists</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
            <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;subDirName)</code>
<div class="block">Return TRUE if BioModule sub-directory exists</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static boolean</code></td>
<td class="colLast"><span class="typeNameLabel">RMetaUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/RMetaUtil.html#updateRConfig-biolockj.module.BioModule-">updateRConfig</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">Get updated R config props</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Boolean.html?is-external=true" title="class or interface in java.lang">Boolean</a></code></td>
<td class="colLast"><span class="typeNameLabel">PathwayUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/PathwayUtil.html#useHumann2RawCount-biolockj.module.BioModule-">useHumann2RawCount</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">Determine if humann2 provided most recent raw count data, used to determine getPreReq modules.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static void</code></td>
<td class="colLast"><span class="typeNameLabel">PathwayUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/PathwayUtil.html#verifyConfig-biolockj.module.BioModule-">verifyConfig</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module)</code>
<div class="block">Verify the HumanN2 Config contains at least one of the following reports are enabled:<br>
 
 <a href="../../../biolockj/Constants.html#HN2_DISABLE_GENE_FAMILIES">"humann2.disableGeneFamilies"</a>
 <a href="../../../biolockj/Constants.html#HN2_DISABLE_PATH_COVERAGE">"humann2.disablePathCoverage"</a>
 <a href="../../../biolockj/Constants.html#HN2_DISABLE_GENE_FAMILIES">"humann2.disableGeneFamilies"</a>
 </div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static void</code></td>
<td class="colLast"><span class="typeNameLabel">RMetaUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/RMetaUtil.html#verifyMetadataFieldsExist-biolockj.module.BioModule-java.lang.String-java.util.Collection-">verifyMetadataFieldsExist</a></span>(<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&nbsp;module,
                         <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&nbsp;prop,
                         <a href="https://docs.oracle.com/javase/8/docs/api/java/util/Collection.html?is-external=true" title="class or interface in java.util">Collection</a>&lt;<a href="https://docs.oracle.com/javase/8/docs/api/java/lang/String.html?is-external=true" title="class or interface in java.lang">String</a>&gt;&nbsp;fields)</code>
<div class="block">This method verifies the fields given exist in the metadata file.</div>
</td>
</tr>
</tbody>
</table>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing methods, and an explanation">
<caption><span>Method parameters in <a href="../../../biolockj/util/package-summary.html">biolockj.util</a> with type arguments of type <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>static <a href="https://docs.oracle.com/javase/8/docs/api/java/io/File.html?is-external=true" title="class or interface in java.io">File</a></code></td>
<td class="colLast"><span class="typeNameLabel">NextflowUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/NextflowUtil.html#buildNextflowMain-java.util.List-">buildNextflowMain</a></span>(<a href="https://docs.oracle.com/javase/8/docs/api/java/util/List.html?is-external=true" title="class or interface in java.util">List</a>&lt;<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&gt;&nbsp;modules)</code>
<div class="block">Call this method to build the Nextflow main.nf for the current pipeline.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>protected static <a href="https://docs.oracle.com/javase/8/docs/api/java/io/File.html?is-external=true" title="class or interface in java.io">File</a></code></td>
<td class="colLast"><span class="typeNameLabel">DownloadUtil.</span><code><span class="memberNameLink"><a href="../../../biolockj/util/DownloadUtil.html#makeRunAllScript-java.util.List-">makeRunAllScript</a></span>(<a href="https://docs.oracle.com/javase/8/docs/api/java/util/List.html?is-external=true" title="class or interface in java.util">List</a>&lt;<a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a>&gt;&nbsp;modules)</code>
<div class="block">This script allows a user to run all R scripts together from a single script.</div>
</td>
</tr>
</tbody>
</table>
</li>
</ul>
</li>
</ul>
</div>
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar.bottom">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.bottom.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../overview-summary.html">Overview</a></li>
<li><a href="../package-summary.html">Package</a></li>
<li><a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module">Class</a></li>
<li class="navBarCell1Rev">Use</li>
<li><a href="../package-tree.html">Tree</a></li>
<li><a href="../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../index-files/index-1.html">Index</a></li>
<li><a href="../../../help-doc.html">Help</a></li>
<li>&nbsp;&nbsp;</li>
<li><a href="https://github.com/msioda/BioLockJ/wiki" target="_top"><strong>RETURN to WIKI</strong></a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li>Prev</li>
<li>Next</li>
</ul>
<ul class="navList">
<li><a href="../../../index.html?biolockj/module/class-use/BioModule.html" target="_top">Frames</a></li>
<li><a href="BioModule.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip.navbar.bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
