<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (1.8.0_144) on Wed Apr 24 00:13:52 EDT 2019 -->
<title>Uses of Interface biolockj.module.JavaModule</title>
<meta name="date" content="2019-04-24">
<link rel="stylesheet" type="text/css" href="../../../stylesheet.css" title="Style">
<script type="text/javascript" src="../../../script.js"></script>
</head>
<body>
<script type="text/javascript"><!--
    try {
        if (location.href.indexOf('is-external=true') == -1) {
            parent.document.title="Uses of Interface biolockj.module.JavaModule";
        }
    }
    catch(err) {
    }
//-->
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar.top">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.top.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../overview-summary.html">Overview</a></li>
<li><a href="../package-summary.html">Package</a></li>
<li><a href="../../../biolockj/module/JavaModule.html" title="interface in biolockj.module">Class</a></li>
<li class="navBarCell1Rev">Use</li>
<li><a href="../package-tree.html">Tree</a></li>
<li><a href="../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../index-files/index-1.html">Index</a></li>
<li><a href="../../../help-doc.html">Help</a></li>
<li>&nbsp;&nbsp;</li>
<li><a href="https://github.com/msioda/BioLockJ/wiki" target="_top"><strong>RETURN to WIKI</strong></a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li>Prev</li>
<li>Next</li>
</ul>
<ul class="navList">
<li><a href="../../../index.html?biolockj/module/class-use/JavaModule.html" target="_top">Frames</a></li>
<li><a href="JavaModule.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip.navbar.top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<div class="header">
<h2 title="Uses of Interface biolockj.module.JavaModule" class="title">Uses of Interface<br>biolockj.module.JavaModule</h2>
</div>
<div class="classUseContainer">
<ul class="blockList">
<li class="blockList">
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing packages, and an explanation">
<caption><span>Packages that use <a href="../../../biolockj/module/JavaModule.html" title="interface in biolockj.module">JavaModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Package</th>
<th class="colLast" scope="col">Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><a href="#biolockj.module">biolockj.module</a></td>
<td class="colLast">
<div class="block">Pipeline modules are loaded based on the <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s listed in the <a href="../../../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>
 file.<br>
 All modules (i.e., implement <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>) are contained within this package or its
 sub-packages.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><a href="#biolockj.module.implicit">biolockj.module.implicit</a></td>
<td class="colLast">
<div class="block">The modules in this package are implicitly added to pipelines as needed.<br>
 These modules cannot be directly added to any pipeline unless overridden via pipeline.disableImplicitModules=Y.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><a href="#biolockj.module.implicit.parser">biolockj.module.implicit.parser</a></td>
<td class="colLast">
<div class="block">This package contains Parser BioModules in the r16s and WGS sub-packages that are paired with a
 <a href="../../../biolockj/module/classifier/ClassifierModule.html" title="interface in biolockj.module.classifier"><code>ClassifierModule</code></a> via
 <a href="../../../biolockj/module/BioModule.html#getPostRequisiteModules--"><code>BioModule.getPostRequisiteModules()</code></a> to run immediately after the
 classifier.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><a href="#biolockj.module.implicit.parser.r16s">biolockj.module.implicit.parser.r16s</a></td>
<td class="colLast">
<div class="block">This package contains Parser <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s that convert the 16S taxonomy reports generated by
 16S classifiers (such as RDP and QIIME) into standardized OTU abundance tables.<br></div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><a href="#biolockj.module.implicit.parser.wgs">biolockj.module.implicit.parser.wgs</a></td>
<td class="colLast">
<div class="block">This package contains Parser <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s that convert the WGS taxonomy reports generated by
 WGS classifiers (such as Kraken, Kraken2 Metaphlan2, or Humann2) into standardized OTU abundance tables.<br></div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><a href="#biolockj.module.report">biolockj.module.report</a></td>
<td class="colLast">
<div class="block">This package contains <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s that normalize OTU abundance tables output by Parser
 modules, merges them with the metadata, and generates various reports and notifications.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><a href="#biolockj.module.report.humann2">biolockj.module.report.humann2</a></td>
<td class="colLast">&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><a href="#biolockj.module.report.otu">biolockj.module.report.otu</a></td>
<td class="colLast">&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><a href="#biolockj.module.report.taxa">biolockj.module.report.taxa</a></td>
<td class="colLast">&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><a href="#biolockj.module.seq">biolockj.module.seq</a></td>
<td class="colLast">
<div class="block"><a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s used to prepare sequence files or update the metadata prior to classification.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList">
<ul class="blockList">
<li class="blockList"><a name="biolockj.module">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/JavaModule.html" title="interface in biolockj.module">JavaModule</a> in <a href="../../../biolockj/module/package-summary.html">biolockj.module</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing classes, and an explanation">
<caption><span>Classes in <a href="../../../biolockj/module/package-summary.html">biolockj.module</a> that implement <a href="../../../biolockj/module/JavaModule.html" title="interface in biolockj.module">JavaModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/JavaModuleImpl.html" title="class in biolockj.module">JavaModuleImpl</a></span></code>
<div class="block">Superclass for Java BioModules that will be called in separate instances of the application.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module.implicit">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/JavaModule.html" title="interface in biolockj.module">JavaModule</a> in <a href="../../../biolockj/module/implicit/package-summary.html">biolockj.module.implicit</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing classes, and an explanation">
<caption><span>Classes in <a href="../../../biolockj/module/implicit/package-summary.html">biolockj.module.implicit</a> that implement <a href="../../../biolockj/module/JavaModule.html" title="interface in biolockj.module">JavaModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/Demultiplexer.html" title="class in biolockj.module.implicit">Demultiplexer</a></span></code>
<div class="block">This BioModule splits multiplexed data into a separate file or pair of files (for paired reads) for each sample.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/RegisterNumReads.html" title="class in biolockj.module.implicit">RegisterNumReads</a></span></code>
<div class="block">This BioModule parses sequence file to count the number of reads per sample.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module.implicit.parser">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/JavaModule.html" title="interface in biolockj.module">JavaModule</a> in <a href="../../../biolockj/module/implicit/parser/package-summary.html">biolockj.module.implicit.parser</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing subinterfaces, and an explanation">
<caption><span>Subinterfaces of <a href="../../../biolockj/module/JavaModule.html" title="interface in biolockj.module">JavaModule</a> in <a href="../../../biolockj/module/implicit/parser/package-summary.html">biolockj.module.implicit.parser</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Interface and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>interface&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/parser/ParserModule.html" title="interface in biolockj.module.implicit.parser">ParserModule</a></span></code>
<div class="block">This interface defines the required methods to parse ClassifierModule output.</div>
</td>
</tr>
</tbody>
</table>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing classes, and an explanation">
<caption><span>Classes in <a href="../../../biolockj/module/implicit/parser/package-summary.html">biolockj.module.implicit.parser</a> that implement <a href="../../../biolockj/module/JavaModule.html" title="interface in biolockj.module">JavaModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/parser/ParserModuleImpl.html" title="class in biolockj.module.implicit.parser">ParserModuleImpl</a></span></code>
<div class="block">Parser <a href="../../../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s read <a href="../../../biolockj/module/classifier/ClassifierModule.html" title="interface in biolockj.module.classifier"><code>ClassifierModule</code></a> output to build
 standardized OTU count tables.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module.implicit.parser.r16s">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/JavaModule.html" title="interface in biolockj.module">JavaModule</a> in <a href="../../../biolockj/module/implicit/parser/r16s/package-summary.html">biolockj.module.implicit.parser.r16s</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing classes, and an explanation">
<caption><span>Classes in <a href="../../../biolockj/module/implicit/parser/r16s/package-summary.html">biolockj.module.implicit.parser.r16s</a> that implement <a href="../../../biolockj/module/JavaModule.html" title="interface in biolockj.module">JavaModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/parser/r16s/QiimeParser.html" title="class in biolockj.module.implicit.parser.r16s">QiimeParser</a></span></code>
<div class="block">This BioModules parses QiimeClassifier output reports to build standard OTU abundance tables.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/parser/r16s/RdpParser.html" title="class in biolockj.module.implicit.parser.r16s">RdpParser</a></span></code>
<div class="block">This BioModule parses RDP output files to build standard OTU abundance tables.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module.implicit.parser.wgs">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/JavaModule.html" title="interface in biolockj.module">JavaModule</a> in <a href="../../../biolockj/module/implicit/parser/wgs/package-summary.html">biolockj.module.implicit.parser.wgs</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing classes, and an explanation">
<caption><span>Classes in <a href="../../../biolockj/module/implicit/parser/wgs/package-summary.html">biolockj.module.implicit.parser.wgs</a> that implement <a href="../../../biolockj/module/JavaModule.html" title="interface in biolockj.module">JavaModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/parser/wgs/Humann2Parser.html" title="class in biolockj.module.implicit.parser.wgs">Humann2Parser</a></span></code>
<div class="block">This BioModules parses Humann2Classifier output reports to build standard OTU abundance tables.<br>
 Samples IDs are found in the column headers starting with the 2nd column.<br>
 The count type depends on the HumanN2 config properties.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/parser/wgs/Kraken2Parser.html" title="class in biolockj.module.implicit.parser.wgs">Kraken2Parser</a></span></code>
<div class="block">This BioModules parses KrakenClassifier output reports to build standard OTU abundance tables.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/parser/wgs/KrakenParser.html" title="class in biolockj.module.implicit.parser.wgs">KrakenParser</a></span></code>
<div class="block">This BioModules parses KrakenClassifier output reports to build standard OTU abundance tables.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/implicit/parser/wgs/Metaphlan2Parser.html" title="class in biolockj.module.implicit.parser.wgs">Metaphlan2Parser</a></span></code>
<div class="block">This BioModules parses Metaphlan2Classifier output reports to build standard OTU abundance tables.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module.report">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/JavaModule.html" title="interface in biolockj.module">JavaModule</a> in <a href="../../../biolockj/module/report/package-summary.html">biolockj.module.report</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing classes, and an explanation">
<caption><span>Classes in <a href="../../../biolockj/module/report/package-summary.html">biolockj.module.report</a> that implement <a href="../../../biolockj/module/JavaModule.html" title="interface in biolockj.module">JavaModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/JsonReport.html" title="class in biolockj.module.report">JsonReport</a></span></code>
<div class="block">This BioModule is used to build a JSON file (summary.json) compiled from all OTUs in the dataset.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module.report.humann2">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/JavaModule.html" title="interface in biolockj.module">JavaModule</a> in <a href="../../../biolockj/module/report/humann2/package-summary.html">biolockj.module.report.humann2</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing classes, and an explanation">
<caption><span>Classes in <a href="../../../biolockj/module/report/humann2/package-summary.html">biolockj.module.report.humann2</a> that implement <a href="../../../biolockj/module/JavaModule.html" title="interface in biolockj.module">JavaModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/humann2/AddMetadataToPathwayTables.html" title="class in biolockj.module.report.humann2">AddMetadataToPathwayTables</a></span></code>
<div class="block">This BioModule is used to add metadata columns to the HumanN2 pathway abundance, pathway coverage, and gene family
 tables.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/humann2/Humann2CountModule.html" title="class in biolockj.module.report.humann2">Humann2CountModule</a></span></code>
<div class="block">This abstract superclass is extended by all other modules in this package.<br>
 Shared method implementations are defined to ensure uniform adoption of dependencies and prerequisites.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/humann2/RemoveLowPathwayCounts.html" title="class in biolockj.module.report.humann2">RemoveLowPathwayCounts</a></span></code>
<div class="block">This BioModule set low Pathway counts below a configured threshold to zero.<br>
 These low sample counts are assumed to be miscategorized or genomic contamination.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/humann2/RemoveScarcePathwayCounts.html" title="class in biolockj.module.report.humann2">RemoveScarcePathwayCounts</a></span></code>
<div class="block">This BioModule removes scarce pathways not found in enough samples.<br>
 Each pathway must be found in a configurable percentage of samples to be retained.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module.report.otu">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/JavaModule.html" title="interface in biolockj.module">JavaModule</a> in <a href="../../../biolockj/module/report/otu/package-summary.html">biolockj.module.report.otu</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing classes, and an explanation">
<caption><span>Classes in <a href="../../../biolockj/module/report/otu/package-summary.html">biolockj.module.report.otu</a> that implement <a href="../../../biolockj/module/JavaModule.html" title="interface in biolockj.module">JavaModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/otu/CompileOtuCounts.html" title="class in biolockj.module.report.otu">CompileOtuCounts</a></span></code>
<div class="block">This BioModule compiles the counts from all OTU count files into a single summary OTU count file containing OTU
 counts for the entire dataset.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/otu/OtuCountModule.html" title="class in biolockj.module.report.otu">OtuCountModule</a></span></code>
<div class="block">OtuCount modules reads OTU count assignment tables (1 file/sample) with 2 columns.<br>
 Col1: Full OTU pathway spanning top to bottom level Col2: Count (# of reads) for the sample.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/otu/RarefyOtuCounts.html" title="class in biolockj.module.report.otu">RarefyOtuCounts</a></span></code>
<div class="block">This BioModule applies a mean iterative post-OTU classification rarefication algorithm so that each output sample
 will have approximately the same number of OTUs.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/otu/RemoveLowOtuCounts.html" title="class in biolockj.module.report.otu">RemoveLowOtuCounts</a></span></code>
<div class="block">This BioModule set low OTU counts below a configured threshold to zero.<br>
 These low sample counts are assumed to be miscategorized or contaminents.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/otu/RemoveScarceOtuCounts.html" title="class in biolockj.module.report.otu">RemoveScarceOtuCounts</a></span></code>
<div class="block">This BioModule removes scarce OTUs not found in enough samples.<br>
 The OTU must be found in a configurable percentage of samples.</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module.report.taxa">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/JavaModule.html" title="interface in biolockj.module">JavaModule</a> in <a href="../../../biolockj/module/report/taxa/package-summary.html">biolockj.module.report.taxa</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing classes, and an explanation">
<caption><span>Classes in <a href="../../../biolockj/module/report/taxa/package-summary.html">biolockj.module.report.taxa</a> that implement <a href="../../../biolockj/module/JavaModule.html" title="interface in biolockj.module">JavaModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/taxa/AddMetadataToTaxaTables.html" title="class in biolockj.module.report.taxa">AddMetadataToTaxaTables</a></span></code>
<div class="block">This BioModule is used to add metadata columns to the OTU abundance tables.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/taxa/BuildTaxaTables.html" title="class in biolockj.module.report.taxa">BuildTaxaTables</a></span></code>
<div class="block">Many R BioModules expect separate tables containing log-normalized taxa counts for each taxonomy level.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/taxa/LogTransformTaxaTables.html" title="class in biolockj.module.report.taxa">LogTransformTaxaTables</a></span></code>
<div class="block">This utility is used to log-transform the raw OTU counts on Log10 or Log-e scales.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/taxa/NormalizeTaxaTables.html" title="class in biolockj.module.report.taxa">NormalizeTaxaTables</a></span></code>
<div class="block">This utility is used to normalize and/or log-transform the raw OTU counts using the formulas:
 
 Normalized OTU count formula = (RC/n)*((SUM(x))/N)+1
 Relative abundance formula = Log(log_base) [ (RC/n)*((SUM(x))/N)+1 ]
 
 The code implementation supports (log_base = e) and (log_base = 10) which is configured via
 <a href="../../../biolockj/Constants.html#REPORT_LOG_BASE"><code>Constants.REPORT_LOG_BASE</code></a> property.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/report/taxa/TaxaCountModule.html" title="class in biolockj.module.report.taxa">TaxaCountModule</a></span></code>
<div class="block">TBD</div>
</td>
</tr>
</tbody>
</table>
</li>
<li class="blockList"><a name="biolockj.module.seq">
<!--   -->
</a>
<h3>Uses of <a href="../../../biolockj/module/JavaModule.html" title="interface in biolockj.module">JavaModule</a> in <a href="../../../biolockj/module/seq/package-summary.html">biolockj.module.seq</a></h3>
<table class="useSummary" border="0" cellpadding="3" cellspacing="0" summary="Use table, listing classes, and an explanation">
<caption><span>Classes in <a href="../../../biolockj/module/seq/package-summary.html">biolockj.module.seq</a> that implement <a href="../../../biolockj/module/JavaModule.html" title="interface in biolockj.module">JavaModule</a></span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Class and Description</th>
</tr>
<tbody>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/seq/Multiplexer.html" title="class in biolockj.module.seq">Multiplexer</a></span></code>
<div class="block">This BioModule will merge sequence files into a single combined sequence file, with either the sample ID or an
 identifying bar-code (if defined in the metatata) is stored in the sequence header.<br>
 BioLockJ is designed to run on demultiplexed data so this must be the last module to run in its branch.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/seq/RarefySeqs.html" title="class in biolockj.module.seq">RarefySeqs</a></span></code>
<div class="block">This BioModule imposes a minimum and/or maximum number of reads per sample.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/seq/SeqFileValidator.html" title="class in biolockj.module.seq">SeqFileValidator</a></span></code>
<div class="block">This BioModule validates fasta/fastq file formats are valid and enforces min/max read lengths.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>class&nbsp;</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../../../biolockj/module/seq/TrimPrimers.html" title="class in biolockj.module.seq">TrimPrimers</a></span></code>
<div class="block">This BioModule removes sequence primers from demultiplexed files.<br>
 The primers are defined using regular expressions in a separate file.</div>
</td>
</tr>
</tbody>
</table>
</li>
</ul>
</li>
</ul>
</div>
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar.bottom">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.bottom.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../../overview-summary.html">Overview</a></li>
<li><a href="../package-summary.html">Package</a></li>
<li><a href="../../../biolockj/module/JavaModule.html" title="interface in biolockj.module">Class</a></li>
<li class="navBarCell1Rev">Use</li>
<li><a href="../package-tree.html">Tree</a></li>
<li><a href="../../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../../index-files/index-1.html">Index</a></li>
<li><a href="../../../help-doc.html">Help</a></li>
<li>&nbsp;&nbsp;</li>
<li><a href="https://github.com/msioda/BioLockJ/wiki" target="_top"><strong>RETURN to WIKI</strong></a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li>Prev</li>
<li>Next</li>
</ul>
<ul class="navList">
<li><a href="../../../index.html?biolockj/module/class-use/JavaModule.html" target="_top">Frames</a></li>
<li><a href="JavaModule.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../../../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip.navbar.bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
