<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (1.8.0_144) on Tue Mar 12 06:27:48 EDT 2019 -->
<title>I-Index</title>
<meta name="date" content="2019-03-12">
<link rel="stylesheet" type="text/css" href="../stylesheet.css" title="Style">
<script type="text/javascript" src="../script.js"></script>
</head>
<body>
<script type="text/javascript"><!--
    try {
        if (location.href.indexOf('is-external=true') == -1) {
            parent.document.title="I-Index";
        }
    }
    catch(err) {
    }
//-->
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar.top">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.top.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../overview-summary.html">Overview</a></li>
<li>Package</li>
<li>Class</li>
<li>Use</li>
<li><a href="../overview-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li class="navBarCell1Rev">Index</li>
<li><a href="../help-doc.html">Help</a></li>
<li>&nbsp;&nbsp;</li>
<li><a href="https://github.com/msioda/BioLockJ/wiki" target="_top"><strong>RETURN to WIKI</strong></a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="index-8.html">Prev Letter</a></li>
<li><a href="index-10.html">Next Letter</a></li>
</ul>
<ul class="navList">
<li><a href="../index.html?index-files/index-9.html" target="_top">Frames</a></li>
<li><a href="index-9.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip.navbar.top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<div class="contentContainer"><a href="index-1.html">A</a>&nbsp;<a href="index-2.html">B</a>&nbsp;<a href="index-3.html">C</a>&nbsp;<a href="index-4.html">D</a>&nbsp;<a href="index-5.html">E</a>&nbsp;<a href="index-6.html">F</a>&nbsp;<a href="index-7.html">G</a>&nbsp;<a href="index-8.html">H</a>&nbsp;<a href="index-9.html">I</a>&nbsp;<a href="index-10.html">J</a>&nbsp;<a href="index-11.html">K</a>&nbsp;<a href="index-12.html">L</a>&nbsp;<a href="index-13.html">M</a>&nbsp;<a href="index-14.html">N</a>&nbsp;<a href="index-15.html">O</a>&nbsp;<a href="index-16.html">P</a>&nbsp;<a href="index-17.html">Q</a>&nbsp;<a href="index-18.html">R</a>&nbsp;<a href="index-19.html">S</a>&nbsp;<a href="index-20.html">T</a>&nbsp;<a href="index-21.html">U</a>&nbsp;<a href="index-22.html">V</a>&nbsp;<a href="index-23.html">W</a>&nbsp;<a name="I:I">
<!--   -->
</a>
<h2 class="title">I</h2>
<dl>
<dt><span class="memberNameLink"><a href="../biolockj/util/BioLockJUtil.html#ignoreFile-java.io.File-">ignoreFile(File)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BioLockJUtil.html" title="class in biolockj.util">BioLockJUtil</a></dt>
<dd>
<div class="block">Method used to add a file to the ignore file list property.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#ILLUMINA_FW_READ_IND">ILLUMINA_FW_READ_IND</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Illumina forward read indicator found in sequence header</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#ILLUMINA_RV_READ_IND">ILLUMINA_RV_READ_IND</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Illumina reverse read indicator found in sequence header</div>
</dd>
<dt><a href="../biolockj/module/implicit/ImportMetadata.html" title="class in biolockj.module.implicit"><span class="typeNameLink">ImportMetadata</span></a> - Class in <a href="../biolockj/module/implicit/package-summary.html">biolockj.module.implicit</a></dt>
<dd>
<div class="block">This BioModule validates the contents/format of the project metadata file and the related Config properties.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/ImportMetadata.html#ImportMetadata--">ImportMetadata()</a></span> - Constructor for class biolockj.module.implicit.<a href="../biolockj/module/implicit/ImportMetadata.html" title="class in biolockj.module.implicit">ImportMetadata</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DockerUtil.html#inAwsEnv--">inAwsEnv()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/DockerUtil.html" title="class in biolockj.util">DockerUtil</a></dt>
<dd>
<div class="block">Return TRUE if running on AWS (based on Config props).</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#INDENT">INDENT</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block">Standard indent = 4 spaces.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Log.html#info-java.lang.Class-java.lang.String-">info(Class&lt;?&gt;, String)</a></span> - Static method in class biolockj.<a href="../biolockj/Log.html" title="class in biolockj">Log</a></dt>
<dd>
<div class="block">Print log level INFO message.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModule.html#init--">init()</a></span> - Method in interface biolockj.module.<a href="../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></dt>
<dd>
<div class="block">Initialize a new module to generate a unique ID and module directory.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModuleImpl.html#init--">init()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a></dt>
<dd>
<div class="block">This method must be called immediately upon instantiation.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DockerUtil.html#initAwsCloudManager--">initAwsCloudManager()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/DockerUtil.html" title="class in biolockj.util">DockerUtil</a></dt>
<dd>
<div class="block">Boolean to determine if should initialize Docker aws_manager</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/BioLockJ.html#initAwsManager--">initAwsManager()</a></span> - Static method in class biolockj.<a href="../biolockj/BioLockJ.html" title="class in biolockj">BioLockJ</a></dt>
<dd>
<div class="block">Initialize AWS manager pipeline:<br>
 
 Build Nexflow main.nf
 Run ImportMetadata module
 Set files editable
 Update pipeline root directory to EFS directory
 Update EFS MASTER <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> with new pipeline root directory path
 Save pipeline input files to EFS for faster processing
 </div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/BioLockJ.html#initBioLockJ-java.lang.String:A-">initBioLockJ(String[])</a></span> - Static method in class biolockj.<a href="../biolockj/BioLockJ.html" title="class in biolockj">BioLockJ</a></dt>
<dd>
<div class="block">Execution summary:<br>
 
 Call <a href="../biolockj/util/MemoryUtil.html#reportMemoryUsage-java.lang.String-"><code>MemoryUtil.reportMemoryUsage(String)</code></a> for baseline memory info
 Call <a href="../biolockj/util/RuntimeParamUtil.html#registerRuntimeParameters-java.lang.String:A-"><code>registerRuntimeParameters(String[])</code></a>
 Call <a href="../biolockj/util/MetaUtil.html#initialize--"><code>MetaUtil.initialize()</code></a> to verify metadata dependencies
 Call <a href="../biolockj/Config.html#initialize--"><code>Config.initialize()</code></a> to create pipeline root dir and load properties
 Initialize <a href="../biolockj/Log.html" title="class in biolockj"><code>Log</code></a> with /resources/log4J.properties
 Copy initial metadata file into the pipeline root directory
 Call <a href="../biolockj/util/SeqUtil.html#initialize--"><code>SeqUtil.initialize()</code></a> to set Config parameters based on sequence files
 </div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/BioModuleFactory.html#initFactory--">initFactory()</a></span> - Static method in class biolockj.<a href="../biolockj/BioModuleFactory.html" title="class in biolockj">BioModuleFactory</a></dt>
<dd>
<div class="block">Register the complete list of modules to run.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#initialize--">initialize()</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Initialize Config by reading in props from file and setting taxonomy as an ordered list.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Log.html#initialize-java.lang.String-">initialize(String)</a></span> - Static method in class biolockj.<a href="../biolockj/Log.html" title="class in biolockj">Log</a></dt>
<dd>
<div class="block">Called by <a href="../biolockj/BioLockJ.html#main-java.lang.String:A-"><code>BioLockJ.main()</code></a> after <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> is initialized.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/MetaUtil.html#initialize--">initialize()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/MetaUtil.html" title="class in biolockj.util">MetaUtil</a></dt>
<dd>
<div class="block">Check dependencies used to handle the metadata file are defined and unique.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#initialize--">initialize()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Initialize Config params set by SeqUtil.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/r16s/QiimeParser.html#initializeMaps--">initializeMaps()</a></span> - Method in class biolockj.module.implicit.parser.r16s.<a href="../biolockj/module/implicit/parser/r16s/QiimeParser.html" title="class in biolockj.module.implicit.parser.r16s">QiimeParser</a></dt>
<dd>
<div class="block">Sample IDs in the original metadata file may contain restricted characters that will be replaced with a "." via
 {link biolockj.module.implicit.qiime.BuildQiimeMapping}.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Pipeline.html#initializeModules--">initializeModules()</a></span> - Static method in class biolockj.<a href="../biolockj/Pipeline.html" title="class in biolockj">Pipeline</a></dt>
<dd>
<div class="block">Initialization occurs by calling <a href="../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a> methods on configured modules<br>
 
 Create module sub-directories under <a href="../biolockj/Constants.html#PIPELINE_DIR">"pipeline.pipelineDir"</a> as ordered in
 <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> file.<br>
 Reset the <a href="../biolockj/util/SummaryUtil.html" title="class in biolockj.util"><code>SummaryUtil</code></a> module so previous summary descriptions can be used for completed
 modules
 Delete incomplete module contents if restarting a failed pipeline
 <a href="../biolockj/module/BioModule.html#OUTPUT_DIR">"output"</a> directory<br>
 Call <a href="../biolockj/Pipeline.html#refreshOutputMetadata-biolockj.module.BioModule-"><code>Pipeline.refreshOutputMetadata(BioModule)</code></a> to cache metadata if output by a complete module<br>
 Call <a href="../biolockj/Pipeline.html#refreshRCacheIfNeeded-biolockj.module.BioModule-"><code>Pipeline.refreshRCacheIfNeeded(BioModule)</code></a> to cache R fields after 1st R module runs<br>
 Verify dependencies with <a href="../biolockj/module/BioModule.html#checkDependencies--"><code>BioModule.checkDependencies()</code></a><br>
 </div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Pipeline.html#initializePipeline--">initializePipeline()</a></span> - Static method in class biolockj.<a href="../biolockj/Pipeline.html" title="class in biolockj">Pipeline</a></dt>
<dd>
<div class="block">This method initializes the Pipeline by building the modules and checking module dependencies.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BashScriptBuilder.html#initMainScript-biolockj.module.ScriptModule-">initMainScript(ScriptModule)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BashScriptBuilder.html" title="class in biolockj.util">BashScriptBuilder</a></dt>
<dd>
<div class="block">Create the ScriptModule main script that calls all worker scripts.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#initProjectProps--">initProjectProps()</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Set  Create a pipeline root directory if the pipeline is new.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/BioLockJ.html#initRestart--">initRestart()</a></span> - Static method in class biolockj.<a href="../biolockj/BioLockJ.html" title="class in biolockj">BioLockJ</a></dt>
<dd>
<div class="block">Initialize restarted pipeline by:
 
 Initialize <a href="../biolockj/Log.html" title="class in biolockj"><code>Log</code></a> file using the name of the pipeline root directory
 Update summary #Attempts count
 Delete status file <a href="../biolockj/Constants.html#BLJ_FAILED">"biolockjFailed"</a> in pipeline root directory
 If pipeline status = <a href="../biolockj/Constants.html#BLJ_COMPLETE">"biolockjComplete"</a>
 Delete file <a href="../biolockj/util/DownloadUtil.html#DOWNLOAD_LIST">"downloadList.txt"</a> in pipeline root directory
 </div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#initSeqParams--">initSeqParams()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Set <a href="../biolockj/Constants.html#INPUT_IGNORE_FILES">"input.ignoreFiles"</a>, , and
 
 
 Ignore the metadata file <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/util/MetaUtil.html#META_FILE_PATH">"metadata.filePath"</a>
 If <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> pipeline contains <a href="../biolockj/module/seq/package-summary.html"><code>biolockj.module.seq</code></a> or
 <a href="../biolockj/module/classifier/package-summary.html"><code>biolockj.module.classifier</code></a> <a href="../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s, ignore non-(fasta/fastq) files found in
 <a href="../biolockj/Constants.html#INPUT_DIRS"><code>Constants.INPUT_DIRS</code></a>
 </div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/TaxaUtil.html#initTaxaLevels--">initTaxaLevels()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/TaxaUtil.html" title="class in biolockj.util">TaxaUtil</a></dt>
<dd>
<div class="block">Set taxonomy levels ordered by level, from highest to lowest.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BashScriptBuilder.html#initWorkerScript-biolockj.module.ScriptModule-java.lang.String-">initWorkerScript(ScriptModule, String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BashScriptBuilder.html" title="class in biolockj.util">BashScriptBuilder</a></dt>
<dd>
<div class="block">Create the numbered worker scripts.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#INPUT_DIR_FLAG">INPUT_DIR_FLAG</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">Input directory file-path runtime parameter switch: "-i"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#INPUT_DIRS">INPUT_DIRS</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> List property: "input.dirPaths"<br>
 Set sequence file directories</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#INPUT_FORWARD_READ_SUFFIX">INPUT_FORWARD_READ_SUFFIX</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> String property: "input.suffixFw"<br>
 Set file suffix used to identify forward reads in <a href="../biolockj/Constants.html#INPUT_DIRS">"input.dirPaths"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#INPUT_IGNORE_FILES">INPUT_IGNORE_FILES</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> List property: "input.ignoreFiles"<br>
 Set file names to ignore if found in <a href="../biolockj/Constants.html#INPUT_DIRS">"input.dirPaths"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/RarefySeqs.html#INPUT_RAREFYING_MAX">INPUT_RAREFYING_MAX</a></span> - Static variable in class biolockj.module.seq.<a href="../biolockj/module/seq/RarefySeqs.html" title="class in biolockj.module.seq">RarefySeqs</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> property "rarefySeqs.max" defines the maximum number of reads per file</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/RarefySeqs.html#INPUT_RAREFYING_MIN">INPUT_RAREFYING_MIN</a></span> - Static variable in class biolockj.module.seq.<a href="../biolockj/module/seq/RarefySeqs.html" title="class in biolockj.module.seq">RarefySeqs</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> property "rarefySeqs.min" defines the minimum number of reads per file</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#INPUT_REQUIRE_COMPLETE_PAIRS">INPUT_REQUIRE_COMPLETE_PAIRS</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> Boolean property: "input.requireCompletePairs"<br>
 Require 100% sequence input files are matching paired reads</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/TrimPrimers.html#INPUT_REQUIRE_PRIMER">INPUT_REQUIRE_PRIMER</a></span> - Static variable in class biolockj.module.seq.<a href="../biolockj/module/seq/TrimPrimers.html" title="class in biolockj.module.seq">TrimPrimers</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> property "trimPrimers.requirePrimer" is a boolean used to determine if sequences
 without a primer should be kept or discarded</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#INPUT_REVERSE_READ_SUFFIX">INPUT_REVERSE_READ_SUFFIX</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> String property: "input.suffixRv"<br>
 Set file suffix used to identify forward reads in <a href="../biolockj/Constants.html#INPUT_DIRS">"input.dirPaths"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/SeqFileValidator.html#INPUT_SEQ_MAX">INPUT_SEQ_MAX</a></span> - Static variable in class biolockj.module.seq.<a href="../biolockj/module/seq/SeqFileValidator.html" title="class in biolockj.module.seq">SeqFileValidator</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> Integer property "seqFileValidator.seqMaxLen" defines the maximum number of bases per read</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/SeqFileValidator.html#INPUT_SEQ_MIN">INPUT_SEQ_MIN</a></span> - Static variable in class biolockj.module.seq.<a href="../biolockj/module/seq/SeqFileValidator.html" title="class in biolockj.module.seq">SeqFileValidator</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> Integer property "seqFileValidator.seqMinLen" defines the minimum number of bases per read</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#INPUT_TRIM_PREFIX">INPUT_TRIM_PREFIX</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> String property: "input.trimPrefix"<br>
 Set value of prefix to trim from sequence file names or headers to obtain Sample ID.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/TrimPrimers.html#INPUT_TRIM_SEQ_FILE">INPUT_TRIM_SEQ_FILE</a></span> - Static variable in class biolockj.module.seq.<a href="../biolockj/module/seq/TrimPrimers.html" title="class in biolockj.module.seq">TrimPrimers</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> property "trimPrimers.filePath" defines the file path to the file that defines the
 primers as regular expressions.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#INPUT_TRIM_SUFFIX">INPUT_TRIM_SUFFIX</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> String property: "input.trimSuffix"<br>
 Set value of suffix to trim from sequence file names or headers to obtain Sample ID.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/ImportMetadata.html#inQuotes-java.lang.String-">inQuotes(String)</a></span> - Method in class biolockj.module.implicit.<a href="../biolockj/module/implicit/ImportMetadata.html" title="class in biolockj.module.implicit">ImportMetadata</a></dt>
<dd>
<div class="block">Method called each time a line from metadata contains the
 <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/util/MetaUtil.html#META_COLUMN_DELIM">"metadata.columnDelim"</a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#INTERNAL_ALL_MODULES">INTERNAL_ALL_MODULES</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block">Internal <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> List property: "internal.allModules"<br>
 List of all configured, implicit, and pre/post-requisite modules for the pipeline.<br>
 Example: biolockj.module.ImportMetadata, etc.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#INTERNAL_BLJ_MODULE">INTERNAL_BLJ_MODULE</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block">Set BioModule tag in <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> file to include in pipeline: "#BioModule"<br>
 Example: #BioModule biolockj.module.ImportMetadata</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#INTERNAL_DEFAULT_CONFIG">INTERNAL_DEFAULT_CONFIG</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block">Internal <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> List property: "internal.defaultConfig"<br>
 List of all nested default config files.<br></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#INTERNAL_IS_MULTI_LINE_SEQ">INTERNAL_IS_MULTI_LINE_SEQ</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> Internal Boolean property: "internal.isMultiLineSeq"<br>
 Store TRUE if <a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util"><code>SeqUtil</code></a> determines input sequences are multi-line format.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#INTERNAL_MULTIPLEXED">INTERNAL_MULTIPLEXED</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> Boolean property: "internal.multiplexed"<br>
 Set to true if multiplexed reads are found, set by the application runtime code.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#INTERNAL_PAIRED_READS">INTERNAL_PAIRED_READS</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> Boolean property: "internal.pairedReads"<br>
 Set to true if paired reads are found, set by the application runtime code.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BioLockJUtil.html#INTERNAL_PIPELINE_INPUT_TYPES">INTERNAL_PIPELINE_INPUT_TYPES</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/BioLockJUtil.html" title="class in biolockj.util">BioLockJUtil</a></dt>
<dd>
<div class="block">Internal <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> String property: "internal.pipelineInputTypes"<br>

 This value is set after parsing the input files from <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> property:
 <a href="../biolockj/Constants.html#INPUT_DIRS">"input.dirPaths"</a> in the method: <a href="../biolockj/util/BioLockJUtil.html#getPipelineInputFiles--"><code>BioLockJUtil.getPipelineInputFiles()</code></a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#INTERNAL_SEQ_HEADER_CHAR">INTERNAL_SEQ_HEADER_CHAR</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> property: "internal.seqHeaderChar"<br>
 The property holds the 1st character used in the sequence header for the given dataset</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#INTERNAL_SEQ_TYPE">INTERNAL_SEQ_TYPE</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> Internal property: "internal.seqType"<br>
 The sequence type requires either <a href="../biolockj/Constants.html#FASTA">"fasta"</a> or <a href="../biolockj/Constants.html#FASTQ">"fastq"</a><br>
 System will auto-detect if not configured</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DockerUtil.html#isBljManager--">isBljManager()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/DockerUtil.html" title="class in biolockj.util">DockerUtil</a></dt>
<dd>
<div class="block">Boolean to determine if running Docker blj_manager</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/ModuleUtil.html#isComplete-biolockj.module.BioModule-">isComplete(BioModule)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/ModuleUtil.html" title="class in biolockj.util">ModuleUtil</a></dt>
<dd>
<div class="block">Return TRUE if module completed successfully.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#isDirectMode--">isDirectMode()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">Return TRUE if runtime parameters indicate attempt to run in direct mode</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#isDockerMode--">isDockerMode()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">Return TRUE if runtime parameter <a href="../biolockj/util/RuntimeParamUtil.html#DOCKER_FLAG">"-docker"</a> was found</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#isFastA--">isFastA()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Return TRUE if input files are in FastA format.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#isFastQ--">isFastQ()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Return TRUE if input files are in FastQ format.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/ModuleUtil.html#isFirstRModule-biolockj.module.BioModule-">isFirstRModule(BioModule)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/ModuleUtil.html" title="class in biolockj.util">ModuleUtil</a></dt>
<dd>
<div class="block">Test if module is the first <a href="../biolockj/module/report/r/R_Module.html" title="class in biolockj.module.report.r"><code>R_Module</code></a> configured in the pipeline.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#isForwardRead-java.lang.String-">isForwardRead(String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Return TRUE if reads are unpaired or if name does not contain the reverse read file suffix:
 </div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#isGzipped-java.lang.String-">isGzipped(String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Determine if file is gzipped based on its file extension.<br>
 Any file ending with <a href="../biolockj/Constants.html#GZIP_EXT">".gz"</a> is treated as a gzipped file.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/humann2/Humann2CountModule.html#isHumann2CountModule-biolockj.module.BioModule-">isHumann2CountModule(BioModule)</a></span> - Method in class biolockj.module.report.humann2.<a href="../biolockj/module/report/humann2/Humann2CountModule.html" title="class in biolockj.module.report.humann2">Humann2CountModule</a></dt>
<dd>
<div class="block">Check the module to determine if it generated OTU count files.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/ModuleUtil.html#isIncomplete-biolockj.module.BioModule-">isIncomplete(BioModule)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/ModuleUtil.html" title="class in biolockj.util">ModuleUtil</a></dt>
<dd>
<div class="block">Return TRUE if module started execution but is not complete.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/TaxaUtil.html#isLogNormalizedTaxaFile-java.io.File-">isLogNormalizedTaxaFile(File)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/TaxaUtil.html" title="class in biolockj.util">TaxaUtil</a></dt>
<dd>
<div class="block">Check the file name to determine if it is a Log normalized taxonomy table file.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/ModuleUtil.html#isMetadataModule-biolockj.module.BioModule-">isMetadataModule(BioModule)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/ModuleUtil.html" title="class in biolockj.util">ModuleUtil</a></dt>
<dd>
<div class="block">Method determines if the given module is a metadata-module (which does not use/modify sequence data.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RMetaUtil.html#isMetaMergeTable-java.io.File-">isMetaMergeTable(File)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RMetaUtil.html" title="class in biolockj.util">RMetaUtil</a></dt>
<dd>
<div class="block">Method analyzes the file name to determine if the file could be output from the BioModule
 <a href="../biolockj/module/report/taxa/AddMetadataToTaxaTables.html" title="class in biolockj.module.report.taxa"><code>AddMetadataToTaxaTables</code></a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/TaxaUtil.html#isNormalizedTaxaFile-java.io.File-">isNormalizedTaxaFile(File)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/TaxaUtil.html" title="class in biolockj.util">TaxaUtil</a></dt>
<dd>
<div class="block">Check the file name to determine if it is a normalized taxonomy table file.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#isOnCluster--">isOnCluster()</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Check if running on cluster</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/OtuUtil.html#isOtuFile-java.io.File-">isOtuFile(File)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/OtuUtil.html" title="class in biolockj.util">OtuUtil</a></dt>
<dd>
<div class="block">Check the file name and contents to determine if file is an OTU count file.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/otu/OtuCountModule.html#isOtuModule-biolockj.module.BioModule-">isOtuModule(BioModule)</a></span> - Method in class biolockj.module.report.otu.<a href="../biolockj/module/report/otu/OtuCountModule.html" title="class in biolockj.module.report.otu">OtuCountModule</a></dt>
<dd>
<div class="block">Check the module to determine if it generated OTU count files.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/PathwayUtil.html#isPathwayFile-java.io.File-">isPathwayFile(File)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/PathwayUtil.html" title="class in biolockj.util">PathwayUtil</a></dt>
<dd>
<div class="block">Check the file name to determine if it is a pathway abundance table file.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/PathwayUtil.html#isPathwayModule-biolockj.module.BioModule-">isPathwayModule(BioModule)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/PathwayUtil.html" title="class in biolockj.util">PathwayUtil</a></dt>
<dd>
<div class="block">Check the module to determine if it generated OTU count files.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/BioLockJ.html#isPipelineComplete--">isPipelineComplete()</a></span> - Static method in class biolockj.<a href="../biolockj/BioLockJ.html" title="class in biolockj">BioLockJ</a></dt>
<dd>
<div class="block">Determine project status based on existence of <a href="../biolockj/Constants.html#BLJ_COMPLETE">"biolockjComplete"</a> in pipeline root
 directory.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#isSeqFile-java.io.File-">isSeqFile(File)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Verify 1st character of sequence header and mask 1st sequence for valid DNA/RNA bases "acgtu"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#isSeqModule-biolockj.module.BioModule-">isSeqModule(BioModule)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Check the module to determine if it generated sequence file output.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/r/R_CalculateStats.html#isStatsFile-java.io.File-">isStatsFile(File)</a></span> - Static method in class biolockj.module.report.r.<a href="../biolockj/module/report/r/R_CalculateStats.html" title="class in biolockj.module.report.r">R_CalculateStats</a></dt>
<dd>
<div class="block">Analyze file name for key strings to determine if file is a stats file output by this module.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/TaxaUtil.html#isTaxaFile-java.io.File-">isTaxaFile(File)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/TaxaUtil.html" title="class in biolockj.util">TaxaUtil</a></dt>
<dd>
<div class="block">Check the file name to determine if it is a taxonomy table file.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/taxa/TaxaCountModule.html#isTaxaModule-biolockj.module.BioModule-">isTaxaModule(BioModule)</a></span> - Method in class biolockj.module.report.taxa.<a href="../biolockj/module/report/taxa/TaxaCountModule.html" title="class in biolockj.module.report.taxa">TaxaCountModule</a></dt>
<dd>
<div class="block">Check the module output directory for taxonomy table files generated by BioLockJ.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/ParserModuleImpl.html#isValid-biolockj.node.OtuNode-">isValid(OtuNode)</a></span> - Method in class biolockj.module.implicit.parser.<a href="../biolockj/module/implicit/parser/ParserModuleImpl.html" title="class in biolockj.module.implicit.parser">ParserModuleImpl</a></dt>
<dd>
<div class="block">Some <a href="../biolockj/module/classifier/ClassifierModule.html" title="interface in biolockj.module.classifier"><code>ClassifierModule</code></a>s can include taxonomy level identifiers without an OTU
 name in the sample report files.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/r16s/RdpParser.html#isValid-biolockj.node.OtuNode-">isValid(OtuNode)</a></span> - Method in class biolockj.module.implicit.parser.r16s.<a href="../biolockj/module/implicit/parser/r16s/RdpParser.html" title="class in biolockj.module.implicit.parser.r16s">RdpParser</a></dt>
<dd>
<div class="block">If <a href="../biolockj/node/r16s/RdpNode.html#getScore--"><code>RdpNode.getScore()</code></a> is above the
 <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>., continue with the standard
 <a href="../biolockj/node/OtuNode.html" title="interface in biolockj.node"><code>OtuNode</code></a> validation.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModule.html#isValidInputModule-biolockj.module.BioModule-">isValidInputModule(BioModule)</a></span> - Method in interface biolockj.module.<a href="../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></dt>
<dd>
<div class="block">BioModules <a href="../biolockj/module/BioModule.html#getInputFiles--"><code>BioModule.getInputFiles()</code></a> method typically, but not always, return the previousModule output files.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModuleImpl.html#isValidInputModule-biolockj.module.BioModule-">isValidInputModule(BioModule)</a></span> - Method in class biolockj.module.<a href="../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a></dt>
<dd>
<div class="block">In the early stages of the pipeline, starting with the very 1st module
 <a href="../biolockj/module/implicit/ImportMetadata.html" title="class in biolockj.module.implicit"><code>ImportMetadata</code></a>, most modules expect sequence files as input.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/wgs/Humann2Parser.html#isValidInputModule-biolockj.module.BioModule-">isValidInputModule(BioModule)</a></span> - Method in class biolockj.module.implicit.parser.wgs.<a href="../biolockj/module/implicit/parser/wgs/Humann2Parser.html" title="class in biolockj.module.implicit.parser.wgs">Humann2Parser</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#isValidInputModule-biolockj.module.BioModule-">isValidInputModule(BioModule)</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime">QiimeClassifier</a></dt>
<dd>
<div class="block">If superclass is fed by another QiimeClassifier, it must be a subclass with biom output.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/JavaModuleImpl.html#isValidInputModule-biolockj.module.BioModule-">isValidInputModule(BioModule)</a></span> - Method in class biolockj.module.<a href="../biolockj/module/JavaModuleImpl.html" title="class in biolockj.module">JavaModuleImpl</a></dt>
<dd>
<div class="block">If module is a <a href="../biolockj/module/SeqModule.html" title="interface in biolockj.module"><code>SeqModule</code></a> input must contain sequence data.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/humann2/Humann2CountModule.html#isValidInputModule-biolockj.module.BioModule-">isValidInputModule(BioModule)</a></span> - Method in class biolockj.module.report.humann2.<a href="../biolockj/module/report/humann2/Humann2CountModule.html" title="class in biolockj.module.report.humann2">Humann2CountModule</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/JsonReport.html#isValidInputModule-biolockj.module.BioModule-">isValidInputModule(BioModule)</a></span> - Method in class biolockj.module.report.<a href="../biolockj/module/report/JsonReport.html" title="class in biolockj.module.report">JsonReport</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/otu/OtuCountModule.html#isValidInputModule-biolockj.module.BioModule-">isValidInputModule(BioModule)</a></span> - Method in class biolockj.module.report.otu.<a href="../biolockj/module/report/otu/OtuCountModule.html" title="class in biolockj.module.report.otu">OtuCountModule</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/taxa/TaxaCountModule.html#isValidInputModule-biolockj.module.BioModule-">isValidInputModule(BioModule)</a></span> - Method in class biolockj.module.report.taxa.<a href="../biolockj/module/report/taxa/TaxaCountModule.html" title="class in biolockj.module.report.taxa">TaxaCountModule</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/SeqModuleImpl.html#isValidInputModule-biolockj.module.BioModule-">isValidInputModule(BioModule)</a></span> - Method in class biolockj.module.<a href="../biolockj/module/SeqModuleImpl.html" title="class in biolockj.module">SeqModuleImpl</a></dt>
<dd>&nbsp;</dd>
</dl>
<a href="index-1.html">A</a>&nbsp;<a href="index-2.html">B</a>&nbsp;<a href="index-3.html">C</a>&nbsp;<a href="index-4.html">D</a>&nbsp;<a href="index-5.html">E</a>&nbsp;<a href="index-6.html">F</a>&nbsp;<a href="index-7.html">G</a>&nbsp;<a href="index-8.html">H</a>&nbsp;<a href="index-9.html">I</a>&nbsp;<a href="index-10.html">J</a>&nbsp;<a href="index-11.html">K</a>&nbsp;<a href="index-12.html">L</a>&nbsp;<a href="index-13.html">M</a>&nbsp;<a href="index-14.html">N</a>&nbsp;<a href="index-15.html">O</a>&nbsp;<a href="index-16.html">P</a>&nbsp;<a href="index-17.html">Q</a>&nbsp;<a href="index-18.html">R</a>&nbsp;<a href="index-19.html">S</a>&nbsp;<a href="index-20.html">T</a>&nbsp;<a href="index-21.html">U</a>&nbsp;<a href="index-22.html">V</a>&nbsp;<a href="index-23.html">W</a>&nbsp;</div>
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar.bottom">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.bottom.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../overview-summary.html">Overview</a></li>
<li>Package</li>
<li>Class</li>
<li>Use</li>
<li><a href="../overview-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li class="navBarCell1Rev">Index</li>
<li><a href="../help-doc.html">Help</a></li>
<li>&nbsp;&nbsp;</li>
<li><a href="https://github.com/msioda/BioLockJ/wiki" target="_top"><strong>RETURN to WIKI</strong></a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="index-8.html">Prev Letter</a></li>
<li><a href="index-10.html">Next Letter</a></li>
</ul>
<ul class="navList">
<li><a href="../index.html?index-files/index-9.html" target="_top">Frames</a></li>
<li><a href="index-9.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip.navbar.bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
