<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (1.8.0_144) on Tue Mar 12 06:27:48 EDT 2019 -->
<title>C-Index</title>
<meta name="date" content="2019-03-12">
<link rel="stylesheet" type="text/css" href="../stylesheet.css" title="Style">
<script type="text/javascript" src="../script.js"></script>
</head>
<body>
<script type="text/javascript"><!--
    try {
        if (location.href.indexOf('is-external=true') == -1) {
            parent.document.title="C-Index";
        }
    }
    catch(err) {
    }
//-->
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar.top">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.top.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../overview-summary.html">Overview</a></li>
<li>Package</li>
<li>Class</li>
<li>Use</li>
<li><a href="../overview-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li class="navBarCell1Rev">Index</li>
<li><a href="../help-doc.html">Help</a></li>
<li>&nbsp;&nbsp;</li>
<li><a href="https://github.com/msioda/BioLockJ/wiki" target="_top"><strong>RETURN to WIKI</strong></a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="index-2.html">Prev Letter</a></li>
<li><a href="index-4.html">Next Letter</a></li>
</ul>
<ul class="navList">
<li><a href="../index.html?index-files/index-3.html" target="_top">Frames</a></li>
<li><a href="index-3.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip.navbar.top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<div class="contentContainer"><a href="index-1.html">A</a>&nbsp;<a href="index-2.html">B</a>&nbsp;<a href="index-3.html">C</a>&nbsp;<a href="index-4.html">D</a>&nbsp;<a href="index-5.html">E</a>&nbsp;<a href="index-6.html">F</a>&nbsp;<a href="index-7.html">G</a>&nbsp;<a href="index-8.html">H</a>&nbsp;<a href="index-9.html">I</a>&nbsp;<a href="index-10.html">J</a>&nbsp;<a href="index-11.html">K</a>&nbsp;<a href="index-12.html">L</a>&nbsp;<a href="index-13.html">M</a>&nbsp;<a href="index-14.html">N</a>&nbsp;<a href="index-15.html">O</a>&nbsp;<a href="index-16.html">P</a>&nbsp;<a href="index-17.html">Q</a>&nbsp;<a href="index-18.html">R</a>&nbsp;<a href="index-19.html">S</a>&nbsp;<a href="index-20.html">T</a>&nbsp;<a href="index-21.html">U</a>&nbsp;<a href="index-22.html">V</a>&nbsp;<a href="index-23.html">W</a>&nbsp;<a name="I:C">
<!--   -->
</a>
<h2 class="title">C</h2>
<dl>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModuleImpl.html#cacheInputFiles-java.util.Collection-">cacheInputFiles(Collection&lt;File&gt;)</a></span> - Method in class biolockj.module.<a href="../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a></dt>
<dd>
<div class="block">Cache the input files for quick access on subsequent calls to <a href="../biolockj/module/BioModuleImpl.html#getInputFiles--"><code>BioModuleImpl.getInputFiles()</code></a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModule.html#checkDependencies--">checkDependencies()</a></span> - Method in interface biolockj.module.<a href="../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></dt>
<dd>
<div class="block">During pipeline initialization, all configured BioModules will run this method to validate dependencies.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModuleImpl.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a></dt>
<dd>
<div class="block">If restarting or running a direct pipeline execute the cleanup for completed modules.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/ClassifierModuleImpl.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.classifier.<a href="../biolockj/module/classifier/ClassifierModuleImpl.html" title="class in biolockj.module.classifier">ClassifierModuleImpl</a></dt>
<dd>
<div class="block">Validate module dependencies:
 
 Call <a href="../biolockj/module/classifier/ClassifierModuleImpl.html#getClassifierExe--"><code>ClassifierModuleImpl.getClassifierExe()</code></a> to verify the executable
 Call <a href="../biolockj/module/classifier/ClassifierModuleImpl.html#getClassifierParams--"><code>ClassifierModuleImpl.getClassifierParams()</code></a> to verify the runtime parameters are valid
 Call <a href="../biolockj/module/classifier/ClassifierModuleImpl.html#validateModuleOrder--"><code>ClassifierModuleImpl.validateModuleOrder()</code></a> to validate module configuration order.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/r16s/QiimeClosedRefClassifier.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.classifier.r16s.<a href="../biolockj/module/classifier/r16s/QiimeClosedRefClassifier.html" title="class in biolockj.module.classifier.r16s">QiimeClosedRefClassifier</a></dt>
<dd>
<div class="block">Call <a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime"><code>QiimeClassifier</code></a> checkOtuPickingDependencies() method to verify OTU
 picking script parameters.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/r16s/QiimeDeNovoClassifier.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.classifier.r16s.<a href="../biolockj/module/classifier/r16s/QiimeDeNovoClassifier.html" title="class in biolockj.module.classifier.r16s">QiimeDeNovoClassifier</a></dt>
<dd>
<div class="block">Call <a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime"><code>QiimeClassifier</code></a> checkOtuPickingDependencies() method to verify OTU
 picking script parameters.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/r16s/QiimeOpenRefClassifier.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.classifier.r16s.<a href="../biolockj/module/classifier/r16s/QiimeOpenRefClassifier.html" title="class in biolockj.module.classifier.r16s">QiimeOpenRefClassifier</a></dt>
<dd>
<div class="block">Call <a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime"><code>QiimeClassifier</code></a> checkOtuPickingDependencies() method to verify OTU
 picking script parameters.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/r16s/RdpClassifier.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.classifier.r16s.<a href="../biolockj/module/classifier/r16s/RdpClassifier.html" title="class in biolockj.module.classifier.r16s">RdpClassifier</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/Humann2Classifier.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/Humann2Classifier.html" title="class in biolockj.module.classifier.wgs">Humann2Classifier</a></dt>
<dd>
<div class="block">Verify that none of the derived command line parameters are included in
 <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/module/classifier/wgs/Humann2Classifier.html#EXE_HUMANN2">"exe.humann2"</a><a href="../biolockj/Constants.html#PARAMS">"Params"</a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/Kraken2Classifier.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/Kraken2Classifier.html" title="class in biolockj.module.classifier.wgs">Kraken2Classifier</a></dt>
<dd>
<div class="block">Verify that none of the derived command line parameters are included in
 <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/module/classifier/wgs/Kraken2Classifier.html#EXE_KRAKEN2">"exe.kraken2"</a><a href="../biolockj/Constants.html#PARAMS">"Params"</a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/KrakenClassifier.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/KrakenClassifier.html" title="class in biolockj.module.classifier.wgs">KrakenClassifier</a></dt>
<dd>
<div class="block">Verify that none of the derived command line parameters are included in
 <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/module/classifier/wgs/KrakenClassifier.html#EXE_KRAKEN">"exe.kraken"</a><a href="../biolockj/Constants.html#PARAMS">"Params"</a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/Metaphlan2Classifier.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/Metaphlan2Classifier.html" title="class in biolockj.module.classifier.wgs">Metaphlan2Classifier</a></dt>
<dd>
<div class="block">Verify none of the derived command line parameters are included in
 <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/module/classifier/wgs/Metaphlan2Classifier.html#EXE_METAPHLAN">"exe.metaphlan2"</a><a href="../biolockj/Constants.html#PARAMS">"Params"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/Demultiplexer.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.implicit.<a href="../biolockj/module/implicit/Demultiplexer.html" title="class in biolockj.module.implicit">Demultiplexer</a></dt>
<dd>
<div class="block">Validate module dependencies:
 
 If <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/util/DemuxUtil.html#DEMUX_STRATEGY">"demultiplexer.strategy"</a> indicates use of barcodes to
 demultiplexer, validate metadata column named
 <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/util/MetaUtil.html#META_BARCODE_COLUMN">"metadata.barcodeColumn"</a> exists
 Call <a href="../biolockj/module/implicit/Demultiplexer.html#setMultiplexedConfig--"><code>Demultiplexer.setMultiplexedConfig()</code></a> to set multiplexed Config if needed
 If <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/util/DemuxUtil.html#BARCODE_CUTOFF">"demultiplexer.barcodeCutoff"</a> defined, validate between 0.0 -
 1.0
 </div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/ImportMetadata.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.implicit.<a href="../biolockj/module/implicit/ImportMetadata.html" title="class in biolockj.module.implicit">ImportMetadata</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/ParserModuleImpl.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.implicit.parser.<a href="../biolockj/module/implicit/parser/ParserModuleImpl.html" title="class in biolockj.module.implicit.parser">ParserModuleImpl</a></dt>
<dd>
<div class="block">Execute <a href="../biolockj/module/implicit/parser/ParserModuleImpl.html#validateModuleOrder--"><code>ParserModuleImpl.validateModuleOrder()</code></a> to validate module configuration order.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/wgs/Humann2Parser.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.implicit.parser.wgs.<a href="../biolockj/module/implicit/parser/wgs/Humann2Parser.html" title="class in biolockj.module.implicit.parser.wgs">Humann2Parser</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/MergeQiimeOtuTables.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/MergeQiimeOtuTables.html" title="class in biolockj.module.implicit.qiime">MergeQiimeOtuTables</a></dt>
<dd>
<div class="block">Verify the previous module = <a href="../biolockj/module/classifier/r16s/QiimeClosedRefClassifier.html" title="class in biolockj.module.classifier.r16s"><code>QiimeClosedRefClassifier</code></a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime">QiimeClassifier</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/Email.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.report.<a href="../biolockj/module/report/Email.html" title="class in biolockj.module.report">Email</a></dt>
<dd>
<div class="block">Verify required email <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> properties exist and are properly formatted.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/humann2/Humann2CountModule.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.report.humann2.<a href="../biolockj/module/report/humann2/Humann2CountModule.html" title="class in biolockj.module.report.humann2">Humann2CountModule</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/humann2/RemoveLowPathwayCounts.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.report.humann2.<a href="../biolockj/module/report/humann2/RemoveLowPathwayCounts.html" title="class in biolockj.module.report.humann2">RemoveLowPathwayCounts</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/humann2/RemoveScarcePathwayCounts.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.report.humann2.<a href="../biolockj/module/report/humann2/RemoveScarcePathwayCounts.html" title="class in biolockj.module.report.humann2">RemoveScarcePathwayCounts</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/otu/RarefyOtuCounts.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.report.otu.<a href="../biolockj/module/report/otu/RarefyOtuCounts.html" title="class in biolockj.module.report.otu">RarefyOtuCounts</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/otu/RemoveLowOtuCounts.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.report.otu.<a href="../biolockj/module/report/otu/RemoveLowOtuCounts.html" title="class in biolockj.module.report.otu">RemoveLowOtuCounts</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/otu/RemoveScarceOtuCounts.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.report.otu.<a href="../biolockj/module/report/otu/RemoveScarceOtuCounts.html" title="class in biolockj.module.report.otu">RemoveScarceOtuCounts</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/r/R_CalculateStats.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.report.r.<a href="../biolockj/module/report/r/R_CalculateStats.html" title="class in biolockj.module.report.r">R_CalculateStats</a></dt>
<dd>
<div class="block">Validate configuration file properties used to build the R report:
 
 super.checkDependencies()
 Require <a href="../biolockj/module/report/r/R_CalculateStats.html#R_ADJ_PVALS_SCOPE">"r_CalculateStats.pAdjustScope"</a>
 Require <a href="../biolockj/module/report/r/R_CalculateStats.html#R_PVAL_ADJ_METHOD">"r_CalculateStats.pAdjustMethod"</a>
 </div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/r/R_Module.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.report.r.<a href="../biolockj/module/report/r/R_Module.html" title="class in biolockj.module.report.r">R_Module</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/r/R_PlotEffectSize.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.report.r.<a href="../biolockj/module/report/r/R_PlotEffectSize.html" title="class in biolockj.module.report.r">R_PlotEffectSize</a></dt>
<dd>
<div class="block">At least one of the available plot types should NOT be disabled.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/r/R_PlotMds.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.report.r.<a href="../biolockj/module/report/r/R_PlotMds.html" title="class in biolockj.module.report.r">R_PlotMds</a></dt>
<dd>
<div class="block">Require <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/module/report/r/R_PlotMds.html#R_MDS_NUM_AXIS">"r_PlotMds.numAxis"</a> set to integer greater than 2</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/r/R_PlotOtus.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.report.r.<a href="../biolockj/module/report/r/R_PlotOtus.html" title="class in biolockj.module.report.r">R_PlotOtus</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/r/R_PlotPvalHistograms.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.report.r.<a href="../biolockj/module/report/r/R_PlotPvalHistograms.html" title="class in biolockj.module.report.r">R_PlotPvalHistograms</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/taxa/LogTransformTaxaTables.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.report.taxa.<a href="../biolockj/module/report/taxa/LogTransformTaxaTables.html" title="class in biolockj.module.report.taxa">LogTransformTaxaTables</a></dt>
<dd>
<div class="block">Verify <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/Constants.html#REPORT_LOG_BASE">"report.logBase"</a> property is valid (if defined) with a
 value = (e or 10).</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/taxa/NormalizeTaxaTables.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.report.taxa.<a href="../biolockj/module/report/taxa/NormalizeTaxaTables.html" title="class in biolockj.module.report.taxa">NormalizeTaxaTables</a></dt>
<dd>
<div class="block">Verify <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/Constants.html#REPORT_LOG_BASE">"report.logBase"</a> property is valid (if defined) with a
 value = (e or 10).</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/ScriptModuleImpl.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/ScriptModuleImpl.html" title="class in biolockj.module">ScriptModuleImpl</a></dt>
<dd>
<div class="block">Validate module dependencies:
 
 Require <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>. exists
 Require <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>. is positive integer
 Require <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>. is positive integer
 Verify <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>. is positive integer if set
 </div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/KneadData.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/KneadData.html" title="class in biolockj.module.seq">KneadData</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/Multiplexer.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/Multiplexer.html" title="class in biolockj.module.seq">Multiplexer</a></dt>
<dd>
<div class="block">Validate module dependencies:
 
 Validate this is the last module to run (excluding <a href="../biolockj/module/report/Email.html" title="class in biolockj.module.report"><code>Email</code></a>)
 
 If this module was completed on a previous run, update the property:
 <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/Constants.html#INTERNAL_MULTIPLEXED">"internal.multiplexed"</a> = <a href="../biolockj/Constants.html#TRUE">"Y"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/PearMergeReads.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/PearMergeReads.html" title="class in biolockj.module.seq">PearMergeReads</a></dt>
<dd>
<div class="block">Validate module dependencies
 
 Verify input files are in fastq format
 Verify matching paired reads are found
 Validate required <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/module/seq/PearMergeReads.html#EXE_PEAR"><code>PearMergeReads.EXE_PEAR</code></a> property
 If running a restarted pipeline that already merged reads, set paired indicator to false
 </div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/RarefySeqs.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/RarefySeqs.html" title="class in biolockj.module.seq">RarefySeqs</a></dt>
<dd>
<div class="block">Validate module dependencies
 
 Validate <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/module/seq/RarefySeqs.html#INPUT_RAREFYING_MIN"><code>RarefySeqs.INPUT_RAREFYING_MIN</code></a> is a non-negative integer
 Validate <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/module/seq/RarefySeqs.html#INPUT_RAREFYING_MAX"><code>RarefySeqs.INPUT_RAREFYING_MAX</code></a> is a positive integer that is greater than or
 equal to <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/module/seq/RarefySeqs.html#INPUT_RAREFYING_MIN"><code>RarefySeqs.INPUT_RAREFYING_MIN</code></a> (if defined)
 </div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/TrimPrimers.html#checkDependencies--">checkDependencies()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/TrimPrimers.html" title="class in biolockj.module.seq">TrimPrimers</a></dt>
<dd>
<div class="block">Validates the file that defines the REGEX primers.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#CLASS">CLASS</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> option for <a href="../biolockj/Constants.html#REPORT_TAXONOMY_LEVELS">"report.taxonomyLevels"</a>: "class"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/OtuNodeImpl.html#CLASS_DELIM">CLASS_DELIM</a></span> - Static variable in class biolockj.node.<a href="../biolockj/node/OtuNodeImpl.html" title="class in biolockj.node">OtuNodeImpl</a></dt>
<dd>
<div class="block">Standard classifier output level delimiter for CLASS</div>
</dd>
<dt><a href="../biolockj/module/classifier/ClassifierModule.html" title="interface in biolockj.module.classifier"><span class="typeNameLink">ClassifierModule</span></a> - Interface in <a href="../biolockj/module/classifier/package-summary.html">biolockj.module.classifier</a></dt>
<dd>
<div class="block">Classifier <a href="../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s build one or more bash scripts to call the application on sequence
 files.</div>
</dd>
<dt><a href="../biolockj/module/classifier/ClassifierModuleImpl.html" title="class in biolockj.module.classifier"><span class="typeNameLink">ClassifierModuleImpl</span></a> - Class in <a href="../biolockj/module/classifier/package-summary.html">biolockj.module.classifier</a></dt>
<dd>
<div class="block">This is the superclass for all WGS and 16S biolockj.module.classifier BioModules.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/ClassifierModuleImpl.html#ClassifierModuleImpl--">ClassifierModuleImpl()</a></span> - Constructor for class biolockj.module.classifier.<a href="../biolockj/module/classifier/ClassifierModuleImpl.html" title="class in biolockj.module.classifier">ClassifierModuleImpl</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RMetaUtil.html#classifyReportableMetadata-biolockj.module.BioModule-">classifyReportableMetadata(BioModule)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RMetaUtil.html" title="class in biolockj.util">RMetaUtil</a></dt>
<dd>
<div class="block">Classify and verify the R filter and reportable metadata fields listed in the <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> file.<br>
 All metadata fields are reported unless specific fields are listed in:
 <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/util/RMetaUtil.html#R_REPORT_FIELDS">"r.reportFields"</a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModule.html#cleanUp--">cleanUp()</a></span> - Method in interface biolockj.module.<a href="../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></dt>
<dd>
<div class="block">This method executes after execution to update Config modified by the module or other cleanup operations.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModuleImpl.html#cleanUp--">cleanUp()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a></dt>
<dd>
<div class="block">By default, no cleanUp code is required.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/Humann2Classifier.html#cleanUp--">cleanUp()</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/Humann2Classifier.html" title="class in biolockj.module.classifier.wgs">Humann2Classifier</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/Demultiplexer.html#cleanUp--">cleanUp()</a></span> - Method in class biolockj.module.implicit.<a href="../biolockj/module/implicit/Demultiplexer.html" title="class in biolockj.module.implicit">Demultiplexer</a></dt>
<dd>
<div class="block">Update SeqUtil to indicate data has been demultiplexed.<br></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/ImportMetadata.html#cleanUp--">cleanUp()</a></span> - Method in class biolockj.module.implicit.<a href="../biolockj/module/implicit/ImportMetadata.html" title="class in biolockj.module.implicit">ImportMetadata</a></dt>
<dd>
<div class="block">Verify the metadata fields configured for R reports.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/r16s/QiimeParser.html#cleanUp--">cleanUp()</a></span> - Method in class biolockj.module.implicit.parser.r16s.<a href="../biolockj/module/implicit/parser/r16s/QiimeParser.html" title="class in biolockj.module.implicit.parser.r16s">QiimeParser</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#cleanUp--">cleanUp()</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime">QiimeClassifier</a></dt>
<dd>
<div class="block">The cleanUp operation builds a new metadata file if alpha diversity metrics were generated by this module.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/humann2/AddMetadataToPathwayTables.html#cleanUp--">cleanUp()</a></span> - Method in class biolockj.module.report.humann2.<a href="../biolockj/module/report/humann2/AddMetadataToPathwayTables.html" title="class in biolockj.module.report.humann2">AddMetadataToPathwayTables</a></dt>
<dd>
<div class="block">For R to report HumanN2 reports instead of taxa levels</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/otu/RarefyOtuCounts.html#cleanUp--">cleanUp()</a></span> - Method in class biolockj.module.report.otu.<a href="../biolockj/module/report/otu/RarefyOtuCounts.html" title="class in biolockj.module.report.otu">RarefyOtuCounts</a></dt>
<dd>
<div class="block">Update <a href="../biolockj/module/implicit/parser/ParserModuleImpl.html" title="class in biolockj.module.implicit.parser"><code>ParserModuleImpl</code></a> OTU_COUNT field name.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/otu/RemoveLowOtuCounts.html#cleanUp--">cleanUp()</a></span> - Method in class biolockj.module.report.otu.<a href="../biolockj/module/report/otu/RemoveLowOtuCounts.html" title="class in biolockj.module.report.otu">RemoveLowOtuCounts</a></dt>
<dd>
<div class="block">Update <a href="../biolockj/module/implicit/parser/ParserModuleImpl.html" title="class in biolockj.module.implicit.parser"><code>ParserModuleImpl</code></a> OTU_COUNT field name.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/otu/RemoveScarceOtuCounts.html#cleanUp--">cleanUp()</a></span> - Method in class biolockj.module.report.otu.<a href="../biolockj/module/report/otu/RemoveScarceOtuCounts.html" title="class in biolockj.module.report.otu">RemoveScarceOtuCounts</a></dt>
<dd>
<div class="block">Set the number of hits field.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/taxa/AddMetadataToTaxaTables.html#cleanUp--">cleanUp()</a></span> - Method in class biolockj.module.report.taxa.<a href="../biolockj/module/report/taxa/AddMetadataToTaxaTables.html" title="class in biolockj.module.report.taxa">AddMetadataToTaxaTables</a></dt>
<dd>
<div class="block">For R to report taxa levels (not HumanN2 reports)</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/AwkFastaConverter.html#cleanUp--">cleanUp()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/AwkFastaConverter.html" title="class in biolockj.module.seq">AwkFastaConverter</a></dt>
<dd>
<div class="block">Set <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/Constants.html#INTERNAL_SEQ_TYPE">"internal.seqType"</a> = <a href="../biolockj/Constants.html#FASTA">"fasta"</a><br>
 Set <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/Constants.html#INTERNAL_SEQ_HEADER_CHAR">"internal.seqHeaderChar"</a> =
 <a href="../biolockj/util/SeqUtil.html#FASTA_HEADER_DEFAULT_DELIM"><code>SeqUtil.FASTA_HEADER_DEFAULT_DELIM</code></a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/PearMergeReads.html#cleanUp--">cleanUp()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/PearMergeReads.html" title="class in biolockj.module.seq">PearMergeReads</a></dt>
<dd>
<div class="block">Set <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/Constants.html#INTERNAL_PAIRED_READS">"internal.pairedReads"</a> = <a href="../biolockj/Constants.html#FALSE">"N"</a>
 and register number of reads.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/RarefySeqs.html#cleanUp--">cleanUp()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/RarefySeqs.html" title="class in biolockj.module.seq">RarefySeqs</a></dt>
<dd>
<div class="block">Set <a href="../biolockj/module/seq/RarefySeqs.html#NUM_RAREFIED_READS">"Num_Rarefied_Reads"</a> as the number of reads field.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/SeqFileValidator.html#cleanUp--">cleanUp()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/SeqFileValidator.html" title="class in biolockj.module.seq">SeqFileValidator</a></dt>
<dd>
<div class="block">Set <a href="../biolockj/module/seq/SeqFileValidator.html#NUM_VALID_READS">"Num_Valid_Reads"</a> as the number of reads field.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/TrimPrimers.html#cleanUp--">cleanUp()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/TrimPrimers.html" title="class in biolockj.module.seq">TrimPrimers</a></dt>
<dd>
<div class="block">Set <a href="../biolockj/module/seq/TrimPrimers.html#NUM_TRIMMED_READS">"Num_Trimmed_Reads"</a> as the number of reads field.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DemuxUtil.html#clearDemuxConfig--">clearDemuxConfig()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/DemuxUtil.html" title="class in biolockj.util">DemuxUtil</a></dt>
<dd>
<div class="block">Clear demultiplexer related fields</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BashScriptBuilder.html#CLUSTER_BATCH_COMMAND">CLUSTER_BATCH_COMMAND</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/BashScriptBuilder.html" title="class in biolockj.util">BashScriptBuilder</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> String property: "cluster.batchCommand"<br>
 Terminal command used to submit jobs on the cluster.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/Email.html#CLUSTER_HOST">CLUSTER_HOST</a></span> - Static variable in class biolockj.module.report.<a href="../biolockj/module/report/Email.html" title="class in biolockj.module.report">Email</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> String property: "cluster.host"<br>
 The cluster host URL used for SSH and SCP connections.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BashScriptBuilder.html#CLUSTER_MODULES">CLUSTER_MODULES</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/BashScriptBuilder.html" title="class in biolockj.util">BashScriptBuilder</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> List property: "cluster.modules"<br>
 List of cluster modules to load at start of worker scripts.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BashScriptBuilder.html#CLUSTER_NUM_PROCESSORS">CLUSTER_NUM_PROCESSORS</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/BashScriptBuilder.html" title="class in biolockj.util">BashScriptBuilder</a></dt>
<dd>
<div class="block">One parameter of the <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> String property <a href="../biolockj/util/BashScriptBuilder.html#SCRIPT_JOB_HEADER">"cluster.jobHeader"</a> to set number of cores.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BashScriptBuilder.html#CLUSTER_PROLOGUE">CLUSTER_PROLOGUE</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/BashScriptBuilder.html" title="class in biolockj.util">BashScriptBuilder</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> String property: "cluster.prologue"<br>
 To run at the start of every script after loading cluster modules (if any)</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#CLUSTER_RUN_JAVA_AS_SCRIPT">CLUSTER_RUN_JAVA_AS_SCRIPT</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> Boolean property "cluster.runJavaAsScript" if set =
 <a href="../biolockj/Constants.html#TRUE">"Y"</a> will run Java module as a script instead of running on the head node.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BashScriptBuilder.html#CLUSTER_VALIDATE_PARAMS">CLUSTER_VALIDATE_PARAMS</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/BashScriptBuilder.html" title="class in biolockj.util">BashScriptBuilder</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> Boolean property: "cluster.validateParams"<br>
 If set to <a href="../biolockj/Constants.html#TRUE">"Y"</a>, validate <a href="../biolockj/util/BashScriptBuilder.html#SCRIPT_JOB_HEADER">"cluster.jobHeader"</a> param cluster number of
 processors = <a href="../biolockj/module/ScriptModule.html#SCRIPT_NUM_THREADS">"script.numThreads"</a> or
 <a href="../biolockj/module/ScriptModule.html#SCRIPT_NUM_THREADS"><code>ScriptModule.SCRIPT_NUM_THREADS</code></a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#COMBINED_FNA">COMBINED_FNA</a></span> - Static variable in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime">QiimeClassifier</a></dt>
<dd>
<div class="block">Multiplexed fasta file produced by QIIME <a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#SCRIPT_ADD_LABELS">"add_qiime_labels.py"</a> script: "combined_seqs.fna"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModuleImpl.html#compareTo-biolockj.module.BioModule-">compareTo(BioModule)</a></span> - Method in class biolockj.module.<a href="../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/JsonNode.html#compareTo-biolockj.node.JsonNode-">compareTo(JsonNode)</a></span> - Method in class biolockj.node.<a href="../biolockj/node/JsonNode.html" title="class in biolockj.node">JsonNode</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/OtuNodeImpl.html#compareTo-biolockj.node.OtuNode-">compareTo(OtuNode)</a></span> - Method in class biolockj.node.<a href="../biolockj/node/OtuNodeImpl.html" title="class in biolockj.node">OtuNodeImpl</a></dt>
<dd>
<div class="block">The sort order of the TreeSet(<a href="../biolockj/node/OtuNode.html" title="interface in biolockj.node"><code>OtuNode</code></a>) is absolutely required for proper processing in
 <a href="../biolockj/node/OtuNodeImpl.html#addTaxa-java.lang.String-java.lang.String-"><code>OtuNodeImpl.addTaxa(String, String)</code></a> and <a href="../biolockj/node/OtuNodeImpl.html#getTaxaMap--"><code>OtuNodeImpl.getTaxaMap()</code></a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/ParsedSample.html#compareTo-biolockj.node.ParsedSample-">compareTo(ParsedSample)</a></span> - Method in class biolockj.node.<a href="../biolockj/node/ParsedSample.html" title="class in biolockj.node">ParsedSample</a></dt>
<dd>&nbsp;</dd>
<dt><a href="../biolockj/module/report/otu/CompileOtuCounts.html" title="class in biolockj.module.report.otu"><span class="typeNameLink">CompileOtuCounts</span></a> - Class in <a href="../biolockj/module/report/otu/package-summary.html">biolockj.module.report.otu</a></dt>
<dd>
<div class="block">This BioModule compiles the counts from all OTU count files into a single summary OTU count file containing OTU
 counts for the entire dataset.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/otu/CompileOtuCounts.html#CompileOtuCounts--">CompileOtuCounts()</a></span> - Constructor for class biolockj.module.report.otu.<a href="../biolockj/module/report/otu/CompileOtuCounts.html" title="class in biolockj.module.report.otu">CompileOtuCounts</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/otu/CompileOtuCounts.html#compileOtuCounts-java.util.Collection-">compileOtuCounts(Collection&lt;File&gt;)</a></span> - Method in class biolockj.module.report.otu.<a href="../biolockj/module/report/otu/CompileOtuCounts.html" title="class in biolockj.module.report.otu">CompileOtuCounts</a></dt>
<dd>
<div class="block">Compile OTU counts from the individual sample OTU count files</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/OtuUtil.html#compileSampleOtuCounts-java.io.File-">compileSampleOtuCounts(File)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/OtuUtil.html" title="class in biolockj.util">OtuUtil</a></dt>
<dd>
<div class="block">Compile OTU counts from an individual sample OTU count file</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DockerUtil.html#COMPUTE_SCRIPT">COMPUTE_SCRIPT</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/DockerUtil.html" title="class in biolockj.util">DockerUtil</a></dt>
<dd>
<div class="block">Docker environment variable holding the name of the compute script file: "COMPUTE_SCRIPT"</div>
</dd>
<dt><a href="../biolockj/Config.html" title="class in biolockj"><span class="typeNameLink">Config</span></a> - Class in <a href="../biolockj/package-summary.html">biolockj</a></dt>
<dd>
<div class="block">Provides type-safe, validated methods for storing/accessing system properties.<br>
 Initially populated by the properties in the Config file, several additional properties are created and stored in the
 the Config (to save system determined info such as: pipeline directory and name, has paired reads?, has multiplexed
 reads?, etc.).</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#Config--">Config()</a></span> - Constructor for class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#CONFIG_DIR_FLAG">CONFIG_DIR_FLAG</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> file directory path runtime parameter switch: "-C"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#CONFIG_FLAG">CONFIG_FLAG</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> file path runtime parameter switch: "-c"</div>
</dd>
<dt><a href="../biolockj/exception/ConfigException.html" title="class in biolockj.exception"><span class="typeNameLink">ConfigException</span></a> - Exception in <a href="../biolockj/exception/package-summary.html">biolockj.exception</a></dt>
<dd>
<div class="block">ConfigException is the superclass for all BioLockJ configuration file Exceptions used to ensure message uniformity.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/exception/ConfigException.html#ConfigException-java.lang.String-">ConfigException(String)</a></span> - Constructor for exception biolockj.exception.<a href="../biolockj/exception/ConfigException.html" title="class in biolockj.exception">ConfigException</a></dt>
<dd>
<div class="block">Abstract <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> exception calls super to instantiate using the superclass
 <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Exception.html?is-external=true" title="class or interface in java.lang"><code>Exception</code></a> implementation</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/exception/ConfigException.html#ConfigException-java.lang.String-java.lang.String-">ConfigException(String, String)</a></span> - Constructor for exception biolockj.exception.<a href="../biolockj/exception/ConfigException.html" title="class in biolockj.exception">ConfigException</a></dt>
<dd>
<div class="block">Abstract <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> exception calls <a href="../biolockj/exception/ConfigException.html#buildMessage-java.lang.String-java.lang.String-"><code>ConfigException.buildMessage(String, String)</code></a> to generate a standard
 error message for Configuration file errors, passing an empty string for the msg parameter.</div>
</dd>
<dt><a href="../biolockj/exception/ConfigFormatException.html" title="class in biolockj.exception"><span class="typeNameLink">ConfigFormatException</span></a> - Exception in <a href="../biolockj/exception/package-summary.html">biolockj.exception</a></dt>
<dd>
<div class="block">ConfigFormatException is thrown if property is defined but the format is invalid.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/exception/ConfigFormatException.html#ConfigFormatException-java.lang.String-java.lang.String-">ConfigFormatException(String, String)</a></span> - Constructor for exception biolockj.exception.<a href="../biolockj/exception/ConfigFormatException.html" title="class in biolockj.exception">ConfigFormatException</a></dt>
<dd>
<div class="block">ConfigFormatException is thrown if property is defined but the format is invalid.</div>
</dd>
<dt><a href="../biolockj/exception/ConfigNotFoundException.html" title="class in biolockj.exception"><span class="typeNameLink">ConfigNotFoundException</span></a> - Exception in <a href="../biolockj/exception/package-summary.html">biolockj.exception</a></dt>
<dd>
<div class="block">ConfigNotFoundException is thrown if a required <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> property is undefined.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/exception/ConfigNotFoundException.html#ConfigNotFoundException-java.lang.String-">ConfigNotFoundException(String)</a></span> - Constructor for exception biolockj.exception.<a href="../biolockj/exception/ConfigNotFoundException.html" title="class in biolockj.exception">ConfigNotFoundException</a></dt>
<dd>
<div class="block">ConfigNotFoundException is thrown if a required <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> property is undefined.</div>
</dd>
<dt><a href="../biolockj/exception/ConfigPathException.html" title="class in biolockj.exception"><span class="typeNameLink">ConfigPathException</span></a> - Exception in <a href="../biolockj/exception/package-summary.html">biolockj.exception</a></dt>
<dd>
<div class="block">ConfigPathException is thrown if property is invalid file paths are encountered when processing the BioLockJ
 configuration file.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/exception/ConfigPathException.html#ConfigPathException-java.lang.String-">ConfigPathException(String)</a></span> - Constructor for exception biolockj.exception.<a href="../biolockj/exception/ConfigPathException.html" title="class in biolockj.exception">ConfigPathException</a></dt>
<dd>
<div class="block">ConfigPathException is thrown if the filePath parameter does not exist on the file system.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/exception/ConfigPathException.html#ConfigPathException-java.lang.String-java.lang.String-">ConfigPathException(String, String)</a></span> - Constructor for exception biolockj.exception.<a href="../biolockj/exception/ConfigPathException.html" title="class in biolockj.exception">ConfigPathException</a></dt>
<dd>
<div class="block">ConfigPathException is thrown if the path does not exist or is not the proper file type: <a href="../biolockj/exception/ConfigPathException.html#FILE">"file"</a> or
 <a href="../biolockj/exception/ConfigPathException.html#DIRECTORY">"directory"</a>.</div>
</dd>
<dt><a href="../biolockj/exception/ConfigViolationException.html" title="class in biolockj.exception"><span class="typeNameLink">ConfigViolationException</span></a> - Exception in <a href="../biolockj/exception/package-summary.html">biolockj.exception</a></dt>
<dd>
<div class="block">ConfigViolationException is thrown at runtime due to application state violations caused by invalid module I/O
 (sequences or metadata) as defined by <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> properties.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/exception/ConfigViolationException.html#ConfigViolationException-java.lang.String-">ConfigViolationException(String)</a></span> - Constructor for exception biolockj.exception.<a href="../biolockj/exception/ConfigViolationException.html" title="class in biolockj.exception">ConfigViolationException</a></dt>
<dd>
<div class="block">Generic exception message.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/exception/ConfigViolationException.html#ConfigViolationException-java.lang.String-java.lang.String-">ConfigViolationException(String, String)</a></span> - Constructor for exception biolockj.exception.<a href="../biolockj/exception/ConfigViolationException.html" title="class in biolockj.exception">ConfigViolationException</a></dt>
<dd>
<div class="block">ConfigViolationException is thrown application detects a violation due to a <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> property
 setting, such as finding Sample IDs with no corresponding sequence file enforced by
 <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/util/MetaUtil.html#USE_EVERY_ROW">"metadata.useEveryRow"</a> = <a href="../biolockj/Constants.html#TRUE">"Y"</a></div>
</dd>
<dt><a href="../biolockj/Constants.html" title="class in biolockj"><span class="typeNameLink">Constants</span></a> - Class in <a href="../biolockj/package-summary.html">biolockj</a></dt>
<dd>
<div class="block">Single Java class to hold shared constant values referenced my multiple classes.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#Constants--">Constants()</a></span> - Constructor for class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DockerUtil.html#CONTAINER_BLJ_DIR">CONTAINER_BLJ_DIR</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/DockerUtil.html" title="class in biolockj.util">DockerUtil</a></dt>
<dd>
<div class="block">Docker container BioLockJ installation directory: "/app/biolockj"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DockerUtil.html#CONTAINER_CONFIG_DIR">CONTAINER_CONFIG_DIR</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/DockerUtil.html" title="class in biolockj.util">DockerUtil</a></dt>
<dd>
<div class="block">All containers mount the host <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> directory to the container"config" volume</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DockerUtil.html#CONTAINER_DB_DIR">CONTAINER_DB_DIR</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/DockerUtil.html" title="class in biolockj.util">DockerUtil</a></dt>
<dd>
<div class="block">Some containers mount a database to the containers "db" volume.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DockerUtil.html#CONTAINER_INPUT_DIR">CONTAINER_INPUT_DIR</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/DockerUtil.html" title="class in biolockj.util">DockerUtil</a></dt>
<dd>
<div class="block">All containers mount the host <a href="../biolockj/Constants.html#INPUT_DIRS">"input.dirPaths"</a> to the container "input" volume</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DockerUtil.html#CONTAINER_META_DIR">CONTAINER_META_DIR</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/DockerUtil.html" title="class in biolockj.util">DockerUtil</a></dt>
<dd>
<div class="block">Some containers mount the <a href="../biolockj/util/MetaUtil.html#META_FILE_PATH">"metadata.filePath"</a> to the container "meta" volume</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DockerUtil.html#CONTAINER_OUTPUT_DIR">CONTAINER_OUTPUT_DIR</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/DockerUtil.html" title="class in biolockj.util">DockerUtil</a></dt>
<dd>
<div class="block">All containers mount <a href="../biolockj/Constants.html#PIPELINE_DIR">"pipeline.pipelineDir"</a> to the container "pipelines" volume</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DockerUtil.html#CONTAINER_PRIMER_DIR">CONTAINER_PRIMER_DIR</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/DockerUtil.html" title="class in biolockj.util">DockerUtil</a></dt>
<dd>
<div class="block">Some containers mount the <a href="../biolockj/module/seq/TrimPrimers.html#INPUT_TRIM_SEQ_FILE">"trimPrimers.filePath"</a> to the containers "primer"
 volume.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/PathwayUtil.html#containsHn2ParserOutput-java.util.Collection-">containsHn2ParserOutput(Collection&lt;File&gt;)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/PathwayUtil.html" title="class in biolockj.util">PathwayUtil</a></dt>
<dd>
<div class="block">Check pipeline input contains Humann2Parser module output.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/r16s/QiimeClosedRefClassifier.html#copyBatchOtuTableToOutputDir-java.io.File-java.lang.Integer-">copyBatchOtuTableToOutputDir(File, Integer)</a></span> - Method in class biolockj.module.classifier.r16s.<a href="../biolockj/module/classifier/r16s/QiimeClosedRefClassifier.html" title="class in biolockj.module.classifier.r16s">QiimeClosedRefClassifier</a></dt>
<dd>
<div class="block">The method returns 1 bash script line that will copy the batch
 <a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#OTU_TABLE">"otu_table.biom"</a> from the batchDir to the output directory.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/BioLockJ.html#copyFileToPipelineRoot-java.io.File-">copyFileToPipelineRoot(File)</a></span> - Static method in class biolockj.<a href="../biolockj/BioLockJ.html" title="class in biolockj">BioLockJ</a></dt>
<dd>
<div class="block">Copy file to pipeline root directory.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/BioLockJ.html#copyInputData--">copyInputData()</a></span> - Static method in class biolockj.<a href="../biolockj/BioLockJ.html" title="class in biolockj">BioLockJ</a></dt>
<dd>
<div class="block">Create a copy of the sequence files in property <a href="../biolockj/Constants.html#INPUT_DIRS">"input.dirPaths"</a>, output to a directory
 named <a href="../biolockj/Constants.html#PIPELINE_DIR">"pipeline.pipelineDir"</a>/input.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/NextflowUtil.html#copyPipelineToEfs--">copyPipelineToEfs()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/NextflowUtil.html" title="class in biolockj.util">NextflowUtil</a></dt>
<dd>
<div class="block">Copy the initialized pipeline to EFS.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/r16s/QiimeDeNovoClassifier.html#copyTempOtuTableToOutputDir--">copyTempOtuTableToOutputDir()</a></span> - Method in class biolockj.module.classifier.r16s.<a href="../biolockj/module/classifier/r16s/QiimeDeNovoClassifier.html" title="class in biolockj.module.classifier.r16s">QiimeDeNovoClassifier</a></dt>
<dd>
<div class="block">The method returns 1 bash script line that will copy the batch
 <a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#OTU_TABLE">"otu_table.biom"</a> from the batchDir to the output directory.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#countNumReads-java.io.File-">countNumReads(File)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Method counts number of reads in the given sequence file by counting the number of lines and dividing by the
 number of lines/sample (fasta=2, fastq=4)</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/BioLockJ.html#createPipelineDirectory--">createPipelineDirectory()</a></span> - Static method in class biolockj.<a href="../biolockj/BioLockJ.html" title="class in biolockj">BioLockJ</a></dt>
<dd>
<div class="block">Create the pipeline root directory under $DOCKER_PROJ and save the path to
 <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/Constants.html#PIPELINE_DIR">"pipeline.pipelineDir"</a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BashScriptBuilder.html#createScript-java.lang.String-java.util.List-">createScript(String, List&lt;String&gt;)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BashScriptBuilder.html" title="class in biolockj.util">BashScriptBuilder</a></dt>
<dd>
<div class="block">Create the script.</div>
</dd>
</dl>
<a href="index-1.html">A</a>&nbsp;<a href="index-2.html">B</a>&nbsp;<a href="index-3.html">C</a>&nbsp;<a href="index-4.html">D</a>&nbsp;<a href="index-5.html">E</a>&nbsp;<a href="index-6.html">F</a>&nbsp;<a href="index-7.html">G</a>&nbsp;<a href="index-8.html">H</a>&nbsp;<a href="index-9.html">I</a>&nbsp;<a href="index-10.html">J</a>&nbsp;<a href="index-11.html">K</a>&nbsp;<a href="index-12.html">L</a>&nbsp;<a href="index-13.html">M</a>&nbsp;<a href="index-14.html">N</a>&nbsp;<a href="index-15.html">O</a>&nbsp;<a href="index-16.html">P</a>&nbsp;<a href="index-17.html">Q</a>&nbsp;<a href="index-18.html">R</a>&nbsp;<a href="index-19.html">S</a>&nbsp;<a href="index-20.html">T</a>&nbsp;<a href="index-21.html">U</a>&nbsp;<a href="index-22.html">V</a>&nbsp;<a href="index-23.html">W</a>&nbsp;</div>
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar.bottom">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.bottom.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../overview-summary.html">Overview</a></li>
<li>Package</li>
<li>Class</li>
<li>Use</li>
<li><a href="../overview-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li class="navBarCell1Rev">Index</li>
<li><a href="../help-doc.html">Help</a></li>
<li>&nbsp;&nbsp;</li>
<li><a href="https://github.com/msioda/BioLockJ/wiki" target="_top"><strong>RETURN to WIKI</strong></a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="index-2.html">Prev Letter</a></li>
<li><a href="index-4.html">Next Letter</a></li>
</ul>
<ul class="navList">
<li><a href="../index.html?index-files/index-3.html" target="_top">Frames</a></li>
<li><a href="index-3.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip.navbar.bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
