<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (1.8.0_144) on Tue Mar 12 06:27:48 EDT 2019 -->
<title>A-Index</title>
<meta name="date" content="2019-03-12">
<link rel="stylesheet" type="text/css" href="../stylesheet.css" title="Style">
<script type="text/javascript" src="../script.js"></script>
</head>
<body>
<script type="text/javascript"><!--
    try {
        if (location.href.indexOf('is-external=true') == -1) {
            parent.document.title="A-Index";
        }
    }
    catch(err) {
    }
//-->
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar.top">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.top.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../overview-summary.html">Overview</a></li>
<li>Package</li>
<li>Class</li>
<li>Use</li>
<li><a href="../overview-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li class="navBarCell1Rev">Index</li>
<li><a href="../help-doc.html">Help</a></li>
<li>&nbsp;&nbsp;</li>
<li><a href="https://github.com/msioda/BioLockJ/wiki" target="_top"><strong>RETURN to WIKI</strong></a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li>Prev Letter</li>
<li><a href="index-2.html">Next Letter</a></li>
</ul>
<ul class="navList">
<li><a href="../index.html?index-files/index-1.html" target="_top">Frames</a></li>
<li><a href="index-1.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip.navbar.top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<div class="contentContainer"><a href="index-1.html">A</a>&nbsp;<a href="index-2.html">B</a>&nbsp;<a href="index-3.html">C</a>&nbsp;<a href="index-4.html">D</a>&nbsp;<a href="index-5.html">E</a>&nbsp;<a href="index-6.html">F</a>&nbsp;<a href="index-7.html">G</a>&nbsp;<a href="index-8.html">H</a>&nbsp;<a href="index-9.html">I</a>&nbsp;<a href="index-10.html">J</a>&nbsp;<a href="index-11.html">K</a>&nbsp;<a href="index-12.html">L</a>&nbsp;<a href="index-13.html">M</a>&nbsp;<a href="index-14.html">N</a>&nbsp;<a href="index-15.html">O</a>&nbsp;<a href="index-16.html">P</a>&nbsp;<a href="index-17.html">Q</a>&nbsp;<a href="index-18.html">R</a>&nbsp;<a href="index-19.html">S</a>&nbsp;<a href="index-20.html">T</a>&nbsp;<a href="index-21.html">U</a>&nbsp;<a href="index-22.html">V</a>&nbsp;<a href="index-23.html">W</a>&nbsp;<a name="I:A">
<!--   -->
</a>
<h2 class="title">A</h2>
<dl>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/TrimPrimers.html#addBadFilesToSummary--">addBadFilesToSummary()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/TrimPrimers.html" title="class in biolockj.module.seq">TrimPrimers</a></dt>
<dd>
<div class="block">Update summaryMsgs with list of seq files that contained no reads with a valid primer.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/MetaUtil.html#addColumn-java.lang.String-java.util.Map-java.io.File-boolean-">addColumn(String, Map&lt;String, String&gt;, File, boolean)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/MetaUtil.html" title="class in biolockj.util">MetaUtil</a></dt>
<dd>
<div class="block">Adds a column to the metadata file.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/JsonNode.html#addCount-java.lang.Integer-">addCount(Integer)</a></span> - Method in class biolockj.node.<a href="../biolockj/node/JsonNode.html" title="class in biolockj.node">JsonNode</a></dt>
<dd>
<div class="block">Add taxa count</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/taxa/AddMetadataToTaxaTables.html#addHitRatioToMetadata--">addHitRatioToMetadata()</a></span> - Method in class biolockj.module.report.taxa.<a href="../biolockj/module/report/taxa/AddMetadataToTaxaTables.html" title="class in biolockj.module.report.taxa">AddMetadataToTaxaTables</a></dt>
<dd>
<div class="block">Add Num_Hits/Num_Reads as Hit_Ratio column to the metadata file</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BioLockJUtil.html#addLeadingSpaces-java.lang.String-int-">addLeadingSpaces(String, int)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BioLockJUtil.html" title="class in biolockj.util">BioLockJUtil</a></dt>
<dd>
<div class="block">Add leading spaces until the val is padded to given length</div>
</dd>
<dt><a href="../biolockj/module/report/humann2/AddMetadataToPathwayTables.html" title="class in biolockj.module.report.humann2"><span class="typeNameLink">AddMetadataToPathwayTables</span></a> - Class in <a href="../biolockj/module/report/humann2/package-summary.html">biolockj.module.report.humann2</a></dt>
<dd>
<div class="block">This BioModule is used to add metadata columns to the HumanN2 pathway abundance, pathway coverage, and gene family tables.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/humann2/AddMetadataToPathwayTables.html#AddMetadataToPathwayTables--">AddMetadataToPathwayTables()</a></span> - Constructor for class biolockj.module.report.humann2.<a href="../biolockj/module/report/humann2/AddMetadataToPathwayTables.html" title="class in biolockj.module.report.humann2">AddMetadataToPathwayTables</a></dt>
<dd>&nbsp;</dd>
<dt><a href="../biolockj/module/report/taxa/AddMetadataToTaxaTables.html" title="class in biolockj.module.report.taxa"><span class="typeNameLink">AddMetadataToTaxaTables</span></a> - Class in <a href="../biolockj/module/report/taxa/package-summary.html">biolockj.module.report.taxa</a></dt>
<dd>
<div class="block">This BioModule is used to add metadata columns to the OTU abundance tables.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/taxa/AddMetadataToTaxaTables.html#AddMetadataToTaxaTables--">AddMetadataToTaxaTables()</a></span> - Constructor for class biolockj.module.report.taxa.<a href="../biolockj/module/report/taxa/AddMetadataToTaxaTables.html" title="class in biolockj.module.report.taxa">AddMetadataToTaxaTables</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/BuildQiimeMapping.html#addMissingFields--">addMissingFields()</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/BuildQiimeMapping.html" title="class in biolockj.module.implicit.qiime">BuildQiimeMapping</a></dt>
<dd>
<div class="block">Create QIIME mapping based on metadata file, output to temp dir.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/ParsedSample.html#addNode-biolockj.node.OtuNode-">addNode(OtuNode)</a></span> - Method in class biolockj.node.<a href="../biolockj/node/ParsedSample.html" title="class in biolockj.node">ParsedSample</a></dt>
<dd>
<div class="block">Add the OtuNode to the ParsedSample.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/ParserModule.html#addOtuNode-biolockj.node.OtuNode-">addOtuNode(OtuNode)</a></span> - Method in interface biolockj.module.implicit.parser.<a href="../biolockj/module/implicit/parser/ParserModule.html" title="interface in biolockj.module.implicit.parser">ParserModule</a></dt>
<dd>
<div class="block">Each sample with taxonomic assignments from a <a href="../biolockj/module/classifier/ClassifierModule.html" title="interface in biolockj.module.classifier"><code>ClassifierModule</code></a> is parsed by
 <a href="../biolockj/module/implicit/parser/ParserModule.html#parseSamples--"><code>ParserModule.parseSamples()</code></a> to produce a <a href="../biolockj/node/ParsedSample.html" title="class in biolockj.node"><code>ParsedSample</code></a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/ParserModuleImpl.html#addOtuNode-biolockj.node.OtuNode-">addOtuNode(OtuNode)</a></span> - Method in class biolockj.module.implicit.parser.<a href="../biolockj/module/implicit/parser/ParserModuleImpl.html" title="class in biolockj.module.implicit.parser">ParserModuleImpl</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SummaryUtil.html#addSummaryFooterForFailedPipeline-java.lang.String-">addSummaryFooterForFailedPipeline(String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SummaryUtil.html" title="class in biolockj.util">SummaryUtil</a></dt>
<dd>
<div class="block">Called when pipeline fails to add summary details to summary file, if possible.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/OtuNode.html#addTaxa-java.lang.String-java.lang.String-">addTaxa(String, String)</a></span> - Method in interface biolockj.node.<a href="../biolockj/node/OtuNode.html" title="interface in biolockj.node">OtuNode</a></dt>
<dd>
<div class="block">Add <a href="../biolockj/node/OtuNode.html" title="interface in biolockj.node"><code>OtuNode</code></a> taxa names for the
 <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/Constants.html#REPORT_TAXONOMY_LEVELS">"report.taxonomyLevels"</a> level mapped by the classifier
 specific levelDelim parameter: .</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/OtuNodeImpl.html#addTaxa-java.lang.String-java.lang.String-">addTaxa(String, String)</a></span> - Method in class biolockj.node.<a href="../biolockj/node/OtuNodeImpl.html" title="class in biolockj.node">OtuNodeImpl</a></dt>
<dd>
<div class="block">If called for level not included in <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/Constants.html#REPORT_TAXONOMY_LEVELS">"report.taxonomyLevels"</a>,
 check if the top level taxonomy level is already assigned.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BioLockJUtil.html#addTrailingSpaces-java.lang.String-int-">addTrailingSpaces(String, int)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BioLockJUtil.html" title="class in biolockj.util">BioLockJUtil</a></dt>
<dd>
<div class="block">Add trailing spaces until the val is padded to given length</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/r/R_CalculateStats.html#ADJ_PVAL_ATTRIBUTE">ADJ_PVAL_ATTRIBUTE</a></span> - Static variable in class biolockj.module.report.r.<a href="../biolockj/module/report/r/R_CalculateStats.html" title="class in biolockj.module.report.r">R_CalculateStats</a></dt>
<dd>
<div class="block">This <a href="../biolockj/module/report/r/R_CalculateStats.html#R_PVAL_ADJ_METHOD">"r_CalculateStats.pAdjustMethod"</a> option can be set in <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> file: "ATTRIBUTE"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/r/R_CalculateStats.html#ADJ_PVAL_GLOBAL">ADJ_PVAL_GLOBAL</a></span> - Static variable in class biolockj.module.report.r.<a href="../biolockj/module/report/r/R_CalculateStats.html" title="class in biolockj.module.report.r">R_CalculateStats</a></dt>
<dd>
<div class="block">This <a href="../biolockj/module/report/r/R_CalculateStats.html#R_PVAL_ADJ_METHOD">"r_CalculateStats.pAdjustMethod"</a> option can be set in <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> file: "GLOBAL"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/r/R_CalculateStats.html#ADJ_PVAL_LOCAL">ADJ_PVAL_LOCAL</a></span> - Static variable in class biolockj.module.report.r.<a href="../biolockj/module/report/r/R_CalculateStats.html" title="class in biolockj.module.report.r">R_CalculateStats</a></dt>
<dd>
<div class="block">This <a href="../biolockj/module/report/r/R_CalculateStats.html#R_PVAL_ADJ_METHOD">"r_CalculateStats.pAdjustMethod"</a> option can be set in <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> file: "LOCAL"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/r/R_CalculateStats.html#ADJ_PVAL_TAXA">ADJ_PVAL_TAXA</a></span> - Static variable in class biolockj.module.report.r.<a href="../biolockj/module/report/r/R_CalculateStats.html" title="class in biolockj.module.report.r">R_CalculateStats</a></dt>
<dd>
<div class="block">This <a href="../biolockj/module/report/r/R_CalculateStats.html#R_PVAL_ADJ_METHOD">"r_CalculateStats.pAdjustMethod"</a> option can be set in <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> file: "TAXA"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/TaxaUtil.html#allTaxonomyLevels--">allTaxonomyLevels()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/TaxaUtil.html" title="class in biolockj.util">TaxaUtil</a></dt>
<dd>
<div class="block">Returns a list of all taxonomy levels, not only the levels configured via
 <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/Constants.html#REPORT_TAXONOMY_LEVELS">"report.taxonomyLevels"</a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#ALPHA_DIV_NULL_VALUE">ALPHA_DIV_NULL_VALUE</a></span> - Static variable in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime">QiimeClassifier</a></dt>
<dd>
<div class="block">Value output by <a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#SCRIPT_CALC_ALPHA_DIVERSITY">"alpha_diversity.py"</a> for null values: "N/A"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#ALPHA_DIVERSITY_TABLE">ALPHA_DIVERSITY_TABLE</a></span> - Static variable in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime">QiimeClassifier</a></dt>
<dd>
<div class="block">File produced by QIIME <a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#SCRIPT_CALC_ALPHA_DIVERSITY">"alpha_diversity.py"</a> script: "alphaDiversity.txt"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#APP_START_TIME">APP_START_TIME</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block">Captures the application start time</div>
</dd>
<dt><a href="../biolockj/module/seq/AwkFastaConverter.html" title="class in biolockj.module.seq"><span class="typeNameLink">AwkFastaConverter</span></a> - Class in <a href="../biolockj/module/seq/package-summary.html">biolockj.module.seq</a></dt>
<dd>
<div class="block">This BioModule uses awk and gzip to convert input sequence files into a decompressed fasta file format.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/AwkFastaConverter.html#AwkFastaConverter--">AwkFastaConverter()</a></span> - Constructor for class biolockj.module.seq.<a href="../biolockj/module/seq/AwkFastaConverter.html" title="class in biolockj.module.seq">AwkFastaConverter</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#AWS_FLAG">AWS_FLAG</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">AWS pipeline runtime parameter switch: "-aws"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#AWS_RAM">AWS_RAM</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> String property: AWS memory set in Nextflow main.nf: "aws.ram"</div>
</dd>
</dl>
<a href="index-1.html">A</a>&nbsp;<a href="index-2.html">B</a>&nbsp;<a href="index-3.html">C</a>&nbsp;<a href="index-4.html">D</a>&nbsp;<a href="index-5.html">E</a>&nbsp;<a href="index-6.html">F</a>&nbsp;<a href="index-7.html">G</a>&nbsp;<a href="index-8.html">H</a>&nbsp;<a href="index-9.html">I</a>&nbsp;<a href="index-10.html">J</a>&nbsp;<a href="index-11.html">K</a>&nbsp;<a href="index-12.html">L</a>&nbsp;<a href="index-13.html">M</a>&nbsp;<a href="index-14.html">N</a>&nbsp;<a href="index-15.html">O</a>&nbsp;<a href="index-16.html">P</a>&nbsp;<a href="index-17.html">Q</a>&nbsp;<a href="index-18.html">R</a>&nbsp;<a href="index-19.html">S</a>&nbsp;<a href="index-20.html">T</a>&nbsp;<a href="index-21.html">U</a>&nbsp;<a href="index-22.html">V</a>&nbsp;<a href="index-23.html">W</a>&nbsp;</div>
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar.bottom">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.bottom.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../overview-summary.html">Overview</a></li>
<li>Package</li>
<li>Class</li>
<li>Use</li>
<li><a href="../overview-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li class="navBarCell1Rev">Index</li>
<li><a href="../help-doc.html">Help</a></li>
<li>&nbsp;&nbsp;</li>
<li><a href="https://github.com/msioda/BioLockJ/wiki" target="_top"><strong>RETURN to WIKI</strong></a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li>Prev Letter</li>
<li><a href="index-2.html">Next Letter</a></li>
</ul>
<ul class="navList">
<li><a href="../index.html?index-files/index-1.html" target="_top">Frames</a></li>
<li><a href="index-1.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip.navbar.bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
