<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (1.8.0_144) on Fri Oct 26 16:45:56 EDT 2018 -->
<title>B-Index</title>
<meta name="date" content="2018-10-26">
<link rel="stylesheet" type="text/css" href="../stylesheet.css" title="Style">
<script type="text/javascript" src="../script.js"></script>
</head>
<body>
<script type="text/javascript"><!--
    try {
        if (location.href.indexOf('is-external=true') == -1) {
            parent.document.title="B-Index";
        }
    }
    catch(err) {
    }
//-->
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar.top">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.top.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../overview-summary.html">Overview</a></li>
<li>Package</li>
<li>Class</li>
<li>Use</li>
<li><a href="../overview-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li class="navBarCell1Rev">Index</li>
<li><a href="../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="index-1.html">Prev Letter</a></li>
<li><a href="index-3.html">Next Letter</a></li>
</ul>
<ul class="navList">
<li><a href="../index.html?index-files/index-2.html" target="_top">Frames</a></li>
<li><a href="index-2.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip.navbar.top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<div class="contentContainer"><a href="index-1.html">A</a>&nbsp;<a href="index-2.html">B</a>&nbsp;<a href="index-3.html">C</a>&nbsp;<a href="index-4.html">D</a>&nbsp;<a href="index-5.html">E</a>&nbsp;<a href="index-6.html">F</a>&nbsp;<a href="index-7.html">G</a>&nbsp;<a href="index-8.html">H</a>&nbsp;<a href="index-9.html">I</a>&nbsp;<a href="index-10.html">J</a>&nbsp;<a href="index-11.html">K</a>&nbsp;<a href="index-12.html">L</a>&nbsp;<a href="index-13.html">M</a>&nbsp;<a href="index-14.html">N</a>&nbsp;<a href="index-15.html">O</a>&nbsp;<a href="index-16.html">P</a>&nbsp;<a href="index-17.html">Q</a>&nbsp;<a href="index-18.html">R</a>&nbsp;<a href="index-19.html">S</a>&nbsp;<a href="index-20.html">T</a>&nbsp;<a href="index-21.html">U</a>&nbsp;<a href="index-22.html">V</a>&nbsp;<a href="index-23.html">W</a>&nbsp;<a name="I:B">
<!--   -->
</a>
<h2 class="title">B</h2>
<dl>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/BuildQiimeMapping.html#BARCODE_SEQUENCE">BARCODE_SEQUENCE</a></span> - Static variable in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/BuildQiimeMapping.html" title="class in biolockj.module.implicit.qiime">BuildQiimeMapping</a></dt>
<dd>
<div class="block">QIIME mapping file required 2nd column name</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DemuxUtil.html#BARCODE_USE_REV_COMP">BARCODE_USE_REV_COMP</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/DemuxUtil.html" title="class in biolockj.util">DemuxUtil</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> boolean property "demux.barcodeUseReverseCompliment" will use the reverse compliment of
 <a href="../biolockj/util/MetaUtil.html#META_BARCODE_COLUMN">"metadata.barcodeColumn"</a> to match sample IDs to sequences.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DemuxUtil.html#barcodeInHeader--">barcodeInHeader()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/DemuxUtil.html" title="class in biolockj.util">DemuxUtil</a></dt>
<dd>
<div class="block">Return TRUE if Config is setup to demux the sequence data based on barcodes in the sequence headers.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DemuxUtil.html#barcodeInSeq--">barcodeInSeq()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/DemuxUtil.html" title="class in biolockj.util">DemuxUtil</a></dt>
<dd>
<div class="block">Return TRUE if Config is setup to demux the sequence data based on barcodes in the sequence itself.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#BASE_DIR_FLAG">BASE_DIR_FLAG</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">Pipeline parent directory file-path runtime parameter switch: "-b"</div>
</dd>
<dt><a href="../biolockj/util/BashScriptBuilder.html" title="class in biolockj.util"><span class="typeNameLink">BashScriptBuilder</span></a> - Class in <a href="../biolockj/util/package-summary.html">biolockj.util</a></dt>
<dd>
<div class="block">This utility class generates the bash script files using the lines provided and <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> script
 properties.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/r16s/QiimeClosedRefClassifier.html#BATCH_MAPPING">BATCH_MAPPING</a></span> - Static variable in class biolockj.module.classifier.r16s.<a href="../biolockj/module/classifier/r16s/QiimeClosedRefClassifier.html" title="class in biolockj.module.classifier.r16s">QiimeClosedRefClassifier</a></dt>
<dd>
<div class="block">Name of each batch mapping file (each in its own batch directory)</div>
</dd>
<dt><a href="../biolockj/package-summary.html">biolockj</a> - package biolockj</dt>
<dd>
<div class="block">The root biolockj package contains core classes used by all BioLockJ pipelines.</div>
</dd>
<dt><a href="../biolockj/BioLockJ.html" title="class in biolockj"><span class="typeNameLink">BioLockJ</span></a> - Class in <a href="../biolockj/package-summary.html">biolockj</a></dt>
<dd>
<div class="block">This is the primary BioLockJ class - its main() method is executed when the jar is run.<br>
 This class validates the runtime parameters to run a new pipeline or restart a failed pipeline.<br>
 The Java log file is initialized and the configuration file is processed before starting the pipeline.<br>
 If the pipeline is successful, the program executes clean up operations (if configured) and creates a status-complete
 indicator file in the pipeline root directory.<br></div>
</dd>
<dt><a href="../biolockj/exception/package-summary.html">biolockj.exception</a> - package biolockj.exception</dt>
<dd>
<div class="block">This package contains <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Exception.html?is-external=true" title="class or interface in java.lang"><code>Exception</code></a> subclasses specific to BioLockJ.</div>
</dd>
<dt><a href="../biolockj/module/package-summary.html">biolockj.module</a> - package biolockj.module</dt>
<dd>
<div class="block">Pipeline modules are loaded based on the <a href="../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s listed in the <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>
 file.<br>
 All modules (i.e., implement <a href="../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>) are contained within this package or its
 sub-packages.</div>
</dd>
<dt><a href="../biolockj/module/classifier/package-summary.html">biolockj.module.classifier</a> - package biolockj.module.classifier</dt>
<dd>
<div class="block">This package contains Classifier specific <a href="../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s that build bash scripts to generate
 taxonomy reports from the sequences files.<br></div>
</dd>
<dt><a href="../biolockj/module/classifier/r16s/package-summary.html">biolockj.module.classifier.r16s</a> - package biolockj.module.classifier.r16s</dt>
<dd>
<div class="block">This package contains RDP and QIIME classifier modules that assign taxonomy to 16s sequences.<br>
 Output file formats are classifier specific so classifier requires a unique post-requisite
 <a href="../biolockj/module/implicit/parser/ParserModule.html" title="interface in biolockj.module.implicit.parser"><code>ParserModule</code></a>.</div>
</dd>
<dt><a href="../biolockj/module/classifier/wgs/package-summary.html">biolockj.module.classifier.wgs</a> - package biolockj.module.classifier.wgs</dt>
<dd>
<div class="block">This package contains Kraken, MetaPhlAn, and SLIMM classifier modules that assign taxonomy to WGS data.<br>
 As with all BioLockJ classifiers, these modules generate raw classifier output files from the sequence data.<br>
 Include the corresponding <a href="../biolockj/module/implicit/parser/ParserModule.html" title="interface in biolockj.module.implicit.parser"><code>ParserModule</code></a> in the <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> file to
 generate standardized OTU abundance tables.</div>
</dd>
<dt><a href="../biolockj/module/implicit/package-summary.html">biolockj.module.implicit</a> - package biolockj.module.implicit</dt>
<dd>
<div class="block">The modules in this package are implicitly added to pipelines as needed.<br>
 These modules cannot be directly added to any pipeline unless overridden via project.allowImplicitModules=Y.</div>
</dd>
<dt><a href="../biolockj/module/implicit/parser/package-summary.html">biolockj.module.implicit.parser</a> - package biolockj.module.implicit.parser</dt>
<dd>
<div class="block">This package contains Parser BioModules in the r16s and WGS sub-packages that are paired with a
 <a href="../biolockj/module/classifier/ClassifierModule.html" title="interface in biolockj.module.classifier"><code>ClassifierModule</code></a> via
 <a href="../biolockj/module/BioModule.html#getPostRequisiteModules--"><code>BioModule.getPostRequisiteModules()</code></a> to run immediately after the
 classifier.</div>
</dd>
<dt><a href="../biolockj/module/implicit/parser/r16s/package-summary.html">biolockj.module.implicit.parser.r16s</a> - package biolockj.module.implicit.parser.r16s</dt>
<dd>
<div class="block">This package contains Parser <a href="../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s that convert the 16S taxonomy reports generated by
 16S classifiers (such as RDP and QIIME) into standardized OTU abundance tables.<br></div>
</dd>
<dt><a href="../biolockj/module/implicit/parser/wgs/package-summary.html">biolockj.module.implicit.parser.wgs</a> - package biolockj.module.implicit.parser.wgs</dt>
<dd>
<div class="block">This package contains Parser <a href="../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s that convert the WGS taxonomy reports generated by
 WGS classifiers (such as Kraken, MetaPhlAn, or SLIMM) into standardized OTU abundance tables.<br></div>
</dd>
<dt><a href="../biolockj/module/implicit/qiime/package-summary.html">biolockj.module.implicit.qiime</a> - package biolockj.module.implicit.qiime</dt>
<dd>
<div class="block">This package contains <a href="../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s that are implicitly added to QIIME pipeline as needed.</div>
</dd>
<dt><a href="../biolockj/module/r/package-summary.html">biolockj.module.r</a> - package biolockj.module.r</dt>
<dd>
<div class="block">This package contains <a href="../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s that build pipeline reports from the standard OTU abundance
 tables by generating R scripts to produce the summary statistics and data visualizations output to PDF files.<br></div>
</dd>
<dt><a href="../biolockj/module/report/package-summary.html">biolockj.module.report</a> - package biolockj.module.report</dt>
<dd>
<div class="block">This package contains <a href="../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s that normalize OTU abundance tables output by Parser
 modules, merges them with the metadata, and generates various reports and notifications.</div>
</dd>
<dt><a href="../biolockj/module/seq/package-summary.html">biolockj.module.seq</a> - package biolockj.module.seq</dt>
<dd>
<div class="block"><a href="../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s used to prepare sequence files or update the metadata prior to classification.</div>
</dd>
<dt><a href="../biolockj/node/package-summary.html">biolockj.node</a> - package biolockj.node</dt>
<dd>
<div class="block">Package <a href="../biolockj/node/package-summary.html"><code>biolockj.node</code></a> contains core classes instantiated by the
 <a href="../biolockj/module/implicit/parser/ParserModule.html" title="interface in biolockj.module.implicit.parser"><code>ParserModule</code></a> to hold Sample ID-(OTU name/count) data read in from
 <a href="../biolockj/module/classifier/ClassifierModule.html" title="interface in biolockj.module.classifier"><code>ClassifierModule</code></a> output.</div>
</dd>
<dt><a href="../biolockj/node/r16s/package-summary.html">biolockj.node.r16s</a> - package biolockj.node.r16s</dt>
<dd>
<div class="block">Package contains RDP and QIIME specific <a href="../biolockj/node/OtuNode.html" title="interface in biolockj.node"><code>OtuNode</code></a>s that represent the OTUs (operational taxonomy
 units) assigned by 16S <a href="../biolockj/module/classifier/ClassifierModule.html" title="interface in biolockj.module.classifier"><code>ClassifierModule</code></a>s</div>
</dd>
<dt><a href="../biolockj/node/wgs/package-summary.html">biolockj.node.wgs</a> - package biolockj.node.wgs</dt>
<dd>
<div class="block">Package contains Kraken, MetaPhlAn, and SLIMM specific <a href="../biolockj/node/OtuNode.html" title="interface in biolockj.node"><code>OtuNode</code></a>s that represent the OTUs
 (operational taxonomy units) assigned by WGS <a href="../biolockj/module/classifier/ClassifierModule.html" title="interface in biolockj.module.classifier"><code>ClassifierModule</code></a>s</div>
</dd>
<dt><a href="../biolockj/util/package-summary.html">biolockj.util</a> - package biolockj.util</dt>
<dd>
<div class="block">Static utilities centralize and organize reusable core methods.</div>
</dd>
<dt><a href="../biolockj/util/BioLockJUtil.html" title="class in biolockj.util"><span class="typeNameLink">BioLockJUtil</span></a> - Class in <a href="../biolockj/util/package-summary.html">biolockj.util</a></dt>
<dd>
<div class="block">Simple utility containing String manipulation and formatting functions.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BioLockJUtil.html#BioLockJUtil--">BioLockJUtil()</a></span> - Constructor for class biolockj.util.<a href="../biolockj/util/BioLockJUtil.html" title="class in biolockj.util">BioLockJUtil</a></dt>
<dd>&nbsp;</dd>
<dt><a href="../biolockj/module/BioModule.html" title="interface in biolockj.module"><span class="typeNameLink">BioModule</span></a> - Interface in <a href="../biolockj/module/package-summary.html">biolockj.module</a></dt>
<dd>
<div class="block">Classes that implement this interface are eligible to be included in a BioLockJ pipeline.<br>
 Use the <b>#BioModule</b> tag with the class name in the <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> file to include a module.<br>
 The <a href="../biolockj/Pipeline.html" title="class in biolockj"><code>Pipeline</code></a> class executes BioModules in the order provided in the <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>
 file.<br></div>
</dd>
<dt><a href="../biolockj/BioModuleFactory.html" title="class in biolockj"><span class="typeNameLink">BioModuleFactory</span></a> - Class in <a href="../biolockj/package-summary.html">biolockj</a></dt>
<dd>
<div class="block">This class initializes pipeline modules, startign with those in the Config file and adding the prerequisite and
 post-requisite modules.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/BioModuleFactory.html#BioModuleFactory--">BioModuleFactory()</a></span> - Constructor for class biolockj.<a href="../biolockj/BioModuleFactory.html" title="class in biolockj">BioModuleFactory</a></dt>
<dd>&nbsp;</dd>
<dt><a href="../biolockj/module/BioModuleImpl.html" title="class in biolockj.module"><span class="typeNameLink">BioModuleImpl</span></a> - Class in <a href="../biolockj/module/package-summary.html">biolockj.module</a></dt>
<dd>
<div class="block">Superclass for standard BioModules (classifiers, parsers, etc).</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModuleImpl.html#BioModuleImpl--">BioModuleImpl()</a></span> - Constructor for class biolockj.module.<a href="../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Pipeline.html#BLJ_COMPLETE">BLJ_COMPLETE</a></span> - Static variable in class biolockj.<a href="../biolockj/Pipeline.html" title="class in biolockj">Pipeline</a></dt>
<dd>
<div class="block">Name of the file created in the BioModule or <a href="../biolockj/Config.html#INTERNAL_PIPELINE_DIR">"internal.pipelineDir"</a> root directory to
 indicate execution was successful: "biolockjComplete"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Pipeline.html#BLJ_FAILED">BLJ_FAILED</a></span> - Static variable in class biolockj.<a href="../biolockj/Pipeline.html" title="class in biolockj">Pipeline</a></dt>
<dd>
<div class="block">Name of the file created in the <a href="../biolockj/Config.html#INTERNAL_PIPELINE_DIR">"internal.pipelineDir"</a> root directory to indicate fatal
 application errors halted execution: "biolockjFailed"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DockerUtil.html#BLJ_OPTIONS">BLJ_OPTIONS</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/DockerUtil.html" title="class in biolockj.util">DockerUtil</a></dt>
<dd>
<div class="block">Docker environment variable holding the Docker program switches: "BLJ_OPTIONS"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Pipeline.html#BLJ_STARTED">BLJ_STARTED</a></span> - Static variable in class biolockj.<a href="../biolockj/Pipeline.html" title="class in biolockj">Pipeline</a></dt>
<dd>
<div class="block">Name of the file created in the BioModule root directory to indicate execution has started: "biolockjStarted"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/MetaphlanClassifier.html#bowtie2ext">bowtie2ext</a></span> - Static variable in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/MetaphlanClassifier.html" title="class in biolockj.module.classifier.wgs">MetaphlanClassifier</a></dt>
<dd>
<div class="block">File suffix for intermediate bowtie2 index created by metaphlan</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/Demultiplexer.html#breakUpFiles--">breakUpFiles()</a></span> - Method in class biolockj.module.implicit.<a href="../biolockj/module/implicit/Demultiplexer.html" title="class in biolockj.module.implicit">Demultiplexer</a></dt>
<dd>
<div class="block">Some multiplexed files can be very large.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Properties.html#buildConfig-java.io.File-">buildConfig(File)</a></span> - Static method in class biolockj.<a href="../biolockj/Properties.html" title="class in biolockj">Properties</a></dt>
<dd>
<div class="block">Recursive method handles nested default Config files.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/r/R_Module.html#buildDockerBashScript--">buildDockerBashScript()</a></span> - Method in class biolockj.module.r.<a href="../biolockj/module/r/R_Module.html" title="class in biolockj.module.r">R_Module</a></dt>
<dd>
<div class="block">Run R script in docker.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DockerUtil.html#buildDockerScript--">buildDockerScript()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/DockerUtil.html" title="class in biolockj.util">DockerUtil</a></dt>
<dd>
<div class="block">Build a docker bash script.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BashScriptBuilder.html#buildExecuteFunction--">buildExecuteFunction()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BashScriptBuilder.html" title="class in biolockj.util">BashScriptBuilder</a></dt>
<dd>
<div class="block">Create bash worker script function: <a href="../biolockj/util/BashScriptBuilder.html#FUNCTION_EXECUTE">"execute"</a>.<br>
 Failure details written to the failure script file if any occur.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/r/R_Module.html#buildGetModuleScriptDirFunction--">buildGetModuleScriptDirFunction()</a></span> - Static method in class biolockj.module.r.<a href="../biolockj/module/r/R_Module.html" title="class in biolockj.module.r">R_Module</a></dt>
<dd>
<div class="block">Build R script function to return the R script parent directory: <a href="../biolockj/module/r/R_Module.html#R_FUNCTION_GET_MOD_SCRIPT">"getModuleScript"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BashScriptBuilder.html#buildMainScript-biolockj.module.ScriptModule-">buildMainScript(ScriptModule)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BashScriptBuilder.html" title="class in biolockj.util">BashScriptBuilder</a></dt>
<dd>
<div class="block">Build the MIAN script.</div>
</dd>
<dt><a href="../biolockj/module/r/BuildMdsPlots.html" title="class in biolockj.module.r"><span class="typeNameLink">BuildMdsPlots</span></a> - Class in <a href="../biolockj/module/r/package-summary.html">biolockj.module.r</a></dt>
<dd>
<div class="block">This BioModule is used to build the R script used to generate MDS plots for each report field and each taxonomy level
 configured.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/r/BuildMdsPlots.html#BuildMdsPlots--">BuildMdsPlots()</a></span> - Constructor for class biolockj.module.r.<a href="../biolockj/module/r/BuildMdsPlots.html" title="class in biolockj.module.r">BuildMdsPlots</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/exception/ConfigException.html#buildMessage-java.lang.String-java.lang.String-java.lang.String-">buildMessage(String, String, String)</a></span> - Static method in exception biolockj.exception.<a href="../biolockj/exception/ConfigException.html" title="class in biolockj.exception">ConfigException</a></dt>
<dd>
<div class="block">Build a standard error message for Configuration file errors.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/BioModuleFactory.html#buildModules--">buildModules()</a></span> - Static method in class biolockj.<a href="../biolockj/BioModuleFactory.html" title="class in biolockj">BioModuleFactory</a></dt>
<dd>
<div class="block">The method returns the ordered list of BioModules required as part of the runtime <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>
 file.<br>
 Each line the begins with <a href="../biolockj/Config.html#INTERNAL_BLJ_MODULE">"#BioModule"</a> should be followed by the full class name
 of a Java class that implements the <a href="../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a> interface.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/ImportMetadata.html#buildNewMetadataFile--">buildNewMetadataFile()</a></span> - Method in class biolockj.module.implicit.<a href="../biolockj/module/implicit/ImportMetadata.html" title="class in biolockj.module.implicit">ImportMetadata</a></dt>
<dd>
<div class="block">Create a simple metadata file in the module output directory, with only the 1st column populated with Sample IDs.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/ParserModuleImpl.html#buildOtuCounts--">buildOtuCounts()</a></span> - Method in class biolockj.module.implicit.parser.<a href="../biolockj/module/implicit/parser/ParserModuleImpl.html" title="class in biolockj.module.implicit.parser">ParserModuleImpl</a></dt>
<dd>
<div class="block">Build OTU counts after all classifier output has been parsed into OtuNode objects.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/ParsedSample.html#buildOtuCounts--">buildOtuCounts()</a></span> - Method in class biolockj.node.<a href="../biolockj/node/ParsedSample.html" title="class in biolockj.node">ParsedSample</a></dt>
<dd>
<div class="block">Build OTU counts after all otuNodes have been built.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/OtuNode.html#buildOtuNode-java.lang.String-java.lang.String-">buildOtuNode(String, String)</a></span> - Method in interface biolockj.node.<a href="../biolockj/node/OtuNode.html" title="interface in biolockj.node">OtuNode</a></dt>
<dd>
<div class="block">Build the <a href="../biolockj/node/OtuNode.html" title="interface in biolockj.node"><code>OtuNode</code></a> if level is configured in:
 <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/Config.html#REPORT_TAXONOMY_LEVELS">"report.taxonomyLevels"</a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/OtuNodeImpl.html#buildOtuNode-java.lang.String-java.lang.String-">buildOtuNode(String, String)</a></span> - Method in class biolockj.node.<a href="../biolockj/node/OtuNodeImpl.html" title="class in biolockj.node">OtuNodeImpl</a></dt>
<dd>&nbsp;</dd>
<dt><a href="../biolockj/module/r/BuildOtuPlots.html" title="class in biolockj.module.r"><span class="typeNameLink">BuildOtuPlots</span></a> - Class in <a href="../biolockj/module/r/package-summary.html">biolockj.module.r</a></dt>
<dd>
<div class="block">This BioModule is used to build the R script used to generate OTU-metadata box-plots and scatter-plots for each
 report field and taxonomy level.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/r/BuildOtuPlots.html#BuildOtuPlots--">BuildOtuPlots()</a></span> - Constructor for class biolockj.module.r.<a href="../biolockj/module/r/BuildOtuPlots.html" title="class in biolockj.module.r">BuildOtuPlots</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/ParserModule.html#buildOtuTables--">buildOtuTables()</a></span> - Method in interface biolockj.module.implicit.parser.<a href="../biolockj/module/implicit/parser/ParserModule.html" title="interface in biolockj.module.implicit.parser">ParserModule</a></dt>
<dd>
<div class="block">After <a href="../biolockj/module/implicit/parser/ParserModule.html#parseSamples--"><code>ParserModule.parseSamples()</code></a> completes, this method builds the OTU tables.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/ParserModuleImpl.html#buildOtuTables--">buildOtuTables()</a></span> - Method in class biolockj.module.implicit.parser.<a href="../biolockj/module/implicit/parser/ParserModuleImpl.html" title="class in biolockj.module.implicit.parser">ParserModuleImpl</a></dt>
<dd>
<div class="block">Method executes 2 core OTU table building functions:
 
 <a href="../biolockj/module/implicit/parser/ParserModuleImpl.html#createTaxaSparseThreeColFiles--"><code>ParserModuleImpl.createTaxaSparseThreeColFiles()</code></a>
 <a href="../biolockj/module/implicit/parser/ParserModuleImpl.html#buildRawCountTables-java.util.List-"><code>ParserModuleImpl.buildRawCountTables(List)</code></a>
 </div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/r/R_Module.html#buildPrimaryScript--">buildPrimaryScript()</a></span> - Method in class biolockj.module.r.<a href="../biolockj/module/r/R_Module.html" title="class in biolockj.module.r">R_Module</a></dt>
<dd>
<div class="block">Build the R module primary script.</div>
</dd>
<dt><a href="../biolockj/module/r/BuildPvalHistograms.html" title="class in biolockj.module.r"><span class="typeNameLink">BuildPvalHistograms</span></a> - Class in <a href="../biolockj/module/r/package-summary.html">biolockj.module.r</a></dt>
<dd>
<div class="block">This BioModule is used to build the R script used to generate p-value histograms for each report field and each
 taxonomy level configured.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/r/BuildPvalHistograms.html#BuildPvalHistograms--">BuildPvalHistograms()</a></span> - Constructor for class biolockj.module.r.<a href="../biolockj/module/r/BuildPvalHistograms.html" title="class in biolockj.module.r">BuildPvalHistograms</a></dt>
<dd>&nbsp;</dd>
<dt><a href="../biolockj/module/implicit/qiime/BuildQiimeMapping.html" title="class in biolockj.module.implicit.qiime"><span class="typeNameLink">BuildQiimeMapping</span></a> - Class in <a href="../biolockj/module/implicit/qiime/package-summary.html">biolockj.module.implicit.qiime</a></dt>
<dd>
<div class="block">This BioModule converts the metadata file into a tab delimited QIIME mapping file (if provided).<br>
 The QIIME mapping file is validated by calling QIIME script validate_mapping_file.py<br></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/BuildQiimeMapping.html#BuildQiimeMapping--">BuildQiimeMapping()</a></span> - Constructor for class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/BuildQiimeMapping.html" title="class in biolockj.module.implicit.qiime">BuildQiimeMapping</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/Rarefier.html#buildRarefiedFile-java.io.File-java.util.List-">buildRarefiedFile(File, List&lt;Long&gt;)</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/Rarefier.html" title="class in biolockj.module.seq">Rarefier</a></dt>
<dd>
<div class="block">Build the rarefied file for the input file, keeping only the given indexes</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/ParserModuleImpl.html#buildRawCountTables-java.util.List-">buildRawCountTables(List&lt;File&gt;)</a></span> - Method in class biolockj.module.implicit.parser.<a href="../biolockj/module/implicit/parser/ParserModuleImpl.html" title="class in biolockj.module.implicit.parser">ParserModuleImpl</a></dt>
<dd>
<div class="block">This method reads the temp 3-col-table to build the raw count table populated with raw OTU counts for OTUs that
 meet the threshold value: <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/module/implicit/parser/ParserModuleImpl.html#REPORT_MINIMUM_OTU_COUNT">"report.minOtuCount"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/r16s/RdpNode.html#buildRdpNode-java.lang.String-java.lang.String-">buildRdpNode(String, String)</a></span> - Method in class biolockj.node.r16s.<a href="../biolockj/node/r16s/RdpNode.html" title="class in biolockj.node.r16s">RdpNode</a></dt>
<dd>
<div class="block">Build the OtuNode by extracting the OTU names for each level from the line.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DockerUtil.html#buildRunDockerFunction-biolockj.module.BioModule-">buildRunDockerFunction(BioModule)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/DockerUtil.html" title="class in biolockj.util">DockerUtil</a></dt>
<dd>
<div class="block">Build the <a href="../biolockj/util/DockerUtil.html#SPAWN_DOCKER_CONTAINER">"spawnDockerContainer"</a> method, which takes container name, in/out port, and optionally script path
 parameters.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/r16s/QiimeClosedRefClassifier.html#buildScript-java.util.List-">buildScript(List&lt;File&gt;)</a></span> - Method in class biolockj.module.classifier.r16s.<a href="../biolockj/module/classifier/r16s/QiimeClosedRefClassifier.html" title="class in biolockj.module.classifier.r16s">QiimeClosedRefClassifier</a></dt>
<dd>
<div class="block">Create bash script lines to split up the QIIME mapping and fasta files into batches of size
 <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/module/ScriptModule.html#SCRIPT_BATCH_SIZE">"script.batchSize"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/r16s/QiimeDeNovoClassifier.html#buildScript-java.util.List-">buildScript(List&lt;File&gt;)</a></span> - Method in class biolockj.module.classifier.r16s.<a href="../biolockj/module/classifier/r16s/QiimeDeNovoClassifier.html" title="class in biolockj.module.classifier.r16s">QiimeDeNovoClassifier</a></dt>
<dd>
<div class="block">Return bash script lines to pick de novo OTUs by calling <a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime"><code>QiimeClassifier</code></a>
 getPickOtuLines() method.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/r16s/QiimeOpenRefClassifier.html#buildScript-java.util.List-">buildScript(List&lt;File&gt;)</a></span> - Method in class biolockj.module.classifier.r16s.<a href="../biolockj/module/classifier/r16s/QiimeOpenRefClassifier.html" title="class in biolockj.module.classifier.r16s">QiimeOpenRefClassifier</a></dt>
<dd>
<div class="block">Return bash script lines to pick open reference OTUs by passing the newest biom file in the temp directory [via
 bash code "$(ls -t $tempDir/*.biom | head -1)"] to <a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime"><code>QiimeClassifier</code></a>
 getPickOtuLines() method.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/r16s/RdpClassifier.html#buildScript-java.util.List-">buildScript(List&lt;File&gt;)</a></span> - Method in class biolockj.module.classifier.r16s.<a href="../biolockj/module/classifier/r16s/RdpClassifier.html" title="class in biolockj.module.classifier.r16s">RdpClassifier</a></dt>
<dd>
<div class="block">Build bash script lines to classify unpaired reads with RDP.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/Kraken2Classifier.html#buildScript-java.util.List-">buildScript(List&lt;File&gt;)</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/Kraken2Classifier.html" title="class in biolockj.module.classifier.wgs">Kraken2Classifier</a></dt>
<dd>
<div class="block">Build bash script lines to classify unpaired WGS reads with Kraken2.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/KrakenClassifier.html#buildScript-java.util.List-">buildScript(List&lt;File&gt;)</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/KrakenClassifier.html" title="class in biolockj.module.classifier.wgs">KrakenClassifier</a></dt>
<dd>
<div class="block">Build bash script lines to classify unpaired WGS reads with Kraken.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/MetaphlanClassifier.html#buildScript-java.util.List-">buildScript(List&lt;File&gt;)</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/MetaphlanClassifier.html" title="class in biolockj.module.classifier.wgs">MetaphlanClassifier</a></dt>
<dd>
<div class="block">Build bash script lines to classify unpaired WGS reads with Metaphlan.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/SlimmClassifier.html#buildScript-java.util.List-">buildScript(List&lt;File&gt;)</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/SlimmClassifier.html" title="class in biolockj.module.classifier.wgs">SlimmClassifier</a></dt>
<dd>
<div class="block">Build 2 bash script lines per sample to classify unpaired WGS reads.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/BuildQiimeMapping.html#buildScript-java.util.List-">buildScript(List&lt;File&gt;)</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/BuildQiimeMapping.html" title="class in biolockj.module.implicit.qiime">BuildQiimeMapping</a></dt>
<dd>
<div class="block">Create the Qiime corrected mapping file by building a new metadata file with ordered columns, and verify the
 format via the <a href="../biolockj/module/implicit/qiime/BuildQiimeMapping.html#SCRIPT_VALIDATE_MAPPING">"validate_mapping_file.py"</a> script.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/MergeOtuTables.html#buildScript-java.util.List-">buildScript(List&lt;File&gt;)</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/MergeOtuTables.html" title="class in biolockj.module.implicit.qiime">MergeOtuTables</a></dt>
<dd>
<div class="block">Build a single bash script line to call <a href="../biolockj/module/implicit/qiime/MergeOtuTables.html#SCRIPT_MERGE_OTU_TABLES">"merge_otu_tables.py"</a> to create the single
 <a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#OTU_TABLE">"otu_table.biom"</a>, as required by
 <a href="../biolockj/module/implicit/parser/r16s/QiimeParser.html" title="class in biolockj.module.implicit.parser.r16s"><code>QiimeParser</code></a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#buildScript-java.util.List-">buildScript(List&lt;File&gt;)</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime">QiimeClassifier</a></dt>
<dd>
<div class="block">Generate bash script lines to summarize QIIME results, build taxonomy reports, and add alpha diversity metrics.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/JavaModuleImpl.html#buildScript-java.util.List-">buildScript(List&lt;File&gt;)</a></span> - Method in class biolockj.module.<a href="../biolockj/module/JavaModuleImpl.html" title="class in biolockj.module">JavaModuleImpl</a></dt>
<dd>
<div class="block">Java script only require 2 lines, one to run the blj_config to update our $PATH and gain access to environment
 variables, and then the direct call to the BioLockJ.jar.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/r/R_Module.html#buildScript-java.util.List-">buildScript(List&lt;File&gt;)</a></span> - Method in class biolockj.module.r.<a href="../biolockj/module/r/R_Module.html" title="class in biolockj.module.r">R_Module</a></dt>
<dd>
<div class="block">Not needed for R script modules.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/ScriptModule.html#buildScript-java.util.List-">buildScript(List&lt;File&gt;)</a></span> - Method in interface biolockj.module.<a href="../biolockj/module/ScriptModule.html" title="interface in biolockj.module">ScriptModule</a></dt>
<dd>
<div class="block">Method returns a nested list of bash script lines to classify samples containing forward reads only.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/ScriptModuleImpl.html#buildScript-java.util.List-">buildScript(List&lt;File&gt;)</a></span> - Method in class biolockj.module.<a href="../biolockj/module/ScriptModuleImpl.html" title="class in biolockj.module">ScriptModuleImpl</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/AwkFastaConverter.html#buildScript-java.util.List-">buildScript(List&lt;File&gt;)</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/AwkFastaConverter.html" title="class in biolockj.module.seq">AwkFastaConverter</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/Gunzipper.html#buildScript-java.util.List-">buildScript(List&lt;File&gt;)</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/Gunzipper.html" title="class in biolockj.module.seq">Gunzipper</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/PearMergeReads.html#buildScript-java.util.List-">buildScript(List&lt;File&gt;)</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/PearMergeReads.html" title="class in biolockj.module.seq">PearMergeReads</a></dt>
<dd>
<div class="block">Build the script lines for each sample as a nested list.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/Kraken2Classifier.html#buildScriptForPairedReads-java.util.List-">buildScriptForPairedReads(List&lt;File&gt;)</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/Kraken2Classifier.html" title="class in biolockj.module.classifier.wgs">Kraken2Classifier</a></dt>
<dd>
<div class="block">Build bash script lines to classify paired WGS reads with Kraken.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/KrakenClassifier.html#buildScriptForPairedReads-java.util.List-">buildScriptForPairedReads(List&lt;File&gt;)</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/KrakenClassifier.html" title="class in biolockj.module.classifier.wgs">KrakenClassifier</a></dt>
<dd>
<div class="block">Build bash script lines to classify paired WGS reads with Kraken.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/MetaphlanClassifier.html#buildScriptForPairedReads-java.util.List-">buildScriptForPairedReads(List&lt;File&gt;)</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/MetaphlanClassifier.html" title="class in biolockj.module.classifier.wgs">MetaphlanClassifier</a></dt>
<dd>
<div class="block">Build bash script lines to classify paired WGS reads with Metaphlan.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/SlimmClassifier.html#buildScriptForPairedReads-java.util.List-">buildScriptForPairedReads(List&lt;File&gt;)</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/SlimmClassifier.html" title="class in biolockj.module.classifier.wgs">SlimmClassifier</a></dt>
<dd>
<div class="block">Build 2 bash script lines per sample to classify paired WGS reads.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#buildScriptForPairedReads-java.util.List-">buildScriptForPairedReads(List&lt;File&gt;)</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime">QiimeClassifier</a></dt>
<dd>
<div class="block">QIIME does not support paired reads</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/ScriptModule.html#buildScriptForPairedReads-java.util.List-">buildScriptForPairedReads(List&lt;File&gt;)</a></span> - Method in interface biolockj.module.<a href="../biolockj/module/ScriptModule.html" title="interface in biolockj.module">ScriptModule</a></dt>
<dd>
<div class="block">Method returns a nested list of bash script lines to classify samples containing paired reads.<br>
 The inner list contains the bash script lines required to classify 1 sample.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/ScriptModuleImpl.html#buildScriptForPairedReads-java.util.List-">buildScriptForPairedReads(List&lt;File&gt;)</a></span> - Method in class biolockj.module.<a href="../biolockj/module/ScriptModuleImpl.html" title="class in biolockj.module">ScriptModuleImpl</a></dt>
<dd>
<div class="block">The default behavior is the same for paired or unpaired data.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BashScriptBuilder.html#buildScripts-biolockj.module.ScriptModule-java.util.List-int-">buildScripts(ScriptModule, List&lt;List&lt;String&gt;&gt;, int)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BashScriptBuilder.html" title="class in biolockj.util">BashScriptBuilder</a></dt>
<dd>
<div class="block">This method builds the bash scripts (MAIN + worker scripts) for the given module.</div>
</dd>
</dl>
<a href="index-1.html">A</a>&nbsp;<a href="index-2.html">B</a>&nbsp;<a href="index-3.html">C</a>&nbsp;<a href="index-4.html">D</a>&nbsp;<a href="index-5.html">E</a>&nbsp;<a href="index-6.html">F</a>&nbsp;<a href="index-7.html">G</a>&nbsp;<a href="index-8.html">H</a>&nbsp;<a href="index-9.html">I</a>&nbsp;<a href="index-10.html">J</a>&nbsp;<a href="index-11.html">K</a>&nbsp;<a href="index-12.html">L</a>&nbsp;<a href="index-13.html">M</a>&nbsp;<a href="index-14.html">N</a>&nbsp;<a href="index-15.html">O</a>&nbsp;<a href="index-16.html">P</a>&nbsp;<a href="index-17.html">Q</a>&nbsp;<a href="index-18.html">R</a>&nbsp;<a href="index-19.html">S</a>&nbsp;<a href="index-20.html">T</a>&nbsp;<a href="index-21.html">U</a>&nbsp;<a href="index-22.html">V</a>&nbsp;<a href="index-23.html">W</a>&nbsp;</div>
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar.bottom">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.bottom.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../overview-summary.html">Overview</a></li>
<li>Package</li>
<li>Class</li>
<li>Use</li>
<li><a href="../overview-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li class="navBarCell1Rev">Index</li>
<li><a href="../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="index-1.html">Prev Letter</a></li>
<li><a href="index-3.html">Next Letter</a></li>
</ul>
<ul class="navList">
<li><a href="../index.html?index-files/index-2.html" target="_top">Frames</a></li>
<li><a href="index-2.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip.navbar.bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
