<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (1.8.0_144) on Mon Feb 18 11:38:21 EST 2019 -->
<title>P-Index</title>
<meta name="date" content="2019-02-18">
<link rel="stylesheet" type="text/css" href="../stylesheet.css" title="Style">
<script type="text/javascript" src="../script.js"></script>
</head>
<body>
<script type="text/javascript"><!--
    try {
        if (location.href.indexOf('is-external=true') == -1) {
            parent.document.title="P-Index";
        }
    }
    catch(err) {
    }
//-->
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar.top">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.top.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../overview-summary.html">Overview</a></li>
<li>Package</li>
<li>Class</li>
<li>Use</li>
<li><a href="../overview-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li class="navBarCell1Rev">Index</li>
<li><a href="../help-doc.html">Help</a></li>
<li>&nbsp;&nbsp;</li>
<li><a href="https://github.com/msioda/BioLockJ/wiki" target="_top"><strong>RETURN to WIKI</strong></a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="index-15.html">Prev Letter</a></li>
<li><a href="index-17.html">Next Letter</a></li>
</ul>
<ul class="navList">
<li><a href="../index.html?index-files/index-16.html" target="_top">Frames</a></li>
<li><a href="index-16.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip.navbar.top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<div class="contentContainer"><a href="index-1.html">A</a>&nbsp;<a href="index-2.html">B</a>&nbsp;<a href="index-3.html">C</a>&nbsp;<a href="index-4.html">D</a>&nbsp;<a href="index-5.html">E</a>&nbsp;<a href="index-6.html">F</a>&nbsp;<a href="index-7.html">G</a>&nbsp;<a href="index-8.html">H</a>&nbsp;<a href="index-9.html">I</a>&nbsp;<a href="index-10.html">J</a>&nbsp;<a href="index-11.html">K</a>&nbsp;<a href="index-12.html">L</a>&nbsp;<a href="index-13.html">M</a>&nbsp;<a href="index-14.html">N</a>&nbsp;<a href="index-15.html">O</a>&nbsp;<a href="index-16.html">P</a>&nbsp;<a href="index-17.html">Q</a>&nbsp;<a href="index-18.html">R</a>&nbsp;<a href="index-19.html">S</a>&nbsp;<a href="index-20.html">T</a>&nbsp;<a href="index-21.html">U</a>&nbsp;<a href="index-22.html">V</a>&nbsp;<a href="index-23.html">W</a>&nbsp;<a name="I:P">
<!--   -->
</a>
<h2 class="title">P</h2>
<dl>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#PARAMS">PARAMS</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> property suffix for exe.* properties, used to set optional parameters: "Params"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/TaxaUtil.html#parentTaxaLevel-java.lang.String-">parentTaxaLevel(String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/TaxaUtil.html" title="class in biolockj.util">TaxaUtil</a></dt>
<dd>
<div class="block">Get parent level name of the given level parameter.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BioLockJUtil.html#parseCountTable-java.io.File-">parseCountTable(File)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BioLockJUtil.html" title="class in biolockj.util">BioLockJUtil</a></dt>
<dd>
<div class="block">Read in BioLockJ count table, each inner lists represents 1 line from the file.<br>
 Each cell in the tab delimited file is stored as 1 element in the inner lists.</div>
</dd>
<dt><a href="../biolockj/node/ParsedSample.html" title="class in biolockj.node"><span class="typeNameLink">ParsedSample</span></a> - Class in <a href="../biolockj/node/package-summary.html">biolockj.node</a></dt>
<dd>
<div class="block"><a href="../biolockj/module/implicit/parser/ParserModule.html" title="interface in biolockj.module.implicit.parser"><code>ParserModule</code></a>s create and store one <a href="../biolockj/node/ParsedSample.html" title="class in biolockj.node"><code>ParsedSample</code></a>, with
 OTU assignments and counts, for each sample.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/ParsedSample.html#ParsedSample-biolockj.node.OtuNode-">ParsedSample(OtuNode)</a></span> - Constructor for class biolockj.node.<a href="../biolockj/node/ParsedSample.html" title="class in biolockj.node">ParsedSample</a></dt>
<dd>
<div class="block">Construct a new ParsedSample with it's 1st OtuNode.</div>
</dd>
<dt><a href="../biolockj/module/implicit/parser/ParserModule.html" title="interface in biolockj.module.implicit.parser"><span class="typeNameLink">ParserModule</span></a> - Interface in <a href="../biolockj/module/implicit/parser/package-summary.html">biolockj.module.implicit.parser</a></dt>
<dd>
<div class="block">This interface defines the required methods to parse ClassifierModule output.</div>
</dd>
<dt><a href="../biolockj/module/implicit/parser/ParserModuleImpl.html" title="class in biolockj.module.implicit.parser"><span class="typeNameLink">ParserModuleImpl</span></a> - Class in <a href="../biolockj/module/implicit/parser/package-summary.html">biolockj.module.implicit.parser</a></dt>
<dd>
<div class="block">Parser <a href="../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s read <a href="../biolockj/module/classifier/ClassifierModule.html" title="interface in biolockj.module.classifier"><code>ClassifierModule</code></a> output to build
 standardized OTU count tables.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/ParserModuleImpl.html#ParserModuleImpl--">ParserModuleImpl()</a></span> - Constructor for class biolockj.module.implicit.parser.<a href="../biolockj/module/implicit/parser/ParserModuleImpl.html" title="class in biolockj.module.implicit.parser">ParserModuleImpl</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/ImportMetadata.html#parseRow-java.lang.String-boolean-">parseRow(String, boolean)</a></span> - Method in class biolockj.module.implicit.<a href="../biolockj/module/implicit/ImportMetadata.html" title="class in biolockj.module.implicit">ImportMetadata</a></dt>
<dd>
<div class="block">Method called to parse a row from the metadata file, where
 <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/util/MetaUtil.html#META_COLUMN_DELIM">"metadata.columnDelim"</a> separates columns.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/ParserModule.html#parseSamples--">parseSamples()</a></span> - Method in interface biolockj.module.implicit.parser.<a href="../biolockj/module/implicit/parser/ParserModule.html" title="interface in biolockj.module.implicit.parser">ParserModule</a></dt>
<dd>
<div class="block">Parse <a href="../biolockj/module/classifier/ClassifierModule.html" title="interface in biolockj.module.classifier"><code>ClassifierModule</code></a> output to build <a href="../biolockj/node/ParsedSample.html" title="class in biolockj.node"><code>ParsedSample</code></a>s.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/ParserModuleImpl.html#parseSamples--">parseSamples()</a></span> - Method in class biolockj.module.implicit.parser.<a href="../biolockj/module/implicit/parser/ParserModuleImpl.html" title="class in biolockj.module.implicit.parser">ParserModuleImpl</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/r16s/QiimeParser.html#parseSamples--">parseSamples()</a></span> - Method in class biolockj.module.implicit.parser.r16s.<a href="../biolockj/module/implicit/parser/r16s/QiimeParser.html" title="class in biolockj.module.implicit.parser.r16s">QiimeParser</a></dt>
<dd>
<div class="block">Parse lowest level <a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime"><code>QiimeClassifier</code></a> report in the input directory.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/r16s/RdpParser.html#parseSamples--">parseSamples()</a></span> - Method in class biolockj.module.implicit.parser.r16s.<a href="../biolockj/module/implicit/parser/r16s/RdpParser.html" title="class in biolockj.module.implicit.parser.r16s">RdpParser</a></dt>
<dd>
<div class="block">Parse all <a href="../biolockj/module/classifier/r16s/RdpClassifier.html" title="class in biolockj.module.classifier.r16s"><code>RdpClassifier</code></a> reports in the input directory.<br>
 Build an <a href="../biolockj/node/r16s/RdpNode.html" title="class in biolockj.node.r16s"><code>RdpNode</code></a> for each line.<br>
 If <a href="../biolockj/module/implicit/parser/r16s/RdpParser.html#isValid-biolockj.node.OtuNode-"><code>RdpParser.isValid(OtuNode)</code></a>,<br>
 
 Create <a href="../biolockj/node/ParsedSample.html" title="class in biolockj.node"><code>ParsedSample</code></a> for the <a href="../biolockj/node/OtuNodeImpl.html#getSampleId--"><code>OtuNodeImpl.getSampleId()</code></a> if not yet
 created.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/wgs/Humann2Parser.html#parseSamples--">parseSamples()</a></span> - Method in class biolockj.module.implicit.parser.wgs.<a href="../biolockj/module/implicit/parser/wgs/Humann2Parser.html" title="class in biolockj.module.implicit.parser.wgs">Humann2Parser</a></dt>
<dd>
<div class="block">To parse the taxonomy level reports output by <a href="../biolockj/module/classifier/wgs/Humann2Classifier.html" title="class in biolockj.module.classifier.wgs"><code>Humann2Classifier</code></a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/wgs/Kraken2Parser.html#parseSamples--">parseSamples()</a></span> - Method in class biolockj.module.implicit.parser.wgs.<a href="../biolockj/module/implicit/parser/wgs/Kraken2Parser.html" title="class in biolockj.module.implicit.parser.wgs">Kraken2Parser</a></dt>
<dd>
<div class="block">Parse all <a href="../biolockj/module/classifier/wgs/Kraken2Classifier.html" title="class in biolockj.module.classifier.wgs"><code>Kraken2Classifier</code></a> reports in the input directory.<br>
 Build an <a href="../biolockj/node/wgs/Kraken2Node.html" title="class in biolockj.node.wgs"><code>Kraken2Node</code></a> for each line.<br>
 If <a href="../biolockj/module/implicit/parser/ParserModuleImpl.html#isValid-biolockj.node.OtuNode-"><code>ParserModuleImpl.isValid(OtuNode)</code></a>: <br>
 
 Create <a href="../biolockj/node/ParsedSample.html" title="class in biolockj.node"><code>ParsedSample</code></a> for the <a href="../biolockj/node/OtuNodeImpl.html#getSampleId--"><code>OtuNodeImpl.getSampleId()</code></a> if not
 yet created.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/wgs/KrakenParser.html#parseSamples--">parseSamples()</a></span> - Method in class biolockj.module.implicit.parser.wgs.<a href="../biolockj/module/implicit/parser/wgs/KrakenParser.html" title="class in biolockj.module.implicit.parser.wgs">KrakenParser</a></dt>
<dd>
<div class="block">Parse all <a href="../biolockj/module/classifier/wgs/KrakenClassifier.html" title="class in biolockj.module.classifier.wgs"><code>KrakenClassifier</code></a> reports in the input directory.<br>
 Build an <a href="../biolockj/node/wgs/KrakenNode.html" title="class in biolockj.node.wgs"><code>KrakenNode</code></a> for each line.<br>
 If <a href="../biolockj/module/implicit/parser/ParserModuleImpl.html#isValid-biolockj.node.OtuNode-"><code>ParserModuleImpl.isValid(OtuNode)</code></a>,<br>
 
 Create <a href="../biolockj/node/ParsedSample.html" title="class in biolockj.node"><code>ParsedSample</code></a> for the <a href="../biolockj/node/OtuNodeImpl.html#getSampleId--"><code>OtuNodeImpl.getSampleId()</code></a> if not
 yet created.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/wgs/Metaphlan2Parser.html#parseSamples--">parseSamples()</a></span> - Method in class biolockj.module.implicit.parser.wgs.<a href="../biolockj/module/implicit/parser/wgs/Metaphlan2Parser.html" title="class in biolockj.module.implicit.parser.wgs">Metaphlan2Parser</a></dt>
<dd>
<div class="block">To parse the taxonomy level reports output by <a href="../biolockj/module/classifier/wgs/Metaphlan2Classifier.html" title="class in biolockj.module.classifier.wgs"><code>Metaphlan2Classifier</code></a>:
 
 Create <a href="../biolockj/node/ParsedSample.html" title="class in biolockj.node"><code>ParsedSample</code></a> for the <a href="../biolockj/node/OtuNodeImpl.html#getSampleId--"><code>OtuNodeImpl.getSampleId()</code></a> if
 not yet created.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/wgs/SlimmParser.html#parseSamples--">parseSamples()</a></span> - Method in class biolockj.module.implicit.parser.wgs.<a href="../biolockj/module/implicit/parser/wgs/SlimmParser.html" title="class in biolockj.module.implicit.parser.wgs">SlimmParser</a></dt>
<dd>
<div class="block">To parse the taxonomy level reports output by <a href="../biolockj/module/classifier/wgs/SlimmClassifier.html" title="class in biolockj.module.classifier.wgs"><code>SlimmClassifier</code></a>:
 
 Create <a href="../biolockj/node/ParsedSample.html" title="class in biolockj.node"><code>ParsedSample</code></a> for the <a href="../biolockj/node/OtuNodeImpl.html#getSampleId--"><code>OtuNodeImpl.getSampleId()</code></a> if not
 yet created.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#PASSWORD_FLAG">PASSWORD_FLAG</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">Change password runtime parameter switch: "-p"</div>
</dd>
<dt><a href="../biolockj/util/PathwayUtil.html" title="class in biolockj.util"><span class="typeNameLink">PathwayUtil</span></a> - Class in <a href="../biolockj/util/package-summary.html">biolockj.util</a></dt>
<dd>
<div class="block">This utility helps work HumanN2 pathway abundance data.<br></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#PDF_EXT">PDF_EXT</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block">BioLockJ PDF file extension constant: ".pdf"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModuleImpl.html#PDF_EXT">PDF_EXT</a></span> - Static variable in class biolockj.module.<a href="../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a></dt>
<dd>
<div class="block">BioLockJ PDF file extension constant: ".pdf"</div>
</dd>
<dt><a href="../biolockj/module/seq/PearMergeReads.html" title="class in biolockj.module.seq"><span class="typeNameLink">PearMergeReads</span></a> - Class in <a href="../biolockj/module/seq/package-summary.html">biolockj.module.seq</a></dt>
<dd>
<div class="block">This BioModule will merge forward and reverse fastq files using PEAR.<br>
 For more informations, see the online PEAR manual:
 <a href="https://sco.h-its.org/exelixis/web/software/pear/doc.html" target=
 "_top">https://sco.h-its.org/exelixis/web/software/pear/doc.html</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/PearMergeReads.html#PearMergeReads--">PearMergeReads()</a></span> - Constructor for class biolockj.module.seq.<a href="../biolockj/module/seq/PearMergeReads.html" title="class in biolockj.module.seq">PearMergeReads</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/TaxaUtil.html#PHYLUM">PHYLUM</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/TaxaUtil.html" title="class in biolockj.util">TaxaUtil</a></dt>
<dd>
<div class="block">One of the <a href="../biolockj/util/TaxaUtil.html#REPORT_TAXONOMY_LEVELS">"report.taxonomyLevels"</a> options: "phylum"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/OtuNodeImpl.html#PHYLUM_DELIM">PHYLUM_DELIM</a></span> - Static variable in class biolockj.node.<a href="../biolockj/node/OtuNodeImpl.html" title="class in biolockj.node">OtuNodeImpl</a></dt>
<dd>
<div class="block">Standard classifier output level delimiter for PHYLUM</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/r16s/QiimeClosedRefClassifier.html#PICK_OTU_SCRIPT">PICK_OTU_SCRIPT</a></span> - Static variable in class biolockj.module.classifier.r16s.<a href="../biolockj/module/classifier/r16s/QiimeClosedRefClassifier.html" title="class in biolockj.module.classifier.r16s">QiimeClosedRefClassifier</a></dt>
<dd>
<div class="block">Closed reference OTU picking script</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/r16s/QiimeDeNovoClassifier.html#PICK_OTU_SCRIPT">PICK_OTU_SCRIPT</a></span> - Static variable in class biolockj.module.classifier.r16s.<a href="../biolockj/module/classifier/r16s/QiimeDeNovoClassifier.html" title="class in biolockj.module.classifier.r16s">QiimeDeNovoClassifier</a></dt>
<dd>
<div class="block">De novo OTU picking script</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/r16s/QiimeOpenRefClassifier.html#PICK_OTU_SCRIPT">PICK_OTU_SCRIPT</a></span> - Static variable in class biolockj.module.classifier.r16s.<a href="../biolockj/module/classifier/r16s/QiimeOpenRefClassifier.html" title="class in biolockj.module.classifier.r16s">QiimeOpenRefClassifier</a></dt>
<dd>
<div class="block">Open reference OTU picking script</div>
</dd>
<dt><a href="../biolockj/Pipeline.html" title="class in biolockj"><span class="typeNameLink">Pipeline</span></a> - Class in <a href="../biolockj/package-summary.html">biolockj</a></dt>
<dd>
<div class="block">This class initializes and executes each <a href="../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a> configured for execution.<br>
 <a href="../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s that generate scripts are monitored until all scripts are complete, before moving
 on to the next module.<br></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BioLockJUtil.html#PIPELINE_HUMANN2_COUNT_TABLE_INPUT_TYPE">PIPELINE_HUMANN2_COUNT_TABLE_INPUT_TYPE</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/BioLockJUtil.html" title="class in biolockj.util">BioLockJUtil</a></dt>
<dd>
<div class="block">Pipeline input file type indicating the file is Humann2 generated</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BioLockJUtil.html#PIPELINE_OTU_COUNT_TABLE_INPUT_TYPE">PIPELINE_OTU_COUNT_TABLE_INPUT_TYPE</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/BioLockJUtil.html" title="class in biolockj.util">BioLockJUtil</a></dt>
<dd>
<div class="block">Internal <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> String property: "otu_count"<br>
 Set as the value of <a href="../biolockj/util/BioLockJUtil.html#INTERNAL_PIPELINE_INPUT_TYPES">"internal.pipelineInputTypes"</a> for OTU count files that meet the file requirements
 to pass <a href="../biolockj/util/OtuUtil.html#isOtuFile-java.io.File-"><code>OtuUtil.isOtuFile(File)</code></a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BioLockJUtil.html#PIPELINE_PARSER_INPUT_TYPE">PIPELINE_PARSER_INPUT_TYPE</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/BioLockJUtil.html" title="class in biolockj.util">BioLockJUtil</a></dt>
<dd>
<div class="block">Internal <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> String property: "classifier_output"<br>
 Set as the value of <a href="../biolockj/util/BioLockJUtil.html#INTERNAL_PIPELINE_INPUT_TYPES">"internal.pipelineInputTypes"</a> for classifier output files.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BioLockJUtil.html#PIPELINE_R_INPUT_TYPE">PIPELINE_R_INPUT_TYPE</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/BioLockJUtil.html" title="class in biolockj.util">BioLockJUtil</a></dt>
<dd>
<div class="block">Internal <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> String property: "R"<br>
 Set as the value of <a href="../biolockj/util/BioLockJUtil.html#INTERNAL_PIPELINE_INPUT_TYPES">"internal.pipelineInputTypes"</a> if input files are some type of count table merged
 with the metadata such as those output by <code>AddMetadataToTaxaTables</code>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BioLockJUtil.html#PIPELINE_SEQ_INPUT_TYPE">PIPELINE_SEQ_INPUT_TYPE</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/BioLockJUtil.html" title="class in biolockj.util">BioLockJUtil</a></dt>
<dd>
<div class="block">Internal <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> String property: "seq"<br>
 Set as the value of <a href="../biolockj/util/BioLockJUtil.html#INTERNAL_PIPELINE_INPUT_TYPES">"internal.pipelineInputTypes"</a> for sequence input files.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BioLockJUtil.html#PIPELINE_STATS_TABLE_INPUT_TYPE">PIPELINE_STATS_TABLE_INPUT_TYPE</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/BioLockJUtil.html" title="class in biolockj.util">BioLockJUtil</a></dt>
<dd>
<div class="block">Internal <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> String property: "stats"<br>
 Set as the value of <a href="../biolockj/util/BioLockJUtil.html#INTERNAL_PIPELINE_INPUT_TYPES">"internal.pipelineInputTypes"</a> if input files are tables of statistics such as those
 output by <code>R_CalculateStats</code>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BioLockJUtil.html#PIPELINE_TAXA_COUNT_TABLE_INPUT_TYPE">PIPELINE_TAXA_COUNT_TABLE_INPUT_TYPE</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/BioLockJUtil.html" title="class in biolockj.util">BioLockJUtil</a></dt>
<dd>
<div class="block">Internal <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> String property: "taxa_count"<br>
 Set as the value of <a href="../biolockj/util/BioLockJUtil.html#INTERNAL_PIPELINE_INPUT_TYPES">"internal.pipelineInputTypes"</a> for OTU count files that meet the file requirements
 to pass <a href="../biolockj/util/TaxaUtil.html#isTaxaFile-java.io.File-"><code>TaxaUtil.isTaxaFile(File)</code></a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/JsonReport.html#pipelineInputContainsOtuSummary--">pipelineInputContainsOtuSummary()</a></span> - Method in class biolockj.module.report.<a href="../biolockj/module/report/JsonReport.html" title="class in biolockj.module.report">JsonReport</a></dt>
<dd>
<div class="block">Check pipeline input to see if OTU summary file is the only pipeline input file.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BioLockJUtil.html#pipelineInputType-java.lang.String-">pipelineInputType(String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BioLockJUtil.html" title="class in biolockj.util">BioLockJUtil</a></dt>
<dd>
<div class="block">Convenience method to check pipeline input file type.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#pipelineName--">pipelineName()</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Get the current pipeline name (root folder name)</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#pipelinePath--">pipelinePath()</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Get the current pipeline absolute directory path (root folder path)</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#piplineHasSeqInput--">piplineHasSeqInput()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Return TRUE if pipeline input files are sequence files.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Pipeline.html#poll-biolockj.module.ScriptModule-">poll(ScriptModule)</a></span> - Static method in class biolockj.<a href="../biolockj/Pipeline.html" title="class in biolockj">Pipeline</a></dt>
<dd>
<div class="block">The <a href="../biolockj/module/ScriptModule.html#getScriptDir--"><code>ScriptModule.getScriptDir()</code></a> will contain one main script and one ore more worker
 scripts.<br>
 An empty file with <a href="../biolockj/Pipeline.html#SCRIPT_STARTED">"Started"</a> appended to the script name is created when execution begins.<br>
 If successful, an empty file with <a href="../biolockj/Pipeline.html#SCRIPT_SUCCESS">"Success"</a> appended to the script name is created.<br>
 Upon failure, an empty file with <a href="../biolockj/Pipeline.html#SCRIPT_FAILURES">"Failures"</a> appended to the script name is created.<br>
 Script status is polled each minute, determining status by counting indicator files.<br>
 <a href="../biolockj/Log.html" title="class in biolockj"><code>Log</code></a> outputs the # of started, failed, and successful scripts (if any change).<br>
 <a href="../biolockj/Log.html" title="class in biolockj"><code>Log</code></a> repeats the previous message every 10 minutes if no status change is detected.<br></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/OtuNodeImpl.html#populateInBetweenTaxa--">populateInBetweenTaxa()</a></span> - Method in class biolockj.node.<a href="../biolockj/node/OtuNodeImpl.html" title="class in biolockj.node">OtuNodeImpl</a></dt>
<dd>
<div class="block">Populate missing OTUs if top level taxa is defined and there is a level gap between the top level and the bottom
 level.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#PRIMER_DIR_FLAG">PRIMER_DIR_FLAG</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">Primer file directory path runtime parameter switch: "-t"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BioLockJUtil.html#printLongFormList-java.util.Collection-">printLongFormList(Collection&lt;?&gt;)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BioLockJUtil.html" title="class in biolockj.util">BioLockJUtil</a></dt>
<dd>
<div class="block">Print collection one item per line.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#printRuntimeArgs-java.lang.String:A-boolean-">printRuntimeArgs(String[], boolean)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">Print the runtime args to System.out or using the Logger based on useSysOut parameter.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#PROCESSED">PROCESSED</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block">File suffix appended to processed samples in the module output directory: "_reported.tsv"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#PROJECT_COPY_FILES">PROJECT_COPY_FILES</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> property set to copy input files into pipeline root directory:
 "project.copyInput"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#PROJECT_DEFAULT_PROPS">PROJECT_DEFAULT_PROPS</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> String property: "project.defaultProps"<br>
 Set file path of default property file.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#PROJECT_DELETE_TEMP_FILES">PROJECT_DELETE_TEMP_FILES</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> property set to delete <a href="../biolockj/module/BioModule.html#getTempDir--"><code>BioModule.getTempDir()</code></a> files:
 "project.deleteTempFiles"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#PROJECT_ENV">PROJECT_ENV</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> Boolean property: "project.env"<br>
 Options: <a href="../biolockj/Constants.html#PROJECT_ENV_CLUSTER">"cluster"</a>, <a href="../biolockj/Constants.html#PROJECT_ENV_AWS">"aws"</a>, <a href="../biolockj/Constants.html#PROJECT_ENV_LOCAL">"local"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#PROJECT_ENV_AWS">PROJECT_ENV_AWS</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> option for property: <a href="../biolockj/Constants.html#PROJECT_ENV">"project.env"</a><br>
 Used to indicate running as an Amazon web service: "aws"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#PROJECT_ENV_CLUSTER">PROJECT_ENV_CLUSTER</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> option for property: <a href="../biolockj/Constants.html#PROJECT_ENV">"project.env"</a><br>
 Used to indicate running on the cluster: "cluster"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#PROJECT_ENV_LOCAL">PROJECT_ENV_LOCAL</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> option for property: <a href="../biolockj/Constants.html#PROJECT_ENV">"project.env"</a><br>
 Used to indicate running on a local machine (laptop, etc): "local"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#PROJECT_PERMISSIONS">PROJECT_PERMISSIONS</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> property to define permission settings when running chmod on pipeline root dir:
 "project.permissions"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#PROJECT_PIPELINE_DIR">PROJECT_PIPELINE_DIR</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> String property: "project.pipelineDir"<br>
 Stores the path of the pipeline root directory path set by the application runtime code.</div>
</dd>
<dt><a href="../biolockj/Properties.html" title="class in biolockj"><span class="typeNameLink">Properties</span></a> - Class in <a href="../biolockj/package-summary.html">biolockj</a></dt>
<dd>
<div class="block">Load properties defined in the BioLockJ configuration file, including inherited properties from project.defaultProps</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Properties.html#Properties--">Properties()</a></span> - Constructor for class biolockj.<a href="../biolockj/Properties.html" title="class in biolockj">Properties</a></dt>
<dd>
<div class="block">Default constructor.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Properties.html#Properties-biolockj.Properties-">Properties(Properties)</a></span> - Constructor for class biolockj.<a href="../biolockj/Properties.html" title="class in biolockj">Properties</a></dt>
<dd>
<div class="block">Constructor called when <a href="../biolockj/Constants.html#PROJECT_DEFAULT_PROPS">"project.defaultProps"</a> contains a valid file-path</div>
</dd>
<dt><a href="../biolockj/util/PropUtil.html" title="class in biolockj.util"><span class="typeNameLink">PropUtil</span></a> - Class in <a href="../biolockj/util/package-summary.html">biolockj.util</a></dt>
<dd>
<div class="block">Simple utility containing String manipulation and formatting functions.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/PropUtil.html#PropUtil--">PropUtil()</a></span> - Constructor for class biolockj.util.<a href="../biolockj/util/PropUtil.html" title="class in biolockj.util">PropUtil</a></dt>
<dd>&nbsp;</dd>
</dl>
<a href="index-1.html">A</a>&nbsp;<a href="index-2.html">B</a>&nbsp;<a href="index-3.html">C</a>&nbsp;<a href="index-4.html">D</a>&nbsp;<a href="index-5.html">E</a>&nbsp;<a href="index-6.html">F</a>&nbsp;<a href="index-7.html">G</a>&nbsp;<a href="index-8.html">H</a>&nbsp;<a href="index-9.html">I</a>&nbsp;<a href="index-10.html">J</a>&nbsp;<a href="index-11.html">K</a>&nbsp;<a href="index-12.html">L</a>&nbsp;<a href="index-13.html">M</a>&nbsp;<a href="index-14.html">N</a>&nbsp;<a href="index-15.html">O</a>&nbsp;<a href="index-16.html">P</a>&nbsp;<a href="index-17.html">Q</a>&nbsp;<a href="index-18.html">R</a>&nbsp;<a href="index-19.html">S</a>&nbsp;<a href="index-20.html">T</a>&nbsp;<a href="index-21.html">U</a>&nbsp;<a href="index-22.html">V</a>&nbsp;<a href="index-23.html">W</a>&nbsp;</div>
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar.bottom">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.bottom.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../overview-summary.html">Overview</a></li>
<li>Package</li>
<li>Class</li>
<li>Use</li>
<li><a href="../overview-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li class="navBarCell1Rev">Index</li>
<li><a href="../help-doc.html">Help</a></li>
<li>&nbsp;&nbsp;</li>
<li><a href="https://github.com/msioda/BioLockJ/wiki" target="_top"><strong>RETURN to WIKI</strong></a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="index-15.html">Prev Letter</a></li>
<li><a href="index-17.html">Next Letter</a></li>
</ul>
<ul class="navList">
<li><a href="../index.html?index-files/index-16.html" target="_top">Frames</a></li>
<li><a href="index-16.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip.navbar.bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
