<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (1.8.0_144) on Tue Mar 12 06:27:48 EDT 2019 -->
<title>G-Index</title>
<meta name="date" content="2019-03-12">
<link rel="stylesheet" type="text/css" href="../stylesheet.css" title="Style">
<script type="text/javascript" src="../script.js"></script>
</head>
<body>
<script type="text/javascript"><!--
    try {
        if (location.href.indexOf('is-external=true') == -1) {
            parent.document.title="G-Index";
        }
    }
    catch(err) {
    }
//-->
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar.top">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.top.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../overview-summary.html">Overview</a></li>
<li>Package</li>
<li>Class</li>
<li>Use</li>
<li><a href="../overview-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li class="navBarCell1Rev">Index</li>
<li><a href="../help-doc.html">Help</a></li>
<li>&nbsp;&nbsp;</li>
<li><a href="https://github.com/msioda/BioLockJ/wiki" target="_top"><strong>RETURN to WIKI</strong></a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="index-6.html">Prev Letter</a></li>
<li><a href="index-8.html">Next Letter</a></li>
</ul>
<ul class="navList">
<li><a href="../index.html?index-files/index-7.html" target="_top">Frames</a></li>
<li><a href="index-7.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip.navbar.top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<div class="contentContainer"><a href="index-1.html">A</a>&nbsp;<a href="index-2.html">B</a>&nbsp;<a href="index-3.html">C</a>&nbsp;<a href="index-4.html">D</a>&nbsp;<a href="index-5.html">E</a>&nbsp;<a href="index-6.html">F</a>&nbsp;<a href="index-7.html">G</a>&nbsp;<a href="index-8.html">H</a>&nbsp;<a href="index-9.html">I</a>&nbsp;<a href="index-10.html">J</a>&nbsp;<a href="index-11.html">K</a>&nbsp;<a href="index-12.html">L</a>&nbsp;<a href="index-13.html">M</a>&nbsp;<a href="index-14.html">N</a>&nbsp;<a href="index-15.html">O</a>&nbsp;<a href="index-16.html">P</a>&nbsp;<a href="index-17.html">Q</a>&nbsp;<a href="index-18.html">R</a>&nbsp;<a href="index-19.html">S</a>&nbsp;<a href="index-20.html">T</a>&nbsp;<a href="index-21.html">U</a>&nbsp;<a href="index-22.html">V</a>&nbsp;<a href="index-23.html">W</a>&nbsp;<a name="I:G">
<!--   -->
</a>
<h2 class="title">G</h2>
<dl>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/humann2/AddMetadataToPathwayTables.html#generateMergedTables--">generateMergedTables()</a></span> - Method in class biolockj.module.report.humann2.<a href="../biolockj/module/report/humann2/AddMetadataToPathwayTables.html" title="class in biolockj.module.report.humann2">AddMetadataToPathwayTables</a></dt>
<dd>
<div class="block">Create the merged metadata tables.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/taxa/AddMetadataToTaxaTables.html#generateMergedTables--">generateMergedTables()</a></span> - Method in class biolockj.module.report.taxa.<a href="../biolockj/module/report/taxa/AddMetadataToTaxaTables.html" title="class in biolockj.module.report.taxa">AddMetadataToTaxaTables</a></dt>
<dd>
<div class="block">Create the merged metadata tables.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/otu/RarefyOtuCounts.html#generateOtuput-java.io.File-java.util.TreeMap-">generateOtuput(File, TreeMap&lt;String, Integer&gt;)</a></span> - Method in class biolockj.module.report.otu.<a href="../biolockj/module/report/otu/RarefyOtuCounts.html" title="class in biolockj.module.report.otu">RarefyOtuCounts</a></dt>
<dd>
<div class="block">Print the output file wit rarefied counts.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#GENUS">GENUS</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> option for <a href="../biolockj/Constants.html#REPORT_TAXONOMY_LEVELS">"report.taxonomyLevels"</a>: "genus"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/OtuNodeImpl.html#GENUS_DELIM">GENUS_DELIM</a></span> - Static variable in class biolockj.node.<a href="../biolockj/node/OtuNodeImpl.html" title="class in biolockj.node">OtuNodeImpl</a></dt>
<dd>
<div class="block">Standard classifier output level delimiter for GENUS</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#getAdminEmailPassword--">getAdminEmailPassword()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">Runtime property getter for <a href="../biolockj/util/RuntimeParamUtil.html#PASSWORD_FLAG">"-p"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#getBaseDir--">getBaseDir()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">Runtime property getter for <a href="../biolockj/util/RuntimeParamUtil.html#BASE_DIR_FLAG">"-b"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#getBaseDirParam--">getBaseDirParam()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/r16s/QiimeClosedRefClassifier.html#getBatch-java.util.List-int-int-">getBatch(List&lt;String&gt;, int, int)</a></span> - Method in class biolockj.module.classifier.r16s.<a href="../biolockj/module/classifier/r16s/QiimeClosedRefClassifier.html" title="class in biolockj.module.classifier.r16s">QiimeClosedRefClassifier</a></dt>
<dd>
<div class="block">Method returns batch script lines for one batch.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/r16s/QiimeClosedRefClassifier.html#getBatchDir-int-">getBatchDir(int)</a></span> - Method in class biolockj.module.classifier.r16s.<a href="../biolockj/module/classifier/r16s/QiimeClosedRefClassifier.html" title="class in biolockj.module.classifier.r16s">QiimeClosedRefClassifier</a></dt>
<dd>
<div class="block">Get the pipeline/module/temp/batch_# dir.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/r16s/QiimeClosedRefClassifier.html#getBatchFastaDir-int-">getBatchFastaDir(int)</a></span> - Method in class biolockj.module.classifier.r16s.<a href="../biolockj/module/classifier/r16s/QiimeClosedRefClassifier.html" title="class in biolockj.module.classifier.r16s">QiimeClosedRefClassifier</a></dt>
<dd>
<div class="block">Get the pipeline/module/temp/batch_#/fasta directory.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RMetaUtil.html#getBinaryFields-biolockj.module.BioModule-">getBinaryFields(BioModule)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RMetaUtil.html" title="class in biolockj.util">RMetaUtil</a></dt>
<dd>
<div class="block">Get the <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/util/RMetaUtil.html#BINARY_FIELDS">"R_internal.binaryFields"</a> fields containing only 2 non-numeric values.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BioLockJUtil.html#getBljDir--">getBljDir()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BioLockJUtil.html" title="class in biolockj.util">BioLockJUtil</a></dt>
<dd>
<div class="block">Get the program source (either the jar path or main class biolockj.BioLockJ);</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getBoolean-biolockj.module.BioModule-java.lang.String-">getBoolean(BioModule, String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Parse property value (Y or N) to return boolean, if not found, return false;</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/ModuleUtil.html#getClassifier-biolockj.module.BioModule-boolean-">getClassifier(BioModule, boolean)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/ModuleUtil.html" title="class in biolockj.util">ModuleUtil</a></dt>
<dd>
<div class="block">Get a classifier module<br>
 Use checkAhead parameter to determine if we look forward or backwards starting from the given module.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/ClassifierModule.html#getClassifierExe--">getClassifierExe()</a></span> - Method in interface biolockj.module.classifier.<a href="../biolockj/module/classifier/ClassifierModule.html" title="interface in biolockj.module.classifier">ClassifierModule</a></dt>
<dd>
<div class="block">Get the executable required to classify your samples.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/ClassifierModuleImpl.html#getClassifierExe--">getClassifierExe()</a></span> - Method in class biolockj.module.classifier.<a href="../biolockj/module/classifier/ClassifierModuleImpl.html" title="class in biolockj.module.classifier">ClassifierModuleImpl</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/r16s/RdpClassifier.html#getClassifierExe--">getClassifierExe()</a></span> - Method in class biolockj.module.classifier.r16s.<a href="../biolockj/module/classifier/r16s/RdpClassifier.html" title="class in biolockj.module.classifier.r16s">RdpClassifier</a></dt>
<dd>
<div class="block">RDP uses java to run a JAR file, so no special command is required</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/Humann2Classifier.html#getClassifierExe--">getClassifierExe()</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/Humann2Classifier.html" title="class in biolockj.module.classifier.wgs">Humann2Classifier</a></dt>
<dd>
<div class="block">Get kraken executable command: <a href="../biolockj/module/classifier/wgs/Humann2Classifier.html#EXE_HUMANN2">"exe.humann2"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/Kraken2Classifier.html#getClassifierExe--">getClassifierExe()</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/Kraken2Classifier.html" title="class in biolockj.module.classifier.wgs">Kraken2Classifier</a></dt>
<dd>
<div class="block">Get kraken executable command: <a href="../biolockj/module/classifier/wgs/Kraken2Classifier.html#EXE_KRAKEN2">"exe.kraken2"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/KrakenClassifier.html#getClassifierExe--">getClassifierExe()</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/KrakenClassifier.html" title="class in biolockj.module.classifier.wgs">KrakenClassifier</a></dt>
<dd>
<div class="block">Get kraken executable command: <a href="../biolockj/module/classifier/wgs/KrakenClassifier.html#EXE_KRAKEN">"exe.kraken"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/Metaphlan2Classifier.html#getClassifierExe--">getClassifierExe()</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/Metaphlan2Classifier.html" title="class in biolockj.module.classifier.wgs">Metaphlan2Classifier</a></dt>
<dd>
<div class="block">Metaphlan runs python scripts, so no special command is required</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#getClassifierExe--">getClassifierExe()</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime">QiimeClassifier</a></dt>
<dd>
<div class="block">QIIME calls python scripts, so no special command is required</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/ClassifierModule.html#getClassifierParams--">getClassifierParams()</a></span> - Method in interface biolockj.module.classifier.<a href="../biolockj/module/classifier/ClassifierModule.html" title="interface in biolockj.module.classifier">ClassifierModule</a></dt>
<dd>
<div class="block">Get optional list of parameters to append whenever the classifier executable is called.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/ClassifierModuleImpl.html#getClassifierParams--">getClassifierParams()</a></span> - Method in class biolockj.module.classifier.<a href="../biolockj/module/classifier/ClassifierModuleImpl.html" title="class in biolockj.module.classifier">ClassifierModuleImpl</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/r16s/RdpClassifier.html#getClassifierParams--">getClassifierParams()</a></span> - Method in class biolockj.module.classifier.r16s.<a href="../biolockj/module/classifier/r16s/RdpClassifier.html" title="class in biolockj.module.classifier.r16s">RdpClassifier</a></dt>
<dd>
<div class="block">Do not accept -t to define a database, since that instead requires the specific property: <a href="../biolockj/module/classifier/r16s/RdpClassifier.html#RDP_DB">"rdp.db"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/Humann2Classifier.html#getClassifierParams--">getClassifierParams()</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/Humann2Classifier.html" title="class in biolockj.module.classifier.wgs">Humann2Classifier</a></dt>
<dd>
<div class="block">Obtain the humann2 runtime params</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/Kraken2Classifier.html#getClassifierParams--">getClassifierParams()</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/Kraken2Classifier.html" title="class in biolockj.module.classifier.wgs">Kraken2Classifier</a></dt>
<dd>
<div class="block">Obtain the kraken2 runtime params</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/KrakenClassifier.html#getClassifierParams--">getClassifierParams()</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/KrakenClassifier.html" title="class in biolockj.module.classifier.wgs">KrakenClassifier</a></dt>
<dd>
<div class="block">Obtain the kraken runtime params</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/Metaphlan2Classifier.html#getClassifierParams--">getClassifierParams()</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/Metaphlan2Classifier.html" title="class in biolockj.module.classifier.wgs">Metaphlan2Classifier</a></dt>
<dd>
<div class="block">Obtain the metaphlan2 runtime params</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#getClassifierParams--">getClassifierParams()</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime">QiimeClassifier</a></dt>
<dd>
<div class="block">Obtain the QIIME runtime params</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/ClassifierModuleImpl.html#getClassifierType--">getClassifierType()</a></span> - Method in class biolockj.module.classifier.<a href="../biolockj/module/classifier/ClassifierModuleImpl.html" title="class in biolockj.module.classifier">ClassifierModuleImpl</a></dt>
<dd>
<div class="block">This method returns the classifier class name in lower case, after "classifier" is removed.<br>
 The remaining text should uniquely identify the name of the program.<br>
 The basic deployment will return one of: (rdp, qiime, kraken, metaphlan2, or humann2).<br></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BioLockJUtil.html#getClassNames-java.util.Collection-">getClassNames(Collection&lt;?&gt;)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BioLockJUtil.html" title="class in biolockj.util">BioLockJUtil</a></dt>
<dd>
<div class="block">Return an ordered list of the class names from the input collection.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DownloadUtil.html#getClusterUser--">getClusterUser()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/DownloadUtil.html" title="class in biolockj.util">DownloadUtil</a></dt>
<dd>
<div class="block">If System.user.home is a file path (like /usr/home/johnDoe), return last directory name (johnDoe).</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BioLockJUtil.html#getCollectionAsString-java.util.Collection-">getCollectionAsString(Collection&lt;?&gt;)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BioLockJUtil.html" title="class in biolockj.util">BioLockJUtil</a></dt>
<dd>
<div class="block">Concatenate data and return as a comma separated String.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BioLockJUtil.html#getCommonParent-java.io.File-java.io.File-">getCommonParent(File, File)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BioLockJUtil.html" title="class in biolockj.util">BioLockJUtil</a></dt>
<dd>
<div class="block">Compare files to return a common parent directory (if any)</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#getConfigFile--">getConfigFile()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">Runtime property getter for <a href="../biolockj/util/RuntimeParamUtil.html#CONFIG_FLAG">"-c"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getConfigFileExt--">getConfigFileExt()</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Gets the configuration file extension (often ".properties")</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getConfigFileName--">getConfigFileName()</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Gets the configuration file name</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#getConfigFileParam--">getConfigFileParam()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getConfigFilePath--">getConfigFilePath()</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Gets the full Config file path passed to BioLockJ as a runtime parameter.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/JsonNode.html#getCount--">getCount()</a></span> - Method in class biolockj.node.<a href="../biolockj/node/JsonNode.html" title="class in biolockj.node">JsonNode</a></dt>
<dd>
<div class="block">Getter method for count</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/OtuNode.html#getCount--">getCount()</a></span> - Method in interface biolockj.node.<a href="../biolockj/node/OtuNode.html" title="interface in biolockj.node">OtuNode</a></dt>
<dd>
<div class="block">Gets the OTU count.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/OtuNodeImpl.html#getCount--">getCount()</a></span> - Method in class biolockj.node.<a href="../biolockj/node/OtuNodeImpl.html" title="class in biolockj.node">OtuNodeImpl</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/OtuUtil.OtuCountLine.html#getCount--">getCount()</a></span> - Method in class biolockj.util.<a href="../biolockj/util/OtuUtil.OtuCountLine.html" title="class in biolockj.util">OtuUtil.OtuCountLine</a></dt>
<dd>
<div class="block">Get OTU count</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SummaryUtil.html#getCountSummary-java.util.Map-java.lang.String-boolean-">getCountSummary(Map&lt;String, String&gt;, String, boolean)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SummaryUtil.html" title="class in biolockj.util">SummaryUtil</a></dt>
<dd>
<div class="block">Return the min/max/mean/median summary stats for the given metadata numeric column</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/otu/RarefyOtuCounts.html#getData-java.lang.String-java.util.TreeMap-">getData(String, TreeMap&lt;String, Integer&gt;)</a></span> - Method in class biolockj.module.report.otu.<a href="../biolockj/module/report/otu/RarefyOtuCounts.html" title="class in biolockj.module.report.otu">RarefyOtuCounts</a></dt>
<dd>
<div class="block">Get OTU count data for the given sampleId.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/ClassifierModuleImpl.html#getDB--">getDB()</a></span> - Method in class biolockj.module.classifier.<a href="../biolockj/module/classifier/ClassifierModuleImpl.html" title="class in biolockj.module.classifier">ClassifierModuleImpl</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/r16s/RdpClassifier.html#getDB--">getDB()</a></span> - Method in class biolockj.module.classifier.r16s.<a href="../biolockj/module/classifier/r16s/RdpClassifier.html" title="class in biolockj.module.classifier.r16s">RdpClassifier</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/Humann2Classifier.html#getDB--">getDB()</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/Humann2Classifier.html" title="class in biolockj.module.classifier.wgs">Humann2Classifier</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/Kraken2Classifier.html#getDB--">getDB()</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/Kraken2Classifier.html" title="class in biolockj.module.classifier.wgs">Kraken2Classifier</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/KrakenClassifier.html#getDB--">getDB()</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/KrakenClassifier.html" title="class in biolockj.module.classifier.wgs">KrakenClassifier</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/Metaphlan2Classifier.html#getDB--">getDB()</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/Metaphlan2Classifier.html" title="class in biolockj.module.classifier.wgs">Metaphlan2Classifier</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/DatabaseModule.html#getDB--">getDB()</a></span> - Method in interface biolockj.module.<a href="../biolockj/module/DatabaseModule.html" title="interface in biolockj.module">DatabaseModule</a></dt>
<dd>
<div class="block">Return database directory, if multiple databases are configured, they must share a common parent directory
 and the common parent directory is returned by this method.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#getDB--">getDB()</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime">QiimeClassifier</a></dt>
<dd>
<div class="block">Check DB parameters for the comment parent directory path, there are 3 parameters:
 
 <a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#QIIME_PYNAST_ALIGN_DB">"qiime.pynastAlignDB"</a>
 <a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#QIIME_REF_SEQ_DB">"qiime.refSeqDB"</a>
 <a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#QIIME_TAXA_DB">"qiime.taxaDB"</a>
 </div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#getDB-java.lang.String-">getDB(String)</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime">QiimeClassifier</a></dt>
<dd>
<div class="block">Return the Docker container database directory (starting with /db/...)</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/KneadData.html#getDB--">getDB()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/KneadData.html" title="class in biolockj.module.seq">KneadData</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/KneadData.html#getDBs--">getDBs()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/KneadData.html" title="class in biolockj.module.seq">KneadData</a></dt>
<dd>
<div class="block">Get the configured database parameters</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Properties.html#getDefaultConfigProp-java.io.File-">getDefaultConfigProp(File)</a></span> - Static method in class biolockj.<a href="../biolockj/Properties.html" title="class in biolockj">Properties</a></dt>
<dd>
<div class="block">Parse property file for the property <a href="../biolockj/Constants.html#PIPELINE_DEFAULT_PROPS">"pipeline.defaultProps"</a>.<br>
 pipeline.defaultProps=$BLJ/resources/config/default/docker.properties</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/ModuleUtil.html#getDefaultDemultiplexer--">getDefaultDemultiplexer()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/ModuleUtil.html" title="class in biolockj.util">ModuleUtil</a></dt>
<dd>
<div class="block">Get the Java Class name for the default Demultiplexer module</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/ModuleUtil.html#getDefaultFastaConverter--">getDefaultFastaConverter()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/ModuleUtil.html" title="class in biolockj.util">ModuleUtil</a></dt>
<dd>
<div class="block">Get the Java Class name for the default Fasta converter module</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/ModuleUtil.html#getDefaultMergePairedReadsConverter--">getDefaultMergePairedReadsConverter()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/ModuleUtil.html" title="class in biolockj.util">ModuleUtil</a></dt>
<dd>
<div class="block">Get the Java Class name for the default Merge paired read module</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/ModuleUtil.html#getDefaultStatsModule--">getDefaultStatsModule()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/ModuleUtil.html" title="class in biolockj.util">ModuleUtil</a></dt>
<dd>
<div class="block">Return the class name of the default R statistics BioModule</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/ParserModuleImpl.html#getDepricatedOtuCountFields--">getDepricatedOtuCountFields()</a></span> - Static method in class biolockj.module.implicit.parser.<a href="../biolockj/module/implicit/parser/ParserModuleImpl.html" title="class in biolockj.module.implicit.parser">ParserModuleImpl</a></dt>
<dd>
<div class="block">Getter for depricatedOtuCountFields</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/RegisterNumReads.html#getDepricatedReadFields--">getDepricatedReadFields()</a></span> - Static method in class biolockj.module.implicit.<a href="../biolockj/module/implicit/RegisterNumReads.html" title="class in biolockj.module.implicit">RegisterNumReads</a></dt>
<dd>
<div class="block">Getter for depricatedReadFields</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#getDirectModuleDir--">getDirectModuleDir()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">Runtime property getter for <a href="../biolockj/util/RuntimeParamUtil.html#DIRECT_FLAG">"-d"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#getDirectModuleParam-biolockj.module.BioModule-">getDirectModuleParam(BioModule)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">Direct module parameters contain 2 parts separated by a colon: (pipeline directory name):(module name)</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#getDirectPipelineDir--">getDirectPipelineDir()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">Runtime property getter for direct module pipeline directory</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DownloadUtil.html#getDirFilter-boolean-boolean-boolean-">getDirFilter(boolean, boolean, boolean)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/DownloadUtil.html" title="class in biolockj.util">DownloadUtil</a></dt>
<dd>
<div class="block">Get a directory name filter to include module sub-directories.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#getDockerHostBLJ--">getDockerHostBLJ()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">Runtime property getter for Docker host BioLockJ dir</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#getDockerHostConfigDir--">getDockerHostConfigDir()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">Runtime property getter for <a href="../biolockj/util/RuntimeParamUtil.html#CONFIG_DIR_FLAG">"-C"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#getDockerHostHomeUserDir--">getDockerHostHomeUserDir()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">Runtime property getter for Docker host $USER $HOME dir</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#getDockerHostInputDir--">getDockerHostInputDir()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">Runtime property getter for <a href="../biolockj/util/RuntimeParamUtil.html#INPUT_DIR_FLAG">"-i"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#getDockerHostMetaDir--">getDockerHostMetaDir()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">Runtime property getter for <a href="../biolockj/util/RuntimeParamUtil.html#META_DIR_FLAG">"-m"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#getDockerHostPipelineDir--">getDockerHostPipelineDir()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">Runtime property getter for Docker host pipeline dir</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DockerUtil.html#getDockerImage-biolockj.module.BioModule-">getDockerImage(BioModule)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/DockerUtil.html" title="class in biolockj.util">DockerUtil</a></dt>
<dd>
<div class="block">Return the name of the Docker image needed for the given module.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#getDockerJavaModuleParams--">getDockerJavaModuleParams()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">Get Docker runtime arguments passed to BioLockJ from dockblj script.<br>
 These are used to to populate BLJ_OPTIONS in Docker java_module scripts.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DockerUtil.html#getDockerUser-java.lang.String-">getDockerUser(String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/DockerUtil.html" title="class in biolockj.util">DockerUtil</a></dt>
<dd>
<div class="block">Return the Docker Hub user ID.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DockerUtil.html#getDockerVolumeFile-java.lang.String-java.lang.String-">getDockerVolumeFile(String, String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/DockerUtil.html" title="class in biolockj.util">DockerUtil</a></dt>
<dd>
<div class="block">Get mapped Docker system File from <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> property by replacing the host system path with the
 mapped container path.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getDoubleVal-biolockj.module.BioModule-java.lang.String-">getDoubleVal(BioModule, String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Parse property for numeric (double) value</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DownloadUtil.html#getDownloadCmd--">getDownloadCmd()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/DownloadUtil.html" title="class in biolockj.util">DownloadUtil</a></dt>
<dd>
<div class="block">If running on cluster, build command for user to download pipeline analysis.<br></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DownloadUtil.html#getDownloadDirPath--">getDownloadDirPath()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/DownloadUtil.html" title="class in biolockj.util">DownloadUtil</a></dt>
<dd>
<div class="block">Get validated <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/util/DownloadUtil.html#DOWNLOAD_DIR">"pipeline.downloadDir"</a> if running on cluster, otherwise return null</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DownloadUtil.html#getDownloadListFile--">getDownloadListFile()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/DownloadUtil.html" title="class in biolockj.util">DownloadUtil</a></dt>
<dd>
<div class="block">Get the download list file.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DownloadUtil.html#getDownloadModules--">getDownloadModules()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/DownloadUtil.html" title="class in biolockj.util">DownloadUtil</a></dt>
<dd>
<div class="block">Get the modules to download.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DownloadUtil.html#getDownloadSize-java.util.List-">getDownloadSize(List&lt;File&gt;)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/DownloadUtil.html" title="class in biolockj.util">DownloadUtil</a></dt>
<dd>
<div class="block">Get the total size of all files included for download.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getExe-biolockj.module.BioModule-java.lang.String-">getExe(BioModule, String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Get exe.* property name.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getExeParams-biolockj.module.BioModule-java.lang.String-">getExeParams(BioModule, String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Call this function to get the parameters configured for this property.<br>
 Make sure the last character for non-null resultss is an empty character for use in bash scripts calling the
 corresponding executable.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getExistingDir-biolockj.module.BioModule-java.lang.String-">getExistingDir(BioModule, String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Get a valid File directory or return null</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getExistingFile-biolockj.module.BioModule-java.lang.String-">getExistingFile(BioModule, String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Get a valid File or return null.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getExistingFileObject-java.lang.String-">getExistingFileObject(String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Build File using filePath.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/MetaUtil.html#getField-java.lang.String-java.lang.String-">getField(String, String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/MetaUtil.html" title="class in biolockj.util">MetaUtil</a></dt>
<dd>
<div class="block">Get metadata field value for given sampleId.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/MetaUtil.html#getFieldNames--">getFieldNames()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/MetaUtil.html" title="class in biolockj.util">MetaUtil</a></dt>
<dd>
<div class="block">Get a list of all metadata fields (metadata file column names except the 1st).</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/MetaUtil.html#getFieldValues-java.lang.String-boolean-">getFieldValues(String, boolean)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/MetaUtil.html" title="class in biolockj.util">MetaUtil</a></dt>
<dd>
<div class="block">Get metadata column for given field name.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Log.html#getFile--">getFile()</a></span> - Static method in class biolockj.<a href="../biolockj/Log.html" title="class in biolockj">Log</a></dt>
<dd>
<div class="block">Returns the log file.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/MetaUtil.html#getFile--">getFile()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/MetaUtil.html" title="class in biolockj.util">MetaUtil</a></dt>
<dd>
<div class="block">Get the current metadata file.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModuleImpl.html#getFileCache--">getFileCache()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a></dt>
<dd>
<div class="block">Get cached input files</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/r16s/QiimeParser.html#getFileNameColumn-java.lang.String-">getFileNameColumn(String)</a></span> - Method in class biolockj.module.implicit.parser.r16s.<a href="../biolockj/module/implicit/parser/r16s/QiimeParser.html" title="class in biolockj.module.implicit.parser.r16s">QiimeParser</a></dt>
<dd>
<div class="block">Examines the header line to find the column index for <a href="../biolockj/Constants.html#QIIME_DEMUX_COL">"BioLockJFileName"</a> which holds the
 demultiplexed sample file name.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BioLockJUtil.html#getFilePaths-java.util.Collection-">getFilePaths(Collection&lt;File&gt;)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BioLockJUtil.html" title="class in biolockj.util">BioLockJUtil</a></dt>
<dd>
<div class="block">Return an ordered list of absolute file paths from the input collection.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BioLockJUtil.html#getFileReader-java.io.File-">getFileReader(File)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BioLockJUtil.html" title="class in biolockj.util">BioLockJUtil</a></dt>
<dd>
<div class="block">Get a <a href="https://docs.oracle.com/javase/8/docs/api/java/io/BufferedReader.html?is-external=true" title="class or interface in java.io"><code>BufferedReader</code></a> for standard text file or <a href="https://docs.oracle.com/javase/8/docs/api/java/util/zip/GZIPInputStream.html?is-external=true" title="class or interface in java.util.zip"><code>GZIPInputStream</code></a> for gzipped files ending in ".gz"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SummaryUtil.html#getFooter--">getFooter()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SummaryUtil.html" title="class in biolockj.util">SummaryUtil</a></dt>
<dd>
<div class="block">Print the final lines of the summary with overall status, runtime, and the download scp command if applicable.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/MetaUtil.html#getForcedColumnName-java.lang.String-">getForcedColumnName(String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/MetaUtil.html" title="class in biolockj.util">MetaUtil</a></dt>
<dd>
<div class="block">Used to generate a guaranteed to be unique column name.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/Multiplexer.html#getHeader-java.io.File-java.util.List-">getHeader(File, List&lt;String&gt;)</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/Multiplexer.html" title="class in biolockj.module.seq">Multiplexer</a></dt>
<dd>
<div class="block">Get the header for the sequence.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#getHeader-java.lang.String-">getHeader(String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Extract the header from the first line of a read (parse to first space).</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#getHeaders-java.io.File-">getHeaders(File)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Return the header of each read in the sequence file.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#getHeaders-java.io.File-java.io.File-">getHeaders(File, File)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Get valid headers found in both forward and reverse read.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/BioLockJ.html#getHelpInfo-java.io.File-">getHelpInfo(File)</a></span> - Static method in class biolockj.<a href="../biolockj/BioLockJ.html" title="class in biolockj">BioLockJ</a></dt>
<dd>
<div class="block">Print error file path, restart instructions, and link to the BioLockJ Wiki</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/PathwayUtil.html#getHn2ClassifierOutput-java.lang.String-">getHn2ClassifierOutput(String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/PathwayUtil.html" title="class in biolockj.util">PathwayUtil</a></dt>
<dd>
<div class="block">Humann2 classifier outputs 3 types of files:
 
 <a href="../biolockj/Constants.html#HN2_PATH_ABUND_SUM">"pAbund"</a>
 <a href="../biolockj/Constants.html#HN2_PATH_COVG_SUM">"pCovg"</a>
 <a href="../biolockj/Constants.html#HN2_GENE_FAM_SUM">"geneFam"</a>
 
 Given the type, this method retuns the output file name.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/PathwayUtil.html#getHn2Type-java.io.File-">getHn2Type(File)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/PathwayUtil.html" title="class in biolockj.util">PathwayUtil</a></dt>
<dd>
<div class="block">Humann2 classifier outputs 3 types of files:
 
 <a href="../biolockj/Constants.html#HN2_PATH_ABUND_SUM">"pAbund"</a>
 <a href="../biolockj/Constants.html#HN2_PATH_COVG_SUM">"pCovg"</a>
 <a href="../biolockj/Constants.html#HN2_GENE_FAM_SUM">"geneFam"</a>
 
 This method returns the type included in the given file name.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModule.html#getID--">getID()</a></span> - Method in interface biolockj.module.<a href="../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></dt>
<dd>
<div class="block">Some BioModules may be added to a pipeline multiple times so must be identified by an ID.<br>
 This is the same value as the directory folder prefix when run.<br>
 The 1st module ID is 0 (or 00 if there are more than 10 modules.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModuleImpl.html#getID--">getID()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/MetaUtil.html#getID--">getID()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/MetaUtil.html" title="class in biolockj.util">MetaUtil</a></dt>
<dd>
<div class="block">Get the metadata file ID column name.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DemuxUtil.html#getIdMap--">getIdMap()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/DemuxUtil.html" title="class in biolockj.util">DemuxUtil</a></dt>
<dd>
<div class="block">Return the ID map (key=barcodeID, value=sampleId) Print the map to the log file.<br>
 Barcodes, if used, are located in the metadata column defined by property:
 <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/util/MetaUtil.html#META_BARCODE_COLUMN">"metadata.barcodeColumn"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DockerUtil.html#getImageName-java.lang.String-">getImageName(String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/DockerUtil.html" title="class in biolockj.util">DockerUtil</a></dt>
<dd>
<div class="block">Return the Docker Image name for the given class name.<br>
 Return <a href="../biolockj/util/DockerUtil.html#BLJ_BASH">"blj_bash"</a> for simple bash script modules that don't rely on special software<br>
 Class names contain no spaces, words are separated via CamelCaseConvension.<br>
 Docker image names cannot contain upper case letters, so this method substitutes "_" before the lower-case
 version of each capital letter.<br>
 <br>
 Example: JavaModule becomes java_module</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DockerUtil.html#getImageVersion-java.lang.String-">getImageVersion(String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/DockerUtil.html" title="class in biolockj.util">DockerUtil</a></dt>
<dd>
<div class="block">Get the Docker image version if defined in the <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> file<br>
 If not found, return the default version "latest"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getInitialProperties--">getInitialProperties()</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Get initial properties ordered by property</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BioLockJUtil.html#getInputDirs--">getInputDirs()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BioLockJUtil.html" title="class in biolockj.util">BioLockJUtil</a></dt>
<dd>
<div class="block">Get the list of input directories for the pipeline.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#getInputFileDir--">getInputFileDir()</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime">QiimeClassifier</a></dt>
<dd>
<div class="block">Module input directories are set to the previous module output directory.<br>
 To ensure we use the correct path, get path from <a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#getInputFiles--"><code>QiimeClassifier.getInputFiles()</code></a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModule.html#getInputFiles--">getInputFiles()</a></span> - Method in interface biolockj.module.<a href="../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></dt>
<dd>
<div class="block">Each BioModule takes the previous BioModule output as input:<br>
 BioModule[ n ].getInputFiles() = BioModule[ n - 1 ].getOutputDir().listFiles()<br>
 Special cases:<br>
 
 The 1st BioModule return all files in <a href="../biolockj/Constants.html#INPUT_DIRS">"input.dirPaths"</a><br>
 If previous BioModule BioModule[ n - 1 ] is a MetadataModule, forward it's input + output file: BioModule[ n
 ].getInputFiles() = BioModule[ n -1 ].getInputFiles() + BioModule[ n -1 ].getOutputFiles()
 </div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModuleImpl.html#getInputFiles--">getInputFiles()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a></dt>
<dd>
<div class="block">BioModule <a href="../biolockj/module/BioModuleImpl.html#getInputFiles--"><code>BioModuleImpl.getInputFiles()</code></a> is called to initialize upon first call and cached.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/r16s/QiimeParser.html#getInputFiles--">getInputFiles()</a></span> - Method in class biolockj.module.implicit.parser.r16s.<a href="../biolockj/module/implicit/parser/r16s/QiimeParser.html" title="class in biolockj.module.implicit.parser.r16s">QiimeParser</a></dt>
<dd>
<div class="block">Get only the lowest level report since it contains taxa info for all higher OTU reports.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#getInputFiles--">getInputFiles()</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime">QiimeClassifier</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/humann2/Humann2CountModule.html#getInputFiles--">getInputFiles()</a></span> - Method in class biolockj.module.report.humann2.<a href="../biolockj/module/report/humann2/Humann2CountModule.html" title="class in biolockj.module.report.humann2">Humann2CountModule</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/otu/OtuCountModule.html#getInputFiles--">getInputFiles()</a></span> - Method in class biolockj.module.report.otu.<a href="../biolockj/module/report/otu/OtuCountModule.html" title="class in biolockj.module.report.otu">OtuCountModule</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/taxa/TaxaCountModule.html#getInputFiles--">getInputFiles()</a></span> - Method in class biolockj.module.report.taxa.<a href="../biolockj/module/report/taxa/TaxaCountModule.html" title="class in biolockj.module.report.taxa">TaxaCountModule</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/SeqModuleImpl.html#getInputFiles--">getInputFiles()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/SeqModuleImpl.html" title="class in biolockj.module">SeqModuleImpl</a></dt>
<dd>
<div class="block">Return <a href="../biolockj/module/SeqModuleImpl.html#getSeqFiles-java.util.Collection-"><code>SeqModuleImpl.getSeqFiles(Collection)</code></a> to filter standard module input files.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SummaryUtil.html#getInputSummary-biolockj.module.BioModule-">getInputSummary(BioModule)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SummaryUtil.html" title="class in biolockj.util">SummaryUtil</a></dt>
<dd>
<div class="block">Build a summary of the input files for the given module</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#getIupacBase-java.lang.String-">getIupacBase(String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Return regex version of IUPAC DNA substitution bases so that only ACGT values are used.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/r/R_Module.html#getJobParams--">getJobParams()</a></span> - Method in class biolockj.module.report.r.<a href="../biolockj/module/report/r/R_Module.html" title="class in biolockj.module.report.r">R_Module</a></dt>
<dd>
<div class="block">If running Docker, run the Docker bash script, otherwise:<br>
 Run <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/module/report/r/R_Module.html#EXE_RSCRIPT">"exe.Rscript"</a> command on the generated R Script:
 <a href="../biolockj/module/ScriptModuleImpl.html#getMainScript--"><code>ScriptModuleImpl.getMainScript()</code></a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/ScriptModule.html#getJobParams--">getJobParams()</a></span> - Method in interface biolockj.module.<a href="../biolockj/module/ScriptModule.html" title="interface in biolockj.module">ScriptModule</a></dt>
<dd>
<div class="block">Many ScriptModules generate bash or R scripts for the Operating System to run as a <a href="../biolockj/Job.html" title="class in biolockj"><code>Job</code></a>.<br>
 The Job creates a <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Process.html?is-external=true" title="class or interface in java.lang"><code>Process</code></a> using the parameters supplied by this method and executed in the
 <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Runtime.html?is-external=true" title="class or interface in java.lang"><code>Runtime</code></a> environment.<br>
 Parameters typically contain the full script path and script parameters, if needed.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/ScriptModuleImpl.html#getJobParams--">getJobParams()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/ScriptModuleImpl.html" title="class in biolockj.module">ScriptModuleImpl</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/MetaUtil.html#getLatestColumnName-java.lang.String-">getLatestColumnName(String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/MetaUtil.html" title="class in biolockj.util">MetaUtil</a></dt>
<dd>
<div class="block">Returns the latest version of the given column.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/JsonNode.html#getLevel--">getLevel()</a></span> - Method in class biolockj.node.<a href="../biolockj/node/JsonNode.html" title="class in biolockj.node">JsonNode</a></dt>
<dd>
<div class="block">Getter method for level</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/r16s/QiimeNode.html#getLevel-java.lang.String-">getLevel(String)</a></span> - Method in class biolockj.node.r16s.<a href="../biolockj/node/r16s/QiimeNode.html" title="class in biolockj.node.r16s">QiimeNode</a></dt>
<dd>
<div class="block">Get the first taxonomy level delim.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/TaxaUtil.html#getLevelTaxaCounts-java.util.TreeMap-java.lang.String-">getLevelTaxaCounts(TreeMap&lt;String, TreeMap&lt;String, Integer&gt;&gt;, String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/TaxaUtil.html" title="class in biolockj.util">TaxaUtil</a></dt>
<dd>
<div class="block">Return TreeMap keyed on Sample ID, each sample maps to an inner map(taxa, count).<br>
 Input param sampleOtuCounts contains OTUs for which many can share the same level taxonomy assignment.<br>
 For example if sample42 contained the following 3 OTU values:
 
 phylum__Actinobacteria|class__Actinobacteria|order__Actinomycetales|family__Actinomycetaceae|genus__Actinomyces
 1000
 phylum__Bacteroidetes|class__Bacteroidia|order__Bacteroidales|family__Porphyromonadaceae|genus__Barnesiella
 500
 phylum__Bacteroidetes|class__Bacteroidia|order__Bacteroidales|family__Porphyromonadaceae|genus__Parabacteroides
 77
 
 The phylum level output would contain entries:
 
 TreeMap(sample42, TreeMap( Actinobacteria=1000, Bacteroidetes=577 ) )
 </div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/OtuNode.html#getLine--">getLine()</a></span> - Method in interface biolockj.node.<a href="../biolockj/node/OtuNode.html" title="interface in biolockj.node">OtuNode</a></dt>
<dd>
<div class="block">Get the line from classifier output file used to create this OtuNode.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/OtuNodeImpl.html#getLine--">getLine()</a></span> - Method in class biolockj.node.<a href="../biolockj/node/OtuNodeImpl.html" title="class in biolockj.node">OtuNodeImpl</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getList-biolockj.module.BioModule-java.lang.String-">getList(BioModule, String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Parse comma delimited property value to return list</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/otu/RarefyOtuCounts.html#getLowAbundantSamples--">getLowAbundantSamples()</a></span> - Method in class biolockj.module.report.otu.<a href="../biolockj/module/report/otu/RarefyOtuCounts.html" title="class in biolockj.module.report.otu">RarefyOtuCounts</a></dt>
<dd>
<div class="block">Return a list of low abundant samples, as defined by <a href="../biolockj/module/report/otu/RarefyOtuCounts.html#LOW_ABUNDANT_CUTOFF">"rarefyOtuCounts.lowAbundantCutoff"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/NextflowUtil.html#getMainNf--">getMainNf()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/NextflowUtil.html" title="class in biolockj.util">NextflowUtil</a></dt>
<dd>
<div class="block">Get the Nextflow main.nf file path.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/r/R_Module.html#getMainR--">getMainR()</a></span> - Method in class biolockj.module.report.r.<a href="../biolockj/module/report/r/R_Module.html" title="class in biolockj.module.report.r">R_Module</a></dt>
<dd>
<div class="block">Get the main R script</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/ScriptModule.html#getMainScript--">getMainScript()</a></span> - Method in interface biolockj.module.<a href="../biolockj/module/ScriptModule.html" title="interface in biolockj.module">ScriptModule</a></dt>
<dd>
<div class="block">Method returns MAIN module script that calls each of the worker scripts.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/ScriptModuleImpl.html#getMainScript--">getMainScript()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/ScriptModuleImpl.html" title="class in biolockj.module">ScriptModuleImpl</a></dt>
<dd>
<div class="block">Get the main script file in the bioModule script directory, with prefix:
 <a href="../biolockj/module/BioModule.html#MAIN_SCRIPT_PREFIX">"MAIN_"</a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BashScriptBuilder.html#getMainScriptExecuteWorkerLine-biolockj.module.ScriptModule-java.lang.String-java.lang.String-">getMainScriptExecuteWorkerLine(ScriptModule, String, String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BashScriptBuilder.html" title="class in biolockj.util">BashScriptBuilder</a></dt>
<dd>
<div class="block">Call <a href="../biolockj/util/BashScriptBuilder.html#FUNCTION_EXECUTE">"execute"</a> on the worker script</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DemuxUtil.html#getMapping--">getMapping()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/DemuxUtil.html" title="class in biolockj.util">DemuxUtil</a></dt>
<dd>
<div class="block">Return the demultiplexer mapping file used to map sequence files to barcode.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/PropUtil.html#getMasterConfig--">getMasterConfig()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/PropUtil.html" title="class in biolockj.util">PropUtil</a></dt>
<dd>
<div class="block">Return the MASTER config file.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/otu/CompileOtuCounts.html#getMaxOtusPerSample--">getMaxOtusPerSample()</a></span> - Method in class biolockj.module.report.otu.<a href="../biolockj/module/report/otu/CompileOtuCounts.html" title="class in biolockj.module.report.otu">CompileOtuCounts</a></dt>
<dd>
<div class="block">Find the maximum OTU count per sample.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SummaryUtil.html#getMean-java.util.Collection-boolean-">getMean(Collection&lt;?&gt;, boolean)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SummaryUtil.html" title="class in biolockj.util">SummaryUtil</a></dt>
<dd>
<div class="block">Return the mean value</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SummaryUtil.html#getMedian-java.util.Collection-boolean-">getMedian(Collection&lt;?&gt;, boolean)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SummaryUtil.html" title="class in biolockj.util">SummaryUtil</a></dt>
<dd>
<div class="block">Return the median value</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/humann2/AddMetadataToPathwayTables.html#getMergedLine-java.lang.String-">getMergedLine(String)</a></span> - Method in class biolockj.module.report.humann2.<a href="../biolockj/module/report/humann2/AddMetadataToPathwayTables.html" title="class in biolockj.module.report.humann2">AddMetadataToPathwayTables</a></dt>
<dd>
<div class="block">Return OTU table line with metadata row appended (both have PK = sample ID)</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/taxa/AddMetadataToTaxaTables.html#getMergedLine-java.lang.String-">getMergedLine(String)</a></span> - Method in class biolockj.module.report.taxa.<a href="../biolockj/module/report/taxa/AddMetadataToTaxaTables.html" title="class in biolockj.module.report.taxa">AddMetadataToTaxaTables</a></dt>
<dd>
<div class="block">Return OTU table line with metadata row appended (both have PK = sample ID)</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Pipeline.html#getMetadata-biolockj.module.BioModule-">getMetadata(BioModule)</a></span> - Static method in class biolockj.<a href="../biolockj/Pipeline.html" title="class in biolockj">Pipeline</a></dt>
<dd>
<div class="block">If the bioModule is complete and contains a metadata file in its output directory, return the metadata file,
 since it must be a new version.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/MetaUtil.html#getMetadata--">getMetadata()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/MetaUtil.html" title="class in biolockj.util">MetaUtil</a></dt>
<dd>
<div class="block">Metadata file getter.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/MetaUtil.html#getMetadataFileName--">getMetadataFileName()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/MetaUtil.html" title="class in biolockj.util">MetaUtil</a></dt>
<dd>
<div class="block">Get the metadata file name, if it exists, otherwise return projectName.tsv</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/MetaUtil.html#getMetadataRecord-java.lang.String-">getMetadataRecord(String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/MetaUtil.html" title="class in biolockj.util">MetaUtil</a></dt>
<dd>
<div class="block">Get metadata row for a given Sample ID.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/r/R_Module.html#getMetaMergedModule--">getMetaMergedModule()</a></span> - Method in class biolockj.module.report.r.<a href="../biolockj/module/report/r/R_Module.html" title="class in biolockj.module.report.r">R_Module</a></dt>
<dd>
<div class="block">Get correct meta-merged BioModule type for the give module.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/otu/CompileOtuCounts.html#getMinOtusPerSample--">getMinOtusPerSample()</a></span> - Method in class biolockj.module.report.otu.<a href="../biolockj/module/report/otu/CompileOtuCounts.html" title="class in biolockj.module.report.otu">CompileOtuCounts</a></dt>
<dd>
<div class="block">Find the minimum OTU count per sample.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/ModuleUtil.html#getModule-biolockj.module.BioModule-java.lang.String-boolean-">getModule(BioModule, String, boolean)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/ModuleUtil.html" title="class in biolockj.util">ModuleUtil</a></dt>
<dd>
<div class="block">Get a module with given className unless a classifier module is found 1st.<br>
 Use checkAhead parameter to determine if we look forward or backwards starting from the given module.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/ModuleUtil.html#getModule-java.lang.String-">getModule(String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/ModuleUtil.html" title="class in biolockj.util">ModuleUtil</a></dt>
<dd>
<div class="block">Construct a BioModule based on its className.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModule.html#getModuleDir--">getModuleDir()</a></span> - Method in interface biolockj.module.<a href="../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></dt>
<dd>
<div class="block">Each BioModule generates sub-directory under $DOCKER_PROJ</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModuleImpl.html#getModuleDir--">getModuleDir()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a></dt>
<dd>
<div class="block">All BioModule work must be contained within the scope of its root directory.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SummaryUtil.html#getModuleNumber--">getModuleNumber()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SummaryUtil.html" title="class in biolockj.util">SummaryUtil</a></dt>
<dd>
<div class="block">Read the summary file to find the last completed module number and return the next number.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getModuleProp-biolockj.module.BioModule-java.lang.String-">getModuleProp(BioModule, String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Return module specific property if configured, otherwise use the given prop.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getModuleProp-java.lang.String-java.lang.String-">getModuleProp(String, String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Return module specific property if configured, otherwise use the given prop.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SummaryUtil.html#getModuleRunTime-biolockj.module.BioModule-">getModuleRunTime(BioModule)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SummaryUtil.html" title="class in biolockj.util">SummaryUtil</a></dt>
<dd>
<div class="block">Return duration module ran based on modified data of started file, formatted for display (as hours, minutes,
 seconds).</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Pipeline.html#getModules--">getModules()</a></span> - Static method in class biolockj.<a href="../biolockj/Pipeline.html" title="class in biolockj">Pipeline</a></dt>
<dd>
<div class="block">Return a list of <a href="../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s constructed by the <a href="../biolockj/BioModuleFactory.html" title="class in biolockj"><code>BioModuleFactory</code></a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/ModuleUtil.html#getModules-biolockj.module.BioModule-java.lang.Boolean-">getModules(BioModule, Boolean)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/ModuleUtil.html" title="class in biolockj.util">ModuleUtil</a></dt>
<dd>
<div class="block">Return pipeline modules after the given module if checkAhead = TRUE<br>
 Otherwise return pipeline modules before the given module.<br>
 If returning the prior modules, return the pipeline modules in reverse order, so the 1st item in the list is the
 module immediately preceding the given module.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/r/R_Module.html#getModuleScript--">getModuleScript()</a></span> - Method in class biolockj.module.report.r.<a href="../biolockj/module/report/r/R_Module.html" title="class in biolockj.module.report.r">R_Module</a></dt>
<dd>
<div class="block">Get the Module script</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Log.html#getMsgs--">getMsgs()</a></span> - Static method in class biolockj.<a href="../biolockj/Log.html" title="class in biolockj">Log</a></dt>
<dd>
<div class="block">Get the logMesseges created prior to creating the actual <a href="../biolockj/Log.html#LOG_FILE"><code>Log.LOG_FILE</code></a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/NextflowUtil.html#getNextflowLines-java.io.File-">getNextflowLines(File)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/NextflowUtil.html" title="class in biolockj.util">NextflowUtil</a></dt>
<dd>
<div class="block">Build the main.nf lines from the template file by replacing several parameters.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/ModuleUtil.html#getNextModule-biolockj.module.BioModule-">getNextModule(BioModule)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/ModuleUtil.html" title="class in biolockj.util">ModuleUtil</a></dt>
<dd>
<div class="block">BioModules are run in the order configured.<br>
 Return the module configured to run after the given module.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getNonNegativeInteger-biolockj.module.BioModule-java.lang.String-">getNonNegativeInteger(BioModule, String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Parse property as non-negative integer value</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#getNumLinesPerRead--">getNumLinesPerRead()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Return number of lines per read (for fasta return 2, for fastq return 4)</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/otu/RarefyOtuCounts.html#getNumOtusForQuantile-java.util.TreeMap-">getNumOtusForQuantile(TreeMap&lt;String, TreeMap&lt;String, Integer&gt;&gt;)</a></span> - Method in class biolockj.module.report.otu.<a href="../biolockj/module/report/otu/RarefyOtuCounts.html" title="class in biolockj.module.report.otu">RarefyOtuCounts</a></dt>
<dd>
<div class="block">Get the quantile number of OTUs.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/RegisterNumReads.html#getNumReadFieldName--">getNumReadFieldName()</a></span> - Static method in class biolockj.module.implicit.<a href="../biolockj/module/implicit/RegisterNumReads.html" title="class in biolockj.module.implicit">RegisterNumReads</a></dt>
<dd>
<div class="block">Getter for numReadFieldName</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#getNumReads-long-">getNumReads(long)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Return number of reads given number of lines in a sample.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/OtuUtil.OtuCountLine.html#getOtu--">getOtu()</a></span> - Method in class biolockj.util.<a href="../biolockj/util/OtuUtil.OtuCountLine.html" title="class in biolockj.util">OtuUtil.OtuCountLine</a></dt>
<dd>
<div class="block">Get OTU name</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/ParserModuleImpl.html#getOtuCountField--">getOtuCountField()</a></span> - Static method in class biolockj.module.implicit.parser.<a href="../biolockj/module/implicit/parser/ParserModuleImpl.html" title="class in biolockj.module.implicit.parser">ParserModuleImpl</a></dt>
<dd>
<div class="block">Getter for otuCountField</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/OtuUtil.html#getOtuCountFile-java.io.File-java.lang.String-java.lang.String-">getOtuCountFile(File, String, String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/OtuUtil.html" title="class in biolockj.util">OtuUtil</a></dt>
<dd>
<div class="block">Build OTU count file using a standard format in the directory given.<br>
 Format: pipeline_name + prefix + <a href="../biolockj/Constants.html#OTU_COUNT">"otuCount"</a> + sampleId +
 <a href="../biolockj/Constants.html#TSV_EXT">".tsv"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/ParsedSample.html#getOtuCounts--">getOtuCounts()</a></span> - Method in class biolockj.node.<a href="../biolockj/node/ParsedSample.html" title="class in biolockj.node">ParsedSample</a></dt>
<dd>
<div class="block">Get the streamlined taxonomy tree with counts, each OTU listed only 1 time with num occurrences in the
 sample.<br>
 Example:
 d__Bacteria;p__Bacteroidetes;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Bacteroides;s__Bacteroides_vulgatus
 87342</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/OtuNode.html#getOtuName--">getOtuName()</a></span> - Method in interface biolockj.node.<a href="../biolockj/node/OtuNode.html" title="interface in biolockj.node">OtuNode</a></dt>
<dd>
<div class="block">Build the OTU name from the taxaMap.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/OtuNodeImpl.html#getOtuName--">getOtuName()</a></span> - Method in class biolockj.node.<a href="../biolockj/node/OtuNodeImpl.html" title="class in biolockj.node">OtuNodeImpl</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/r16s/QiimeParser.html#getOtuTableRowId-java.lang.String-">getOtuTableRowId(String)</a></span> - Method in class biolockj.module.implicit.parser.r16s.<a href="../biolockj/module/implicit/parser/r16s/QiimeParser.html" title="class in biolockj.module.implicit.parser.r16s">QiimeParser</a></dt>
<dd>
<div class="block">This method is used to get an R-friendly Sample ID value by stripping out the quotes AND by replacing any Qiime
 ID found with the original Sample ID.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModule.html#getOutputDir--">getOutputDir()</a></span> - Method in interface biolockj.module.<a href="../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></dt>
<dd>
<div class="block">Output files destined as input for the next BioModule is created in this directory.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModuleImpl.html#getOutputDir--">getOutputDir()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a></dt>
<dd>
<div class="block">Returns moduleDir/output which will be used as the next module's input.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SummaryUtil.html#getOutputDirSummary-biolockj.module.BioModule-">getOutputDirSummary(BioModule)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SummaryUtil.html" title="class in biolockj.util">SummaryUtil</a></dt>
<dd>
<div class="block">Return summary of <a href="../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a> output directory, with metrics:
 
 Number of output files
 Mean output file size
 Path of new metadata file if any created
 </div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SummaryUtil.html#getPad-java.lang.String-">getPad(String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SummaryUtil.html" title="class in biolockj.util">SummaryUtil</a></dt>
<dd>
<div class="block">Summary count label padding as label length + 18</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#getPairedReads-java.util.Collection-">getPairedReads(Collection&lt;File&gt;)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Paired reads must have a unique file suffix to identify forward and reverse reads.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/Metaphlan2Classifier.html#getParams--">getParams()</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/Metaphlan2Classifier.html" title="class in biolockj.module.classifier.wgs">Metaphlan2Classifier</a></dt>
<dd>
<div class="block">Metaphlan queries require standard parameters: --input_type, --nproc, -t<br>
 Verify no invalid runtime params are passed and add rankSwitch if needed.<br>
 Set the rankSwitch based on the <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/Constants.html#REPORT_TAXONOMY_LEVELS">"report.taxonomyLevels"</a> if
 only one taxonomy level is to be reported, otherwise report all levels.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#getParams--">getParams()</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime">QiimeClassifier</a></dt>
<dd>
<div class="block">Subclasses call this method to check dependencies before picking OTUs to validate
 <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#QIIME_PARAMS">"qiime.params"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/JsonNode.html#getParent--">getParent()</a></span> - Method in class biolockj.node.<a href="../biolockj/node/JsonNode.html" title="class in biolockj.node">JsonNode</a></dt>
<dd>
<div class="block">Getter method for parent</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/ParserModule.html#getParsedSample-java.lang.String-">getParsedSample(String)</a></span> - Method in interface biolockj.module.implicit.parser.<a href="../biolockj/module/implicit/parser/ParserModule.html" title="interface in biolockj.module.implicit.parser">ParserModule</a></dt>
<dd>
<div class="block">After <a href="../biolockj/module/implicit/parser/ParserModule.html#parseSamples--"><code>ParserModule.parseSamples()</code></a> completes, this method can be called to get a <a href="../biolockj/node/ParsedSample.html" title="class in biolockj.node"><code>ParsedSample</code></a> by
 its sample ID.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/ParserModuleImpl.html#getParsedSample-java.lang.String-">getParsedSample(String)</a></span> - Method in class biolockj.module.implicit.parser.<a href="../biolockj/module/implicit/parser/ParserModuleImpl.html" title="class in biolockj.module.implicit.parser">ParserModuleImpl</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/MetaUtil.html#getPath--">getPath()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/MetaUtil.html" title="class in biolockj.util">MetaUtil</a></dt>
<dd>
<div class="block">Return the metadata file path</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/PathwayUtil.html#getPathwayCountFile-java.io.File-java.io.File-java.lang.String-">getPathwayCountFile(File, File, String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/PathwayUtil.html" title="class in biolockj.util">PathwayUtil</a></dt>
<dd>
<div class="block">Return a pathway abundance file path in the given dir with the given prefix (if provided).</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#getPickOtuLines-java.lang.String-java.io.File-java.lang.String-java.io.File-">getPickOtuLines(String, File, String, File)</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime">QiimeClassifier</a></dt>
<dd>
<div class="block">Subclasses call this method to add OTU picking lines by calling <a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#SCRIPT_ADD_LABELS">"add_qiime_labels.py"</a> via OTU picking
 script.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BioLockJUtil.html#getPipelineInputFiles--">getPipelineInputFiles()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BioLockJUtil.html" title="class in biolockj.util">BioLockJUtil</a></dt>
<dd>
<div class="block">Basic input files may be sequences, or any other file type acceptable in a pipeline module.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getPositiveDoubleVal-biolockj.module.BioModule-java.lang.String-">getPositiveDoubleVal(BioModule, String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Parse property as positive double value</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getPositiveInteger-biolockj.module.BioModule-java.lang.String-">getPositiveInteger(BioModule, String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Parse property as positive integer value</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModule.html#getPostRequisiteModules--">getPostRequisiteModules()</a></span> - Method in interface biolockj.module.<a href="../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></dt>
<dd>
<div class="block"><a href="../biolockj/Pipeline.html" title="class in biolockj"><code>Pipeline</code></a> calls this method when building the list of pipeline BioModules to execute.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModuleImpl.html#getPostRequisiteModules--">getPostRequisiteModules()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a></dt>
<dd>
<div class="block">By default, no post-requisites are required.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/ClassifierModuleImpl.html#getPostRequisiteModules--">getPostRequisiteModules()</a></span> - Method in class biolockj.module.classifier.<a href="../biolockj/module/classifier/ClassifierModuleImpl.html" title="class in biolockj.module.classifier">ClassifierModuleImpl</a></dt>
<dd>
<div class="block">This method returns the corresponding Parser module associated with the classifier.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/r16s/QiimeClosedRefClassifier.html#getPostRequisiteModules--">getPostRequisiteModules()</a></span> - Method in class biolockj.module.classifier.r16s.<a href="../biolockj/module/classifier/r16s/QiimeClosedRefClassifier.html" title="class in biolockj.module.classifier.r16s">QiimeClosedRefClassifier</a></dt>
<dd>
<div class="block">If paired reads found, return prerequisite module: <a href="../biolockj/module/seq/PearMergeReads.html" title="class in biolockj.module.seq"><code>PearMergeReads</code></a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#getPostRequisiteModules--">getPostRequisiteModules()</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime">QiimeClassifier</a></dt>
<dd>
<div class="block">Subclasses of QiimeClassifier add post-requisite module: <a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime"><code>QiimeClassifier</code></a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/BioModuleFactory.html#getPostRequisites-biolockj.module.BioModule-">getPostRequisites(BioModule)</a></span> - Method in class biolockj.<a href="../biolockj/BioModuleFactory.html" title="class in biolockj">BioModuleFactory</a></dt>
<dd>
<div class="block">This method returns all module post-requisites (including post-requisites of the post-requisites).</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModule.html#getPreRequisiteModules--">getPreRequisiteModules()</a></span> - Method in interface biolockj.module.<a href="../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></dt>
<dd>
<div class="block"><a href="../biolockj/Pipeline.html" title="class in biolockj"><code>Pipeline</code></a> calls this method when building the list of pipeline BioModules to execute.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModuleImpl.html#getPreRequisiteModules--">getPreRequisiteModules()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a></dt>
<dd>
<div class="block">By default, no prerequisites are required.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/r16s/RdpClassifier.html#getPreRequisiteModules--">getPreRequisiteModules()</a></span> - Method in class biolockj.module.classifier.r16s.<a href="../biolockj/module/classifier/r16s/RdpClassifier.html" title="class in biolockj.module.classifier.r16s">RdpClassifier</a></dt>
<dd>
<div class="block">If paired reads found, add prerequisite: <a href="../biolockj/module/seq/PearMergeReads.html" title="class in biolockj.module.seq"><code>PearMergeReads</code></a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#getPreRequisiteModules--">getPreRequisiteModules()</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime">QiimeClassifier</a></dt>
<dd>
<div class="block">If paired reads found, add prerequisite module: <a href="../biolockj/module/seq/PearMergeReads.html" title="class in biolockj.module.seq"><code>PearMergeReads</code></a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/humann2/Humann2CountModule.html#getPreRequisiteModules--">getPreRequisiteModules()</a></span> - Method in class biolockj.module.report.humann2.<a href="../biolockj/module/report/humann2/Humann2CountModule.html" title="class in biolockj.module.report.humann2">Humann2CountModule</a></dt>
<dd>
<div class="block">Module prerequisite: <a href="../biolockj/module/implicit/parser/wgs/Humann2Parser.html" title="class in biolockj.module.implicit.parser.wgs"><code>Humann2Parser</code></a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/JsonReport.html#getPreRequisiteModules--">getPreRequisiteModules()</a></span> - Method in class biolockj.module.report.<a href="../biolockj/module/report/JsonReport.html" title="class in biolockj.module.report">JsonReport</a></dt>
<dd>
<div class="block">Module prerequisite: <a href="../biolockj/module/report/otu/CompileOtuCounts.html" title="class in biolockj.module.report.otu"><code>CompileOtuCounts</code></a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/r/R_Module.html#getPreRequisiteModules--">getPreRequisiteModules()</a></span> - Method in class biolockj.module.report.r.<a href="../biolockj/module/report/r/R_Module.html" title="class in biolockj.module.report.r">R_Module</a></dt>
<dd>
<div class="block">Require combined count-metadata tables as input.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/r/R_PlotEffectSize.html#getPreRequisiteModules--">getPreRequisiteModules()</a></span> - Method in class biolockj.module.report.r.<a href="../biolockj/module/report/r/R_PlotEffectSize.html" title="class in biolockj.module.report.r">R_PlotEffectSize</a></dt>
<dd>
<div class="block">Returns <a href="../biolockj/module/report/r/R_Module.html#getStatPreReqs--"><code>R_Module.getStatPreReqs()</code></a> and if fold change plots are to be generated, add
 <a href="../biolockj/module/report/taxa/NormalizeTaxaTables.html" title="class in biolockj.module.report.taxa"><code>NormalizeTaxaTables</code></a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/r/R_PlotOtus.html#getPreRequisiteModules--">getPreRequisiteModules()</a></span> - Method in class biolockj.module.report.r.<a href="../biolockj/module/report/r/R_PlotOtus.html" title="class in biolockj.module.report.r">R_PlotOtus</a></dt>
<dd>
<div class="block">Returns <a href="../biolockj/module/report/r/R_Module.html#getStatPreReqs--"><code>R_Module.getStatPreReqs()</code></a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/r/R_PlotPvalHistograms.html#getPreRequisiteModules--">getPreRequisiteModules()</a></span> - Method in class biolockj.module.report.r.<a href="../biolockj/module/report/r/R_PlotPvalHistograms.html" title="class in biolockj.module.report.r">R_PlotPvalHistograms</a></dt>
<dd>
<div class="block">Returns <a href="../biolockj/module/report/r/R_Module.html#getStatPreReqs--"><code>R_Module.getStatPreReqs()</code></a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/taxa/TaxaCountModule.html#getPreRequisiteModules--">getPreRequisiteModules()</a></span> - Method in class biolockj.module.report.taxa.<a href="../biolockj/module/report/taxa/TaxaCountModule.html" title="class in biolockj.module.report.taxa">TaxaCountModule</a></dt>
<dd>
<div class="block">Require taxonomy table module as prerequisite</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/RarefySeqs.html#getPreRequisiteModules--">getPreRequisiteModules()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/RarefySeqs.html" title="class in biolockj.module.seq">RarefySeqs</a></dt>
<dd>
<div class="block">This method always requires a prerequisite module with a "number of reads" count such as:
 <a href="../biolockj/module/implicit/RegisterNumReads.html" title="class in biolockj.module.implicit"><code>RegisterNumReads</code></a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/BioModuleFactory.html#getPreRequisites-biolockj.module.BioModule-">getPreRequisites(BioModule)</a></span> - Method in class biolockj.<a href="../biolockj/BioModuleFactory.html" title="class in biolockj">BioModuleFactory</a></dt>
<dd>
<div class="block">This method returns all module prerequisites (including prerequisites of the prerequisites).</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/ModuleUtil.html#getPreviousModule-biolockj.module.BioModule-">getPreviousModule(BioModule)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/ModuleUtil.html" title="class in biolockj.util">ModuleUtil</a></dt>
<dd>
<div class="block">BioModules are run in the order configured.<br>
 Return the module configured to run before the given module.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/r/R_Module.html#getPrimaryScript--">getPrimaryScript()</a></span> - Method in class biolockj.module.report.r.<a href="../biolockj/module/report/r/R_Module.html" title="class in biolockj.module.report.r">R_Module</a></dt>
<dd>
<div class="block">Get the primary R script</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/TrimPrimers.html#getPrimers--">getPrimers()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/TrimPrimers.html" title="class in biolockj.module.seq">TrimPrimers</a></dt>
<dd>
<div class="block">Get the primers listed in the <a href="../biolockj/module/seq/TrimPrimers.html#INPUT_TRIM_SEQ_FILE">"trimPrimers.filePath"</a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getProperties--">getProperties()</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Get current properties ordered by property</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/ImportMetadata.html#getQuotedValue-java.lang.String-">getQuotedValue(String)</a></span> - Method in class biolockj.module.implicit.<a href="../biolockj/module/implicit/ImportMetadata.html" title="class in biolockj.module.implicit">ImportMetadata</a></dt>
<dd>
<div class="block">The member variable quotedText caches the input held within a quoted block.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/otu/RarefyOtuCounts.html#getRandomQuantileOtus-java.util.List-int-">getRandomQuantileOtus(List&lt;String&gt;, int)</a></span> - Method in class biolockj.module.report.otu.<a href="../biolockj/module/report/otu/RarefyOtuCounts.html" title="class in biolockj.module.report.otu">RarefyOtuCounts</a></dt>
<dd>
<div class="block">Select random OTUs based on the quantileNum from the list of OTUs in data.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#getReadDirectionSuffix-java.io.File-">getReadDirectionSuffix(File)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Return read direction indicator for forward or reverse read if found in the file name.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#getReadDirectionSuffix-java.lang.String-">getReadDirectionSuffix(String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Return read direction indicator for forward or reverse read if found in the file name.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#getRestartDir--">getRestartDir()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">Return restart pipeline directory</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/r/R_Module.html#getRTemplateDir--">getRTemplateDir()</a></span> - Static method in class biolockj.module.report.r.<a href="../biolockj/module/report/r/R_Module.html" title="class in biolockj.module.report.r">R_Module</a></dt>
<dd>
<div class="block">Get the BioLockJ resource R directory.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SummaryUtil.html#getRunTime-long-">getRunTime(long)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SummaryUtil.html" title="class in biolockj.util">SummaryUtil</a></dt>
<dd>
<div class="block">Get runtime message (formatted as hours, minutes, seconds) based on startTime passed.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/Humann2Classifier.html#getRuntimeParams--">getRuntimeParams()</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/Humann2Classifier.html" title="class in biolockj.module.classifier.wgs">Humann2Classifier</a></dt>
<dd>
<div class="block">Get formatted runtime parameters and </div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/ScriptModuleImpl.html#getRuntimeParams-java.util.List-java.lang.String-">getRuntimeParams(List&lt;String&gt;, String)</a></span> - Method in class biolockj.module.<a href="../biolockj/module/ScriptModuleImpl.html" title="class in biolockj.module">ScriptModuleImpl</a></dt>
<dd>
<div class="block">Return all collectionProperty values separated by a space.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/r16s/QiimeParser.html#getSampleId-java.lang.String-">getSampleId(String)</a></span> - Method in class biolockj.module.implicit.parser.r16s.<a href="../biolockj/module/implicit/parser/r16s/QiimeParser.html" title="class in biolockj.module.implicit.parser.r16s">QiimeParser</a></dt>
<dd>
<div class="block">Get the Sample ID from <a href="../biolockj/module/implicit/parser/r16s/QiimeParser.html#qiimeIdToSampleIdMap"><code>QiimeParser.qiimeIdToSampleIdMap</code></a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/OtuNode.html#getSampleId--">getSampleId()</a></span> - Method in interface biolockj.node.<a href="../biolockj/node/OtuNode.html" title="interface in biolockj.node">OtuNode</a></dt>
<dd>
<div class="block">Gets the sample ID to which the read belongs</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/OtuNodeImpl.html#getSampleId--">getSampleId()</a></span> - Method in class biolockj.node.<a href="../biolockj/node/OtuNodeImpl.html" title="class in biolockj.node">OtuNodeImpl</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/ParsedSample.html#getSampleId--">getSampleId()</a></span> - Method in class biolockj.node.<a href="../biolockj/node/ParsedSample.html" title="class in biolockj.node">ParsedSample</a></dt>
<dd>
<div class="block">Getter for sampleId.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DemuxUtil.html#getSampleId-java.util.List-">getSampleId(List&lt;String&gt;)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/DemuxUtil.html" title="class in biolockj.util">DemuxUtil</a></dt>
<dd>
<div class="block">Determine Sample Id by examining the sequence lines.<br>
 If =<a href="../biolockj/util/DemuxUtil.html#OPTION_ID_IN_HEADER">"id_in_header"</a>, extract the Sample Id from the sequence header via
 <a href="../biolockj/util/SeqUtil.html#getSampleId-java.lang.String-"><code>SeqUtil.getSampleId(String)</code></a><br>
 If =<a href="../biolockj/util/DemuxUtil.html#OPTION_BARCODE_IN_HEADER">"barcode_in_header"</a> and the sequence header contains a bar-code in
 the idMap, return the corresponding SampleID from the idMap.<br>
 If =<a href="../biolockj/util/DemuxUtil.html#OPTION_BARCODE_IN_MAPPING">"barcode_in_mapping"</a> and the sequence header maps to a bar-code in
 the idMap, return the corresponding SampleID from the idMap.<br>
 If =<a href="../biolockj/util/DemuxUtil.html#OPTION_BARCODE_IN_SEQ">"barcode_in_seq"</a> and the sequence itself begins with a bar-code in
 the idMap, return the corresponding SampleID from the idMap.<br></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/OtuUtil.html#getSampleId-java.io.File-">getSampleId(File)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/OtuUtil.html" title="class in biolockj.util">OtuUtil</a></dt>
<dd>
<div class="block">Extract the sampleId from the OTU count file name.<br>
 Input files should include a file name just before the .tsv file extension.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#getSampleId-java.lang.String-">getSampleId(String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Method extracts Sample ID from the name param.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/r16s/QiimeParser.html#getSampleIdFromMappingFile-java.lang.String-int-">getSampleIdFromMappingFile(String, int)</a></span> - Method in class biolockj.module.implicit.parser.r16s.<a href="../biolockj/module/implicit/parser/r16s/QiimeParser.html" title="class in biolockj.module.implicit.parser.r16s">QiimeParser</a></dt>
<dd>
<div class="block">The mapping file used the original sampleId when generating the demultipled files, as recorded in mapping file
 column - <a href="../biolockj/Constants.html#QIIME_DEMUX_COL">"BioLockJFileName"</a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/ImportMetadata.html#getSampleIds--">getSampleIds()</a></span> - Method in class biolockj.module.implicit.<a href="../biolockj/module/implicit/ImportMetadata.html" title="class in biolockj.module.implicit">ImportMetadata</a></dt>
<dd>
<div class="block">Extract the sample IDs from the file names with <a href="../biolockj/util/SeqUtil.html#getSampleId-java.lang.String-"><code>SeqUtil.getSampleId(String)</code></a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/MetaUtil.html#getSampleIds--">getSampleIds()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/MetaUtil.html" title="class in biolockj.util">MetaUtil</a></dt>
<dd>
<div class="block">Get the first column from the metadata file.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/OtuUtil.html#getSampleOtuCounts-java.util.Collection-">getSampleOtuCounts(Collection&lt;File&gt;)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/OtuUtil.html" title="class in biolockj.util">OtuUtil</a></dt>
<dd>
<div class="block">TreeMap OTU counts for each sample file formatted and named as in
 <a href="../biolockj/module/implicit/parser/ParserModule.html" title="interface in biolockj.module.implicit.parser"><code>ParserModule</code></a> output.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/KneadData.html#getSanatizedFile-java.lang.String-java.lang.Boolean-">getSanatizedFile(String, Boolean)</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/KneadData.html" title="class in biolockj.module.seq">KneadData</a></dt>
<dd>
<div class="block">Return sanitized sequence data file.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/r16s/RdpNode.html#getScore--">getScore()</a></span> - Method in class biolockj.node.r16s.<a href="../biolockj/node/r16s/RdpNode.html" title="class in biolockj.node.r16s">RdpNode</a></dt>
<dd>
<div class="block">Return lowest RDP confidence percentage (0-100) above the configured threshold value.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/ScriptModule.html#getScriptDir--">getScriptDir()</a></span> - Method in interface biolockj.module.<a href="../biolockj/module/ScriptModule.html" title="interface in biolockj.module">ScriptModule</a></dt>
<dd>
<div class="block">ScriptModules that generate scripts to complete their task, create script files in this directory.<br>
 The main script must begin with prefix  and there must only be one main script.<br>
 The main script executes, or submits to the job queue, each of the other scripts in this directory.<br>
 The # subscripts generated by the ScriptModule depends upon <a href="../biolockj/module/ScriptModule.html#SCRIPT_BATCH_SIZE">"script.batchSize"</a>.<br></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/ScriptModuleImpl.html#getScriptDir--">getScriptDir()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/ScriptModuleImpl.html" title="class in biolockj.module">ScriptModuleImpl</a></dt>
<dd>
<div class="block">Returns moduleDir/script which contains all scripts generated by the module.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SummaryUtil.html#getScriptDirSummary-biolockj.module.ScriptModule-">getScriptDirSummary(ScriptModule)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SummaryUtil.html" title="class in biolockj.util">SummaryUtil</a></dt>
<dd>
<div class="block">Return summary of the <a href="../biolockj/module/ScriptModule.html" title="interface in biolockj.module"><code>ScriptModule</code></a> script directory with metrics:
 
 Print main script name
 Number of worker scripts run
 Number of worker scripts successful/failed/incomplete
 Average worker script run time
 Longest running worker script names/duration
 Longest running workers script names/duration
 </div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/ScriptModule.html#getScriptErrors--">getScriptErrors()</a></span> - Method in interface biolockj.module.<a href="../biolockj/module/ScriptModule.html" title="interface in biolockj.module">ScriptModule</a></dt>
<dd>
<div class="block">Return a list of the error messages saved the failure scripts in the module script directory.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/ScriptModuleImpl.html#getScriptErrors--">getScriptErrors()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/ScriptModuleImpl.html" title="class in biolockj.module">ScriptModuleImpl</a></dt>
<dd>
<div class="block">This method returns all of the lines from any failure files found in the script directory.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SummaryUtil.html#getScriptRunTime-java.lang.Long-">getScriptRunTime(Long)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SummaryUtil.html" title="class in biolockj.util">SummaryUtil</a></dt>
<dd>
<div class="block">Get the script runtime in hours and minutes</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/Demultiplexer.html#getSeqFiles-java.util.Collection-">getSeqFiles(Collection&lt;File&gt;)</a></span> - Method in class biolockj.module.implicit.<a href="../biolockj/module/implicit/Demultiplexer.html" title="class in biolockj.module.implicit">Demultiplexer</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/RegisterNumReads.html#getSeqFiles-java.util.Collection-">getSeqFiles(Collection&lt;File&gt;)</a></span> - Method in class biolockj.module.implicit.<a href="../biolockj/module/implicit/RegisterNumReads.html" title="class in biolockj.module.implicit">RegisterNumReads</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/AwkFastaConverter.html#getSeqFiles-java.util.Collection-">getSeqFiles(Collection&lt;File&gt;)</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/AwkFastaConverter.html" title="class in biolockj.module.seq">AwkFastaConverter</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/Multiplexer.html#getSeqFiles-java.util.Collection-">getSeqFiles(Collection&lt;File&gt;)</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/Multiplexer.html" title="class in biolockj.module.seq">Multiplexer</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/RarefySeqs.html#getSeqFiles-java.util.Collection-">getSeqFiles(Collection&lt;File&gt;)</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/RarefySeqs.html" title="class in biolockj.module.seq">RarefySeqs</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/SeqFileValidator.html#getSeqFiles-java.util.Collection-">getSeqFiles(Collection&lt;File&gt;)</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/SeqFileValidator.html" title="class in biolockj.module.seq">SeqFileValidator</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/TrimPrimers.html#getSeqFiles-java.util.Collection-">getSeqFiles(Collection&lt;File&gt;)</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/TrimPrimers.html" title="class in biolockj.module.seq">TrimPrimers</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/SeqModule.html#getSeqFiles-java.util.Collection-">getSeqFiles(Collection&lt;File&gt;)</a></span> - Method in interface biolockj.module.<a href="../biolockj/module/SeqModule.html" title="interface in biolockj.module">SeqModule</a></dt>
<dd>
<div class="block">Return only sequence files for sample IDs found in the metadata file.<br>
 If <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/util/MetaUtil.html#META_REQUIRED">"metadata.required"</a> = <a href="../biolockj/Constants.html#TRUE">"Y"</a>, an
 error is thrown to list the files that cannot be matched to a metadata row.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/SeqModuleImpl.html#getSeqFiles-java.util.Collection-">getSeqFiles(Collection&lt;File&gt;)</a></span> - Method in class biolockj.module.<a href="../biolockj/module/SeqModuleImpl.html" title="class in biolockj.module">SeqModuleImpl</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#getSeqFiles-java.util.Collection-">getSeqFiles(Collection&lt;File&gt;)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Return only sequence files for sample IDs found in the metadata file.<br>
 If <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/util/MetaUtil.html#META_REQUIRED">"metadata.required"</a> = <a href="../biolockj/Constants.html#TRUE">"Y"</a>, an
 error is thrown to list the files that cannot be matched to a metadata row.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#getSeqHeaderChars--">getSeqHeaderChars()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Get all sequence header characters for fasta and fastq files.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/Email.html#getSession--">getSession()</a></span> - Method in class biolockj.module.report.<a href="../biolockj/module/report/Email.html" title="class in biolockj.module.report">Email</a></dt>
<dd>
<div class="block">Build an authenticated javax.mail.Session using <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> email properties</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getSet-biolockj.module.BioModule-java.lang.String-">getSet(BioModule, String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Parse comma-separated property value to build an unordered Set</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/JavaModuleImpl.html#getSource--">getSource()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/JavaModuleImpl.html" title="class in biolockj.module">JavaModuleImpl</a></dt>
<dd>
<div class="block">Get the program source (either the jar path or main class biolockj.BioLockJ);</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BioLockJUtil.html#getSource--">getSource()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BioLockJUtil.html" title="class in biolockj.util">BioLockJUtil</a></dt>
<dd>
<div class="block">Get the source of the java runtime classes ( /bin directory or JAR file ).</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/r/R_Module.html#getStatPreReqs--">getStatPreReqs()</a></span> - Method in class biolockj.module.report.r.<a href="../biolockj/module/report/r/R_Module.html" title="class in biolockj.module.report.r">R_Module</a></dt>
<dd>
<div class="block">Add <a href="../biolockj/module/report/r/R_CalculateStats.html" title="class in biolockj.module.report.r"><code>R_CalculateStats</code></a> to standard <a href="../biolockj/module/report/r/R_Module.html#getPreRequisiteModules--"><code>R_Module.getPreRequisiteModules()</code></a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/JsonNode.html#getStats--">getStats()</a></span> - Method in class biolockj.node.<a href="../biolockj/node/JsonNode.html" title="class in biolockj.node">JsonNode</a></dt>
<dd>
<div class="block">Getter method for stats</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/r/R_CalculateStats.html#getStatsFile-biolockj.module.BioModule-java.lang.String-java.lang.Boolean-java.lang.Boolean-">getStatsFile(BioModule, String, Boolean, Boolean)</a></span> - Static method in class biolockj.module.report.r.<a href="../biolockj/module/report/r/R_CalculateStats.html" title="class in biolockj.module.report.r">R_CalculateStats</a></dt>
<dd>
<div class="block">Get the stats file for the given fileType and taxonomy level.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Pipeline.html#getStatus--">getStatus()</a></span> - Static method in class biolockj.<a href="../biolockj/Pipeline.html" title="class in biolockj">Pipeline</a></dt>
<dd>
<div class="block">Return  if no pipelineException has been thrown, otherwise return
 </div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getString-biolockj.module.BioModule-java.lang.String-">getString(BioModule, String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Get property value as String.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/r/R_CalculateStats.html#getSuffix-java.lang.Boolean-java.lang.Boolean-">getSuffix(Boolean, Boolean)</a></span> - Static method in class biolockj.module.report.r.<a href="../biolockj/module/report/r/R_CalculateStats.html" title="class in biolockj.module.report.r">R_CalculateStats</a></dt>
<dd>
<div class="block">Get the file name suffix used to specify types of statistics.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModule.html#getSummary--">getSummary()</a></span> - Method in interface biolockj.module.<a href="../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></dt>
<dd>
<div class="block">Gets the BioModule execution summary, this is sent as part of the notification email, if configured.<br>
 Summary should not include data content, to avoid unintentional publication of confidential information.<br>
 However, meta-data such as number/size of files can be helpful during debug.<br></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModuleImpl.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a></dt>
<dd>
<div class="block">Returns summary message to be displayed by Email module so must not contain confidential info.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/ClassifierModuleImpl.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.classifier.<a href="../biolockj/module/classifier/ClassifierModuleImpl.html" title="class in biolockj.module.classifier">ClassifierModuleImpl</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/Humann2Classifier.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/Humann2Classifier.html" title="class in biolockj.module.classifier.wgs">Humann2Classifier</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/Demultiplexer.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.implicit.<a href="../biolockj/module/implicit/Demultiplexer.html" title="class in biolockj.module.implicit">Demultiplexer</a></dt>
<dd>
<div class="block">Produces initial count and demultiplexed output count summaries for forward/reverse reads.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/ImportMetadata.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.implicit.<a href="../biolockj/module/implicit/ImportMetadata.html" title="class in biolockj.module.implicit">ImportMetadata</a></dt>
<dd>
<div class="block">The metadata file can be updated several times during pipeline execution.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/ParserModuleImpl.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.implicit.parser.<a href="../biolockj/module/implicit/parser/ParserModuleImpl.html" title="class in biolockj.module.implicit.parser">ParserModuleImpl</a></dt>
<dd>
<div class="block">Produce summary message with min, max, mean, and median number of reads.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/wgs/Humann2Parser.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.implicit.parser.wgs.<a href="../biolockj/module/implicit/parser/wgs/Humann2Parser.html" title="class in biolockj.module.implicit.parser.wgs">Humann2Parser</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/BuildQiimeMapping.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/BuildQiimeMapping.html" title="class in biolockj.module.implicit.qiime">BuildQiimeMapping</a></dt>
<dd>
<div class="block">Message summarizes contents of <a href="../biolockj/util/SummaryUtil.html#getScriptDirSummary-biolockj.module.ScriptModule-"><code>scriptDir</code></a>, and
 status of the QIIME mapping files (exists or not).</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime">QiimeClassifier</a></dt>
<dd>
<div class="block">This method extends the classifier summary by adding the Qiime OTU summary metrics.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/RegisterNumReads.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.implicit.<a href="../biolockj/module/implicit/RegisterNumReads.html" title="class in biolockj.module.implicit">RegisterNumReads</a></dt>
<dd>
<div class="block">Produce summary message with min, max, mean, and median number of reads.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/Email.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.report.<a href="../biolockj/module/report/Email.html" title="class in biolockj.module.report">Email</a></dt>
<dd>
<div class="block">Summary simply reports the status.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/humann2/AddMetadataToPathwayTables.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.report.humann2.<a href="../biolockj/module/report/humann2/AddMetadataToPathwayTables.html" title="class in biolockj.module.report.humann2">AddMetadataToPathwayTables</a></dt>
<dd>
<div class="block">Produce summary message with min, max, mean, and median hit ratios</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/humann2/RemoveLowPathwayCounts.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.report.humann2.<a href="../biolockj/module/report/humann2/RemoveLowPathwayCounts.html" title="class in biolockj.module.report.humann2">RemoveLowPathwayCounts</a></dt>
<dd>
<div class="block">Produce summary message with min, max, mean, and median number of pathways.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/humann2/RemoveScarcePathwayCounts.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.report.humann2.<a href="../biolockj/module/report/humann2/RemoveScarcePathwayCounts.html" title="class in biolockj.module.report.humann2">RemoveScarcePathwayCounts</a></dt>
<dd>
<div class="block">Produce summary message with min, max, mean, and median number of pathways.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/JsonReport.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.report.<a href="../biolockj/module/report/JsonReport.html" title="class in biolockj.module.report">JsonReport</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/otu/CompileOtuCounts.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.report.otu.<a href="../biolockj/module/report/otu/CompileOtuCounts.html" title="class in biolockj.module.report.otu">CompileOtuCounts</a></dt>
<dd>
<div class="block">Add summary with unique OTU counts/level.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/otu/RarefyOtuCounts.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.report.otu.<a href="../biolockj/module/report/otu/RarefyOtuCounts.html" title="class in biolockj.module.report.otu">RarefyOtuCounts</a></dt>
<dd>
<div class="block">Produce summary message with min, max, mean, and median number of reads.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/otu/RemoveLowOtuCounts.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.report.otu.<a href="../biolockj/module/report/otu/RemoveLowOtuCounts.html" title="class in biolockj.module.report.otu">RemoveLowOtuCounts</a></dt>
<dd>
<div class="block">Produce summary message with min, max, mean, and median number of reads.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/otu/RemoveScarceOtuCounts.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.report.otu.<a href="../biolockj/module/report/otu/RemoveScarceOtuCounts.html" title="class in biolockj.module.report.otu">RemoveScarceOtuCounts</a></dt>
<dd>
<div class="block">Produce summary message with min, max, mean, and median number of reads.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/r/R_Module.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.report.r.<a href="../biolockj/module/report/r/R_Module.html" title="class in biolockj.module.report.r">R_Module</a></dt>
<dd>
<div class="block">Produce summary file counts for each file extension in the output directory and the number of log files in the
 temp directory.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/taxa/AddMetadataToTaxaTables.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.report.taxa.<a href="../biolockj/module/report/taxa/AddMetadataToTaxaTables.html" title="class in biolockj.module.report.taxa">AddMetadataToTaxaTables</a></dt>
<dd>
<div class="block">Produce summary message with min, max, mean, and median hit ratios</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/taxa/BuildTaxaTables.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.report.taxa.<a href="../biolockj/module/report/taxa/BuildTaxaTables.html" title="class in biolockj.module.report.taxa">BuildTaxaTables</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/taxa/NormalizeTaxaTables.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.report.taxa.<a href="../biolockj/module/report/taxa/NormalizeTaxaTables.html" title="class in biolockj.module.report.taxa">NormalizeTaxaTables</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/ScriptModuleImpl.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/ScriptModuleImpl.html" title="class in biolockj.module">ScriptModuleImpl</a></dt>
<dd>
<div class="block">Returns summary message to be displayed by Email module so must not contain confidential info.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/KneadData.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/KneadData.html" title="class in biolockj.module.seq">KneadData</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/Multiplexer.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/Multiplexer.html" title="class in biolockj.module.seq">Multiplexer</a></dt>
<dd>
<div class="block">Produce a summary message with counts on total number of reads and number reads multiplexed.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/PearMergeReads.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/PearMergeReads.html" title="class in biolockj.module.seq">PearMergeReads</a></dt>
<dd>
<div class="block">Produce summary message with min, max, mean, and median number of reads.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/RarefySeqs.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/RarefySeqs.html" title="class in biolockj.module.seq">RarefySeqs</a></dt>
<dd>
<div class="block">Produce summary message with min, max, mean, and median number of reads.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/SeqFileValidator.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/SeqFileValidator.html" title="class in biolockj.module.seq">SeqFileValidator</a></dt>
<dd>
<div class="block">Produce a summary message with counts on total number of reads and number of valid reads containing a barcode
 defined in the metadata file.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/TrimPrimers.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/TrimPrimers.html" title="class in biolockj.module.seq">TrimPrimers</a></dt>
<dd>
<div class="block">Output the summary messages generated by the module.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SummaryUtil.html#getSummary--">getSummary()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SummaryUtil.html" title="class in biolockj.util">SummaryUtil</a></dt>
<dd>
<div class="block">Pipeline execution summary.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SummaryUtil.html#getSummaryFile--">getSummaryFile()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SummaryUtil.html" title="class in biolockj.util">SummaryUtil</a></dt>
<dd>
<div class="block">Getter for the summary file.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/otu/CompileOtuCounts.html#getSummaryOtuFile--">getSummaryOtuFile()</a></span> - Method in class biolockj.module.report.otu.<a href="../biolockj/module/report/otu/CompileOtuCounts.html" title="class in biolockj.module.report.otu">CompileOtuCounts</a></dt>
<dd>
<div class="block">Get the summary output file</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getSystemFilePath-java.lang.String-">getSystemFilePath(String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Return filePath with system parameter values replaced.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/MetaUtil.html#getSystemMetaCol-biolockj.module.BioModule-java.lang.String-">getSystemMetaCol(BioModule, String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/MetaUtil.html" title="class in biolockj.util">MetaUtil</a></dt>
<dd>
<div class="block">Return a system generated metadata column name based on the module status.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/JsonNode.html#getTaxa--">getTaxa()</a></span> - Method in class biolockj.node.<a href="../biolockj/node/JsonNode.html" title="class in biolockj.node">JsonNode</a></dt>
<dd>
<div class="block">Getter method for taxa</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/TaxaUtil.html#getTaxaByLevel-java.lang.String-">getTaxaByLevel(String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/TaxaUtil.html" title="class in biolockj.util">TaxaUtil</a></dt>
<dd>
<div class="block">Return a map of the given otu parameter split by level.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/TaxaUtil.html#getTaxaLevels--">getTaxaLevels()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/TaxaUtil.html" title="class in biolockj.util">TaxaUtil</a></dt>
<dd>
<div class="block">Method ensures taxonomy levels are returned in descending order from top to bottom.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/TaxaUtil.html#getTaxaLevelSpan--">getTaxaLevelSpan()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/TaxaUtil.html" title="class in biolockj.util">TaxaUtil</a></dt>
<dd>
<div class="block">Return taxa levels from top to bottom level, including in-between levels not configured as part of
 </div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/OtuNode.html#getTaxaMap--">getTaxaMap()</a></span> - Method in interface biolockj.node.<a href="../biolockj/node/OtuNode.html" title="interface in biolockj.node">OtuNode</a></dt>
<dd>
<div class="block">Gets the map holding level-specific OTU names</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/OtuNodeImpl.html#getTaxaMap--">getTaxaMap()</a></span> - Method in class biolockj.node.<a href="../biolockj/node/OtuNodeImpl.html" title="class in biolockj.node">OtuNodeImpl</a></dt>
<dd>
<div class="block">This implementation ensures all levels between top and bottom taxonomy levels are complete.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/TaxaUtil.html#getTaxaName-java.lang.String-java.lang.String-">getTaxaName(String, String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/TaxaUtil.html" title="class in biolockj.util">TaxaUtil</a></dt>
<dd>
<div class="block">Extract a taxonomy name at the given level from the given OTU.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/TaxaUtil.html#getTaxonomyTableFile-java.io.File-java.lang.String-java.lang.String-">getTaxonomyTableFile(File, String, String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/TaxaUtil.html" title="class in biolockj.util">TaxaUtil</a></dt>
<dd>
<div class="block">Create File object of a taxonomy table at the given level, with the given suffix, in the given directory dir.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/TaxaUtil.html#getTaxonomyTableLevel-java.io.File-">getTaxonomyTableLevel(File)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/TaxaUtil.html" title="class in biolockj.util">TaxaUtil</a></dt>
<dd>
<div class="block">Extract the table taxonomy level from an existing taxonomy table file.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModule.html#getTempDir--">getTempDir()</a></span> - Method in interface biolockj.module.<a href="../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></dt>
<dd>
<div class="block">Contains intermediate files generated by the module but not used by the next BioModule.<br>
 The files may contain supplementary information or data that may be helpful during debug or recovery.<br>
 If <a href="../biolockj/Constants.html#PIPELINE_DELETE_TEMP_FILES">"pipeline.deleteTempFiles"</a> = <a href="../biolockj/Constants.html#TRUE">"Y"</a>, successful pipelines
 delete this directory.<br></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModuleImpl.html#getTempDir--">getTempDir()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a></dt>
<dd>
<div class="block">Returns moduleDir/temp for intermediate files.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/r/R_Module.html#getTimeout--">getTimeout()</a></span> - Method in class biolockj.module.report.r.<a href="../biolockj/module/report/r/R_Module.html" title="class in biolockj.module.report.r">R_Module</a></dt>
<dd>
<div class="block">The R Script should run quickly, timeout = 10 minutes appears to work well.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/ScriptModule.html#getTimeout--">getTimeout()</a></span> - Method in interface biolockj.module.<a href="../biolockj/module/ScriptModule.html" title="interface in biolockj.module">ScriptModule</a></dt>
<dd>
<div class="block">ScriptModule that run scripts can opt to set a timeout (such as used by the R script).</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/ScriptModuleImpl.html#getTimeout--">getTimeout()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/ScriptModuleImpl.html" title="class in biolockj.module">ScriptModuleImpl</a></dt>
<dd>
<div class="block">Default behavior is for scripts to run indefinitely (no timeout).</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getTreeSet-biolockj.module.BioModule-java.lang.String-">getTreeSet(BioModule, String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Parse comma-separated property value to build an ordered Set</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/MetaUtil.html#getUniqueFieldValues-java.lang.String-boolean-">getUniqueFieldValues(String, boolean)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/MetaUtil.html" title="class in biolockj.util">MetaUtil</a></dt>
<dd>
<div class="block">Count the number of unique values in the given field.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/otu/RemoveScarceOtuCounts.html#getUpdatedOtuCounts-java.util.TreeMap-java.util.TreeMap-">getUpdatedOtuCounts(TreeMap&lt;String, TreeMap&lt;String, Integer&gt;&gt;, TreeMap&lt;String, TreeSet&lt;String&gt;&gt;)</a></span> - Method in class biolockj.module.report.otu.<a href="../biolockj/module/report/otu/RemoveScarceOtuCounts.html" title="class in biolockj.module.report.otu">RemoveScarceOtuCounts</a></dt>
<dd>
<div class="block">Remove scarce OTUs from the sampleOtuCounts and return it.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getUsedProps--">getUsedProps()</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Cache of the properties used in this pipeline.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/Demultiplexer.html#getValidFwHeaders--">getValidFwHeaders()</a></span> - Method in class biolockj.module.implicit.<a href="../biolockj/module/implicit/Demultiplexer.html" title="class in biolockj.module.implicit">Demultiplexer</a></dt>
<dd>
<div class="block">Get valid forward read headers that belong to reads with a valid barcode or sample identifier.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/Demultiplexer.html#getValidHeaders--">getValidHeaders()</a></span> - Method in class biolockj.module.implicit.<a href="../biolockj/module/implicit/Demultiplexer.html" title="class in biolockj.module.implicit">Demultiplexer</a></dt>
<dd>
<div class="block">This method obtains all valid headers for the forward reads, and returns only headers that also have a matching
 reverse read</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BioLockJUtil.html#getVersion--">getVersion()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BioLockJUtil.html" title="class in biolockj.util">BioLockJUtil</a></dt>
<dd>
<div class="block">Method returns the current version of BioLockJ.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#getVsearchParams--">getVsearchParams()</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime">QiimeClassifier</a></dt>
<dd>
<div class="block">Return runtime parameters for <a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#EXE_VSEARCH_PARAMS">"exe.vsearchParams"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/r16s/QiimeClosedRefClassifier.html#getWorkerScriptFunctions--">getWorkerScriptFunctions()</a></span> - Method in class biolockj.module.classifier.r16s.<a href="../biolockj/module/classifier/r16s/QiimeClosedRefClassifier.html" title="class in biolockj.module.classifier.r16s">QiimeClosedRefClassifier</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/r16s/RdpClassifier.html#getWorkerScriptFunctions--">getWorkerScriptFunctions()</a></span> - Method in class biolockj.module.classifier.r16s.<a href="../biolockj/module/classifier/r16s/RdpClassifier.html" title="class in biolockj.module.classifier.r16s">RdpClassifier</a></dt>
<dd>
<div class="block">This method generates the required bash functions: <a href="../biolockj/module/classifier/r16s/RdpClassifier.html#FUNCTION_RDP">"runRdp"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/Humann2Classifier.html#getWorkerScriptFunctions--">getWorkerScriptFunctions()</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/Humann2Classifier.html" title="class in biolockj.module.classifier.wgs">Humann2Classifier</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/Kraken2Classifier.html#getWorkerScriptFunctions--">getWorkerScriptFunctions()</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/Kraken2Classifier.html" title="class in biolockj.module.classifier.wgs">Kraken2Classifier</a></dt>
<dd>
<div class="block">This method generates the required bash function: <a href="../biolockj/module/classifier/wgs/Kraken2Classifier.html#FUNCTION_KRAKEN">"runKraken2"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/KrakenClassifier.html#getWorkerScriptFunctions--">getWorkerScriptFunctions()</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/KrakenClassifier.html" title="class in biolockj.module.classifier.wgs">KrakenClassifier</a></dt>
<dd>
<div class="block">This method generates the required bash functions: <a href="../biolockj/module/classifier/wgs/KrakenClassifier.html#FUNCTION_TRANSLATE">"translate"</a> and <a href="../biolockj/module/classifier/wgs/KrakenClassifier.html#FUNCTION_KRAKEN">"runKraken"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/Metaphlan2Classifier.html#getWorkerScriptFunctions--">getWorkerScriptFunctions()</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/Metaphlan2Classifier.html" title="class in biolockj.module.classifier.wgs">Metaphlan2Classifier</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/BuildQiimeMapping.html#getWorkerScriptFunctions--">getWorkerScriptFunctions()</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/BuildQiimeMapping.html" title="class in biolockj.module.implicit.qiime">BuildQiimeMapping</a></dt>
<dd>
<div class="block">This method generates the bash function used to reorder columns as per QIIME requirements:
 <a href="../biolockj/module/implicit/qiime/BuildQiimeMapping.html#FUNCTION_REORDER_FIELDS">"reorderColumns"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#getWorkerScriptFunctions--">getWorkerScriptFunctions()</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime">QiimeClassifier</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/JavaModuleImpl.html#getWorkerScriptFunctions--">getWorkerScriptFunctions()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/JavaModuleImpl.html" title="class in biolockj.module">JavaModuleImpl</a></dt>
<dd>
<div class="block">If in Docker mode, set <a href="../biolockj/module/JavaModuleImpl.html#BLJ_OPTIONS">"BLJ_OPTIONS"</a> which will be reference in every worker script when running BioLockJ
 in direct mode.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/r/R_Module.html#getWorkerScriptFunctions--">getWorkerScriptFunctions()</a></span> - Method in class biolockj.module.report.r.<a href="../biolockj/module/report/r/R_Module.html" title="class in biolockj.module.report.r">R_Module</a></dt>
<dd>
<div class="block">This method generates the bash function that calls the R script: runScript.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/ScriptModule.html#getWorkerScriptFunctions--">getWorkerScriptFunctions()</a></span> - Method in interface biolockj.module.<a href="../biolockj/module/ScriptModule.html" title="interface in biolockj.module">ScriptModule</a></dt>
<dd>
<div class="block">Method returns bash script lines used to build the functions called by the worker scripts.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/ScriptModuleImpl.html#getWorkerScriptFunctions--">getWorkerScriptFunctions()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/ScriptModuleImpl.html" title="class in biolockj.module">ScriptModuleImpl</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/AwkFastaConverter.html#getWorkerScriptFunctions--">getWorkerScriptFunctions()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/AwkFastaConverter.html" title="class in biolockj.module.seq">AwkFastaConverter</a></dt>
<dd>
<div class="block">This method generates the required bash functions used by the module scripts.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/Gunzipper.html#getWorkerScriptFunctions--">getWorkerScriptFunctions()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/Gunzipper.html" title="class in biolockj.module.seq">Gunzipper</a></dt>
<dd>
<div class="block">This method generates the bash function: <a href="../biolockj/module/seq/Gunzipper.html#FUNCTION_GUNZIP">"openZip"</a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/KneadData.html#getWorkerScriptFunctions--">getWorkerScriptFunctions()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/KneadData.html" title="class in biolockj.module.seq">KneadData</a></dt>
<dd>
<div class="block">This method generates the worker script function: <a href="../biolockj/module/seq/KneadData.html#FUNCTION_SANATIZE">"sanatizeData"</a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/PearMergeReads.html#getWorkerScriptFunctions--">getWorkerScriptFunctions()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/PearMergeReads.html" title="class in biolockj.module.seq">PearMergeReads</a></dt>
<dd>
<div class="block">This method generates the required bash functions used by the worker scripts: <a href="../biolockj/module/seq/PearMergeReads.html#FUNCTION_PEAR_MERGE">"mergeReads"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BashScriptBuilder.html#getWorkerScriptLines-java.util.List-">getWorkerScriptLines(List&lt;String&gt;)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BashScriptBuilder.html" title="class in biolockj.util">BashScriptBuilder</a></dt>
<dd>
<div class="block">Pass each line and the current line number to <a href="../biolockj/util/BashScriptBuilder.html#FUNCTION_EXECUTE">"execute"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BashScriptBuilder.html#getWorkerScriptPath-biolockj.module.ScriptModule-java.lang.String-">getWorkerScriptPath(ScriptModule, String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BashScriptBuilder.html" title="class in biolockj.util">BashScriptBuilder</a></dt>
<dd>
<div class="block">Build the file path for the numbered worker script.</div>
</dd>
<dt><a href="../biolockj/module/seq/Gunzipper.html" title="class in biolockj.module.seq"><span class="typeNameLink">Gunzipper</span></a> - Class in <a href="../biolockj/module/seq/package-summary.html">biolockj.module.seq</a></dt>
<dd>
<div class="block">This BioModule uses gzip to decompress input sequence files.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/Gunzipper.html#Gunzipper--">Gunzipper()</a></span> - Constructor for class biolockj.module.seq.<a href="../biolockj/module/seq/Gunzipper.html" title="class in biolockj.module.seq">Gunzipper</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#GZIP_EXT">GZIP_EXT</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block">Gzip compressed file extension constant: ".gz"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModuleImpl.html#GZIP_EXT">GZIP_EXT</a></span> - Static variable in class biolockj.module.<a href="../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a></dt>
<dd>
<div class="block">BioLockJ gzip file extension constant: ".gz"</div>
</dd>
</dl>
<a href="index-1.html">A</a>&nbsp;<a href="index-2.html">B</a>&nbsp;<a href="index-3.html">C</a>&nbsp;<a href="index-4.html">D</a>&nbsp;<a href="index-5.html">E</a>&nbsp;<a href="index-6.html">F</a>&nbsp;<a href="index-7.html">G</a>&nbsp;<a href="index-8.html">H</a>&nbsp;<a href="index-9.html">I</a>&nbsp;<a href="index-10.html">J</a>&nbsp;<a href="index-11.html">K</a>&nbsp;<a href="index-12.html">L</a>&nbsp;<a href="index-13.html">M</a>&nbsp;<a href="index-14.html">N</a>&nbsp;<a href="index-15.html">O</a>&nbsp;<a href="index-16.html">P</a>&nbsp;<a href="index-17.html">Q</a>&nbsp;<a href="index-18.html">R</a>&nbsp;<a href="index-19.html">S</a>&nbsp;<a href="index-20.html">T</a>&nbsp;<a href="index-21.html">U</a>&nbsp;<a href="index-22.html">V</a>&nbsp;<a href="index-23.html">W</a>&nbsp;</div>
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar.bottom">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.bottom.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../overview-summary.html">Overview</a></li>
<li>Package</li>
<li>Class</li>
<li>Use</li>
<li><a href="../overview-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li class="navBarCell1Rev">Index</li>
<li><a href="../help-doc.html">Help</a></li>
<li>&nbsp;&nbsp;</li>
<li><a href="https://github.com/msioda/BioLockJ/wiki" target="_top"><strong>RETURN to WIKI</strong></a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="index-6.html">Prev Letter</a></li>
<li><a href="index-8.html">Next Letter</a></li>
</ul>
<ul class="navList">
<li><a href="../index.html?index-files/index-7.html" target="_top">Frames</a></li>
<li><a href="index-7.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip.navbar.bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
