<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (1.8.0_144) on Wed Feb 27 01:20:49 EST 2019 -->
<title>H-Index</title>
<meta name="date" content="2019-02-27">
<link rel="stylesheet" type="text/css" href="../stylesheet.css" title="Style">
<script type="text/javascript" src="../script.js"></script>
</head>
<body>
<script type="text/javascript"><!--
    try {
        if (location.href.indexOf('is-external=true') == -1) {
            parent.document.title="H-Index";
        }
    }
    catch(err) {
    }
//-->
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar.top">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.top.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../overview-summary.html">Overview</a></li>
<li>Package</li>
<li>Class</li>
<li>Use</li>
<li><a href="../overview-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li class="navBarCell1Rev">Index</li>
<li><a href="../help-doc.html">Help</a></li>
<li>&nbsp;&nbsp;</li>
<li><a href="https://github.com/msioda/BioLockJ/wiki" target="_top"><strong>RETURN to WIKI</strong></a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="index-7.html">Prev Letter</a></li>
<li><a href="index-9.html">Next Letter</a></li>
</ul>
<ul class="navList">
<li><a href="../index.html?index-files/index-8.html" target="_top">Frames</a></li>
<li><a href="index-8.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip.navbar.top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<div class="contentContainer"><a href="index-1.html">A</a>&nbsp;<a href="index-2.html">B</a>&nbsp;<a href="index-3.html">C</a>&nbsp;<a href="index-4.html">D</a>&nbsp;<a href="index-5.html">E</a>&nbsp;<a href="index-6.html">F</a>&nbsp;<a href="index-7.html">G</a>&nbsp;<a href="index-8.html">H</a>&nbsp;<a href="index-9.html">I</a>&nbsp;<a href="index-10.html">J</a>&nbsp;<a href="index-11.html">K</a>&nbsp;<a href="index-12.html">L</a>&nbsp;<a href="index-13.html">M</a>&nbsp;<a href="index-14.html">N</a>&nbsp;<a href="index-15.html">O</a>&nbsp;<a href="index-16.html">P</a>&nbsp;<a href="index-17.html">Q</a>&nbsp;<a href="index-18.html">R</a>&nbsp;<a href="index-19.html">S</a>&nbsp;<a href="index-20.html">T</a>&nbsp;<a href="index-21.html">U</a>&nbsp;<a href="index-22.html">V</a>&nbsp;<a href="index-23.html">W</a>&nbsp;<a name="I:H">
<!--   -->
</a>
<h2 class="title">H</h2>
<dl>
<dt><span class="memberNameLink"><a href="../biolockj/util/MetaUtil.html#hasColumn-java.lang.String-">hasColumn(String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/MetaUtil.html" title="class in biolockj.util">MetaUtil</a></dt>
<dd>
<div class="block">Check if columnName exists in the current metadata file.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/ModuleUtil.html#hasExecuted-biolockj.module.BioModule-">hasExecuted(BioModule)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/ModuleUtil.html" title="class in biolockj.util">ModuleUtil</a></dt>
<dd>
<div class="block">Return TRUE if module has executed.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/JsonNode.html#hashCode--">hashCode()</a></span> - Method in class biolockj.node.<a href="../biolockj/node/JsonNode.html" title="class in biolockj.node">JsonNode</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/ScriptModuleImpl.html#hasScripts--">hasScripts()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/ScriptModuleImpl.html" title="class in biolockj.module">ScriptModuleImpl</a></dt>
<dd>
<div class="block">Check if module produced any scripts</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DemuxUtil.html#hasValidBarcodes--">hasValidBarcodes()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/DemuxUtil.html" title="class in biolockj.util">DemuxUtil</a></dt>
<dd>
<div class="block">Check for the existance of the barcode column.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/taxa/AddMetadataToTaxaTables.html#HIT_RATIO">HIT_RATIO</a></span> - Static variable in class biolockj.module.report.taxa.<a href="../biolockj/module/report/taxa/AddMetadataToTaxaTables.html" title="class in biolockj.module.report.taxa">AddMetadataToTaxaTables</a></dt>
<dd>
<div class="block">Metadata column name for column that stores the calculation for:
 <a href="../biolockj/module/implicit/parser/ParserModuleImpl.html#getOtuCountField--"><code>ParserModuleImpl.getOtuCountField()</code></a>/
 <a href="../biolockj/module/implicit/RegisterNumReads.html#getNumReadFieldName--"><code>RegisterNumReads.getNumReadFieldName()</code></a>: "Hit_Ratio".</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#HN2_DISABLE_GENE_FAMILIES">HN2_DISABLE_GENE_FAMILIES</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> Boolean property to disable HumanN2 Gene Family report:
 "humann2.disableGeneFamilies"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#HN2_DISABLE_PATH_ABUNDANCE">HN2_DISABLE_PATH_ABUNDANCE</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> Boolean property to disable HumanN2 Pathway Abundance report:
 "humann2.disablePathAbundance"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#HN2_DISABLE_PATH_COVERAGE">HN2_DISABLE_PATH_COVERAGE</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> Boolean property to disable HumanN2 Pathway Coverage report:
 "humann2.disablePathCoverage"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#HN2_GENE_FAM_SUM">HN2_GENE_FAM_SUM</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block">HumanN2 file suffix identifier for Gene Family Summary report: "geneFam"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/Humann2Classifier.html#HN2_NUCL_DB">HN2_NUCL_DB</a></span> - Static variable in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/Humann2Classifier.html" title="class in biolockj.module.classifier.wgs">Humann2Classifier</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> Directory property may contain multiple nucleotide database files: "humann2.nuclDB"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#HN2_PATH_ABUND_SUM">HN2_PATH_ABUND_SUM</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block">HumanN2 file suffix identifier for Pathway Abundance Summary report: "pAbund"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#HN2_PATH_COVG_SUM">HN2_PATH_COVG_SUM</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block">HumanN2 file suffix identifier for Pathway Coverage Summary report: "pCovg"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/Humann2Classifier.html#HN2_PROT_DB">HN2_PROT_DB</a></span> - Static variable in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/Humann2Classifier.html" title="class in biolockj.module.classifier.wgs">Humann2Classifier</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> Directory property may contain protein nucleotide database files: "humann2.protDB"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#HN2_TOTAL_PATH_COUNT">HN2_TOTAL_PATH_COUNT</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block">HumanN2 meta column to store the total pathway count/sample: "Total_Pathways"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#HN2_UNIQUE_PATH_COUNT">HN2_UNIQUE_PATH_COUNT</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block">HumanN2 meta column to store the unique pathway count/sample: <a href="../biolockj/Constants.html#HN2_TOTAL_PATH_COUNT">"Total_Pathways"</a></div>
</dd>
<dt><a href="../biolockj/module/classifier/wgs/Humann2Classifier.html" title="class in biolockj.module.classifier.wgs"><span class="typeNameLink">Humann2Classifier</span></a> - Class in <a href="../biolockj/module/classifier/wgs/package-summary.html">biolockj.module.classifier.wgs</a></dt>
<dd>
<div class="block">This BioModule runs biobakery humann2 program to generate the HMP Unified Metabolic Analysis Network<br>
 HUMAnN is a pipeline for efficiently and accurately profiling the presence/absence and abundance of microbial
 pathways in a community from metagenomic or metatranscriptomic sequencing data (typically millions of short DNA/RNA
 reads).</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/Humann2Classifier.html#Humann2Classifier--">Humann2Classifier()</a></span> - Constructor for class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/Humann2Classifier.html" title="class in biolockj.module.classifier.wgs">Humann2Classifier</a></dt>
<dd>&nbsp;</dd>
<dt><a href="../biolockj/module/report/humann2/Humann2CountModule.html" title="class in biolockj.module.report.humann2"><span class="typeNameLink">Humann2CountModule</span></a> - Class in <a href="../biolockj/module/report/humann2/package-summary.html">biolockj.module.report.humann2</a></dt>
<dd>
<div class="block">TBD</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/humann2/Humann2CountModule.html#Humann2CountModule--">Humann2CountModule()</a></span> - Constructor for class biolockj.module.report.humann2.<a href="../biolockj/module/report/humann2/Humann2CountModule.html" title="class in biolockj.module.report.humann2">Humann2CountModule</a></dt>
<dd>&nbsp;</dd>
<dt><a href="../biolockj/module/implicit/parser/wgs/Humann2Parser.html" title="class in biolockj.module.implicit.parser.wgs"><span class="typeNameLink">Humann2Parser</span></a> - Class in <a href="../biolockj/module/implicit/parser/wgs/package-summary.html">biolockj.module.implicit.parser.wgs</a></dt>
<dd>
<div class="block">This BioModules parses Humann2Classifier output reports to build standard OTU abundance tables.<br>
 Samples IDs are found in the column headers starting with the 2nd column.<br>
 The count type depends on the HumanN2 config properties.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/wgs/Humann2Parser.html#Humann2Parser--">Humann2Parser()</a></span> - Constructor for class biolockj.module.implicit.parser.wgs.<a href="../biolockj/module/implicit/parser/wgs/Humann2Parser.html" title="class in biolockj.module.implicit.parser.wgs">Humann2Parser</a></dt>
<dd>&nbsp;</dd>
</dl>
<a href="index-1.html">A</a>&nbsp;<a href="index-2.html">B</a>&nbsp;<a href="index-3.html">C</a>&nbsp;<a href="index-4.html">D</a>&nbsp;<a href="index-5.html">E</a>&nbsp;<a href="index-6.html">F</a>&nbsp;<a href="index-7.html">G</a>&nbsp;<a href="index-8.html">H</a>&nbsp;<a href="index-9.html">I</a>&nbsp;<a href="index-10.html">J</a>&nbsp;<a href="index-11.html">K</a>&nbsp;<a href="index-12.html">L</a>&nbsp;<a href="index-13.html">M</a>&nbsp;<a href="index-14.html">N</a>&nbsp;<a href="index-15.html">O</a>&nbsp;<a href="index-16.html">P</a>&nbsp;<a href="index-17.html">Q</a>&nbsp;<a href="index-18.html">R</a>&nbsp;<a href="index-19.html">S</a>&nbsp;<a href="index-20.html">T</a>&nbsp;<a href="index-21.html">U</a>&nbsp;<a href="index-22.html">V</a>&nbsp;<a href="index-23.html">W</a>&nbsp;</div>
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar.bottom">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.bottom.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../overview-summary.html">Overview</a></li>
<li>Package</li>
<li>Class</li>
<li>Use</li>
<li><a href="../overview-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li class="navBarCell1Rev">Index</li>
<li><a href="../help-doc.html">Help</a></li>
<li>&nbsp;&nbsp;</li>
<li><a href="https://github.com/msioda/BioLockJ/wiki" target="_top"><strong>RETURN to WIKI</strong></a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="index-7.html">Prev Letter</a></li>
<li><a href="index-9.html">Next Letter</a></li>
</ul>
<ul class="navList">
<li><a href="../index.html?index-files/index-8.html" target="_top">Frames</a></li>
<li><a href="index-8.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip.navbar.bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
