<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (1.8.0_144) on Wed Oct 24 07:25:44 EDT 2018 -->
<title>G-Index</title>
<meta name="date" content="2018-10-24">
<link rel="stylesheet" type="text/css" href="../stylesheet.css" title="Style">
<script type="text/javascript" src="../script.js"></script>
</head>
<body>
<script type="text/javascript"><!--
    try {
        if (location.href.indexOf('is-external=true') == -1) {
            parent.document.title="G-Index";
        }
    }
    catch(err) {
    }
//-->
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar.top">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.top.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../overview-summary.html">Overview</a></li>
<li>Package</li>
<li>Class</li>
<li>Use</li>
<li><a href="../overview-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li class="navBarCell1Rev">Index</li>
<li><a href="../help-doc.html">Help</a></li>
<li>&nbsp;&nbsp;</li>
<li><a href="https://github.com/msioda/BioLockJ/wiki" target="_top"><strong>RETURN to WIKI</strong></a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="index-6.html">Prev Letter</a></li>
<li><a href="index-8.html">Next Letter</a></li>
</ul>
<ul class="navList">
<li><a href="../index.html?index-files/index-7.html" target="_top">Frames</a></li>
<li><a href="index-7.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip.navbar.top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<div class="contentContainer"><a href="index-1.html">A</a>&nbsp;<a href="index-2.html">B</a>&nbsp;<a href="index-3.html">C</a>&nbsp;<a href="index-4.html">D</a>&nbsp;<a href="index-5.html">E</a>&nbsp;<a href="index-6.html">F</a>&nbsp;<a href="index-7.html">G</a>&nbsp;<a href="index-8.html">H</a>&nbsp;<a href="index-9.html">I</a>&nbsp;<a href="index-10.html">J</a>&nbsp;<a href="index-11.html">K</a>&nbsp;<a href="index-12.html">L</a>&nbsp;<a href="index-13.html">M</a>&nbsp;<a href="index-14.html">N</a>&nbsp;<a href="index-15.html">O</a>&nbsp;<a href="index-16.html">P</a>&nbsp;<a href="index-17.html">Q</a>&nbsp;<a href="index-18.html">R</a>&nbsp;<a href="index-19.html">S</a>&nbsp;<a href="index-20.html">T</a>&nbsp;<a href="index-21.html">U</a>&nbsp;<a href="index-22.html">V</a>&nbsp;<a href="index-23.html">W</a>&nbsp;<a name="I:G">
<!--   -->
</a>
<h2 class="title">G</h2>
<dl>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/AddMetadataToOtuTables.html#generateMergedTables--">generateMergedTables()</a></span> - Method in class biolockj.module.report.<a href="../biolockj/module/report/AddMetadataToOtuTables.html" title="class in biolockj.module.report">AddMetadataToOtuTables</a></dt>
<dd>
<div class="block">Create the merged metadata tables.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#GENUS">GENUS</a></span> - Static variable in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">One of the <a href="../biolockj/Config.html#REPORT_TAXONOMY_LEVELS">"report.taxonomyLevels"</a> options: "genus"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/OtuNodeImpl.html#GENUS_DELIM">GENUS_DELIM</a></span> - Static variable in class biolockj.node.<a href="../biolockj/node/OtuNodeImpl.html" title="class in biolockj.node">OtuNodeImpl</a></dt>
<dd>
<div class="block">Standard classifier output level delimiter for GENUS</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Log.html#get-java.lang.Class-">get(Class&lt;?&gt;)</a></span> - Static method in class biolockj.<a href="../biolockj/Log.html" title="class in biolockj">Log</a></dt>
<dd>
<div class="block">Returns the Logger for the callingClass.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#getAdminEmailPassword--">getAdminEmailPassword()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">Runtime property getter for <a href="../biolockj/util/RuntimeParamUtil.html#PASSWORD_FLAG">"-p"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#getBaseDir--">getBaseDir()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">Runtime property getter for <a href="../biolockj/util/RuntimeParamUtil.html#BASE_DIR_FLAG">"-b"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/r16s/QiimeClosedRefClassifier.html#getBatch-java.util.List-int-int-">getBatch(List&lt;String&gt;, int, int)</a></span> - Method in class biolockj.module.classifier.r16s.<a href="../biolockj/module/classifier/r16s/QiimeClosedRefClassifier.html" title="class in biolockj.module.classifier.r16s">QiimeClosedRefClassifier</a></dt>
<dd>
<div class="block">Method returns batch script lines for one batch.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/r16s/QiimeClosedRefClassifier.html#getBatchDir-int-">getBatchDir(int)</a></span> - Method in class biolockj.module.classifier.r16s.<a href="../biolockj/module/classifier/r16s/QiimeClosedRefClassifier.html" title="class in biolockj.module.classifier.r16s">QiimeClosedRefClassifier</a></dt>
<dd>
<div class="block">Get the pipeline/module/temp/batch_# dir.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/r16s/QiimeClosedRefClassifier.html#getBatchFastaDir-int-">getBatchFastaDir(int)</a></span> - Method in class biolockj.module.classifier.r16s.<a href="../biolockj/module/classifier/r16s/QiimeClosedRefClassifier.html" title="class in biolockj.module.classifier.r16s">QiimeClosedRefClassifier</a></dt>
<dd>
<div class="block">Get the pipeline/module/temp/batch_#/fasta directory.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RMetaUtil.html#getBinaryFields--">getBinaryFields()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RMetaUtil.html" title="class in biolockj.util">RMetaUtil</a></dt>
<dd>
<div class="block">Get the <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/util/RMetaUtil.html#R_NOMINAL_FIELDS">"r.nominalFields"</a> fields containing only 2 non-numeric values.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getBoolean-java.lang.String-">getBoolean(String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Parse property value (Y or N) to return boolean, if not found, return false;</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/ClassifierModule.html#getClassifierExe--">getClassifierExe()</a></span> - Method in interface biolockj.module.classifier.<a href="../biolockj/module/classifier/ClassifierModule.html" title="interface in biolockj.module.classifier">ClassifierModule</a></dt>
<dd>
<div class="block">Get the executable from <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> <a href="../biolockj/module/classifier/ClassifierModule.html#EXE_CLASSIFIER">"exe.classifier"</a> required to classify the samples.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/ClassifierModuleImpl.html#getClassifierExe--">getClassifierExe()</a></span> - Method in class biolockj.module.classifier.<a href="../biolockj/module/classifier/ClassifierModuleImpl.html" title="class in biolockj.module.classifier">ClassifierModuleImpl</a></dt>
<dd>
<div class="block">Get the classifier executable command from the Config.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#getClassifierExe--">getClassifierExe()</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime">QiimeClassifier</a></dt>
<dd>
<div class="block">QIIME does not use this method, instead specific scripts are called based on OTU picking method.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/ClassifierModule.html#getClassifierParams--">getClassifierParams()</a></span> - Method in interface biolockj.module.classifier.<a href="../biolockj/module/classifier/ClassifierModule.html" title="interface in biolockj.module.classifier">ClassifierModule</a></dt>
<dd>
<div class="block">Get optional list of parameters from <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> <a href="../biolockj/module/classifier/ClassifierModule.html#EXE_CLASSIFIER_PARAMS">"exe.classifierParams"</a> to append whenever
 the classifier executable is called.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/ClassifierModuleImpl.html#getClassifierParams--">getClassifierParams()</a></span> - Method in class biolockj.module.classifier.<a href="../biolockj/module/classifier/ClassifierModuleImpl.html" title="class in biolockj.module.classifier">ClassifierModuleImpl</a></dt>
<dd>
<div class="block">Get optional classifier switches by appending "--" to each value in the list parameter if defined in the Config
 file.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/SlimmClassifier.html#getClassifierParams--">getClassifierParams()</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/SlimmClassifier.html" title="class in biolockj.module.classifier.wgs">SlimmClassifier</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#getClassifierParams--">getClassifierParams()</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime">QiimeClassifier</a></dt>
<dd>
<div class="block">Get optional list of parameters from <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>
 <a href="../biolockj/module/classifier/ClassifierModule.html#EXE_CLASSIFIER_PARAMS">"exe.classifierParams"</a> to append whenever the OTU picking
 script is used.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/ClassifierModuleImpl.html#getClassifierType--">getClassifierType()</a></span> - Method in class biolockj.module.classifier.<a href="../biolockj/module/classifier/ClassifierModuleImpl.html" title="class in biolockj.module.classifier">ClassifierModuleImpl</a></dt>
<dd>
<div class="block">This method returns the classifier class name in lower case, after "classifier" is removed.<br>
 The remaining text should uniquely identify the name of the program.<br>
 The basic deployment will return one of: (rdp, qiime, kraken, metaphlan, or slimm).<br></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BioLockJUtil.html#getClassNames-java.util.Collection-">getClassNames(Collection&lt;?&gt;)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BioLockJUtil.html" title="class in biolockj.util">BioLockJUtil</a></dt>
<dd>
<div class="block">Return an ordered list of the class names from the input collection.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DownloadUtil.html#getClusterUser--">getClusterUser()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/DownloadUtil.html" title="class in biolockj.util">DownloadUtil</a></dt>
<dd>
<div class="block">If System.user.home is a file path (like /usr/home/johnDoe), return last directory name (johnDoe).</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BioLockJUtil.html#getCollectionAsString-java.util.Collection-">getCollectionAsString(Collection&lt;?&gt;)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BioLockJUtil.html" title="class in biolockj.util">BioLockJUtil</a></dt>
<dd>
<div class="block">Concatenate data and return as a comma separated String.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/MetaUtil.html#getCommentChar--">getCommentChar()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/MetaUtil.html" title="class in biolockj.util">MetaUtil</a></dt>
<dd>
<div class="block">Get the comment delimeter (or default empty string)</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#getConfigFile--">getConfigFile()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">Runtime property getter for <a href="../biolockj/util/RuntimeParamUtil.html#CONFIG_FLAG">"-c"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getConfigFileName--">getConfigFileName()</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Gets the configuration file name</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getConfigFilePath--">getConfigFilePath()</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Gets the full Config file path passed to BioLockJ as a runtime parameter.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/JsonNode.html#getCount--">getCount()</a></span> - Method in class biolockj.node.<a href="../biolockj/node/JsonNode.html" title="class in biolockj.node">JsonNode</a></dt>
<dd>
<div class="block">Getter method for count</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/OtuNode.html#getCount--">getCount()</a></span> - Method in interface biolockj.node.<a href="../biolockj/node/OtuNode.html" title="interface in biolockj.node">OtuNode</a></dt>
<dd>
<div class="block">Gets the OTU count.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/OtuNodeImpl.html#getCount--">getCount()</a></span> - Method in class biolockj.node.<a href="../biolockj/node/OtuNodeImpl.html" title="class in biolockj.node">OtuNodeImpl</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Properties.html#getDefaultConfigProp-java.io.File-">getDefaultConfigProp(File)</a></span> - Static method in class biolockj.<a href="../biolockj/Properties.html" title="class in biolockj">Properties</a></dt>
<dd>
<div class="block">Parse property file for the property <a href="../biolockj/Config.html#PROJECT_DEFAULT_PROPS">"project.defaultProps"</a>.<br>
 project.defaultProps=/app/biolockj_v1.0/resources/config/default/docker.properties</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/BioModuleFactory.html#getDefaultDemultiplexer--">getDefaultDemultiplexer()</a></span> - Static method in class biolockj.<a href="../biolockj/BioModuleFactory.html" title="class in biolockj">BioModuleFactory</a></dt>
<dd>
<div class="block">Get the Java Class name for the default Demultiplexer module</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/BioModuleFactory.html#getDefaultFastaConverter--">getDefaultFastaConverter()</a></span> - Static method in class biolockj.<a href="../biolockj/BioModuleFactory.html" title="class in biolockj">BioModuleFactory</a></dt>
<dd>
<div class="block">Get the Java Class name for the default Fasta converter module</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/BioModuleFactory.html#getDefaultMergePairedReadsConverter--">getDefaultMergePairedReadsConverter()</a></span> - Static method in class biolockj.<a href="../biolockj/BioModuleFactory.html" title="class in biolockj">BioModuleFactory</a></dt>
<dd>
<div class="block">Get the Java Class name for the default Merge paired read module</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/RegisterNumReads.html#getDepricatedReadFields--">getDepricatedReadFields()</a></span> - Static method in class biolockj.module.implicit.<a href="../biolockj/module/implicit/RegisterNumReads.html" title="class in biolockj.module.implicit">RegisterNumReads</a></dt>
<dd>
<div class="block">Getter for depricatedReadFields</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#getDirectModule--">getDirectModule()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">Runtime property getter for <a href="../biolockj/util/RuntimeParamUtil.html#DIRECT_FLAG">"-d"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#getDockerContainerName--">getDockerContainerName()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">Returns the name of the Docker container running the current module.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#getDockerHostConfigDir--">getDockerHostConfigDir()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">Runtime property getter for <a href="../biolockj/util/RuntimeParamUtil.html#CONFIG_DIR_FLAG">"-C"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#getDockerHostInputDir--">getDockerHostInputDir()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">Runtime property getter for <a href="../biolockj/util/RuntimeParamUtil.html#INPUT_DIR_FLAG">"-i"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#getDockerHostMetaDir--">getDockerHostMetaDir()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">Runtime property getter for <a href="../biolockj/util/RuntimeParamUtil.html#META_DIR_FLAG">"-m"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#getDockerHostPipelineDir--">getDockerHostPipelineDir()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">Runtime property getter for Docker host pipeline dir</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#getDockerHostPrimerDir--">getDockerHostPrimerDir()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">Runtime property getter for <a href="../biolockj/util/RuntimeParamUtil.html#PRIMER_DIR_FLAG">"-t"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DockerUtil.html#getDockerImage-biolockj.module.BioModule-">getDockerImage(BioModule)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/DockerUtil.html" title="class in biolockj.util">DockerUtil</a></dt>
<dd>
<div class="block">Get the name of the Docker image</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DockerUtil.html#getDockerVolumeFile-java.lang.String-java.lang.String-java.lang.String-">getDockerVolumeFile(String, String, String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/DockerUtil.html" title="class in biolockj.util">DockerUtil</a></dt>
<dd>
<div class="block">Get mapped Docker system path from <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> property by replacing the host system path with the
 mapped container path.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getDoubleVal-java.lang.String-">getDoubleVal(String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Parse property for numeric (double) value</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DownloadUtil.html#getDownloadCmd--">getDownloadCmd()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/DownloadUtil.html" title="class in biolockj.util">DownloadUtil</a></dt>
<dd>
<div class="block">If running on cluster, build scp command for user to download pipeline analysis.<br></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DownloadUtil.html#getDownloadDirPath--">getDownloadDirPath()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/DownloadUtil.html" title="class in biolockj.util">DownloadUtil</a></dt>
<dd>
<div class="block">Get validated <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/util/DownloadUtil.html#DOWNLOAD_DIR">"project.downloadDir"</a> if running on cluster, otherwise return null</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DownloadUtil.html#getDownloadModules--">getDownloadModules()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/DownloadUtil.html" title="class in biolockj.util">DownloadUtil</a></dt>
<dd>
<div class="block">Get the modules to download.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getExe-java.lang.String-">getExe(String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Get exe.* property name.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getExistingDir-java.lang.String-">getExistingDir(String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Get a valid File directory or return null</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getExistingFile-java.lang.String-">getExistingFile(String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Get a valid File or return null.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getExistingFileObject-java.lang.String-">getExistingFileObject(String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Build File using filePath.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DownloadUtil.html#getExts-biolockj.module.r.R_Module-">getExts(R_Module)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/DownloadUtil.html" title="class in biolockj.util">DownloadUtil</a></dt>
<dd>
<div class="block">Get file extensions for each R module to include in scp download command</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/MetaUtil.html#getField-java.lang.String-java.lang.String-">getField(String, String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/MetaUtil.html" title="class in biolockj.util">MetaUtil</a></dt>
<dd>
<div class="block">Get metadata field value for given sampleId.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/MetaUtil.html#getFieldNames--">getFieldNames()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/MetaUtil.html" title="class in biolockj.util">MetaUtil</a></dt>
<dd>
<div class="block">Get a list of all metadata fields (metadata file column names except the 1st).</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/MetaUtil.html#getFieldValues-java.lang.String-">getFieldValues(String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/MetaUtil.html" title="class in biolockj.util">MetaUtil</a></dt>
<dd>
<div class="block">Get metadata column for given field name.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Log.html#getFile--">getFile()</a></span> - Static method in class biolockj.<a href="../biolockj/Log.html" title="class in biolockj">Log</a></dt>
<dd>
<div class="block">Returns the log file.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/MetaUtil.html#getFile--">getFile()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/MetaUtil.html" title="class in biolockj.util">MetaUtil</a></dt>
<dd>
<div class="block">Get the current metadata file.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/r16s/QiimeParser.html#getFileNameColumn-java.lang.String-">getFileNameColumn(String)</a></span> - Method in class biolockj.module.implicit.parser.r16s.<a href="../biolockj/module/implicit/parser/r16s/QiimeParser.html" title="class in biolockj.module.implicit.parser.r16s">QiimeParser</a></dt>
<dd>
<div class="block">Examines the header line to find the column index for
 <a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#DEMUX_COLUMN">"BioLockJFileName"</a> which holds the demultiplexed sample file
 name.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BioLockJUtil.html#getFilePaths-java.util.Collection-">getFilePaths(Collection&lt;File&gt;)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BioLockJUtil.html" title="class in biolockj.util">BioLockJUtil</a></dt>
<dd>
<div class="block">Return an ordered list of absolute file paths from the input collection.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#getFileReader-java.io.File-">getFileReader(File)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Get a <a href="https://docs.oracle.com/javase/8/docs/api/java/io/BufferedReader.html?is-external=true" title="class or interface in java.io"><code>BufferedReader</code></a> for standard text file or <a href="https://docs.oracle.com/javase/8/docs/api/java/util/zip/GZIPInputStream.html?is-external=true" title="class or interface in java.util.zip"><code>GZIPInputStream</code></a> for gzipped files ending in ".gz"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SummaryUtil.html#getFooter--">getFooter()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SummaryUtil.html" title="class in biolockj.util">SummaryUtil</a></dt>
<dd>
<div class="block">Print the final lines of the summary with overall status, runtime, and the download scp command if applicable.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/r/R_Module.html#getFunctionLib--">getFunctionLib()</a></span> - Method in class biolockj.module.r.<a href="../biolockj/module/r/R_Module.html" title="class in biolockj.module.r">R_Module</a></dt>
<dd>
<div class="block">Get the function library script</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/Multiplexer.html#getHeader-java.io.File-java.util.List-">getHeader(File, List&lt;String&gt;)</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/Multiplexer.html" title="class in biolockj.module.seq">Multiplexer</a></dt>
<dd>
<div class="block">Get the header for the sequence.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#getHeader-java.lang.String-">getHeader(String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Extract the header from the first line of a read (parse to first space).</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#getHeaders-java.io.File-">getHeaders(File)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Return the header of each read in the sequence file.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#getHeaders-java.io.File-java.io.File-">getHeaders(File, File)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Get valid headers found in both forward and reverse read.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/MetaUtil.html#getID--">getID()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/MetaUtil.html" title="class in biolockj.util">MetaUtil</a></dt>
<dd>
<div class="block">Get the metadata file ID column name.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DemuxUtil.html#getIdMap--">getIdMap()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/DemuxUtil.html" title="class in biolockj.util">DemuxUtil</a></dt>
<dd>
<div class="block">Return the ID map (key=barcodeID, value=sampleId) Print the map to the log file.<br>
 Barcodes, if used, are located in the metadata column defined by property:
 <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/util/MetaUtil.html#META_BARCODE_COLUMN">"metadata.barcodeColumn"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#getInputDirs--">getInputDirs()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Get the list of input directories for the pipeline.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#getInputFileDir--">getInputFileDir()</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime">QiimeClassifier</a></dt>
<dd>
<div class="block">Module input directories are set to the previous module output directory.<br>
 To ensure we use the correct path, get path from <a href="../biolockj/module/BioModuleImpl.html#getInputFiles--"><code>BioModuleImpl.getInputFiles()</code></a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModule.html#getInputFiles--">getInputFiles()</a></span> - Method in interface biolockj.module.<a href="../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></dt>
<dd>
<div class="block">Each BioModule takes the previous BioModule output as input:<br>
 BioModule[ n ].getInputFiles() = BioModule[ n - 1 ].getOutputDir().listFiles()<br>
 Special cases:<br>
 
 The 1st BioModule return all files in <a href="../biolockj/Config.html#INPUT_DIRS">"input.dirPaths"</a><br>
 If previous BioModule BioModule[ n - 1 ] is a MetadataModule, forward it's input + output file: BioModule[ n
 ].getInputFiles() = BioModule[ n -1 ].getInputFiles() + BioModule[ n -1 ].getOutputFiles()
 </div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModuleImpl.html#getInputFiles--">getInputFiles()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a></dt>
<dd>
<div class="block">BioModule <a href="../biolockj/module/BioModuleImpl.html#initInputFiles--"><code>BioModuleImpl.initInputFiles()</code></a> is called to initialize upon first call and cached.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/r16s/QiimeParser.html#getInputFiles--">getInputFiles()</a></span> - Method in class biolockj.module.implicit.parser.r16s.<a href="../biolockj/module/implicit/parser/r16s/QiimeParser.html" title="class in biolockj.module.implicit.parser.r16s">QiimeParser</a></dt>
<dd>
<div class="block">Get only the lowest level report since it contains taxa info for all higher OTU reports.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/JsonReport.html#getInputFiles--">getInputFiles()</a></span> - Method in class biolockj.module.report.<a href="../biolockj/module/report/JsonReport.html" title="class in biolockj.module.report">JsonReport</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#getInputSequenceType--">getInputSequenceType()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Determine if seqs are fasta or fastq by checking input file format (check 1st character only).</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/r/R_Module.html#getJobParams--">getJobParams()</a></span> - Method in class biolockj.module.r.<a href="../biolockj/module/r/R_Module.html" title="class in biolockj.module.r">R_Module</a></dt>
<dd>
<div class="block">If running Docker, run the Docker bash script, otherwise:<br>
 Run <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/module/r/R_Module.html#EXE_RSCRIPT">"exe.Rscript"</a> command on the generated R Script:
 <a href="../biolockj/util/ModuleUtil.html#getMainScript-biolockj.module.BioModule-"><code>ModuleUtil.getMainScript(BioModule)</code></a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/ScriptModule.html#getJobParams--">getJobParams()</a></span> - Method in interface biolockj.module.<a href="../biolockj/module/ScriptModule.html" title="interface in biolockj.module">ScriptModule</a></dt>
<dd>
<div class="block">Many ScriptModules generate bash or R scripts for the Operating System to run as a <a href="../biolockj/Job.html" title="class in biolockj"><code>Job</code></a>.<br>
 The Job creates a <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Process.html?is-external=true" title="class or interface in java.lang"><code>Process</code></a> using the parameters supplied by this method and executed in the
 <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/Runtime.html?is-external=true" title="class or interface in java.lang"><code>Runtime</code></a> environment.<br>
 Parameters typically contain the full script path and script parameters, if needed.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/ScriptModuleImpl.html#getJobParams--">getJobParams()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/ScriptModuleImpl.html" title="class in biolockj.module">ScriptModuleImpl</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/JsonNode.html#getLevel--">getLevel()</a></span> - Method in class biolockj.node.<a href="../biolockj/node/JsonNode.html" title="class in biolockj.node">JsonNode</a></dt>
<dd>
<div class="block">Getter method for level</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/r16s/QiimeNode.html#getLevelDelim-java.lang.String-">getLevelDelim(String)</a></span> - Method in class biolockj.node.r16s.<a href="../biolockj/node/r16s/QiimeNode.html" title="class in biolockj.node.r16s">QiimeNode</a></dt>
<dd>
<div class="block">Get the first taxonomy level delim.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/OtuNode.html#getLine--">getLine()</a></span> - Method in interface biolockj.node.<a href="../biolockj/node/OtuNode.html" title="interface in biolockj.node">OtuNode</a></dt>
<dd>
<div class="block">Get the line from classifier output file used to create this OtuNode.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/OtuNodeImpl.html#getLine--">getLine()</a></span> - Method in class biolockj.node.<a href="../biolockj/node/OtuNodeImpl.html" title="class in biolockj.node">OtuNodeImpl</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getList-java.lang.String-">getList(String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Parse comma delimited property value to return list</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/r/R_Module.html#getMainRscriptPrefix--">getMainRscriptPrefix()</a></span> - Method in class biolockj.module.r.<a href="../biolockj/module/r/R_Module.html" title="class in biolockj.module.r">R_Module</a></dt>
<dd>
<div class="block">Get the main R script prefix, typically  except when running in Docker mode, which instead
 uses a numeric prefix since  will be reserved for the MAIN bash script.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/ModuleUtil.html#getMainScript-biolockj.module.BioModule-">getMainScript(BioModule)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/ModuleUtil.html" title="class in biolockj.util">ModuleUtil</a></dt>
<dd>
<div class="block">Get the main script file in the bioModule script directory, with prefix:
 <a href="../biolockj/module/BioModule.html#MAIN_SCRIPT_PREFIX">"MAIN_"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BashScriptBuilder.html#getMainScriptExecuteWorkerLine-biolockj.module.ScriptModule-java.lang.String-java.lang.String-">getMainScriptExecuteWorkerLine(ScriptModule, String, String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BashScriptBuilder.html" title="class in biolockj.util">BashScriptBuilder</a></dt>
<dd>
<div class="block">Call <a href="../biolockj/util/BashScriptBuilder.html#FUNCTION_EXECUTE">"execute"</a> on the worker script</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BioLockJUtil.html#getMasterConfig--">getMasterConfig()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BioLockJUtil.html" title="class in biolockj.util">BioLockJUtil</a></dt>
<dd>
<div class="block">Return the MASTER config file.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RMetaUtil.html#getMdsFields--">getMdsFields()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RMetaUtil.html" title="class in biolockj.util">RMetaUtil</a></dt>
<dd>
<div class="block">Get the <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/util/RMetaUtil.html#MDS_REPORT_FIELDS">"rMds.reportFields"</a> fields.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SummaryUtil.html#getMean-java.util.Collection-boolean-">getMean(Collection&lt;?&gt;, boolean)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SummaryUtil.html" title="class in biolockj.util">SummaryUtil</a></dt>
<dd>
<div class="block">Return the mean value</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SummaryUtil.html#getMedian-java.util.Collection-boolean-">getMedian(Collection&lt;?&gt;, boolean)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SummaryUtil.html" title="class in biolockj.util">SummaryUtil</a></dt>
<dd>
<div class="block">Return the median value</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/AddMetadataToOtuTables.html#getMergedLine-java.lang.String-">getMergedLine(String)</a></span> - Method in class biolockj.module.report.<a href="../biolockj/module/report/AddMetadataToOtuTables.html" title="class in biolockj.module.report">AddMetadataToOtuTables</a></dt>
<dd>
<div class="block">Return OTU table line with metadata row appended (both have PK = sample ID)</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Pipeline.html#getMetadata-biolockj.module.BioModule-">getMetadata(BioModule)</a></span> - Static method in class biolockj.<a href="../biolockj/Pipeline.html" title="class in biolockj">Pipeline</a></dt>
<dd>
<div class="block">If the bioModule is complete and contains a metadata file in its output directory, return the metadata file,
 since it must be a new version.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/MetaUtil.html#getMetadata--">getMetadata()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/MetaUtil.html" title="class in biolockj.util">MetaUtil</a></dt>
<dd>
<div class="block">Metadata file getter.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/MetaUtil.html#getMetadataFileName--">getMetadataFileName()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/MetaUtil.html" title="class in biolockj.util">MetaUtil</a></dt>
<dd>
<div class="block">Get the metadata file name, if it exists, otherwise return projectName.tsv</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/MetaUtil.html#getMetadataRecord-java.lang.String-">getMetadataRecord(String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/MetaUtil.html" title="class in biolockj.util">MetaUtil</a></dt>
<dd>
<div class="block">Get metadata row for a given Sample ID.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/MetaphlanClassifier.html#getMetaphlanDerivedSwitches--">getMetaphlanDerivedSwitches()</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/MetaphlanClassifier.html" title="class in biolockj.module.classifier.wgs">MetaphlanClassifier</a></dt>
<dd>
<div class="block">Metaphlan queries require standard parameters: --input_type, --nproc, -t</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/ModuleUtil.html#getModule-java.lang.String-">getModule(String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/ModuleUtil.html" title="class in biolockj.util">ModuleUtil</a></dt>
<dd>
<div class="block">Return module with given name.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModule.html#getModuleDir--">getModuleDir()</a></span> - Method in interface biolockj.module.<a href="../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></dt>
<dd>
<div class="block">Each BioModule generates sub-directory under $DOCKER_PROJ</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModuleImpl.html#getModuleDir--">getModuleDir()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a></dt>
<dd>
<div class="block">All BioModule work must be contained within the scope of its root directory.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/ModuleUtil.html#getModuleNum-biolockj.module.BioModule-">getModuleNum(BioModule)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/ModuleUtil.html" title="class in biolockj.util">ModuleUtil</a></dt>
<dd>
<div class="block">Get the module number, formated as found in the module directory names.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SummaryUtil.html#getModuleNumber--">getModuleNumber()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SummaryUtil.html" title="class in biolockj.util">SummaryUtil</a></dt>
<dd>
<div class="block">Read the summary file to find the last completed module number and return the next number.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/ModuleUtil.html#getModuleRootDir-biolockj.module.BioModule-">getModuleRootDir(BioModule)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/ModuleUtil.html" title="class in biolockj.util">ModuleUtil</a></dt>
<dd>
<div class="block">Get the Module root directory full path.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/ModuleUtil.html#getModuleRunTime-biolockj.module.BioModule-">getModuleRunTime(BioModule)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/ModuleUtil.html" title="class in biolockj.util">ModuleUtil</a></dt>
<dd>
<div class="block">Return duration bioModule ran based on modified data of started file, formatted for display (as hours, minutes,
 seconds).</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Pipeline.html#getModules--">getModules()</a></span> - Static method in class biolockj.<a href="../biolockj/Pipeline.html" title="class in biolockj">Pipeline</a></dt>
<dd>
<div class="block">Return a list of <a href="../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>s constructed by the <a href="../biolockj/BioModuleFactory.html" title="class in biolockj"><code>BioModuleFactory</code></a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/r/R_Module.html#getModuleScript--">getModuleScript()</a></span> - Method in class biolockj.module.r.<a href="../biolockj/module/r/R_Module.html" title="class in biolockj.module.r">R_Module</a></dt>
<dd>
<div class="block">Get the Module script</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Log.html#getMsgs--">getMsgs()</a></span> - Static method in class biolockj.<a href="../biolockj/Log.html" title="class in biolockj">Log</a></dt>
<dd>
<div class="block">Get the logMesseges created prior to creating the actual <a href="../biolockj/Log.html#LOG_FILE"><code>Log.LOG_FILE</code></a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RMetaUtil.html#getNominalFields--">getNominalFields()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RMetaUtil.html" title="class in biolockj.util">RMetaUtil</a></dt>
<dd>
<div class="block">Get the <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/util/RMetaUtil.html#R_NOMINAL_FIELDS">"r.nominalFields"</a> fields.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getNonNegativeInteger-java.lang.String-">getNonNegativeInteger(String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Parse property as non-negative integer value</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RMetaUtil.html#getNumericFields--">getNumericFields()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RMetaUtil.html" title="class in biolockj.util">RMetaUtil</a></dt>
<dd>
<div class="block">Get the <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/util/RMetaUtil.html#R_NUMERIC_FIELDS">"r.numericFields"</a> fields.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/ParsedSample.html#getNumHits--">getNumHits()</a></span> - Method in class biolockj.node.<a href="../biolockj/node/ParsedSample.html" title="class in biolockj.node">ParsedSample</a></dt>
<dd>
<div class="block">Return total OTU count for all OTUs assigned at the highest taxonomy level found in the sample.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#getNumLinesPerRead--">getNumLinesPerRead()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Return number of lines per read (for fasta return 2, for fastq return 4)</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/RegisterNumReads.html#getNumReadFieldName--">getNumReadFieldName()</a></span> - Static method in class biolockj.module.implicit.<a href="../biolockj/module/implicit/RegisterNumReads.html" title="class in biolockj.module.implicit">RegisterNumReads</a></dt>
<dd>
<div class="block">Getter for numReadFieldName</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#getNumReads-long-">getNumReads(long)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Return number of reads given number of lines in a sample.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/ClassifierModuleImpl.html#getNumThreads--">getNumThreads()</a></span> - Static method in class biolockj.module.classifier.<a href="../biolockj/module/classifier/ClassifierModuleImpl.html" title="class in biolockj.module.classifier">ClassifierModuleImpl</a></dt>
<dd>
<div class="block">Get classifier specific number of threads property value, if defined.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/ClassifierModuleImpl.html#getNumThreadsParam--">getNumThreadsParam()</a></span> - Static method in class biolockj.module.classifier.<a href="../biolockj/module/classifier/ClassifierModuleImpl.html" title="class in biolockj.module.classifier">ClassifierModuleImpl</a></dt>
<dd>
<div class="block">Get classifier number of threads property name.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/JsonNode.html#getOtu--">getOtu()</a></span> - Method in class biolockj.node.<a href="../biolockj/node/JsonNode.html" title="class in biolockj.node">JsonNode</a></dt>
<dd>
<div class="block">Getter method for otu</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/ParsedSample.html#getOtuCountMap--">getOtuCountMap()</a></span> - Method in class biolockj.node.<a href="../biolockj/node/ParsedSample.html" title="class in biolockj.node">ParsedSample</a></dt>
<dd>
<div class="block">Getter for otuCountMap, a nested TreeMap with outer key = level, inner key = OTU, inner value = count.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/OtuNode.html#getOtuMap--">getOtuMap()</a></span> - Method in interface biolockj.node.<a href="../biolockj/node/OtuNode.html" title="interface in biolockj.node">OtuNode</a></dt>
<dd>
<div class="block">Gets the map holding level-specific OTU names</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/OtuNodeImpl.html#getOtuMap--">getOtuMap()</a></span> - Method in class biolockj.node.<a href="../biolockj/node/OtuNodeImpl.html" title="class in biolockj.node">OtuNodeImpl</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/ParsedSample.html#getOtuNodes--">getOtuNodes()</a></span> - Method in class biolockj.node.<a href="../biolockj/node/ParsedSample.html" title="class in biolockj.node">ParsedSample</a></dt>
<dd>
<div class="block">Return all otuNodes</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/ParserModuleImpl.html#getOtuTableRowId-java.lang.String-">getOtuTableRowId(String)</a></span> - Method in class biolockj.module.implicit.parser.<a href="../biolockj/module/implicit/parser/ParserModuleImpl.html" title="class in biolockj.module.implicit.parser">ParserModuleImpl</a></dt>
<dd>
<div class="block">This method is used to get an R-friendly OTU table row value by stripping out the quotes.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/r16s/QiimeParser.html#getOtuTableRowId-java.lang.String-">getOtuTableRowId(String)</a></span> - Method in class biolockj.module.implicit.parser.r16s.<a href="../biolockj/module/implicit/parser/r16s/QiimeParser.html" title="class in biolockj.module.implicit.parser.r16s">QiimeParser</a></dt>
<dd>
<div class="block">This method is used to get an R-friendly Sample ID value by stripping out the quotes AND by replacing any Qiime
 ID found with the original Sample ID.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModule.html#getOutputDir--">getOutputDir()</a></span> - Method in interface biolockj.module.<a href="../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></dt>
<dd>
<div class="block">Output files destined as input for the next BioModule is created in this directory.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModuleImpl.html#getOutputDir--">getOutputDir()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a></dt>
<dd>
<div class="block">Returns moduleDir/output which will be used as the next module's input.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SummaryUtil.html#getOutputDirSummary-biolockj.module.BioModule-">getOutputDirSummary(BioModule)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SummaryUtil.html" title="class in biolockj.util">SummaryUtil</a></dt>
<dd>
<div class="block">Return summary of <a href="../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a> output directory, with metrics:
 
 Number of output files
 Mean output file size
 Path of new metadata file if any created
 </div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#getPairedReads-java.util.Collection-">getPairedReads(Collection&lt;File&gt;)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Paired reads must have a unique file suffix to identify forward and reverse reads.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/JsonNode.html#getParent--">getParent()</a></span> - Method in class biolockj.node.<a href="../biolockj/node/JsonNode.html" title="class in biolockj.node">JsonNode</a></dt>
<dd>
<div class="block">Getter method for parent</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/ParserModule.html#getParsedSample-java.lang.String-">getParsedSample(String)</a></span> - Method in interface biolockj.module.implicit.parser.<a href="../biolockj/module/implicit/parser/ParserModule.html" title="interface in biolockj.module.implicit.parser">ParserModule</a></dt>
<dd>
<div class="block">After <a href="../biolockj/module/implicit/parser/ParserModule.html#parseSamples--"><code>ParserModule.parseSamples()</code></a> completes, this method can be called to get a <a href="../biolockj/node/ParsedSample.html" title="class in biolockj.node"><code>ParsedSample</code></a> by
 its sample ID.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/ParserModuleImpl.html#getParsedSample-java.lang.String-">getParsedSample(String)</a></span> - Method in class biolockj.module.implicit.parser.<a href="../biolockj/module/implicit/parser/ParserModuleImpl.html" title="class in biolockj.module.implicit.parser">ParserModuleImpl</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/ParserModule.html#getParsedSamples--">getParsedSamples()</a></span> - Method in interface biolockj.module.implicit.parser.<a href="../biolockj/module/implicit/parser/ParserModule.html" title="interface in biolockj.module.implicit.parser">ParserModule</a></dt>
<dd>
<div class="block">After <a href="../biolockj/module/implicit/parser/ParserModule.html#parseSamples--"><code>ParserModule.parseSamples()</code></a> completes, this method can be called to get all of the
 <a href="../biolockj/node/ParsedSample.html" title="class in biolockj.node"><code>ParsedSample</code></a>s generated from the classifier output</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/ParserModuleImpl.html#getParsedSamples--">getParsedSamples()</a></span> - Method in class biolockj.module.implicit.parser.<a href="../biolockj/module/implicit/parser/ParserModuleImpl.html" title="class in biolockj.module.implicit.parser">ParserModuleImpl</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/ModuleUtil.html#getParserModule--">getParserModule()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/ModuleUtil.html" title="class in biolockj.util">ModuleUtil</a></dt>
<dd>
<div class="block">Returns the PerserModule, if configured.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#getPickOtuLines-java.lang.String-java.io.File-java.lang.String-java.io.File-">getPickOtuLines(String, File, String, File)</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime">QiimeClassifier</a></dt>
<dd>
<div class="block">Subclasses call this method to add OTU picking lines by calling <a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#SCRIPT_ADD_LABELS">"add_qiime_labels.py"</a> via OTU picking
 script.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#getPipelineInputFiles--">getPipelineInputFiles()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Recursively get files located in the directories listed in
 <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/Config.html#INPUT_DIRS">"input.dirPaths"</a> after removing
 <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/Config.html#INPUT_IGNORE_FILES">"input.ignoreFiles"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getPositiveDoubleVal-java.lang.String-">getPositiveDoubleVal(String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Parse property as positive double value</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getPositiveInteger-java.lang.String-">getPositiveInteger(String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Parse property as positive integer value</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModule.html#getPostRequisiteModules--">getPostRequisiteModules()</a></span> - Method in interface biolockj.module.<a href="../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></dt>
<dd>
<div class="block"><a href="../biolockj/Pipeline.html" title="class in biolockj"><code>Pipeline</code></a> calls this method when building the list of pipeline BioModules to execute.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModuleImpl.html#getPostRequisiteModules--">getPostRequisiteModules()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a></dt>
<dd>
<div class="block">By default, no post-requisites are required.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/ClassifierModuleImpl.html#getPostRequisiteModules--">getPostRequisiteModules()</a></span> - Method in class biolockj.module.classifier.<a href="../biolockj/module/classifier/ClassifierModuleImpl.html" title="class in biolockj.module.classifier">ClassifierModuleImpl</a></dt>
<dd>
<div class="block">This method returns the corresponding Parser module associated with the classifier.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/r16s/QiimeClosedRefClassifier.html#getPostRequisiteModules--">getPostRequisiteModules()</a></span> - Method in class biolockj.module.classifier.r16s.<a href="../biolockj/module/classifier/r16s/QiimeClosedRefClassifier.html" title="class in biolockj.module.classifier.r16s">QiimeClosedRefClassifier</a></dt>
<dd>
<div class="block">If paired reads found, return prerequisite module: <a href="../biolockj/module/seq/PearMergeReads.html" title="class in biolockj.module.seq"><code>PearMergeReads</code></a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#getPostRequisiteModules--">getPostRequisiteModules()</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime">QiimeClassifier</a></dt>
<dd>
<div class="block">Subclasses of QiimeClassifier add post-requisite module: <a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime"><code>QiimeClassifier</code></a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/BioModuleFactory.html#getPostRequisites-biolockj.module.BioModule-">getPostRequisites(BioModule)</a></span> - Static method in class biolockj.<a href="../biolockj/BioModuleFactory.html" title="class in biolockj">BioModuleFactory</a></dt>
<dd>
<div class="block">This method returns all module post-requisites (including prerequisites and post-requisites for the
 post-requisites).</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModule.html#getPreRequisiteModules--">getPreRequisiteModules()</a></span> - Method in interface biolockj.module.<a href="../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></dt>
<dd>
<div class="block"><a href="../biolockj/Pipeline.html" title="class in biolockj"><code>Pipeline</code></a> calls this method when building the list of pipeline BioModules to execute.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModuleImpl.html#getPreRequisiteModules--">getPreRequisiteModules()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a></dt>
<dd>
<div class="block">By default, no prerequisites are required.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/r16s/RdpClassifier.html#getPreRequisiteModules--">getPreRequisiteModules()</a></span> - Method in class biolockj.module.classifier.r16s.<a href="../biolockj/module/classifier/r16s/RdpClassifier.html" title="class in biolockj.module.classifier.r16s">RdpClassifier</a></dt>
<dd>
<div class="block">If paired reads found, return prerequisite module: <a href="../biolockj/module/seq/PearMergeReads.html" title="class in biolockj.module.seq"><code>PearMergeReads</code></a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/BuildQiimeMapping.html#getPreRequisiteModules--">getPreRequisiteModules()</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/BuildQiimeMapping.html" title="class in biolockj.module.implicit.qiime">BuildQiimeMapping</a></dt>
<dd>
<div class="block">If paired reads found, add prerequisite module: <a href="../biolockj/module/seq/PearMergeReads.html" title="class in biolockj.module.seq"><code>PearMergeReads</code></a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#getPreRequisiteModules--">getPreRequisiteModules()</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime">QiimeClassifier</a></dt>
<dd>
<div class="block">If paired reads found, add prerequisite module: <a href="../biolockj/module/seq/PearMergeReads.html" title="class in biolockj.module.seq"><code>PearMergeReads</code></a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/r/BuildMdsPlots.html#getPreRequisiteModules--">getPreRequisiteModules()</a></span> - Method in class biolockj.module.r.<a href="../biolockj/module/r/BuildMdsPlots.html" title="class in biolockj.module.r">BuildMdsPlots</a></dt>
<dd>
<div class="block">Add prerequisite module: <a href="../biolockj/module/r/CalculateStats.html" title="class in biolockj.module.r"><code>CalculateStats</code></a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/r/BuildOtuPlots.html#getPreRequisiteModules--">getPreRequisiteModules()</a></span> - Method in class biolockj.module.r.<a href="../biolockj/module/r/BuildOtuPlots.html" title="class in biolockj.module.r">BuildOtuPlots</a></dt>
<dd>
<div class="block">Add prerequisite module: <a href="../biolockj/module/r/CalculateStats.html" title="class in biolockj.module.r"><code>CalculateStats</code></a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/r/BuildPvalHistograms.html#getPreRequisiteModules--">getPreRequisiteModules()</a></span> - Method in class biolockj.module.r.<a href="../biolockj/module/r/BuildPvalHistograms.html" title="class in biolockj.module.r">BuildPvalHistograms</a></dt>
<dd>
<div class="block">Add prerequisite module: <a href="../biolockj/module/r/CalculateStats.html" title="class in biolockj.module.r"><code>CalculateStats</code></a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/r/R_Module.html#getPreRequisiteModules--">getPreRequisiteModules()</a></span> - Method in class biolockj.module.r.<a href="../biolockj/module/r/R_Module.html" title="class in biolockj.module.r">R_Module</a></dt>
<dd>
<div class="block">All R modules require combined OTU-metadata tables.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/JsonReport.html#getPreRequisiteModules--">getPreRequisiteModules()</a></span> - Method in class biolockj.module.report.<a href="../biolockj/module/report/JsonReport.html" title="class in biolockj.module.report">JsonReport</a></dt>
<dd>
<div class="block">Module prerequisites include:
 
 <a href="../biolockj/module/report/AddMetadataToOtuTables.html" title="class in biolockj.module.report"><code>AddMetadataToOtuTables</code></a>
 <a href="../biolockj/module/r/CalculateStats.html" title="class in biolockj.module.r"><code>CalculateStats</code></a>
 </div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/Rarefier.html#getPreRequisiteModules--">getPreRequisiteModules()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/Rarefier.html" title="class in biolockj.module.seq">Rarefier</a></dt>
<dd>
<div class="block">This method always requires the prerequisite module: <a href="../biolockj/module/implicit/RegisterNumReads.html" title="class in biolockj.module.implicit"><code>RegisterNumReads</code></a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/BioModuleFactory.html#getPreRequisites-biolockj.module.BioModule-">getPreRequisites(BioModule)</a></span> - Static method in class biolockj.<a href="../biolockj/BioModuleFactory.html" title="class in biolockj">BioModuleFactory</a></dt>
<dd>
<div class="block">This method returns all module prerequisites (including prerequisites and post-requisites for the prerequisites).</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/ModuleUtil.html#getPreviousModule-biolockj.module.BioModule-">getPreviousModule(BioModule)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/ModuleUtil.html" title="class in biolockj.util">ModuleUtil</a></dt>
<dd>
<div class="block">BioModules are run in the order configured.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/r/R_Module.html#getPrimaryScript--">getPrimaryScript()</a></span> - Method in class biolockj.module.r.<a href="../biolockj/module/r/R_Module.html" title="class in biolockj.module.r">R_Module</a></dt>
<dd>
<div class="block">Get the primary R script</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/TrimPrimers.html#getPrimers-boolean-">getPrimers(boolean)</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/TrimPrimers.html" title="class in biolockj.module.seq">TrimPrimers</a></dt>
<dd>
<div class="block">Get the primers listed in the <a href="../biolockj/module/seq/TrimPrimers.html#INPUT_TRIM_SEQ_FILE">"trimPrimers.filePath"</a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getProperties--">getProperties()</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Get properties ordered by propertyName</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/ImportMetadata.html#getQuotedValue-java.lang.String-">getQuotedValue(String)</a></span> - Method in class biolockj.module.implicit.<a href="../biolockj/module/implicit/ImportMetadata.html" title="class in biolockj.module.implicit">ImportMetadata</a></dt>
<dd>
<div class="block">The member variable quotedText caches the input held within a quoted block.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/BioLockJ.html#getRestartDir--">getRestartDir()</a></span> - Static method in class biolockj.<a href="../biolockj/BioLockJ.html" title="class in biolockj">BioLockJ</a></dt>
<dd>
<div class="block">Method called if restart flag passed to BioLockJ to look in $DOCKER_PROJ for the most recent pipeline sharing the
 same project name.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/r/R_Module.html#getRTemplateDir--">getRTemplateDir()</a></span> - Method in class biolockj.module.r.<a href="../biolockj/module/r/R_Module.html" title="class in biolockj.module.r">R_Module</a></dt>
<dd>
<div class="block">Get the BioLockJ resource R directory.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/ModuleUtil.html#getRunTime-long-">getRunTime(long)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/ModuleUtil.html" title="class in biolockj.util">ModuleUtil</a></dt>
<dd>
<div class="block">Get runtime message (formatted as hours, minutes, seconds) based on startTime passed.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#getRuntimeArgs--">getRuntimeArgs()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">Return all runtime args</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/r16s/QiimeParser.html#getSampleId-java.lang.String-">getSampleId(String)</a></span> - Method in class biolockj.module.implicit.parser.r16s.<a href="../biolockj/module/implicit/parser/r16s/QiimeParser.html" title="class in biolockj.module.implicit.parser.r16s">QiimeParser</a></dt>
<dd>
<div class="block">Get the Sample ID from <a href="../biolockj/module/implicit/parser/r16s/QiimeParser.html#qiimeIdToSampleIdMap"><code>QiimeParser.qiimeIdToSampleIdMap</code></a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/OtuNode.html#getSampleId--">getSampleId()</a></span> - Method in interface biolockj.node.<a href="../biolockj/node/OtuNode.html" title="interface in biolockj.node">OtuNode</a></dt>
<dd>
<div class="block">Gets the sample ID to which the read belongs</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/OtuNodeImpl.html#getSampleId--">getSampleId()</a></span> - Method in class biolockj.node.<a href="../biolockj/node/OtuNodeImpl.html" title="class in biolockj.node">OtuNodeImpl</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/ParsedSample.html#getSampleId--">getSampleId()</a></span> - Method in class biolockj.node.<a href="../biolockj/node/ParsedSample.html" title="class in biolockj.node">ParsedSample</a></dt>
<dd>
<div class="block">Getter for sampleId.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DemuxUtil.html#getSampleId-java.util.List-">getSampleId(List&lt;String&gt;)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/DemuxUtil.html" title="class in biolockj.util">DemuxUtil</a></dt>
<dd>
<div class="block">Determine Sample Id by examining the sequence lines.<br>
 If =, extract the Sample Id from the sequence header via
 <a href="../biolockj/util/SeqUtil.html#getSampleId-java.lang.String-"><code>SeqUtil.getSampleId(String)</code></a><br>
 If = and the sequence header contains a bar-code in
 the idMap, return the corresponding SampleID from the idMap.<br>
 If = and the sequence itself begins with a bar-code in
 the idMap, return the corresponding SampleID from the idMap.<br></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#getSampleId-java.lang.String-">getSampleId(String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Method extracts Sample ID from the name param.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/r16s/QiimeParser.html#getSampleIdFromMappingFile-java.lang.String-int-">getSampleIdFromMappingFile(String, int)</a></span> - Method in class biolockj.module.implicit.parser.r16s.<a href="../biolockj/module/implicit/parser/r16s/QiimeParser.html" title="class in biolockj.module.implicit.parser.r16s">QiimeParser</a></dt>
<dd>
<div class="block">The mapping file used the original sampleId when generating the demultipled files, as recorded in mapping file
 column - <a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#DEMUX_COLUMN">"BioLockJFileName"</a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/ImportMetadata.html#getSampleIds--">getSampleIds()</a></span> - Method in class biolockj.module.implicit.<a href="../biolockj/module/implicit/ImportMetadata.html" title="class in biolockj.module.implicit">ImportMetadata</a></dt>
<dd>
<div class="block">Extract the sample IDs from the file names with <a href="../biolockj/util/SeqUtil.html#getSampleId-java.lang.String-"><code>SeqUtil.getSampleId(String)</code></a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/MetaUtil.html#getSampleIds--">getSampleIds()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/MetaUtil.html" title="class in biolockj.util">MetaUtil</a></dt>
<dd>
<div class="block">Get the first column from the metadata file.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/r16s/RdpNode.html#getScore--">getScore()</a></span> - Method in class biolockj.node.r16s.<a href="../biolockj/node/r16s/RdpNode.html" title="class in biolockj.node.r16s">RdpNode</a></dt>
<dd>
<div class="block">Return lowest RDP confidence percentage (0-100) above the configured threshold value.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/ScriptModule.html#getScriptDir--">getScriptDir()</a></span> - Method in interface biolockj.module.<a href="../biolockj/module/ScriptModule.html" title="interface in biolockj.module">ScriptModule</a></dt>
<dd>
<div class="block">ScriptModules that generate scripts to complete their task, create script files in this directory.<br>
 The main script must begin with prefix  and there must only be one main script.<br>
 The main script executes, or submits to the job queue, each of the other scripts in this directory.<br>
 The # subscripts generated by the ScriptModule depends upon <a href="../biolockj/module/ScriptModule.html#SCRIPT_BATCH_SIZE">"script.batchSize"</a>.<br></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/ScriptModuleImpl.html#getScriptDir--">getScriptDir()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/ScriptModuleImpl.html" title="class in biolockj.module">ScriptModuleImpl</a></dt>
<dd>
<div class="block">Returns moduleDir/script which contains all scripts generated by the module.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SummaryUtil.html#getScriptDirSummary-biolockj.module.ScriptModule-">getScriptDirSummary(ScriptModule)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SummaryUtil.html" title="class in biolockj.util">SummaryUtil</a></dt>
<dd>
<div class="block">Return summary of the <a href="../biolockj/module/ScriptModule.html" title="interface in biolockj.module"><code>ScriptModule</code></a> script directory with metrics:
 
 Print main script name
 Number of worker scripts run
 Number of worker scripts successful/failed/incomplete
 Average worker script run time
 Longest running worker script names/duration
 Longest running workers script names/duration
 </div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/ModuleUtil.html#getScriptErrors-biolockj.module.ScriptModule-">getScriptErrors(ScriptModule)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/ModuleUtil.html" title="class in biolockj.util">ModuleUtil</a></dt>
<dd>
<div class="block">This method returns all of the lines from any failure files found in the script directory.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SummaryUtil.html#getScriptRunTime-java.lang.Long-">getScriptRunTime(Long)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SummaryUtil.html" title="class in biolockj.util">SummaryUtil</a></dt>
<dd>
<div class="block">Get the script runtime in hours and minutes</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#getSeqHeaderChars--">getSeqHeaderChars()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Get all sequence header characters for fasta and fastq files.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/Email.html#getSession--">getSession()</a></span> - Method in class biolockj.module.report.<a href="../biolockj/module/report/Email.html" title="class in biolockj.module.report">Email</a></dt>
<dd>
<div class="block">Build an authenticated javax.mail.Session using <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> email properties</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getSet-java.lang.String-">getSet(String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Parse comma-separated property value to build an ordered Set</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/JavaModuleImpl.html#getSource--">getSource()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/JavaModuleImpl.html" title="class in biolockj.module">JavaModuleImpl</a></dt>
<dd>
<div class="block">Get the program source (either the jar path or main class biolockj.BioLockJ);</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BioLockJUtil.html#getSource--">getSource()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BioLockJUtil.html" title="class in biolockj.util">BioLockJUtil</a></dt>
<dd>
<div class="block">Get the program source (either the jar path or main class biolockj.BioLockJ);</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/JsonNode.html#getStats--">getStats()</a></span> - Method in class biolockj.node.<a href="../biolockj/node/JsonNode.html" title="class in biolockj.node">JsonNode</a></dt>
<dd>
<div class="block">Getter method for stats</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/r/CalculateStats.html#getStatsFile-java.lang.String-java.lang.String-">getStatsFile(String, String)</a></span> - Static method in class biolockj.module.r.<a href="../biolockj/module/r/CalculateStats.html" title="class in biolockj.module.r">CalculateStats</a></dt>
<dd>
<div class="block">Get the stats file for the given fileType and otuLevel.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Pipeline.html#getStatus--">getStatus()</a></span> - Static method in class biolockj.<a href="../biolockj/Pipeline.html" title="class in biolockj">Pipeline</a></dt>
<dd>
<div class="block">Return <a href="../biolockj/Pipeline.html#SCRIPT_SUCCESS">"Success"</a> if no pipelineException has been thrown, otherwise return
 <a href="../biolockj/Pipeline.html#SCRIPT_FAILURES">"Failures"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getString-java.lang.String-">getString(String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Get property value as String.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/ModuleUtil.html#getSubDir-biolockj.module.BioModule-java.lang.String-">getSubDir(BioModule, String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/ModuleUtil.html" title="class in biolockj.util">ModuleUtil</a></dt>
<dd>
<div class="block">Get BioModule subdirectory File object with given name.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModule.html#getSummary--">getSummary()</a></span> - Method in interface biolockj.module.<a href="../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></dt>
<dd>
<div class="block">Gets the BioModule execution summary, this is sent as part of the notification email, if configured.<br>
 Summary should not include data content, to avoid unintentional publication of confidential information.<br>
 However, meta-data such as number/size of files can be helpful during debug.<br></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModuleImpl.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a></dt>
<dd>
<div class="block">Returns summary message to be displayed by Email module so must not contain confidential info.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/ClassifierModuleImpl.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.classifier.<a href="../biolockj/module/classifier/ClassifierModuleImpl.html" title="class in biolockj.module.classifier">ClassifierModuleImpl</a></dt>
<dd>
<div class="block">This method extends the basic summary by adding metrics on the input files.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/Demultiplexer.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.implicit.<a href="../biolockj/module/implicit/Demultiplexer.html" title="class in biolockj.module.implicit">Demultiplexer</a></dt>
<dd>
<div class="block">Produces initial count and demultiplexed output count summaries for forward/reverse reads.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/ImportMetadata.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.implicit.<a href="../biolockj/module/implicit/ImportMetadata.html" title="class in biolockj.module.implicit">ImportMetadata</a></dt>
<dd>
<div class="block">The metadata file can be updated several times during pipeline execution.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/BuildQiimeMapping.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/BuildQiimeMapping.html" title="class in biolockj.module.implicit.qiime">BuildQiimeMapping</a></dt>
<dd>
<div class="block">Message summarizes contents of <a href="../biolockj/util/SummaryUtil.html#getScriptDirSummary-biolockj.module.ScriptModule-"><code>scriptDir</code></a>, and
 status of the QIIME mapping files (exists or not).</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime">QiimeClassifier</a></dt>
<dd>
<div class="block">This method extends the classifier summary by adding the Qiime OTU summary metrics.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/RegisterNumReads.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.implicit.<a href="../biolockj/module/implicit/RegisterNumReads.html" title="class in biolockj.module.implicit">RegisterNumReads</a></dt>
<dd>
<div class="block">Produce summary message with min, max, mean, and median number of reads.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/r/R_Module.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.r.<a href="../biolockj/module/r/R_Module.html" title="class in biolockj.module.r">R_Module</a></dt>
<dd>
<div class="block">Produce summary file counts for each file extension in the output directory and the number of log files in the
 temp directory.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/AddMetadataToOtuTables.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.report.<a href="../biolockj/module/report/AddMetadataToOtuTables.html" title="class in biolockj.module.report">AddMetadataToOtuTables</a></dt>
<dd>
<div class="block">Produce summary message with min, max, mean, and median hit ratios</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/Email.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.report.<a href="../biolockj/module/report/Email.html" title="class in biolockj.module.report">Email</a></dt>
<dd>
<div class="block">Summary simply reports the status.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/JsonReport.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.report.<a href="../biolockj/module/report/JsonReport.html" title="class in biolockj.module.report">JsonReport</a></dt>
<dd>
<div class="block">Add summary with unique OTU counts/level, and for each level, give # reads with a gap in the taxonomy assignment.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/ScriptModuleImpl.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/ScriptModuleImpl.html" title="class in biolockj.module">ScriptModuleImpl</a></dt>
<dd>
<div class="block">Returns summary message to be displayed by Email module so must not contain confidential info.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/Multiplexer.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/Multiplexer.html" title="class in biolockj.module.seq">Multiplexer</a></dt>
<dd>
<div class="block">Produce a summary message with counts on total number of reads and number reads multiplexed.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/SeqFileValidator.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/SeqFileValidator.html" title="class in biolockj.module.seq">SeqFileValidator</a></dt>
<dd>
<div class="block">Produce a summary message with counts on total number of reads and number of valid reads containing a barcode
 defined in the metadata file.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/TrimPrimers.html#getSummary--">getSummary()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/TrimPrimers.html" title="class in biolockj.module.seq">TrimPrimers</a></dt>
<dd>
<div class="block">Output the summary messages generated by the module.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SummaryUtil.html#getSummary--">getSummary()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SummaryUtil.html" title="class in biolockj.util">SummaryUtil</a></dt>
<dd>
<div class="block">Pipeline execution summary.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SummaryUtil.html#getSummaryFile--">getSummaryFile()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SummaryUtil.html" title="class in biolockj.util">SummaryUtil</a></dt>
<dd>
<div class="block">Getter for the summary file.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#getSystemFilePath-java.lang.String-">getSystemFilePath(String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Return filePath with system parameter values replaced.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModule.html#getTempDir--">getTempDir()</a></span> - Method in interface biolockj.module.<a href="../biolockj/module/BioModule.html" title="interface in biolockj.module">BioModule</a></dt>
<dd>
<div class="block">Contains intermediate files generated by the module but not used by the next BioModule.<br>
 The files may contain supplementary information or data that may be helpful during debug or recovery.<br>
 If <a href="../biolockj/BioLockJ.html#PROJECT_DELETE_TEMP_FILES">"project.deleteTempFiles"</a> = <a href="../biolockj/Config.html#TRUE">"Y"</a>, successful pipelines
 delete this directory.<br></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModuleImpl.html#getTempDir--">getTempDir()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a></dt>
<dd>
<div class="block">Returns moduleDir/temp for intermediate files.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/r/R_Module.html#getTimeout--">getTimeout()</a></span> - Method in class biolockj.module.r.<a href="../biolockj/module/r/R_Module.html" title="class in biolockj.module.r">R_Module</a></dt>
<dd>
<div class="block">The R Script should run quickly, timeout = 10 minutes appears to work well.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/ScriptModule.html#getTimeout--">getTimeout()</a></span> - Method in interface biolockj.module.<a href="../biolockj/module/ScriptModule.html" title="interface in biolockj.module">ScriptModule</a></dt>
<dd>
<div class="block">ScriptModule that run scripts can opt to set a timeout (such as used by the R script).</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/ScriptModuleImpl.html#getTimeout--">getTimeout()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/ScriptModuleImpl.html" title="class in biolockj.module">ScriptModuleImpl</a></dt>
<dd>
<div class="block">Default behavior is for scripts to run indefinitely (no timeout).</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/Demultiplexer.html#getValidFwHeaders--">getValidFwHeaders()</a></span> - Method in class biolockj.module.implicit.<a href="../biolockj/module/implicit/Demultiplexer.html" title="class in biolockj.module.implicit">Demultiplexer</a></dt>
<dd>
<div class="block">Get valid forward read headers that belong to reads with a valid barcode or sample identifier.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/Demultiplexer.html#getValidHeaders--">getValidHeaders()</a></span> - Method in class biolockj.module.implicit.<a href="../biolockj/module/implicit/Demultiplexer.html" title="class in biolockj.module.implicit">Demultiplexer</a></dt>
<dd>
<div class="block">This method obtains all valid headers for the forward reads, and returns only headers that also have a matching
 reverse read</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/JsonReport.html#getValidOtus--">getValidOtus()</a></span> - Method in class biolockj.module.report.<a href="../biolockj/module/report/JsonReport.html" title="class in biolockj.module.report">JsonReport</a></dt>
<dd>
<div class="block">Return list of OTUs that passed the filters in <a href="../biolockj/module/r/CalculateStats.html" title="class in biolockj.module.r"><code>CalculateStats</code></a> which has generated
 statistics.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#getVsearchParams--">getVsearchParams()</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime">QiimeClassifier</a></dt>
<dd>
<div class="block">Get Vsearch params from the prop file and format switches for the bash script.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/r16s/QiimeClosedRefClassifier.html#getWorkerScriptFunctions--">getWorkerScriptFunctions()</a></span> - Method in class biolockj.module.classifier.r16s.<a href="../biolockj/module/classifier/r16s/QiimeClosedRefClassifier.html" title="class in biolockj.module.classifier.r16s">QiimeClosedRefClassifier</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/r16s/RdpClassifier.html#getWorkerScriptFunctions--">getWorkerScriptFunctions()</a></span> - Method in class biolockj.module.classifier.r16s.<a href="../biolockj/module/classifier/r16s/RdpClassifier.html" title="class in biolockj.module.classifier.r16s">RdpClassifier</a></dt>
<dd>
<div class="block">This method generates the required bash functions: <a href="../biolockj/module/classifier/r16s/RdpClassifier.html#FUNCTION_RDP">"runRdp"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/Kraken2Classifier.html#getWorkerScriptFunctions--">getWorkerScriptFunctions()</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/Kraken2Classifier.html" title="class in biolockj.module.classifier.wgs">Kraken2Classifier</a></dt>
<dd>
<div class="block">This method generates the required bash function: <a href="../biolockj/module/classifier/wgs/Kraken2Classifier.html#FUNCTION_KRAKEN">"runKraken2"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/KrakenClassifier.html#getWorkerScriptFunctions--">getWorkerScriptFunctions()</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/KrakenClassifier.html" title="class in biolockj.module.classifier.wgs">KrakenClassifier</a></dt>
<dd>
<div class="block">This method generates the required bash functions: <a href="../biolockj/module/classifier/wgs/KrakenClassifier.html#FUNCTION_TRANSLATE">"translate"</a> and <a href="../biolockj/module/classifier/wgs/KrakenClassifier.html#FUNCTION_KRAKEN">"runKraken"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/MetaphlanClassifier.html#getWorkerScriptFunctions--">getWorkerScriptFunctions()</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/MetaphlanClassifier.html" title="class in biolockj.module.classifier.wgs">MetaphlanClassifier</a></dt>
<dd>
<div class="block">This method generates the required bash function: <a href="../biolockj/module/classifier/wgs/MetaphlanClassifier.html#FUNCTION_RUN_METAPHLAN">"runMetaphlan"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/wgs/SlimmClassifier.html#getWorkerScriptFunctions--">getWorkerScriptFunctions()</a></span> - Method in class biolockj.module.classifier.wgs.<a href="../biolockj/module/classifier/wgs/SlimmClassifier.html" title="class in biolockj.module.classifier.wgs">SlimmClassifier</a></dt>
<dd>
<div class="block">This method generates the required bash function: <a href="../biolockj/module/classifier/wgs/SlimmClassifier.html#FUNCTION_ALIGN">"align"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/BuildQiimeMapping.html#getWorkerScriptFunctions--">getWorkerScriptFunctions()</a></span> - Method in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/BuildQiimeMapping.html" title="class in biolockj.module.implicit.qiime">BuildQiimeMapping</a></dt>
<dd>
<div class="block">This method generates the bash function used to reorder columns as per QIIME requirements:
 <a href="../biolockj/module/implicit/qiime/BuildQiimeMapping.html#FUNCTION_REORDER_FIELDS">"reorderColumns"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/r/R_Module.html#getWorkerScriptFunctions--">getWorkerScriptFunctions()</a></span> - Method in class biolockj.module.r.<a href="../biolockj/module/r/R_Module.html" title="class in biolockj.module.r">R_Module</a></dt>
<dd>
<div class="block">This method generates the bash function that calls the R script: runScript.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/ScriptModule.html#getWorkerScriptFunctions--">getWorkerScriptFunctions()</a></span> - Method in interface biolockj.module.<a href="../biolockj/module/ScriptModule.html" title="interface in biolockj.module">ScriptModule</a></dt>
<dd>
<div class="block">Method returns bash functions to add to the worker scripts.<br>
 The default behavior is that no bash functions are needed.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/ScriptModuleImpl.html#getWorkerScriptFunctions--">getWorkerScriptFunctions()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/ScriptModuleImpl.html" title="class in biolockj.module">ScriptModuleImpl</a></dt>
<dd>
<div class="block">Build the docker run command to launch a JavaModule in a Docker container.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/AwkFastaConverter.html#getWorkerScriptFunctions--">getWorkerScriptFunctions()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/AwkFastaConverter.html" title="class in biolockj.module.seq">AwkFastaConverter</a></dt>
<dd>
<div class="block">This method generates the required bash functions used by the module scripts.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/Gunzipper.html#getWorkerScriptFunctions--">getWorkerScriptFunctions()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/Gunzipper.html" title="class in biolockj.module.seq">Gunzipper</a></dt>
<dd>
<div class="block">This method generates the bash function: <a href="../biolockj/module/seq/Gunzipper.html#FUNCTION_GUNZIP">"decompressGzip"</a>.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/PearMergeReads.html#getWorkerScriptFunctions--">getWorkerScriptFunctions()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/PearMergeReads.html" title="class in biolockj.module.seq">PearMergeReads</a></dt>
<dd>
<div class="block">This method generates the required bash functions used by the worker scripts: <a href="../biolockj/module/seq/PearMergeReads.html#FUNCTION_PEAR_MERGE">"mergeReads"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BashScriptBuilder.html#getWorkerScriptLines-java.util.List-">getWorkerScriptLines(List&lt;String&gt;)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BashScriptBuilder.html" title="class in biolockj.util">BashScriptBuilder</a></dt>
<dd>
<div class="block">Pass each line and the current line number to <a href="../biolockj/util/BashScriptBuilder.html#FUNCTION_EXECUTE">"execute"</a></div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BashScriptBuilder.html#getWorkerScriptPath-biolockj.module.ScriptModule-java.lang.String-">getWorkerScriptPath(ScriptModule, String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BashScriptBuilder.html" title="class in biolockj.util">BashScriptBuilder</a></dt>
<dd>
<div class="block">Build the file path for the numbered worker script.</div>
</dd>
<dt><a href="../biolockj/module/seq/Gunzipper.html" title="class in biolockj.module.seq"><span class="typeNameLink">Gunzipper</span></a> - Class in <a href="../biolockj/module/seq/package-summary.html">biolockj.module.seq</a></dt>
<dd>
<div class="block">This BioModule uses gzip to decompress input sequence files.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/Gunzipper.html#Gunzipper--">Gunzipper()</a></span> - Constructor for class biolockj.module.seq.<a href="../biolockj/module/seq/Gunzipper.html" title="class in biolockj.module.seq">Gunzipper</a></dt>
<dd>&nbsp;</dd>
</dl>
<a href="index-1.html">A</a>&nbsp;<a href="index-2.html">B</a>&nbsp;<a href="index-3.html">C</a>&nbsp;<a href="index-4.html">D</a>&nbsp;<a href="index-5.html">E</a>&nbsp;<a href="index-6.html">F</a>&nbsp;<a href="index-7.html">G</a>&nbsp;<a href="index-8.html">H</a>&nbsp;<a href="index-9.html">I</a>&nbsp;<a href="index-10.html">J</a>&nbsp;<a href="index-11.html">K</a>&nbsp;<a href="index-12.html">L</a>&nbsp;<a href="index-13.html">M</a>&nbsp;<a href="index-14.html">N</a>&nbsp;<a href="index-15.html">O</a>&nbsp;<a href="index-16.html">P</a>&nbsp;<a href="index-17.html">Q</a>&nbsp;<a href="index-18.html">R</a>&nbsp;<a href="index-19.html">S</a>&nbsp;<a href="index-20.html">T</a>&nbsp;<a href="index-21.html">U</a>&nbsp;<a href="index-22.html">V</a>&nbsp;<a href="index-23.html">W</a>&nbsp;</div>
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar.bottom">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.bottom.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../overview-summary.html">Overview</a></li>
<li>Package</li>
<li>Class</li>
<li>Use</li>
<li><a href="../overview-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li class="navBarCell1Rev">Index</li>
<li><a href="../help-doc.html">Help</a></li>
<li>&nbsp;&nbsp;</li>
<li><a href="https://github.com/msioda/BioLockJ/wiki" target="_top"><strong>RETURN to WIKI</strong></a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="index-6.html">Prev Letter</a></li>
<li><a href="index-8.html">Next Letter</a></li>
</ul>
<ul class="navList">
<li><a href="../index.html?index-files/index-7.html" target="_top">Frames</a></li>
<li><a href="index-7.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip.navbar.bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
