<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (1.8.0_144) on Sun Feb 24 04:02:32 EST 2019 -->
<title>R-Index</title>
<meta name="date" content="2019-02-24">
<link rel="stylesheet" type="text/css" href="../stylesheet.css" title="Style">
<script type="text/javascript" src="../script.js"></script>
</head>
<body>
<script type="text/javascript"><!--
    try {
        if (location.href.indexOf('is-external=true') == -1) {
            parent.document.title="R-Index";
        }
    }
    catch(err) {
    }
//-->
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar.top">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.top.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../overview-summary.html">Overview</a></li>
<li>Package</li>
<li>Class</li>
<li>Use</li>
<li><a href="../overview-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li class="navBarCell1Rev">Index</li>
<li><a href="../help-doc.html">Help</a></li>
<li>&nbsp;&nbsp;</li>
<li><a href="https://github.com/msioda/BioLockJ/wiki" target="_top"><strong>RETURN to WIKI</strong></a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="index-17.html">Prev Letter</a></li>
<li><a href="index-19.html">Next Letter</a></li>
</ul>
<ul class="navList">
<li><a href="../index.html?index-files/index-18.html" target="_top">Frames</a></li>
<li><a href="index-18.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip.navbar.top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<div class="contentContainer"><a href="index-1.html">A</a>&nbsp;<a href="index-2.html">B</a>&nbsp;<a href="index-3.html">C</a>&nbsp;<a href="index-4.html">D</a>&nbsp;<a href="index-5.html">E</a>&nbsp;<a href="index-6.html">F</a>&nbsp;<a href="index-7.html">G</a>&nbsp;<a href="index-8.html">H</a>&nbsp;<a href="index-9.html">I</a>&nbsp;<a href="index-10.html">J</a>&nbsp;<a href="index-11.html">K</a>&nbsp;<a href="index-12.html">L</a>&nbsp;<a href="index-13.html">M</a>&nbsp;<a href="index-14.html">N</a>&nbsp;<a href="index-15.html">O</a>&nbsp;<a href="index-16.html">P</a>&nbsp;<a href="index-17.html">Q</a>&nbsp;<a href="index-18.html">R</a>&nbsp;<a href="index-19.html">S</a>&nbsp;<a href="index-20.html">T</a>&nbsp;<a href="index-21.html">U</a>&nbsp;<a href="index-22.html">V</a>&nbsp;<a href="index-23.html">W</a>&nbsp;<a name="I:R">
<!--   -->
</a>
<h2 class="title">R</h2>
<dl>
<dt><span class="memberNameLink"><a href="../biolockj/util/RMetaUtil.html#R_EXCLUDE_FIELDS">R_EXCLUDE_FIELDS</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/RMetaUtil.html" title="class in biolockj.util">RMetaUtil</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> List property: "r.excludeFields"<br>
 R reports must contain at least one valid nominal or numeric metadata field.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#R_EXT">R_EXT</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block">File extension of BioLockJ generated R Scripts: ".R"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#R_INTERNAL_RUN_HN2">R_INTERNAL_RUN_HN2</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> Boolean property to signal R scripts to build HumanN2 reports</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RMetaUtil.html#R_MDS_REPORT_FIELDS">R_MDS_REPORT_FIELDS</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/RMetaUtil.html" title="class in biolockj.util">RMetaUtil</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> List property: "r_PlotMds.reportFields"<br>
 Fields listed here must exist in the metadata file.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RMetaUtil.html#R_NOMINAL_FIELDS">R_NOMINAL_FIELDS</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/RMetaUtil.html" title="class in biolockj.util">RMetaUtil</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> List property: "r.nominalFields"<br>
 Override default property type by explicitly listing it as nominal.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RMetaUtil.html#R_NUMERIC_FIELDS">R_NUMERIC_FIELDS</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/RMetaUtil.html" title="class in biolockj.util">RMetaUtil</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> List property: "r.numericFields"<br>
 Override default property type by explicitly listing it as numeric.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#R_PLOT_EFFECT_SIZE_DISABLE_FC">R_PLOT_EFFECT_SIZE_DISABLE_FC</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> Boolean property to disable fold change plots: "r_PlotEffectSize.disableFoldChange"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RMetaUtil.html#R_REPORT_FIELDS">R_REPORT_FIELDS</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/RMetaUtil.html" title="class in biolockj.util">RMetaUtil</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> List property: "r.reportFields"<br>
 R reports must contain at least one valid field.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/RarefySeqs.html#rarefy-java.io.File-">rarefy(File)</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/RarefySeqs.html" title="class in biolockj.module.seq">RarefySeqs</a></dt>
<dd>
<div class="block">Builds the rarefied file if too many seqs found, or adds files with too few samples to the list of bad samples.</div>
</dd>
<dt><a href="../biolockj/module/seq/RarefySeqs.html" title="class in biolockj.module.seq"><span class="typeNameLink">RarefySeqs</span></a> - Class in <a href="../biolockj/module/seq/package-summary.html">biolockj.module.seq</a></dt>
<dd>
<div class="block">This BioModule imposes a minimum and/or maximum number of reads per sample.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/RarefySeqs.html#RarefySeqs--">RarefySeqs()</a></span> - Constructor for class biolockj.module.seq.<a href="../biolockj/module/seq/RarefySeqs.html" title="class in biolockj.module.seq">RarefySeqs</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/r16s/RdpClassifier.html#RDP_JAR">RDP_JAR</a></span> - Static variable in class biolockj.module.classifier.r16s.<a href="../biolockj/module/classifier/r16s/RdpClassifier.html" title="class in biolockj.module.classifier.r16s">RdpClassifier</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> File property for RDP java executable JAR: "rdp.jar"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/r16s/RdpClassifier.html#RDP_PARAMS">RDP_PARAMS</a></span> - Static variable in class biolockj.module.classifier.r16s.<a href="../biolockj/module/classifier/r16s/RdpClassifier.html" title="class in biolockj.module.classifier.r16s">RdpClassifier</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> List property for RDP java executable JAR runtime params: "exe.rdpParams"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#RDP_THRESHOLD_SCORE">RDP_THRESHOLD_SCORE</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> String property: "rdp.minThresholdScore"<br>
 RdpParser will ignore OTU assignments below the threshold score (0-100)</div>
</dd>
<dt><a href="../biolockj/module/classifier/r16s/RdpClassifier.html" title="class in biolockj.module.classifier.r16s"><span class="typeNameLink">RdpClassifier</span></a> - Class in <a href="../biolockj/module/classifier/r16s/package-summary.html">biolockj.module.classifier.r16s</a></dt>
<dd>
<div class="block">This BioModule uses RDP to assign taxonomy to 16s sequences.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/classifier/r16s/RdpClassifier.html#RdpClassifier--">RdpClassifier()</a></span> - Constructor for class biolockj.module.classifier.r16s.<a href="../biolockj/module/classifier/r16s/RdpClassifier.html" title="class in biolockj.module.classifier.r16s">RdpClassifier</a></dt>
<dd>&nbsp;</dd>
<dt><a href="../biolockj/node/r16s/RdpNode.html" title="class in biolockj.node.r16s"><span class="typeNameLink">RdpNode</span></a> - Class in <a href="../biolockj/node/r16s/package-summary.html">biolockj.node.r16s</a></dt>
<dd>
<div class="block">This class represents one line of <a href="../biolockj/module/classifier/r16s/RdpClassifier.html" title="class in biolockj.module.classifier.r16s"><code>RdpClassifier</code></a> output.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/r16s/RdpNode.html#RdpNode-java.lang.String-java.lang.String-">RdpNode(String, String)</a></span> - Constructor for class biolockj.node.r16s.<a href="../biolockj/node/r16s/RdpNode.html" title="class in biolockj.node.r16s">RdpNode</a></dt>
<dd>
<div class="block">Constructor called one line of RDP output.</div>
</dd>
<dt><a href="../biolockj/module/implicit/parser/r16s/RdpParser.html" title="class in biolockj.module.implicit.parser.r16s"><span class="typeNameLink">RdpParser</span></a> - Class in <a href="../biolockj/module/implicit/parser/r16s/package-summary.html">biolockj.module.implicit.parser.r16s</a></dt>
<dd>
<div class="block">This BioModule parses RDP output files to build standard OTU abundance tables.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/r16s/RdpParser.html#RdpParser--">RdpParser()</a></span> - Constructor for class biolockj.module.implicit.parser.r16s.<a href="../biolockj/module/implicit/parser/r16s/RdpParser.html" title="class in biolockj.module.implicit.parser.r16s">RdpParser</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Properties.html#readProps-java.io.File-biolockj.Properties-">readProps(File, Properties)</a></span> - Static method in class biolockj.<a href="../biolockj/Properties.html" title="class in biolockj">Properties</a></dt>
<dd>
<div class="block">Read the properties defined in the required propFile and defaultProps (if included) to build Config.<br>
 Properties in propFile will override the defaultProps.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/MetaUtil.html#refreshCache--">refreshCache()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/MetaUtil.html" title="class in biolockj.util">MetaUtil</a></dt>
<dd>
<div class="block">Refresh the metadata cache.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Pipeline.html#refreshOutputMetadata-biolockj.module.BioModule-">refreshOutputMetadata(BioModule)</a></span> - Static method in class biolockj.<a href="../biolockj/Pipeline.html" title="class in biolockj">Pipeline</a></dt>
<dd>
<div class="block">Call <a href="../biolockj/util/MetaUtil.html" title="class in biolockj.util"><code>MetaUtil</code></a> to refresh the metadata cache if a new metadata file was output by the
 bioModule .</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Pipeline.html#refreshRCacheIfNeeded-biolockj.module.BioModule-">refreshRCacheIfNeeded(BioModule)</a></span> - Static method in class biolockj.<a href="../biolockj/Pipeline.html" title="class in biolockj">Pipeline</a></dt>
<dd>
<div class="block">Refresh R cache if about to run the 1st R module.</div>
</dd>
<dt><a href="../biolockj/module/implicit/RegisterNumReads.html" title="class in biolockj.module.implicit"><span class="typeNameLink">RegisterNumReads</span></a> - Class in <a href="../biolockj/module/implicit/package-summary.html">biolockj.module.implicit</a></dt>
<dd>
<div class="block">This BioModule parses sequence file to count the number of reads per sample.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/RegisterNumReads.html#RegisterNumReads--">RegisterNumReads()</a></span> - Constructor for class biolockj.module.implicit.<a href="../biolockj/module/implicit/RegisterNumReads.html" title="class in biolockj.module.implicit">RegisterNumReads</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#registerPairedReadStatus--">registerPairedReadStatus()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Inspect the pipeline input files to determine if input includes paired reads.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#registerRuntimeParameters-java.lang.String:A-">registerRuntimeParameters(String[])</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">Register and verify the runtime parameters.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/SeqFileValidator.html#removeBadFiles--">removeBadFiles()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/SeqFileValidator.html" title="class in biolockj.module.seq">SeqFileValidator</a></dt>
<dd>
<div class="block">Remove sequence files in which all reads failed validation checks, leaving only an empty file.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/MetaUtil.html#removeColumn-java.lang.String-java.io.File-">removeColumn(String, File)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/MetaUtil.html" title="class in biolockj.util">MetaUtil</a></dt>
<dd>
<div class="block">Remove the metadata column from the metadata file</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BioLockJUtil.html#removeIgnoredAndEmptyFiles-java.util.Collection-">removeIgnoredAndEmptyFiles(Collection&lt;File&gt;)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BioLockJUtil.html" title="class in biolockj.util">BioLockJUtil</a></dt>
<dd>
<div class="block">Remove ignored and empty files from the input files.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BioLockJUtil.html#removeOuterQuotes-java.lang.String-">removeOuterQuotes(String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BioLockJUtil.html" title="class in biolockj.util">BioLockJUtil</a></dt>
<dd>
<div class="block">Remove the outer single or double quotes of the given value.<br>
 Quotes are only removed if quotes are found as very 1st and last characters.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/BioLockJUtil.html#removeQuotes-java.lang.String-">removeQuotes(String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/BioLockJUtil.html" title="class in biolockj.util">BioLockJUtil</a></dt>
<dd>
<div class="block">Remove all single and double quotation marks found in value.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/BioLockJ.html#removeTempFiles--">removeTempFiles()</a></span> - Static method in class biolockj.<a href="../biolockj/BioLockJ.html" title="class in biolockj">BioLockJ</a></dt>
<dd>
<div class="block">Delete all <a href="../biolockj/module/BioModule.html" title="interface in biolockj.module"><code>BioModule</code></a>/<a href="../biolockj/module/BioModule.html#TEMP_DIR">"temp"</a> folders.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/qiime/QiimeClassifier.html#REP_SET">REP_SET</a></span> - Static variable in class biolockj.module.implicit.qiime.<a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime">QiimeClassifier</a></dt>
<dd>
<div class="block">Directory created by <a href="../biolockj/module/classifier/r16s/QiimeDeNovoClassifier.html#PICK_OTU_SCRIPT">"pick_de_novo_otus.py"</a> and
 <a href="../biolockj/module/classifier/r16s/QiimeOpenRefClassifier.html#PICK_OTU_SCRIPT">"pick_open_reference_otus.py"</a>: "rep_set"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/node/JsonNode.html#report--">report()</a></span> - Method in class biolockj.node.<a href="../biolockj/node/JsonNode.html" title="class in biolockj.node">JsonNode</a></dt>
<dd>
<div class="block">Print node info.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#REPORT_LOG_BASE">REPORT_LOG_BASE</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> String property: "report.logBase"<br>
 Required to be set to "e" or "10" to build log normalized reports.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#REPORT_MIN_COUNT">REPORT_MIN_COUNT</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> Positive Integer property "report.minCount" defines the minimum table count
 allowed, if a count less that this value is found, it is set to 0.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#REPORT_NUM_HITS">REPORT_NUM_HITS</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> Boolean property: "report.numHits"<br>
 If set to <a href="../biolockj/Constants.html#TRUE">"Y"</a>, NUM_OTUS will be added to metadata file by
 <a href="../biolockj/module/implicit/parser/ParserModuleImpl.html" title="class in biolockj.module.implicit.parser"><code>ParserModuleImpl</code></a> and included in R reports</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#REPORT_NUM_READS">REPORT_NUM_READS</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> Boolean property: "report.numReads"<br>
 If set to <a href="../biolockj/Constants.html#TRUE">"Y"</a> and NUM_READS exists in metadata file, NUM_READS will be included in the R reports</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#REPORT_SAMPLE_CUTOFF">REPORT_SAMPLE_CUTOFF</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> Positive Double property "report.scarceSampleCutoff" defines minimum percentage of
 data columns must be non-zero to keep the sample.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#REPORT_SCARCE_CUTOFF">REPORT_SCARCE_CUTOFF</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> Positive Double property "report.scarceCountCutoff" defines minimum percentage of
 samples that must contain a count value for it to be kept.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#REPORT_TAXONOMY_LEVELS">REPORT_TAXONOMY_LEVELS</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> List property: "report.taxonomyLevels"<br>
 This property drives a lot of BioLockJ functionality and determines which taxonomy-levels are reported.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RMetaUtil.html#reportAllFields-biolockj.module.BioModule-">reportAllFields(BioModule)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/RMetaUtil.html" title="class in biolockj.util">RMetaUtil</a></dt>
<dd>
<div class="block">The override property: <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/util/RMetaUtil.html#R_REPORT_FIELDS">"r.reportFields"</a> can be used to list the metadata
 reportable fields for use in the R modules.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SummaryUtil.html#reportFailure-java.lang.Exception-">reportFailure(Exception)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SummaryUtil.html" title="class in biolockj.util">SummaryUtil</a></dt>
<dd>
<div class="block">Report exception details in the summary</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/MemoryUtil.html#reportMemoryUsage-java.lang.String-">reportMemoryUsage(String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/MemoryUtil.html" title="class in biolockj.util">MemoryUtil</a></dt>
<dd>
<div class="block">Print stats on memory usage: Max, Total, ∆Total, %Free, Free, ∆Free.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SummaryUtil.html#reportSuccess-biolockj.module.BioModule-">reportSuccess(BioModule)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SummaryUtil.html" title="class in biolockj.util">SummaryUtil</a></dt>
<dd>
<div class="block">After each module completes, this method is called to track the execution summary.<br>
 If module is null, the pipeline is complete/successful.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/SeqFileValidator.html#REQUIRE_EUQL_NUM_PAIRS">REQUIRE_EUQL_NUM_PAIRS</a></span> - Static variable in class biolockj.module.seq.<a href="../biolockj/module/seq/SeqFileValidator.html" title="class in biolockj.module.seq">SeqFileValidator</a></dt>
<dd>
<div class="block"><a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a> Boolean property "seqFileValidator.requireEqualNumPairs" determines if module requires equal
 number of forward and reverse reads (simple check).</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#requireBoolean-biolockj.module.BioModule-java.lang.String-">requireBoolean(BioModule, String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Required to return a valid boolean  or </div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#requireDoubleVal-biolockj.module.BioModule-java.lang.String-">requireDoubleVal(BioModule, String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Requires valid double value</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#requireExistingDir-biolockj.module.BioModule-java.lang.String-">requireExistingDir(BioModule, String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Requires valid existing directory.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#requireExistingDirs-biolockj.module.BioModule-java.lang.String-">requireExistingDirs(BioModule, String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Requires valid list of file directories</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#requireExistingFile-biolockj.module.BioModule-java.lang.String-">requireExistingFile(BioModule, String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Require valid existing file</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#requireInteger-biolockj.module.BioModule-java.lang.String-">requireInteger(BioModule, String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Requires valid integer value</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#requireList-biolockj.module.BioModule-java.lang.String-">requireList(BioModule, String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Require valid list property</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#requirePositiveDouble-biolockj.module.BioModule-java.lang.String-">requirePositiveDouble(BioModule, String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Require valid positive double value</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#requirePositiveInteger-biolockj.module.BioModule-java.lang.String-">requirePositiveInteger(BioModule, String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Require valid positive integer value</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#requireSet-biolockj.module.BioModule-java.lang.String-">requireSet(BioModule, String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Require valid Set value</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Config.html#requireString-biolockj.module.BioModule-java.lang.String-">requireString(BioModule, String)</a></span> - Static method in class biolockj.<a href="../biolockj/Config.html" title="class in biolockj">Config</a></dt>
<dd>
<div class="block">Require valid String value</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/ModuleUtil.html#requireSubDir-biolockj.module.BioModule-java.lang.String-">requireSubDir(BioModule, String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/ModuleUtil.html" title="class in biolockj.util">ModuleUtil</a></dt>
<dd>
<div class="block">Get BioModule subdirectory File object with given name.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SummaryUtil.html#resetModuleSummary-biolockj.module.BioModule-">resetModuleSummary(BioModule)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SummaryUtil.html" title="class in biolockj.util">SummaryUtil</a></dt>
<dd>
<div class="block">Modules can be forced to reset to incomplete status.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#RESTART_FLAG">RESTART_FLAG</a></span> - Static variable in class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>
<div class="block">Restart pipeline runtime parameter switch: "-r"</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Constants.html#RETURN">RETURN</a></span> - Static variable in class biolockj.<a href="../biolockj/Constants.html" title="class in biolockj">Constants</a></dt>
<dd>
<div class="block">Return character constant *backslash-n*</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/BioModuleImpl.html#RETURN">RETURN</a></span> - Static variable in class biolockj.module.<a href="../biolockj/module/BioModuleImpl.html" title="class in biolockj.module">BioModuleImpl</a></dt>
<dd>
<div class="block">Return character constant *backslash-n*</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/SeqUtil.html#reverseComplement-java.lang.String-">reverseComplement(String)</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/SeqUtil.html" title="class in biolockj.util">SeqUtil</a></dt>
<dd>
<div class="block">Return the DNA reverse compliment for the input dna parameter.</div>
</dd>
<dt><a href="../biolockj/util/RMetaUtil.html" title="class in biolockj.util"><span class="typeNameLink">RMetaUtil</span></a> - Class in <a href="../biolockj/util/package-summary.html">biolockj.util</a></dt>
<dd>
<div class="block">This utility is used to validate the metadata to help ensure the format is valid R script input.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/DockerUtil.html#runAws--">runAws()</a></span> - Static method in class biolockj.util.<a href="../biolockj/util/DockerUtil.html" title="class in biolockj.util">DockerUtil</a></dt>
<dd>
<div class="block">Return TRUE if running on AWS (based on Config props).</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Pipeline.html#runDirectModule-java.lang.Integer-">runDirectModule(Integer)</a></span> - Static method in class biolockj.<a href="../biolockj/Pipeline.html" title="class in biolockj">Pipeline</a></dt>
<dd>
<div class="block">If moduleName is null, run all modules, otherwise only run the specified module.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Job.html#runJob-java.lang.String:A-java.lang.String-">runJob(String[], String)</a></span> - Method in class biolockj.<a href="../biolockj/Job.html" title="class in biolockj">Job</a></dt>
<dd>
<div class="block">Execute the command args and log output with label.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/Demultiplexer.html#runModule--">runModule()</a></span> - Method in class biolockj.module.implicit.<a href="../biolockj/module/implicit/Demultiplexer.html" title="class in biolockj.module.implicit">Demultiplexer</a></dt>
<dd>
<div class="block">Module execution summary:<br>
 
 Execute <a href="../biolockj/module/implicit/Demultiplexer.html#breakUpFiles--"><code>Demultiplexer.breakUpFiles()</code></a> to split the multiplex file into smaller size for processing
 Execute <a href="../biolockj/module/implicit/Demultiplexer.html#getValidHeaders--"><code>Demultiplexer.getValidHeaders()</code></a> to obtain list of valid headers matched to sample ID with the metadata
 file and also verifies matching forward and reverse read headers if demuliplexing paired reads.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/ParserModuleImpl.html#runModule--">runModule()</a></span> - Method in class biolockj.module.implicit.parser.<a href="../biolockj/module/implicit/parser/ParserModuleImpl.html" title="class in biolockj.module.implicit.parser">ParserModuleImpl</a></dt>
<dd>
<div class="block">Parsers execute a task with 3 core functions:
 
 <a href="../biolockj/module/implicit/parser/ParserModuleImpl.html#parseSamples--"><code>ParserModuleImpl.parseSamples()</code></a> - generates <a href="../biolockj/node/ParsedSample.html" title="class in biolockj.node"><code>ParsedSample</code></a>s
 <a href="../biolockj/module/implicit/parser/ParserModuleImpl.html#buildOtuCountFiles--"><code>ParserModuleImpl.buildOtuCountFiles()</code></a> - builds OTU tree tables from the <a href="../biolockj/node/ParsedSample.html" title="class in biolockj.node"><code>ParsedSample</code></a>s
 </div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/r16s/QiimeParser.html#runModule--">runModule()</a></span> - Method in class biolockj.module.implicit.parser.r16s.<a href="../biolockj/module/implicit/parser/r16s/QiimeParser.html" title="class in biolockj.module.implicit.parser.r16s">QiimeParser</a></dt>
<dd>
<div class="block">To parse the taxonomy level reports output by <a href="../biolockj/module/implicit/qiime/QiimeClassifier.html" title="class in biolockj.module.implicit.qiime"><code>QiimeClassifier</code></a>:
 
 Call <a href="../biolockj/module/implicit/parser/r16s/QiimeParser.html#getInputFiles--"><code>QiimeParser.getInputFiles()</code></a>
 Call <a href="../biolockj/module/implicit/parser/r16s/QiimeParser.html#initializeMaps--"><code>QiimeParser.initializeMaps()</code></a>
 Call <a href="../biolockj/module/implicit/parser/r16s/QiimeParser.html#setOrderedQiimeIDs-java.io.File-"><code>QiimeParser.setOrderedQiimeIDs(File)</code></a>
 Create <a href="../biolockj/node/ParsedSample.html" title="class in biolockj.node"><code>ParsedSample</code></a> for the <a href="../biolockj/node/OtuNodeImpl.html#getSampleId--"><code>OtuNodeImpl.getSampleId()</code></a> if not
 yet created.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/parser/wgs/Humann2Parser.html#runModule--">runModule()</a></span> - Method in class biolockj.module.implicit.parser.wgs.<a href="../biolockj/module/implicit/parser/wgs/Humann2Parser.html" title="class in biolockj.module.implicit.parser.wgs">Humann2Parser</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/implicit/RegisterNumReads.html#runModule--">runModule()</a></span> - Method in class biolockj.module.implicit.<a href="../biolockj/module/implicit/RegisterNumReads.html" title="class in biolockj.module.implicit">RegisterNumReads</a></dt>
<dd>
<div class="block">Register number of reads for each sample by parsing each file and counting number of lines.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/JavaModule.html#runModule--">runModule()</a></span> - Method in interface biolockj.module.<a href="../biolockj/module/JavaModule.html" title="interface in biolockj.module">JavaModule</a></dt>
<dd>
<div class="block">This method executes the modules primary Java routine.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/JavaModuleImpl.html#runModule--">runModule()</a></span> - Method in class biolockj.module.<a href="../biolockj/module/JavaModuleImpl.html" title="class in biolockj.module">JavaModuleImpl</a></dt>
<dd>&nbsp;</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/report/JsonReport.html#runModule--">runModule()</a></span> - Method in class biolockj.module.report.<a href="../biolockj/module/report/JsonReport.html" title="class in biolockj.module.report">JsonReport</a></dt>
<dd>
<div class="block">Obtain parsed sample data, build root node, and create the jsonMap by passing both to buildMap().</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/Multiplexer.html#runModule--">runModule()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/Multiplexer.html" title="class in biolockj.module.seq">Multiplexer</a></dt>
<dd>
<div class="block">Execute <a href="../biolockj/module/seq/Multiplexer.html#multiplex-java.io.File-"><code>Multiplexer.multiplex(File)</code></a> on each input file</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/RarefySeqs.html#runModule--">runModule()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/RarefySeqs.html" title="class in biolockj.module.seq">RarefySeqs</a></dt>
<dd>
<div class="block">For each file with number reads outside of <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/module/seq/RarefySeqs.html#INPUT_RAREFYING_MIN"><code>RarefySeqs.INPUT_RAREFYING_MIN</code></a> and
 <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/module/seq/RarefySeqs.html#INPUT_RAREFYING_MAX"><code>RarefySeqs.INPUT_RAREFYING_MAX</code></a> values, generate a new sequence file from a shuffled list of
 its sequences.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/SeqFileValidator.html#runModule--">runModule()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/SeqFileValidator.html" title="class in biolockj.module.seq">SeqFileValidator</a></dt>
<dd>
<div class="block">Cache sampleIds to compare to validated sampleIds post-processing.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/module/seq/TrimPrimers.html#runModule--">runModule()</a></span> - Method in class biolockj.module.seq.<a href="../biolockj/module/seq/TrimPrimers.html" title="class in biolockj.module.seq">TrimPrimers</a></dt>
<dd>
<div class="block">Trims primers from fasta or fastq files.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/BioLockJ.html#runPipeline--">runPipeline()</a></span> - Static method in class biolockj.<a href="../biolockj/BioLockJ.html" title="class in biolockj">BioLockJ</a></dt>
<dd>
<div class="block">Execution summary:<br>
 
 Call <a href="../biolockj/Pipeline.html#initializePipeline--"><code>Pipeline.initializePipeline()</code></a> to initialize Pipeline modules
 For direct module execution call <a href="../biolockj/Pipeline.html#runDirectModule-java.lang.Integer-"><code>Pipeline.runDirectModule(Integer)</code></a>
 Otherwise execute <a href="../biolockj/Pipeline.html#runPipeline--"><code>Pipeline.runPipeline()</code></a>
 If <a href="../biolockj/Config.html" title="class in biolockj"><code>Config</code></a>.<a href="../biolockj/Constants.html#PROJECT_DELETE_TEMP_FILES">"project.deleteTempFiles"</a> =
 <a href="../biolockj/Constants.html#TRUE">"Y"</a>, Call <a href="../biolockj/BioLockJ.html#removeTempFiles--"><code>BioLockJ.removeTempFiles()</code></a> to delete tem files
 Call <a href="../biolockj/BioLockJ.html#markProjectStatus-java.lang.String-"><code>BioLockJ.markProjectStatus(String)</code></a> to set the overall pipeline status as successful
 </div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/Pipeline.html#runPipeline--">runPipeline()</a></span> - Static method in class biolockj.<a href="../biolockj/Pipeline.html" title="class in biolockj">Pipeline</a></dt>
<dd>
<div class="block">This method initializes and executes the BioModules set in the BioLockJ configuration file.<br></div>
</dd>
<dt><a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util"><span class="typeNameLink">RuntimeParamUtil</span></a> - Class in <a href="../biolockj/util/package-summary.html">biolockj.util</a></dt>
<dd>
<div class="block">This utility processes the application runtime parameters passed to BioLockJ.</div>
</dd>
<dt><span class="memberNameLink"><a href="../biolockj/util/RuntimeParamUtil.html#RuntimeParamUtil--">RuntimeParamUtil()</a></span> - Constructor for class biolockj.util.<a href="../biolockj/util/RuntimeParamUtil.html" title="class in biolockj.util">RuntimeParamUtil</a></dt>
<dd>&nbsp;</dd>
</dl>
<a href="index-1.html">A</a>&nbsp;<a href="index-2.html">B</a>&nbsp;<a href="index-3.html">C</a>&nbsp;<a href="index-4.html">D</a>&nbsp;<a href="index-5.html">E</a>&nbsp;<a href="index-6.html">F</a>&nbsp;<a href="index-7.html">G</a>&nbsp;<a href="index-8.html">H</a>&nbsp;<a href="index-9.html">I</a>&nbsp;<a href="index-10.html">J</a>&nbsp;<a href="index-11.html">K</a>&nbsp;<a href="index-12.html">L</a>&nbsp;<a href="index-13.html">M</a>&nbsp;<a href="index-14.html">N</a>&nbsp;<a href="index-15.html">O</a>&nbsp;<a href="index-16.html">P</a>&nbsp;<a href="index-17.html">Q</a>&nbsp;<a href="index-18.html">R</a>&nbsp;<a href="index-19.html">S</a>&nbsp;<a href="index-20.html">T</a>&nbsp;<a href="index-21.html">U</a>&nbsp;<a href="index-22.html">V</a>&nbsp;<a href="index-23.html">W</a>&nbsp;</div>
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar.bottom">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.bottom.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../overview-summary.html">Overview</a></li>
<li>Package</li>
<li>Class</li>
<li>Use</li>
<li><a href="../overview-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li class="navBarCell1Rev">Index</li>
<li><a href="../help-doc.html">Help</a></li>
<li>&nbsp;&nbsp;</li>
<li><a href="https://github.com/msioda/BioLockJ/wiki" target="_top"><strong>RETURN to WIKI</strong></a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="index-17.html">Prev Letter</a></li>
<li><a href="index-19.html">Next Letter</a></li>
</ul>
<ul class="navList">
<li><a href="../index.html?index-files/index-18.html" target="_top">Frames</a></li>
<li><a href="index-18.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<a name="skip.navbar.bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
